
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Fast and accurate cell segmentation for single-molecule spatial omics">
      
      
        <meta name="author" content="Elyas Heidari">
      
      
        <link rel="canonical" href="https://EliHei2.github.io/segger_dev/api/data/io/">
      
      
      
      
      <link rel="icon" href="../../../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.11">
    
    
      
        <title>segger.data.io - segger Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=ubuntu:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"ubuntu";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../css/extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css">
    
      <link rel="stylesheet" href="https://unpkg.com/termynal@0.0.1/termynal.css">
    
      <link rel="stylesheet" href="../../../termynal.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#seggerdataio" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="segger Documentation" class="md-header__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            segger Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              segger.data.io
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="segger Documentation" class="md-nav__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    segger Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/data_creation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Dataset Creation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../notebooks/segger_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Segger
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../cli/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    CLI
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../models/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Models
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../prediction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Prediction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="seggerdataio">segger.data.io<a class="headerlink" href="#seggerdataio" title="Permanent link">&para;</a></h1>


<div class="doc doc-object doc-module">



<a id="segger.data.io"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="segger.data.io.MerscopeKeys" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">MerscopeKeys</span>


<a href="#segger.data.io.MerscopeKeys" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


        <p>Keys for <em>MERSCOPE</em> data (Vizgen platform).</p>










  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.MerscopeSample" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">MerscopeSample</span>


<a href="#segger.data.io.MerscopeSample" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">MerscopeSample</span><span class="p">(</span><span class="n">transcripts_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">transcripts_radius</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">boundaries_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">embedding_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="SpatialTranscriptomicsSample (segger.data.io.SpatialTranscriptomicsSample)" href="#segger.data.io.SpatialTranscriptomicsSample">SpatialTranscriptomicsSample</a></code></p>








                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/io.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">transcripts_radius</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">boundaries_graph</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">embedding_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">transcripts_radius</span><span class="p">,</span> <span class="n">boundaries_graph</span><span class="p">,</span> <span class="n">embedding_df</span><span class="p">,</span> <span class="n">MerscopeKeys</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="segger.data.io.MerscopeSample.filter_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">filter_transcripts</span>


<a href="#segger.data.io.MerscopeSample.filter_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">filter_transcripts</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">min_qv</span><span class="o">=</span><span class="mf">20.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Filters transcripts based on specific criteria for Merscope using Dask.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dd.DataFrame
The Dask DataFrame containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_qv</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>float, optional
The minimum quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>20.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dd.DataFrame
The filtered Dask DataFrame.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">min_qv</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">20.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters transcripts based on specific criteria for Merscope using Dask.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        transcripts_df : dd.DataFrame</span>
<span class="sd">            The Dask DataFrame containing transcript data.</span>
<span class="sd">        min_qv : float, optional</span>
<span class="sd">            The minimum quality value threshold for filtering transcripts.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dd.DataFrame</span>
<span class="sd">            The filtered Dask DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Add custom Merscope-specific filtering logic if needed</span>
    <span class="c1"># For now, apply only the quality value filter</span>
    <span class="k">return</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">QUALITY_VALUE</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_qv</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.SpatialDataKeys" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">SpatialDataKeys</span>


<a href="#segger.data.io.SpatialDataKeys" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


        <p>Keys for <em>MERSCOPE</em> data (Vizgen platform).</p>










  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.SpatialDataSample" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">SpatialDataSample</span>


<a href="#segger.data.io.SpatialDataSample" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">SpatialDataSample</span><span class="p">(</span><span class="n">transcripts_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">transcripts_radius</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">boundaries_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">embedding_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">feature_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="SpatialTranscriptomicsSample (segger.data.io.SpatialTranscriptomicsSample)" href="#segger.data.io.SpatialTranscriptomicsSample">SpatialTranscriptomicsSample</a></code></p>








                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/io.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">transcripts_radius</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">boundaries_graph</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">embedding_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">feature_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="k">if</span> <span class="n">feature_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># luca: just a quick hack for now, I propose to use dataclasses instead of enums to address this</span>
        <span class="n">SpatialDataKeys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">_value_</span> <span class="o">=</span> <span class="n">feature_name</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;the automatic determination of a feature_name from a SpatialData object is not enabled yet&quot;</span>
        <span class="p">)</span>

    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
        <span class="n">transcripts_df</span><span class="p">,</span> <span class="n">transcripts_radius</span><span class="p">,</span> <span class="n">boundaries_graph</span><span class="p">,</span> <span class="n">embedding_df</span><span class="p">,</span> <span class="n">SpatialDataKeys</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialDataSample.filter_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">filter_transcripts</span>


<a href="#segger.data.io.SpatialDataSample.filter_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">filter_transcripts</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">min_qv</span><span class="o">=</span><span class="mf">20.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Filters transcripts based on quality value and removes unwanted transcripts for Xenium using Dask.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The Dask DataFrame containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_qv</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>20.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dd.DataFrame: The filtered Dask DataFrame.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">min_qv</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">20.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters transcripts based on quality value and removes unwanted transcripts for Xenium using Dask.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        transcripts_df (dd.DataFrame): The Dask DataFrame containing transcript data.</span>
<span class="sd">        min_qv (float, optional): The minimum quality value threshold for filtering transcripts.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dd.DataFrame: The filtered Dask DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filter_codewords</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;NegControlProbe_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;antisense_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NegControlCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;BLANK_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DeprecatedCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;UnassignedCodeword_&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Ensure FEATURE_NAME is a string type for proper filtering (compatible with Dask)</span>
    <span class="c1"># Handle potential bytes to string conversion for Dask DataFrame</span>
    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">api</span><span class="o">.</span><span class="n">types</span><span class="o">.</span><span class="n">is_object_dtype</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]):</span>
        <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
        <span class="p">)</span>

    <span class="c1"># Apply the quality value filter using Dask</span>
    <span class="n">mask_quality</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">QUALITY_VALUE</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_qv</span>

    <span class="c1"># Apply the filter for unwanted codewords using Dask string functions</span>
    <span class="n">mask_codewords</span> <span class="o">=</span> <span class="o">~</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">filter_codewords</span><span class="p">)</span>

    <span class="c1"># Combine the filters and return the filtered Dask DataFrame</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">mask_quality</span> <span class="o">&amp;</span> <span class="n">mask_codewords</span>

    <span class="c1"># Return the filtered DataFrame lazily</span>
    <span class="k">return</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.SpatialTranscriptomicsDataset" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">SpatialTranscriptomicsDataset</span>


<a href="#segger.data.io.SpatialTranscriptomicsDataset" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">SpatialTranscriptomicsDataset</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">transform</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pre_transform</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pre_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="torch_geometric.data.InMemoryDataset">InMemoryDataset</span></code></p>


        <p>A dataset class for handling SpatialTranscriptomics spatial transcriptomics data.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="segger.data.io.SpatialTranscriptomicsDataset.root">root</span></code></td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The root directory where the dataset is stored.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="segger.data.io.SpatialTranscriptomicsDataset.transform">transform</span></code></td>
            <td>
                  <code><span title="callable">callable</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="segger.data.io.SpatialTranscriptomicsDataset.pre_transform">pre_transform</span></code></td>
            <td>
                  <code><span title="callable">callable</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="segger.data.io.SpatialTranscriptomicsDataset.pre_filter">pre_filter</span></code></td>
            <td>
                  <code><span title="callable">callable</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function that takes in a Data object and returns a boolean indicating whether to keep it.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

        <p>Initialize the SpatialTranscriptomicsDataset.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>root</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Root directory where the dataset is stored.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>transform</code>
            </td>
            <td>
                  <code><span title="callable">callable</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pre_transform</code>
            </td>
            <td>
                  <code><span title="callable">callable</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pre_filter</code>
            </td>
            <td>
                  <code><span title="callable">callable</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function that takes in a Data object and returns a boolean indicating whether to keep it. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>







                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/utils.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span> <span class="n">root</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">transform</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">pre_transform</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">pre_filter</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialize the SpatialTranscriptomicsDataset.</span>

<span class="sd">    Args:</span>
<span class="sd">        root (str): Root directory where the dataset is stored.</span>
<span class="sd">        transform (callable, optional): A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</span>
<span class="sd">        pre_transform (callable, optional): A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</span>
<span class="sd">        pre_filter (callable, optional): A function that takes in a Data object and returns a boolean indicating whether to keep it. Defaults to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">transform</span><span class="p">,</span> <span class="n">pre_transform</span><span class="p">,</span> <span class="n">pre_filter</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="segger.data.io.SpatialTranscriptomicsDataset.processed_file_names" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">processed_file_names</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#segger.data.io.SpatialTranscriptomicsDataset.processed_file_names" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">processed_file_names</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Return a list of processed file names in the processed directory.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of processed file names.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="segger.data.io.SpatialTranscriptomicsDataset.raw_file_names" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">raw_file_names</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#segger.data.io.SpatialTranscriptomicsDataset.raw_file_names" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">raw_file_names</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Return a list of raw file names in the raw directory.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of raw file names.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsDataset.download" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">download</span>


<a href="#segger.data.io.SpatialTranscriptomicsDataset.download" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">download</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Download the raw data. This method should be overridden if you need to download the data.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">download</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Download the raw data. This method should be overridden if you need to download the data.&quot;&quot;&quot;</span>
    <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsDataset.get" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get</span>


<a href="#segger.data.io.SpatialTranscriptomicsDataset.get" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Get a processed data object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>idx</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Index of the data object to retrieve.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Data</code></td>            <td>
                  <code><span title="torch_geometric.data.Data">Data</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The processed data object.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Data</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get a processed data object.</span>

<span class="sd">    Args:</span>
<span class="sd">        idx (int): Index of the data object to retrieve.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Data: The processed data object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">processed_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">processed_file_names</span><span class="p">[</span><span class="n">idx</span><span class="p">]))</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">to_dense</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsDataset.len" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">len</span>


<a href="#segger.data.io.SpatialTranscriptomicsDataset.len" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nb">len</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Return the number of processed files.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of processed files.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the number of processed files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: Number of processed files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">processed_file_names</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsDataset.process" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">process</span>


<a href="#segger.data.io.SpatialTranscriptomicsDataset.process" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">process</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Process the raw data and save it to the processed directory. This method should be overridden if you need to process the data.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process the raw data and save it to the processed directory. This method should be overridden if you need to process the data.&quot;&quot;&quot;</span>
    <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.SpatialTranscriptomicsKeys" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">SpatialTranscriptomicsKeys</span>


<a href="#segger.data.io.SpatialTranscriptomicsKeys" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


        <p>Unified keys for spatial transcriptomics data, supporting multiple platforms.</p>










  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.SpatialTranscriptomicsSample" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">SpatialTranscriptomicsSample</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">SpatialTranscriptomicsSample</span><span class="p">(</span><span class="n">transcripts_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">transcripts_radius</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">boundaries_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">embedding_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="abc.ABC">ABC</span></code></p>


        <p>Initialize the SpatialTranscriptomicsSample class.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
                  <code><span title="segger.data.utils.pd.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A DataFrame containing transcript data.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_radius</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Radius for transcripts in the analysis.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>boundaries_graph</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to include boundaries (e.g., nucleus, cell) graph information.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keys</code>
            </td>
            <td>
                  <code><span title="segger.data.utils.Dict">Dict</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The enum class containing key mappings specific to the dataset.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>







                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/io.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">transcripts_radius</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">boundaries_graph</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">embedding_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">keys</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialize the SpatialTranscriptomicsSample class.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcripts_df (pd.DataFrame, optional): A DataFrame containing transcript data.</span>
<span class="sd">        transcripts_radius (int, optional): Radius for transcripts in the analysis.</span>
<span class="sd">        boundaries_graph (bool, optional): Whether to include boundaries (e.g., nucleus, cell) graph information.</span>
<span class="sd">        keys (Dict, optional): The enum class containing key mappings specific to the dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">transcripts_df</span> <span class="o">=</span> <span class="n">transcripts_df</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">transcripts_radius</span> <span class="o">=</span> <span class="n">transcripts_radius</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">boundaries_graph</span> <span class="o">=</span> <span class="n">boundaries_graph</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">keys</span> <span class="o">=</span> <span class="n">keys</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">embedding_df</span> <span class="o">=</span> <span class="n">embedding_df</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">current_embedding</span> <span class="o">=</span> <span class="s2">&quot;token&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.build_pyg_data_from_tile" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">build_pyg_data_from_tile</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.build_pyg_data_from_tile" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">build_pyg_data_from_tile</span><span class="p">(</span><span class="n">boundaries_df</span><span class="p">,</span> <span class="n">transcripts_df</span><span class="p">,</span> <span class="n">r_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span> <span class="n">k_tx</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;kd_tree&#39;</span><span class="p">,</span> <span class="n">gpu</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">scale_boundaries</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Builds PyG data from a tile of boundaries and transcripts data using Dask utilities for efficient processing.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>boundaries_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dask DataFrame containing boundaries data (e.g., nucleus, cell).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dask DataFrame containing transcripts data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>r_tx</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Radius for building the transcript-to-transcript graph.</p>
              </div>
            </td>
            <td>
                  <code>5.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>k_tx</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for the tx-tx graph.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>method</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Method for computing edge indices (e.g., 'kd_tree', 'faiss').</p>
              </div>
            </td>
            <td>
                  <code>&#39;kd_tree&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gpu</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to use GPU acceleration for edge index computation.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>workers</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of workers to use for parallel processing.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_boundaries</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The factor by which to scale the boundary polygons. Default is 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>HeteroData</code></td>            <td>
                  <code><span title="segger.data.utils.HeteroData">HeteroData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PyG Heterogeneous Data object.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">build_pyg_data_from_tile</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">boundaries_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">r_tx</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5.0</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;kd_tree&quot;</span><span class="p">,</span>
    <span class="n">gpu</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">scale_boundaries</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">HeteroData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Builds PyG data from a tile of boundaries and transcripts data using Dask utilities for efficient processing.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        boundaries_df (dd.DataFrame): Dask DataFrame containing boundaries data (e.g., nucleus, cell).</span>
<span class="sd">        transcripts_df (dd.DataFrame): Dask DataFrame containing transcripts data.</span>
<span class="sd">        r_tx (float): Radius for building the transcript-to-transcript graph.</span>
<span class="sd">        k_tx (int): Number of nearest neighbors for the tx-tx graph.</span>
<span class="sd">        method (str, optional): Method for computing edge indices (e.g., &#39;kd_tree&#39;, &#39;faiss&#39;).</span>
<span class="sd">        gpu (bool, optional): Whether to use GPU acceleration for edge index computation.</span>
<span class="sd">        workers (int, optional): Number of workers to use for parallel processing.</span>
<span class="sd">        scale_boundaries (float, optional): The factor by which to scale the boundary polygons. Default is 1.0.</span>

<span class="sd">    Returns:</span>
<span class="sd">        HeteroData: PyG Heterogeneous Data object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initialize the PyG HeteroData object</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">HeteroData</span><span class="p">()</span>

    <span class="c1"># Lazily compute boundaries geometries using Dask</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing boundaries geometries...&quot;</span><span class="p">)</span>
    <span class="n">bd_gdf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_boundaries_geometries</span><span class="p">(</span><span class="n">boundaries_df</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="n">scale_boundaries</span><span class="p">)</span>
    <span class="n">bd_gdf</span> <span class="o">=</span> <span class="n">bd_gdf</span><span class="p">[</span><span class="n">bd_gdf</span><span class="p">[</span><span class="s2">&quot;geometry&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">notnull</span><span class="p">()]</span>

    <span class="c1"># Add boundary node data to PyG HeteroData lazily</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">bd_gdf</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">bd_gdf</span><span class="p">[[</span><span class="s2">&quot;centroid_x&quot;</span><span class="p">,</span> <span class="s2">&quot;centroid_y&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span><span class="o">.</span><span class="n">isnan</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span><span class="o">.</span><span class="n">isnan</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="mi">1</span><span class="p">)])</span>

    <span class="n">bd_x</span> <span class="o">=</span> <span class="n">bd_gdf</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">4</span><span class="p">:]</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">bd_x</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

    <span class="c1"># Extract the transcript coordinates lazily</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Preparing transcript features and positions...&quot;</span><span class="p">)</span>
    <span class="n">x_xyz</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_Y</span><span class="o">.</span><span class="n">value</span><span class="p">]]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">))</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">x_xyz</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

    <span class="c1"># Lazily prepare transcript embeddings (if available)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Preparing transcript embeddings..&quot;</span><span class="p">)</span>
    <span class="n">token_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tx_encoder</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">])</span>
    <span class="n">transcripts_df</span><span class="p">[</span><span class="s2">&quot;token&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">token_encoding</span>  <span class="c1"># Store the integer tokens in the &#39;token&#39; column</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">token</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">token_encoding</span><span class="p">)</span><span class="o">.</span><span class="n">int</span><span class="p">()</span>
    <span class="c1"># Handle additional embeddings lazily as well</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">embeddings</span> <span class="o">=</span> <span class="n">delayed</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)(</span>
            <span class="n">transcripts_df</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">embeddings</span> <span class="o">=</span> <span class="n">token_encoding</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">embeddings</span><span class="p">,</span> <span class="s2">&quot;compute&quot;</span><span class="p">):</span>
        <span class="n">embeddings</span> <span class="o">=</span> <span class="n">embeddings</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>
    <span class="n">x_features</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">embeddings</span><span class="p">)</span><span class="o">.</span><span class="n">int</span><span class="p">()</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">x_features</span>

    <span class="c1"># Check if the overlap column exists, if not, compute it lazily using Dask</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">transcripts_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Computing overlaps for transcripts...&quot;</span><span class="p">)</span>
        <span class="n">transcripts_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compute_transcript_overlap_with_boundaries</span><span class="p">(</span>
            <span class="n">transcripts_df</span><span class="p">,</span> <span class="n">polygons_gdf</span><span class="o">=</span><span class="n">bd_gdf</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="mf">1.0</span>
        <span class="p">)</span>

    <span class="c1"># Connect transcripts with their corresponding boundaries (e.g., nuclei, cells)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Connecting transcripts with boundaries...&quot;</span><span class="p">)</span>
    <span class="n">overlaps</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">valid_cell_ids</span> <span class="o">=</span> <span class="n">bd_gdf</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">ind</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">overlaps</span> <span class="o">&amp;</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">valid_cell_ids</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">tx_bd_edge_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">column_stack</span><span class="p">(</span>
        <span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">searchsorted</span><span class="p">(</span><span class="n">valid_cell_ids</span><span class="p">,</span> <span class="n">transcripts_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">ind</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]))</span>
    <span class="p">)</span>

    <span class="c1"># Add transcript-boundary edge index to PyG HeteroData</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;belongs&quot;</span><span class="p">,</span> <span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">tx_bd_edge_index</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">)</span>

    <span class="c1"># Compute transcript-to-transcript (tx-tx) edges using Dask (lazy computation)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing tx-tx edges...&quot;</span><span class="p">)</span>
    <span class="n">tx_positions</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_Y</span><span class="o">.</span><span class="n">value</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">delayed_tx_edge_index</span> <span class="o">=</span> <span class="n">delayed</span><span class="p">(</span><span class="n">get_edge_index</span><span class="p">)(</span>
        <span class="n">tx_positions</span><span class="p">,</span> <span class="n">tx_positions</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k_tx</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="n">r_tx</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="n">gpu</span><span class="o">=</span><span class="n">gpu</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="n">workers</span>
    <span class="p">)</span>
    <span class="n">tx_edge_index</span> <span class="o">=</span> <span class="n">delayed_tx_edge_index</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

    <span class="c1"># Add the tx-tx edge index to the PyG HeteroData object</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;neighbors&quot;</span><span class="p">,</span> <span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">tx_edge_index</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Finished building PyG data for the tile.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.compute_boundaries_geometries" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_boundaries_geometries</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.compute_boundaries_geometries" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_boundaries_geometries</span><span class="p">(</span><span class="n">boundaries_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">polygons_gdf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">area</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">convexity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">elongation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">circularity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Computes geometries for boundaries (e.g., nuclei, cells) from the dataframe using Dask.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>boundaries_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing boundaries data. Required if polygons_gdf is not provided.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>polygons_gdf</code>
            </td>
            <td>
                  <code><span title="dask_geopandas.GeoDataFrame">GeoDataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Precomputed Dask GeoDataFrame containing boundary polygons. If None, will compute from boundaries_df.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_factor</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The factor by which to scale the polygons (default is 1.0).</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>area</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to compute area.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>convexity</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to compute convexity.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>elongation</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to compute elongation.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>circularity</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to compute circularity.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask_geopandas.GeoDataFrame">GeoDataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dgpd.GeoDataFrame: A GeoDataFrame containing computed geometries.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_boundaries_geometries</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">boundaries_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">polygons_gdf</span><span class="p">:</span> <span class="n">dgpd</span><span class="o">.</span><span class="n">GeoDataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">scale_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="n">area</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">convexity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">elongation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">circularity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dgpd</span><span class="o">.</span><span class="n">GeoDataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes geometries for boundaries (e.g., nuclei, cells) from the dataframe using Dask.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        boundaries_df (dd.DataFrame, optional): The dataframe containing boundaries data. Required if polygons_gdf is not provided.</span>
<span class="sd">        polygons_gdf (dgpd.GeoDataFrame, optional): Precomputed Dask GeoDataFrame containing boundary polygons. If None, will compute from boundaries_df.</span>
<span class="sd">        scale_factor (float, optional): The factor by which to scale the polygons (default is 1.0).</span>
<span class="sd">        area (bool, optional): Whether to compute area.</span>
<span class="sd">        convexity (bool, optional): Whether to compute convexity.</span>
<span class="sd">        elongation (bool, optional): Whether to compute elongation.</span>
<span class="sd">        circularity (bool, optional): Whether to compute circularity.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dgpd.GeoDataFrame: A GeoDataFrame containing computed geometries.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check if polygons_gdf is provided, otherwise compute from boundaries_df</span>
    <span class="k">if</span> <span class="n">polygons_gdf</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">boundaries_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Both boundaries_df and polygons_gdf cannot be None. Provide at least one.&quot;</span><span class="p">)</span>

        <span class="c1"># Generate polygons from boundaries_df if polygons_gdf is None</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;No precomputed polygons provided. Computing polygons from boundaries with a scale factor of </span><span class="si">{</span><span class="n">scale_factor</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="p">)</span>
        <span class="n">polygons_gdf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_and_scale_polygons</span><span class="p">(</span><span class="n">boundaries_df</span><span class="p">,</span> <span class="n">scale_factor</span><span class="p">)</span>

    <span class="c1"># Check if the generated polygons_gdf is empty</span>
    <span class="k">if</span> <span class="n">polygons_gdf</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No valid polygons were generated from the boundaries.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Polygons are available. Proceeding with geometrical computations.&quot;</span><span class="p">)</span>

    <span class="c1"># Compute additional geometrical properties</span>
    <span class="n">polygons</span> <span class="o">=</span> <span class="n">polygons_gdf</span><span class="o">.</span><span class="n">geometry</span>

    <span class="c1"># Compute additional geometrical properties</span>
    <span class="k">if</span> <span class="n">area</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing area...&quot;</span><span class="p">)</span>
        <span class="n">polygons_gdf</span><span class="p">[</span><span class="s2">&quot;area&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">area</span>
    <span class="k">if</span> <span class="n">convexity</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing convexity...&quot;</span><span class="p">)</span>
        <span class="n">polygons_gdf</span><span class="p">[</span><span class="s2">&quot;convexity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">convex_hull</span><span class="o">.</span><span class="n">area</span> <span class="o">/</span> <span class="n">polygons</span><span class="o">.</span><span class="n">area</span>
    <span class="k">if</span> <span class="n">elongation</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing elongation...&quot;</span><span class="p">)</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">minimum_rotated_rectangle</span><span class="p">()</span>
        <span class="n">polygons_gdf</span><span class="p">[</span><span class="s2">&quot;elongation&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">length</span> <span class="o">*</span> <span class="n">r</span><span class="o">.</span><span class="n">length</span><span class="p">)</span> <span class="o">/</span> <span class="n">r</span><span class="o">.</span><span class="n">area</span>
    <span class="k">if</span> <span class="n">circularity</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Computing circularity...&quot;</span><span class="p">)</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">polygons_gdf</span><span class="o">.</span><span class="n">minimum_bounding_radius</span><span class="p">()</span>
        <span class="n">polygons_gdf</span><span class="p">[</span><span class="s2">&quot;circularity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">area</span> <span class="o">/</span> <span class="p">(</span><span class="n">r</span> <span class="o">*</span> <span class="n">r</span><span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Geometrical computations completed.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">polygons_gdf</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.compute_transcript_overlap_with_boundaries" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_transcript_overlap_with_boundaries</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.compute_transcript_overlap_with_boundaries" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_transcript_overlap_with_boundaries</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">boundaries_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">polygons_gdf</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Computes the overlap of transcript locations with scaled boundary polygons
and assigns corresponding cell IDs to the transcripts using Dask.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dask DataFrame containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>boundaries_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dask DataFrame containing boundary data. Required if polygons_gdf is not provided.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>polygons_gdf</code>
            </td>
            <td>
                  <code><span title="dask_geopandas.GeoDataFrame">GeoDataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Precomputed Dask GeoDataFrame containing boundary polygons. If None, will compute from boundaries_df.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_factor</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The factor by which to scale the boundary polygons. Default is 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dd.DataFrame: The updated DataFrame with overlap information and assigned cell IDs.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_transcript_overlap_with_boundaries</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">boundaries_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">polygons_gdf</span><span class="p">:</span> <span class="n">dgpd</span><span class="o">.</span><span class="n">GeoDataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">scale_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes the overlap of transcript locations with scaled boundary polygons</span>
<span class="sd">    and assigns corresponding cell IDs to the transcripts using Dask.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        transcripts_df (dd.DataFrame): Dask DataFrame containing transcript data.</span>
<span class="sd">        boundaries_df (dd.DataFrame, optional): Dask DataFrame containing boundary data. Required if polygons_gdf is not provided.</span>
<span class="sd">        polygons_gdf (dgpd.GeoDataFrame, optional): Precomputed Dask GeoDataFrame containing boundary polygons. If None, will compute from boundaries_df.</span>
<span class="sd">        scale_factor (float, optional): The factor by which to scale the boundary polygons. Default is 1.0.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dd.DataFrame: The updated DataFrame with overlap information and assigned cell IDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check if polygons_gdf is provided, otherwise compute from boundaries_df</span>
    <span class="k">if</span> <span class="n">polygons_gdf</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">boundaries_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Both boundaries_df and polygons_gdf cannot be None. Provide at least one.&quot;</span><span class="p">)</span>

        <span class="c1"># Generate polygons from boundaries_df if polygons_gdf is None</span>
        <span class="c1"># if self.verbose: print(f&quot;No precomputed polygons provided. Computing polygons from boundaries with a scale factor of {scale_factor}.&quot;)</span>
        <span class="n">polygons_gdf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">generate_and_scale_polygons</span><span class="p">(</span><span class="n">boundaries_df</span><span class="p">,</span> <span class="n">scale_factor</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">polygons_gdf</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No valid polygons were generated from the boundaries.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Polygons are available. Proceeding with overlap computation.&quot;</span><span class="p">)</span>

    <span class="c1"># Create a delayed function to check if a point is within any polygon</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">check_overlap</span><span class="p">(</span><span class="n">transcript</span><span class="p">,</span> <span class="n">polygons_gdf</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_X</span><span class="o">.</span><span class="n">value</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_Y</span><span class="o">.</span><span class="n">value</span><span class="p">]</span>
        <span class="n">point</span> <span class="o">=</span> <span class="n">Point</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>

        <span class="n">overlap</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">cell_id</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Check for point containment lazily within polygons</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">polygon</span> <span class="ow">in</span> <span class="n">polygons_gdf</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">polygon</span><span class="o">.</span><span class="n">geometry</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">point</span><span class="p">):</span>
                <span class="n">overlap</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">cell_id</span> <span class="o">=</span> <span class="n">polygon</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span>
                <span class="k">break</span>

        <span class="k">return</span> <span class="n">overlap</span><span class="p">,</span> <span class="n">cell_id</span>

    <span class="c1"># Apply the check_overlap function in parallel to each row using Dask&#39;s map_partitions</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Starting overlap computation for transcripts with the boundary polygons.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="c1"># luca: I found this bug here</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;BUG! This function expects Dask DataFrames, not Pandas DataFrames.&quot;</span><span class="p">)</span>
        <span class="c1"># if we want to really have the below working in parallel, we need to add n_partitions&gt;1 here</span>
        <span class="n">transcripts_df</span> <span class="o">=</span> <span class="n">dd</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">npartitions</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">transcripts_df</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span><span class="o">.</span><span class="n">columns</span>
    <span class="n">transcripts_df</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="o">.</span><span class="n">map_partitions</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">delayed</span><span class="p">(</span><span class="n">check_overlap</span><span class="p">)(</span><span class="n">row</span><span class="p">,</span> <span class="n">polygons_gdf</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
                <span class="p">),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">delayed</span><span class="p">(</span><span class="n">check_overlap</span><span class="p">)(</span><span class="n">row</span><span class="p">,</span> <span class="n">polygons_gdf</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">transcripts_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.create_scaled_polygon" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">create_scaled_polygon</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#segger.data.io.SpatialTranscriptomicsSample.create_scaled_polygon" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">create_scaled_polygon</span><span class="p">(</span><span class="n">group</span><span class="p">,</span> <span class="n">scale_factor</span><span class="p">,</span> <span class="n">keys</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Static method to create and scale a polygon from boundary vertices and return a GeoDataFrame.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>group</code>
            </td>
            <td>
                  <code><span title="segger.data.utils.pd.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Group of boundary coordinates (for a specific cell).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_factor</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The factor by which to scale the polygons.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keys</code>
            </td>
            <td>
                  <code><span title="segger.data.utils.Dict">Dict</span> or <span title="enum.Enum">Enum</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A collection of keys to access column names for 'cell_id', 'vertex_x', and 'vertex_y'.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="segger.data.utils.gpd.GeoDataFrame">GeoDataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>gpd.GeoDataFrame: A GeoDataFrame containing the scaled Polygon and cell_id.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">create_scaled_polygon</span><span class="p">(</span><span class="n">group</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">scale_factor</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">keys</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoDataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Static method to create and scale a polygon from boundary vertices and return a GeoDataFrame.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        group (pd.DataFrame): Group of boundary coordinates (for a specific cell).</span>
<span class="sd">        scale_factor (float): The factor by which to scale the polygons.</span>
<span class="sd">        keys (Dict or Enum): A collection of keys to access column names for &#39;cell_id&#39;, &#39;vertex_x&#39;, and &#39;vertex_y&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        gpd.GeoDataFrame: A GeoDataFrame containing the scaled Polygon and cell_id.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract coordinates and cell ID from the group using keys</span>
    <span class="n">x_coords</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="n">keys</span><span class="p">[</span><span class="s2">&quot;vertex_x&quot;</span><span class="p">]]</span>
    <span class="n">y_coords</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="n">keys</span><span class="p">[</span><span class="s2">&quot;vertex_y&quot;</span><span class="p">]]</span>
    <span class="n">cell_id</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="n">keys</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Ensure there are at least 3 points to form a polygon</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x_coords</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">3</span><span class="p">:</span>

        <span class="n">polygon</span> <span class="o">=</span> <span class="n">Polygon</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">x_coords</span><span class="p">,</span> <span class="n">y_coords</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">polygon</span><span class="o">.</span><span class="n">is_valid</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">polygon</span><span class="o">.</span><span class="n">is_empty</span><span class="p">:</span>
            <span class="c1"># Scale the polygon by the provided factor</span>
            <span class="n">scaled_polygon</span> <span class="o">=</span> <span class="n">polygon</span><span class="o">.</span><span class="n">buffer</span><span class="p">(</span><span class="n">scale_factor</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">scaled_polygon</span><span class="o">.</span><span class="n">is_valid</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">scaled_polygon</span><span class="o">.</span><span class="n">is_empty</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoDataFrame</span><span class="p">(</span>
                    <span class="p">{</span><span class="s2">&quot;geometry&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">scaled_polygon</span><span class="p">],</span> <span class="n">keys</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">]:</span> <span class="p">[</span><span class="n">cell_id</span><span class="p">]},</span> <span class="n">geometry</span><span class="o">=</span><span class="s2">&quot;geometry&quot;</span><span class="p">,</span> <span class="n">crs</span><span class="o">=</span><span class="s2">&quot;EPSG:4326&quot;</span>
                <span class="p">)</span>
    <span class="c1"># Return an empty GeoDataFrame if no valid polygon is created</span>
    <span class="k">return</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoDataFrame</span><span class="p">({</span><span class="s2">&quot;geometry&quot;</span><span class="p">:</span> <span class="p">[</span><span class="kc">None</span><span class="p">],</span> <span class="n">keys</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">]:</span> <span class="p">[</span><span class="n">cell_id</span><span class="p">]},</span> <span class="n">geometry</span><span class="o">=</span><span class="s2">&quot;geometry&quot;</span><span class="p">,</span> <span class="n">crs</span><span class="o">=</span><span class="s2">&quot;EPSG:4326&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.filter_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">filter_transcripts</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-abstractmethod"><code>abstractmethod</code></small>
  </span>

<a href="#segger.data.io.SpatialTranscriptomicsSample.filter_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">filter_transcripts</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">min_qv</span><span class="o">=</span><span class="mf">20.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Abstract method to filter transcripts based on dataset-specific criteria.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
                  <code><span title="segger.data.utils.pd.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_qv</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>20.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="segger.data.utils.pd.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: The filtered dataframe.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@abstractmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">filter_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">min_qv</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">20.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Abstract method to filter transcripts based on dataset-specific criteria.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        transcripts_df (pd.DataFrame): The dataframe containing transcript data.</span>
<span class="sd">        min_qv (float, optional): The minimum quality value threshold for filtering transcripts.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The filtered dataframe.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.generate_and_scale_polygons" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">generate_and_scale_polygons</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.generate_and_scale_polygons" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">generate_and_scale_polygons</span><span class="p">(</span><span class="n">boundaries_df</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Generate and scale polygons from boundary coordinates using Dask.
Keeps class structure intact by using static method for the core polygon generation.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>boundaries_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing boundary coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_factor</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The factor by which to scale the polygons (default is 1.0).</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask_geopandas.GeoDataFrame">GeoDataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dgpd.GeoDataFrame: A GeoDataFrame containing scaled Polygon objects and their centroids.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">generate_and_scale_polygons</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">boundaries_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">scale_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dgpd</span><span class="o">.</span><span class="n">GeoDataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate and scale polygons from boundary coordinates using Dask.</span>
<span class="sd">    Keeps class structure intact by using static method for the core polygon generation.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        boundaries_df (dd.DataFrame): DataFrame containing boundary coordinates.</span>
<span class="sd">        scale_factor (float, optional): The factor by which to scale the polygons (default is 1.0).</span>

<span class="sd">    Returns:</span>
<span class="sd">        dgpd.GeoDataFrame: A GeoDataFrame containing scaled Polygon objects and their centroids.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># if self.verbose: print(f&quot;No precomputed polygons provided. Computing polygons from boundaries with a scale factor of {scale_factor}.&quot;)</span>

    <span class="c1"># Extract required columns from self.keys</span>
    <span class="n">cell_id_column</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span>
    <span class="n">vertex_x_column</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_X</span><span class="o">.</span><span class="n">value</span>
    <span class="n">vertex_y_column</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_Y</span><span class="o">.</span><span class="n">value</span>

    <span class="n">create_polygon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">create_scaled_polygon</span>
    <span class="c1"># Use a lambda to wrap the static method call and avoid passing the function object directly to Dask</span>
    <span class="n">polygons_ddf</span> <span class="o">=</span> <span class="n">boundaries_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">cell_id_column</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">group</span><span class="p">:</span> <span class="n">create_polygon</span><span class="p">(</span>
            <span class="n">group</span><span class="o">=</span><span class="n">group</span><span class="p">,</span>
            <span class="n">scale_factor</span><span class="o">=</span><span class="n">scale_factor</span><span class="p">,</span>
            <span class="n">keys</span><span class="o">=</span><span class="p">{</span>  <span class="c1"># Pass keys as a dict for the lambda function</span>
                <span class="s2">&quot;vertex_x&quot;</span><span class="p">:</span> <span class="n">vertex_x_column</span><span class="p">,</span>
                <span class="s2">&quot;vertex_y&quot;</span><span class="p">:</span> <span class="n">vertex_y_column</span><span class="p">,</span>
                <span class="s2">&quot;cell_id&quot;</span><span class="p">:</span> <span class="n">cell_id_column</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Lazily compute centroids for each polygon</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Adding centroids to the polygons...&quot;</span><span class="p">)</span>
    <span class="n">polygons_ddf</span><span class="p">[</span><span class="s2">&quot;centroid_x&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons_ddf</span><span class="o">.</span><span class="n">geometry</span><span class="o">.</span><span class="n">centroid</span><span class="o">.</span><span class="n">x</span>
    <span class="n">polygons_ddf</span><span class="p">[</span><span class="s2">&quot;centroid_y&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons_ddf</span><span class="o">.</span><span class="n">geometry</span><span class="o">.</span><span class="n">centroid</span><span class="o">.</span><span class="n">y</span>

    <span class="n">polygons_ddf</span> <span class="o">=</span> <span class="n">polygons_ddf</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="c1"># polygons_ddf = polygons_ddf.to_crs(&quot;EPSG:3857&quot;)</span>

    <span class="k">return</span> <span class="n">polygons_ddf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.load_boundaries" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">load_boundaries</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.load_boundaries" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">load_boundaries</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">file_format</span><span class="o">=</span><span class="s1">&#39;parquet&#39;</span><span class="p">,</span> <span class="n">x_min</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">x_max</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">y_min</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">y_max</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Load boundaries data lazily using Dask, filtering by the specified bounding box.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the boundaries file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>file_format</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Format of the file to load. Only 'parquet' is supported in this refactor.</p>
              </div>
            </td>
            <td>
                  <code>&#39;parquet&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>x_min</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum X-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>x_max</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum X-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>y_min</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum Y-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>y_max</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum Y-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dd.DataFrame: The filtered boundaries DataFrame.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_boundaries</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">file_format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;parquet&quot;</span><span class="p">,</span>
    <span class="n">x_min</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">x_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">y_min</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">y_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load boundaries data lazily using Dask, filtering by the specified bounding box.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        path (Path): Path to the boundaries file.</span>
<span class="sd">        file_format (str, optional): Format of the file to load. Only &#39;parquet&#39; is supported in this refactor.</span>
<span class="sd">        x_min (float, optional): Minimum X-coordinate for the bounding box.</span>
<span class="sd">        x_max (float, optional): Maximum X-coordinate for the bounding box.</span>
<span class="sd">        y_min (float, optional): Minimum Y-coordinate for the bounding box.</span>
<span class="sd">        y_max (float, optional): Maximum Y-coordinate for the bounding box.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dd.DataFrame: The filtered boundaries DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file_format</span> <span class="o">!=</span> <span class="s2">&quot;parquet&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported file format: </span><span class="si">{</span><span class="n">file_format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">boundaries_path</span> <span class="o">=</span> <span class="n">path</span>

    <span class="c1"># Use bounding box values from set_metadata if not explicitly provided</span>
    <span class="n">x_min</span> <span class="o">=</span> <span class="n">x_min</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">x_min</span>
    <span class="n">x_max</span> <span class="o">=</span> <span class="n">x_max</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">x_max</span>
    <span class="n">y_min</span> <span class="o">=</span> <span class="n">y_min</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_min</span>
    <span class="n">y_max</span> <span class="o">=</span> <span class="n">y_max</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_max</span>

    <span class="c1"># Define the list of columns to read</span>
    <span class="n">columns_to_read</span> <span class="o">=</span> <span class="p">[</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_Y</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c1"># Use filters to only load data within the specified bounding box (x_min, x_max, y_min, y_max)</span>
    <span class="n">filters</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&gt;=&quot;</span><span class="p">,</span> <span class="n">x_min</span><span class="p">),</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&lt;=&quot;</span><span class="p">,</span> <span class="n">x_max</span><span class="p">),</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_Y</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&gt;=&quot;</span><span class="p">,</span> <span class="n">y_min</span><span class="p">),</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_Y</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&lt;=&quot;</span><span class="p">,</span> <span class="n">y_max</span><span class="p">),</span>
    <span class="p">]</span>

    <span class="c1"># Load the dataset lazily with filters applied for the bounding box</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">dd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">path</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="k">if</span> <span class="s2">&quot;geometry&quot;</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
        <span class="n">bbox</span> <span class="o">=</span> <span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">y_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">,</span> <span class="n">y_max</span><span class="p">)</span>
        <span class="c1"># TODO: check that SpatialData objects write the &quot;bbox covering metadata&quot; to the parquet file</span>
        <span class="n">gdf</span> <span class="o">=</span> <span class="n">dgpd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">bbox</span><span class="o">=</span><span class="n">bbox</span><span class="p">)</span>
        <span class="n">id_col</span><span class="p">,</span> <span class="n">x_col</span><span class="p">,</span> <span class="n">y_col</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">BOUNDARIES_VERTEX_Y</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># Function to expand each polygon into a list of vertices</span>
        <span class="k">def</span><span class="w"> </span><span class="nf">expand_polygon</span><span class="p">(</span><span class="n">row</span><span class="p">):</span>
            <span class="n">expanded_data</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">polygon</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;geometry&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">polygon</span><span class="o">.</span><span class="n">geom_type</span> <span class="o">==</span> <span class="s2">&quot;Polygon&quot;</span><span class="p">:</span>
                <span class="n">exterior_coords</span> <span class="o">=</span> <span class="n">polygon</span><span class="o">.</span><span class="n">exterior</span><span class="o">.</span><span class="n">coords</span>
                <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">exterior_coords</span><span class="p">:</span>
                    <span class="n">expanded_data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="n">id_col</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">x_col</span><span class="p">:</span> <span class="n">x</span><span class="p">,</span> <span class="n">y_col</span><span class="p">:</span> <span class="n">y</span><span class="p">})</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># Instead of expanding the gdf and then having code later to recreate it (when computing the pyg graph)</span>
                <span class="c1"># we could directly have this function returning a Dask GeoDataFrame. This means that we don&#39;t need</span>
                <span class="c1"># to implement this else black</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported geometry type: </span><span class="si">{</span><span class="n">polygon</span><span class="o">.</span><span class="n">geom_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">expanded_data</span>

        <span class="c1"># Apply the function to each partition and collect results</span>
        <span class="k">def</span><span class="w"> </span><span class="nf">process_partition</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
            <span class="n">expanded_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">expand_polygon</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>
            <span class="c1"># Flatten the list of lists</span>
            <span class="n">flattened_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">expanded_data</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>
            <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">flattened_data</span><span class="p">)</span>

        <span class="c1"># Use map_partitions to apply the function and convert it into a Dask DataFrame</span>
        <span class="n">boundaries_df</span> <span class="o">=</span> <span class="n">gdf</span><span class="o">.</span><span class="n">map_partitions</span><span class="p">(</span><span class="n">process_partition</span><span class="p">,</span> <span class="n">meta</span><span class="o">=</span><span class="p">{</span><span class="n">id_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">x_col</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">y_col</span><span class="p">:</span> <span class="nb">float</span><span class="p">})</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">boundaries_df</span> <span class="o">=</span> <span class="n">dd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns_to_read</span><span class="p">,</span> <span class="n">filters</span><span class="o">=</span><span class="n">filters</span><span class="p">)</span>

        <span class="c1"># Convert the cell IDs to strings lazily</span>
        <span class="n">boundaries_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">=</span> <span class="n">boundaries_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span> <span class="n">meta</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;cell_id&quot;</span><span class="p">,</span> <span class="s2">&quot;object&quot;</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded boundaries from &#39;</span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&#39; within bounding box (</span><span class="si">{</span><span class="n">x_min</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">x_max</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">y_min</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">y_max</span><span class="si">}</span><span class="s2">).&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">boundaries_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.load_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">load_transcripts</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.load_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">load_transcripts</span><span class="p">(</span><span class="n">base_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sample</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">transcripts_filename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">file_format</span><span class="o">=</span><span class="s1">&#39;parquet&#39;</span><span class="p">,</span> <span class="n">x_min</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">x_max</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">y_min</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">y_max</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Load transcripts from a Parquet file using Dask for efficient chunked processing,
only within the specified bounding box, and return the filtered DataFrame with integer token embeddings.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>base_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The base directory path where samples are stored.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sample</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The sample name or identifier.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_filename</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The filename of the transcripts file (default is derived from the dataset keys).</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Specific path to the transcripts file.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>file_format</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Format of the file to load (default is 'parquet').</p>
              </div>
            </td>
            <td>
                  <code>&#39;parquet&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>x_min</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum X-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>x_max</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum X-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>y_min</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum Y-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>y_max</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum Y-coordinate for the bounding box.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dd.DataFrame: The filtered transcripts DataFrame.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_transcripts</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">base_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">transcripts_filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">file_format</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;parquet&quot;</span><span class="p">,</span>
    <span class="n">x_min</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">x_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">y_min</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">y_max</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="c1"># additional_embeddings: Optional[Dict[str, pd.DataFrame]] = None,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load transcripts from a Parquet file using Dask for efficient chunked processing,</span>
<span class="sd">    only within the specified bounding box, and return the filtered DataFrame with integer token embeddings.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        base_path (Path, optional): The base directory path where samples are stored.</span>
<span class="sd">        sample (str, optional): The sample name or identifier.</span>
<span class="sd">        transcripts_filename (str, optional): The filename of the transcripts file (default is derived from the dataset keys).</span>
<span class="sd">        path (Path, optional): Specific path to the transcripts file.</span>
<span class="sd">        file_format (str, optional): Format of the file to load (default is &#39;parquet&#39;).</span>
<span class="sd">        x_min (float, optional): Minimum X-coordinate for the bounding box.</span>
<span class="sd">        x_max (float, optional): Maximum X-coordinate for the bounding box.</span>
<span class="sd">        y_min (float, optional): Minimum Y-coordinate for the bounding box.</span>
<span class="sd">        y_max (float, optional): Maximum Y-coordinate for the bounding box.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dd.DataFrame: The filtered transcripts DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file_format</span> <span class="o">!=</span> <span class="s2">&quot;parquet&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;This version only supports parquet files with Dask.&quot;</span><span class="p">)</span>

    <span class="c1"># Set the file path for transcripts</span>
    <span class="n">transcripts_filename</span> <span class="o">=</span> <span class="n">transcripts_filename</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_FILE</span><span class="o">.</span><span class="n">value</span>
    <span class="n">file_path</span> <span class="o">=</span> <span class="n">path</span> <span class="ow">or</span> <span class="p">(</span><span class="n">base_path</span> <span class="o">/</span> <span class="n">sample</span> <span class="o">/</span> <span class="n">transcripts_filename</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">transcripts_path</span> <span class="o">=</span> <span class="n">file_path</span>

    <span class="c1"># Set metadata</span>
    <span class="c1"># self.set_metadata()</span>

    <span class="c1"># Use bounding box values from set_metadata if not explicitly provided</span>
    <span class="n">x_min</span> <span class="o">=</span> <span class="n">x_min</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">x_min</span>
    <span class="n">x_max</span> <span class="o">=</span> <span class="n">x_max</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">x_max</span>
    <span class="n">y_min</span> <span class="o">=</span> <span class="n">y_min</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_min</span>
    <span class="n">y_max</span> <span class="o">=</span> <span class="n">y_max</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">y_max</span>

    <span class="c1"># Check for available columns in the file&#39;s metadata (without loading the data)</span>
    <span class="n">parquet_metadata</span> <span class="o">=</span> <span class="n">dd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">meta_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">available_columns</span> <span class="o">=</span> <span class="n">parquet_metadata</span><span class="o">.</span><span class="n">columns</span>

    <span class="c1"># Define the list of columns to read</span>
    <span class="n">columns_to_read</span> <span class="o">=</span> <span class="p">[</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_ID</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_Y</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="c1"># Check if the QUALITY_VALUE key exists in the dataset, and add it to the columns list if present</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">QUALITY_VALUE</span><span class="o">.</span><span class="n">value</span> <span class="ow">in</span> <span class="n">available_columns</span><span class="p">:</span>
        <span class="n">columns_to_read</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">QUALITY_VALUE</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span> <span class="ow">in</span> <span class="n">available_columns</span><span class="p">:</span>
        <span class="n">columns_to_read</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>

    <span class="c1"># Use filters to only load data within the specified bounding box (x_min, x_max, y_min, y_max)</span>
    <span class="n">filters</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&gt;=&quot;</span><span class="p">,</span> <span class="n">x_min</span><span class="p">),</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_X</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&lt;=&quot;</span><span class="p">,</span> <span class="n">x_max</span><span class="p">),</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_Y</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&gt;=&quot;</span><span class="p">,</span> <span class="n">y_min</span><span class="p">),</span>
        <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_Y</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="s2">&quot;&lt;=&quot;</span><span class="p">,</span> <span class="n">y_max</span><span class="p">),</span>
    <span class="p">]</span>

    <span class="c1"># Load the dataset lazily with filters applied for the bounding box</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">dd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="n">transcripts_df</span> <span class="o">=</span> <span class="n">dd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns_to_read</span><span class="p">,</span> <span class="n">filters</span><span class="o">=</span><span class="n">filters</span><span class="p">)</span><span class="o">.</span><span class="n">compute</span><span class="p">()</span>

    <span class="c1"># Convert transcript and cell IDs to strings lazily</span>
    <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">CELL_ID</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Convert feature names from bytes to strings if necessary</span>
    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">api</span><span class="o">.</span><span class="n">types</span><span class="o">.</span><span class="n">is_object_dtype</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]):</span>
        <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

    <span class="c1"># Apply dataset-specific filtering (e.g., quality filtering for Xenium)</span>
    <span class="n">transcripts_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">filter_transcripts</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">)</span>

    <span class="c1"># Handle additional embeddings if provided</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">valid_genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">embedding_df</span><span class="o">.</span><span class="n">index</span>
        <span class="c1"># Lazily count the number of rows in the DataFrame before filtering</span>
        <span class="n">initial_count</span> <span class="o">=</span> <span class="n">delayed</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])(</span><span class="n">transcripts_df</span><span class="p">)</span>
        <span class="c1"># Filter the DataFrame lazily based on valid genes from embeddings</span>
        <span class="n">transcripts_df</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">valid_genes</span><span class="p">)]</span>
        <span class="n">final_count</span> <span class="o">=</span> <span class="n">delayed</span><span class="p">(</span><span class="k">lambda</span> <span class="n">df</span><span class="p">:</span> <span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])(</span><span class="n">transcripts_df</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Dropped </span><span class="si">{</span><span class="n">initial_count</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">final_count</span><span class="si">}</span><span class="s2"> transcripts not found in embedding.&quot;</span><span class="p">)</span>

    <span class="c1"># Ensure that the &#39;OVERLAPS_BOUNDARY&#39; column is boolean if it exists</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span> <span class="ow">in</span> <span class="n">transcripts_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">OVERLAPS_BOUNDARY</span><span class="o">.</span><span class="n">value</span>
        <span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">transcripts_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.save_dataset_for_segger" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">save_dataset_for_segger</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.save_dataset_for_segger" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">save_dataset_for_segger</span><span class="p">(</span><span class="n">processed_dir</span><span class="p">,</span> <span class="n">x_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">y_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">d_x</span><span class="o">=</span><span class="mi">900</span><span class="p">,</span> <span class="n">d_y</span><span class="o">=</span><span class="mi">900</span><span class="p">,</span> <span class="n">margin_x</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">margin_y</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">compute_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">r_tx</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">k_tx</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">neg_sampling_ratio_approx</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">sampling_rate</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">scale_boundaries</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;kd_tree&#39;</span><span class="p">,</span> <span class="n">gpu</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Saves the dataset for Segger in a processed format using Dask for parallel and lazy processing.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>processed_dir</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to save the processed dataset.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>x_size</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Width of each tile.</p>
              </div>
            </td>
            <td>
                  <code>1000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>y_size</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Height of each tile.</p>
              </div>
            </td>
            <td>
                  <code>1000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>d_x</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Step size in the x direction for tiles.</p>
              </div>
            </td>
            <td>
                  <code>900</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>d_y</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Step size in the y direction for tiles.</p>
              </div>
            </td>
            <td>
                  <code>900</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>margin_x</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Margin in the x direction to include transcripts.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>margin_y</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Margin in the y direction to include transcripts.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>compute_labels</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to compute edge labels for tx_belongs_bd edges.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>r_tx</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Radius for building the transcript-to-transcript graph.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>k_tx</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for the tx-tx graph.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>val_prob</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Probability of assigning a tile to the validation set.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>test_prob</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Probability of assigning a tile to the test set.</p>
              </div>
            </td>
            <td>
                  <code>0.2</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>neg_sampling_ratio_approx</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Approximate ratio of negative samples.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sampling_rate</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Rate of sampling tiles.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_workers</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of workers to use for parallel processing.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_boundaries</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The factor by which to scale the boundary polygons. Default is 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>method</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Method for computing edge indices (e.g., 'kd_tree', 'faiss').</p>
              </div>
            </td>
            <td>
                  <code>&#39;kd_tree&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gpu</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to use GPU acceleration for edge index computation.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>workers</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of workers to use to compute the neighborhood graph (per tile).</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">save_dataset_for_segger</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">processed_dir</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">x_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">y_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">d_x</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">900</span><span class="p">,</span>
    <span class="n">d_y</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">900</span><span class="p">,</span>
    <span class="n">margin_x</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">margin_y</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">compute_labels</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">r_tx</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">val_prob</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">test_prob</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">neg_sampling_ratio_approx</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">sampling_rate</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">num_workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">scale_boundaries</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;kd_tree&quot;</span><span class="p">,</span>
    <span class="n">gpu</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Saves the dataset for Segger in a processed format using Dask for parallel and lazy processing.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        processed_dir (Path): Directory to save the processed dataset.</span>
<span class="sd">        x_size (float, optional): Width of each tile.</span>
<span class="sd">        y_size (float, optional): Height of each tile.</span>
<span class="sd">        d_x (float, optional): Step size in the x direction for tiles.</span>
<span class="sd">        d_y (float, optional): Step size in the y direction for tiles.</span>
<span class="sd">        margin_x (float, optional): Margin in the x direction to include transcripts.</span>
<span class="sd">        margin_y (float, optional): Margin in the y direction to include transcripts.</span>
<span class="sd">        compute_labels (bool, optional): Whether to compute edge labels for tx_belongs_bd edges.</span>
<span class="sd">        r_tx (float, optional): Radius for building the transcript-to-transcript graph.</span>
<span class="sd">        k_tx (int, optional): Number of nearest neighbors for the tx-tx graph.</span>
<span class="sd">        val_prob (float, optional): Probability of assigning a tile to the validation set.</span>
<span class="sd">        test_prob (float, optional): Probability of assigning a tile to the test set.</span>
<span class="sd">        neg_sampling_ratio_approx (float, optional): Approximate ratio of negative samples.</span>
<span class="sd">        sampling_rate (float, optional): Rate of sampling tiles.</span>
<span class="sd">        num_workers (int, optional): Number of workers to use for parallel processing.</span>
<span class="sd">        scale_boundaries (float, optional): The factor by which to scale the boundary polygons. Default is 1.0.</span>
<span class="sd">        method (str, optional): Method for computing edge indices (e.g., &#39;kd_tree&#39;, &#39;faiss&#39;).</span>
<span class="sd">        gpu (bool, optional): Whether to use GPU acceleration for edge index computation.</span>
<span class="sd">        workers (int, optional): Number of workers to use to compute the neighborhood graph (per tile).</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Prepare directories for storing processed tiles</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_prepare_directories</span><span class="p">(</span><span class="n">processed_dir</span><span class="p">)</span>

    <span class="c1"># Get x and y coordinate ranges for tiling</span>
    <span class="n">x_range</span><span class="p">,</span> <span class="n">y_range</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_ranges</span><span class="p">(</span><span class="n">d_x</span><span class="p">,</span> <span class="n">d_y</span><span class="p">)</span>

    <span class="c1"># Generate parameters for each tile</span>
    <span class="n">tile_params</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_generate_tile_params</span><span class="p">(</span>
        <span class="n">x_range</span><span class="p">,</span>
        <span class="n">y_range</span><span class="p">,</span>
        <span class="n">x_size</span><span class="p">,</span>
        <span class="n">y_size</span><span class="p">,</span>
        <span class="n">margin_x</span><span class="p">,</span>
        <span class="n">margin_y</span><span class="p">,</span>
        <span class="n">compute_labels</span><span class="p">,</span>
        <span class="n">r_tx</span><span class="p">,</span>
        <span class="n">k_tx</span><span class="p">,</span>
        <span class="n">val_prob</span><span class="p">,</span>
        <span class="n">test_prob</span><span class="p">,</span>
        <span class="n">neg_sampling_ratio_approx</span><span class="p">,</span>
        <span class="n">sampling_rate</span><span class="p">,</span>
        <span class="n">processed_dir</span><span class="p">,</span>
        <span class="n">scale_boundaries</span><span class="p">,</span>
        <span class="n">method</span><span class="p">,</span>
        <span class="n">gpu</span><span class="p">,</span>
        <span class="n">workers</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Process each tile using Dask to parallelize the task</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Starting tile processing...&quot;</span><span class="p">)</span>
    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[</span><span class="n">delayed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_process_tile</span><span class="p">)(</span><span class="n">params</span><span class="p">)</span> <span class="k">for</span> <span class="n">params</span> <span class="ow">in</span> <span class="n">tile_params</span><span class="p">]</span>

    <span class="k">with</span> <span class="n">ProgressBar</span><span class="p">():</span>
        <span class="c1"># Use Dask to process all tiles in parallel</span>
        <span class="n">dask</span><span class="o">.</span><span class="n">compute</span><span class="p">(</span><span class="o">*</span><span class="n">tasks</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Tile processing completed.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.set_embedding" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">set_embedding</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.set_embedding" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">set_embedding</span><span class="p">(</span><span class="n">embedding_name</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Set the current embedding type for the transcripts.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>embedding_name</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str
The name of the embedding to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">set_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">embedding_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set the current embedding type for the transcripts.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        embedding_name : str</span>
<span class="sd">            The name of the embedding to use.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">embedding_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">embeddings_dict</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current_embedding</span> <span class="o">=</span> <span class="n">embedding_name</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Embedding </span><span class="si">{</span><span class="n">embedding_name</span><span class="si">}</span><span class="s2"> not found in embeddings_dict.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.set_file_paths" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">set_file_paths</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.set_file_paths" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">set_file_paths</span><span class="p">(</span><span class="n">transcripts_path</span><span class="p">,</span> <span class="n">boundaries_path</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Set the paths for the transcript and boundary files.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the Parquet file containing transcripts data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>boundaries_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the Parquet file containing boundaries data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">set_file_paths</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcripts_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">boundaries_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set the paths for the transcript and boundary files.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        transcripts_path (Path): Path to the Parquet file containing transcripts data.</span>
<span class="sd">        boundaries_path (Path): Path to the Parquet file containing boundaries data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">transcripts_path</span> <span class="o">=</span> <span class="n">transcripts_path</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">boundaries_path</span> <span class="o">=</span> <span class="n">boundaries_path</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Set transcripts file path to </span><span class="si">{</span><span class="n">transcripts_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Set boundaries file path to </span><span class="si">{</span><span class="n">boundaries_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="segger.data.io.SpatialTranscriptomicsSample.set_metadata" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">set_metadata</span>


<a href="#segger.data.io.SpatialTranscriptomicsSample.set_metadata" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">set_metadata</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Set metadata for the transcript dataset, including bounding box limits and unique gene names,
without reading the entire Parquet file. Additionally, return integer tokens for unique gene names
instead of one-hot encodings and store the lookup table for later mapping.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">set_metadata</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set metadata for the transcript dataset, including bounding box limits and unique gene names,</span>
<span class="sd">    without reading the entire Parquet file. Additionally, return integer tokens for unique gene names</span>
<span class="sd">    instead of one-hot encodings and store the lookup table for later mapping.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load the Parquet file metadata</span>
    <span class="n">parquet_file</span> <span class="o">=</span> <span class="n">pq</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">transcripts_path</span><span class="p">)</span>

    <span class="c1"># Get the column names for X, Y, and feature names from the class&#39;s keys</span>
    <span class="n">x_col</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_X</span><span class="o">.</span><span class="n">value</span>
    <span class="n">y_col</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">TRANSCRIPTS_Y</span><span class="o">.</span><span class="n">value</span>
    <span class="n">feature_col</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span>

    <span class="c1"># Initialize variables to track min/max values for X and Y</span>
    <span class="n">x_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">,</span> <span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">),</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;-inf&quot;</span><span class="p">),</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">),</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;-inf&quot;</span><span class="p">)</span>

    <span class="c1"># Extract unique gene names and ensure they&#39;re strings</span>
    <span class="n">gene_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

    <span class="c1"># Define the filter for unwanted codewords</span>
    <span class="n">filter_codewords</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;NegControlProbe_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;antisense_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NegControlCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;BLANK_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DeprecatedCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;UnassignedCodeword_&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">row_group_size</span> <span class="o">=</span> <span class="mi">4_000_000</span>
    <span class="n">start</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parquet_file</span><span class="p">)</span>
    <span class="k">while</span> <span class="n">start</span> <span class="o">&lt;</span> <span class="n">n</span><span class="p">:</span>
        <span class="n">chunk</span> <span class="o">=</span> <span class="n">parquet_file</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">start</span> <span class="o">+</span> <span class="n">row_group_size</span><span class="p">)</span>
        <span class="n">start</span> <span class="o">+=</span> <span class="n">row_group_size</span>

        <span class="c1"># Update the bounding box values (min/max)</span>
        <span class="n">x_values</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">x_col</span><span class="p">]</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
        <span class="n">y_values</span> <span class="o">=</span> <span class="n">chunk</span><span class="p">[</span><span class="n">y_col</span><span class="p">]</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>

        <span class="n">x_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">x_values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">x_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">x_max</span><span class="p">,</span> <span class="n">x_values</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
        <span class="n">y_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">y_values</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>
        <span class="n">y_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">y_max</span><span class="p">,</span> <span class="n">y_values</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>

        <span class="c1"># Convert feature values (gene names) to strings and filter out unwanted codewords</span>
        <span class="n">feature_values</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">chunk</span><span class="p">[</span><span class="n">feature_col</span><span class="p">]</span>
            <span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
            <span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">)</span> <span class="k">else</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Filter out unwanted codewords</span>
        <span class="n">filtered_genes</span> <span class="o">=</span> <span class="n">feature_values</span><span class="p">[</span><span class="o">~</span><span class="n">feature_values</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">filter_codewords</span><span class="p">)]</span>

        <span class="c1"># Update the unique gene set</span>
        <span class="n">gene_set</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">filtered_genes</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="c1"># Set bounding box limits</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">x_min</span> <span class="o">=</span> <span class="n">x_min</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">x_max</span> <span class="o">=</span> <span class="n">x_max</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">y_min</span> <span class="o">=</span> <span class="n">y_min</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">y_max</span> <span class="o">=</span> <span class="n">y_max</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Bounding box limits set: x_min=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">x_min</span><span class="si">}</span><span class="s2">, x_max=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">x_max</span><span class="si">}</span><span class="s2">, y_min=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">y_min</span><span class="si">}</span><span class="s2">, y_max=</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">y_max</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Convert the set of unique genes into a sorted list for consistent ordering</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">unique_genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_set</span><span class="p">)</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Extracted </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">unique_genes</span><span class="p">)</span><span class="si">}</span><span class="s2"> unique gene names for integer tokenization.&quot;</span><span class="p">)</span>

    <span class="c1"># Initialize a LabelEncoder to convert unique genes into integer tokens</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">tx_encoder</span> <span class="o">=</span> <span class="n">LabelEncoder</span><span class="p">()</span>

    <span class="c1"># Fit the LabelEncoder on the unique genes</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">tx_encoder</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">unique_genes</span><span class="p">)</span>

    <span class="c1"># Store the integer tokens mapping to gene names</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_token_map</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tx_encoder</span><span class="o">.</span><span class="n">classes_</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">tx_encoder</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tx_encoder</span><span class="o">.</span><span class="n">classes_</span><span class="p">))</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Integer tokens have been computed and stored based on unique gene names.&quot;</span><span class="p">)</span>

    <span class="c1"># Optional: Create a reverse mapping for lookup purposes (token to gene)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">token_to_gene_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">v</span><span class="p">:</span> <span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_to_token_map</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Lookup tables (gene_to_token_map and token_to_gene_map) have been created.&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.XeniumKeys" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">XeniumKeys</span>


<a href="#segger.data.io.XeniumKeys" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


        <p>Keys for <em>10X Genomics Xenium</em> formatted dataset.</p>










  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="segger.data.io.XeniumSample" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">XeniumSample</span>


<a href="#segger.data.io.XeniumSample" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">XeniumSample</span><span class="p">(</span><span class="n">transcripts_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">transcripts_radius</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">boundaries_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">embedding_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="SpatialTranscriptomicsSample (segger.data.io.SpatialTranscriptomicsSample)" href="#segger.data.io.SpatialTranscriptomicsSample">SpatialTranscriptomicsSample</a></code></p>








                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/io.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">transcripts_radius</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">boundaries_graph</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">embedding_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
        <span class="n">transcripts_df</span><span class="p">,</span> <span class="n">transcripts_radius</span><span class="p">,</span> <span class="n">boundaries_graph</span><span class="p">,</span> <span class="n">embedding_df</span><span class="p">,</span> <span class="n">XeniumKeys</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">verbose</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="segger.data.io.XeniumSample.filter_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">filter_transcripts</span>


<a href="#segger.data.io.XeniumSample.filter_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">filter_transcripts</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">min_qv</span><span class="o">=</span><span class="mf">20.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Filters transcripts based on quality value and removes unwanted transcripts for Xenium using Dask.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The Dask DataFrame containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_qv</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>20.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="dask.dataframe.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dd.DataFrame: The filtered Dask DataFrame.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/io.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">min_qv</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">20.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">dd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters transcripts based on quality value and removes unwanted transcripts for Xenium using Dask.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        transcripts_df (dd.DataFrame): The Dask DataFrame containing transcript data.</span>
<span class="sd">        min_qv (float, optional): The minimum quality value threshold for filtering transcripts.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dd.DataFrame: The filtered Dask DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filter_codewords</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;NegControlProbe_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;antisense_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NegControlCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;BLANK_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DeprecatedCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;UnassignedCodeword_&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Ensure FEATURE_NAME is a string type for proper filtering (compatible with Dask)</span>
    <span class="c1"># Handle potential bytes to string conversion for Dask DataFrame</span>
    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">api</span><span class="o">.</span><span class="n">types</span><span class="o">.</span><span class="n">is_object_dtype</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]):</span>
        <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
        <span class="p">)</span>

    <span class="c1"># Apply the quality value filter using Dask</span>
    <span class="n">mask_quality</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">QUALITY_VALUE</span><span class="o">.</span><span class="n">value</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_qv</span>

    <span class="c1"># Apply the filter for unwanted codewords using Dask string functions</span>
    <span class="n">mask_codewords</span> <span class="o">=</span> <span class="o">~</span><span class="n">transcripts_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="o">.</span><span class="n">FEATURE_NAME</span><span class="o">.</span><span class="n">value</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">filter_codewords</span><span class="p">)</span>

    <span class="c1"># Combine the filters and return the filtered Dask DataFrame</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">mask_quality</span> <span class="o">&amp;</span> <span class="n">mask_codewords</span>

    <span class="c1"># Return the filtered DataFrame lazily</span>
    <span class="k">return</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="segger.data.io.calculate_gene_celltype_abundance_embedding" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">calculate_gene_celltype_abundance_embedding</span>


<a href="#segger.data.io.calculate_gene_celltype_abundance_embedding" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">calculate_gene_celltype_abundance_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">celltype_column</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Calculate the cell type abundance embedding for each gene based on the fraction of cells in each cell type
that express the gene (non-zero expression).</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An AnnData object containing gene expression data and cell type information.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>celltype_column</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The column name in <code>adata.obs</code> that contains the cell type information.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame where rows are genes and columns are cell types, with each value representing
the fraction of cells in that cell type expressing the gene.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>adata = AnnData(...)  # Load your scRNA-seq AnnData object
celltype_column = 'celltype_major'
abundance_df = calculate_gene_celltype_abundance_embedding(adata, celltype_column)
abundance_df.head()</p>
</blockquote>
</blockquote>
</blockquote>
</details>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">calculate_gene_celltype_abundance_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">celltype_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the cell type abundance embedding for each gene based on the fraction of cells in each cell type</span>
<span class="sd">    that express the gene (non-zero expression).</span>

<span class="sd">    Parameters:</span>
<span class="sd">        adata (ad.AnnData): An AnnData object containing gene expression data and cell type information.</span>
<span class="sd">        celltype_column (str): The column name in `adata.obs` that contains the cell type information.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame where rows are genes and columns are cell types, with each value representing</span>
<span class="sd">            the fraction of cells in that cell type expressing the gene.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; adata = AnnData(...)  # Load your scRNA-seq AnnData object</span>
<span class="sd">        &gt;&gt;&gt; celltype_column = &#39;celltype_major&#39;</span>
<span class="sd">        &gt;&gt;&gt; abundance_df = calculate_gene_celltype_abundance_embedding(adata, celltype_column)</span>
<span class="sd">        &gt;&gt;&gt; abundance_df.head()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract expression data (cells x genes) and cell type information (cells)</span>
    <span class="n">expression_data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="s2">&quot;toarray&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
    <span class="n">cell_types</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_column</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="c1"># Create a binary matrix for gene expression (1 if non-zero, 0 otherwise)</span>
    <span class="n">gene_expression_binary</span> <span class="o">=</span> <span class="p">(</span><span class="n">expression_data</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="c1"># Convert the binary matrix to a DataFrame</span>
    <span class="n">gene_expression_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_expression_binary</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
    <span class="c1"># Perform one-hot encoding on the cell types</span>
    <span class="n">encoder</span> <span class="o">=</span> <span class="n">OneHotEncoder</span><span class="p">(</span><span class="n">sparse_output</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">cell_type_encoded</span> <span class="o">=</span> <span class="n">encoder</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">cell_types</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
    <span class="c1"># Calculate the fraction of cells expressing each gene per cell type</span>
    <span class="n">cell_type_abundance_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cell_type_encoded</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="c1"># Extract cells of the current cell type</span>
        <span class="n">cell_type_mask</span> <span class="o">=</span> <span class="n">cell_type_encoded</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="c1"># Calculate the abundance: sum of non-zero expressions in this cell type / total cells in this cell type</span>
        <span class="n">abundance</span> <span class="o">=</span> <span class="n">gene_expression_df</span><span class="p">[</span><span class="n">cell_type_mask</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">cell_type_abundance_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">abundance</span><span class="p">)</span>
    <span class="c1"># Create a DataFrame for the cell type abundance with gene names as rows and cell types as columns</span>
    <span class="n">cell_type_abundance_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">cell_type_abundance_list</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">encoder</span><span class="o">.</span><span class="n">categories_</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="k">return</span> <span class="n">cell_type_abundance_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.data.io.compute_transcript_metrics" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_transcript_metrics</span>


<a href="#segger.data.io.compute_transcript_metrics" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_transcript_metrics</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">qv_threshold</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">cell_id_col</span><span class="o">=</span><span class="s1">&#39;cell_id&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Computes various metrics for a given dataframe of transcript data filtered by quality value threshold.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>qv_threshold</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>30</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_id_col</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the cell ID.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cell_id&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="typing.Any">Any</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Any]: A dictionary containing various transcript metrics:
- 'percent_assigned' (float): The percentage of assigned transcripts.
- 'percent_cytoplasmic' (float): The percentage of cytoplasmic transcripts among assigned transcripts.
- 'percent_nucleus' (float): The percentage of nucleus transcripts among assigned transcripts.
- 'percent_non_assigned_cytoplasmic' (float): The percentage of non-assigned cytoplasmic transcripts.
- 'gene_metrics' (pd.DataFrame): A dataframe containing gene-level metrics.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_transcript_metrics</span><span class="p">(</span>
    <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">qv_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">cell_id_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_id&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes various metrics for a given dataframe of transcript data filtered by quality value threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        df (pd.DataFrame): The dataframe containing transcript data.</span>
<span class="sd">        qv_threshold (float): The quality value threshold for filtering transcripts.</span>
<span class="sd">        cell_id_col (str): The name of the column representing the cell ID.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Any]: A dictionary containing various transcript metrics:</span>
<span class="sd">            - &#39;percent_assigned&#39; (float): The percentage of assigned transcripts.</span>
<span class="sd">            - &#39;percent_cytoplasmic&#39; (float): The percentage of cytoplasmic transcripts among assigned transcripts.</span>
<span class="sd">            - &#39;percent_nucleus&#39; (float): The percentage of nucleus transcripts among assigned transcripts.</span>
<span class="sd">            - &#39;percent_non_assigned_cytoplasmic&#39; (float): The percentage of non-assigned cytoplasmic transcripts.</span>
<span class="sd">            - &#39;gene_metrics&#39; (pd.DataFrame): A dataframe containing gene-level metrics.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;qv&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">qv_threshold</span><span class="p">]</span>
    <span class="n">total_transcripts</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_filtered</span><span class="p">)</span>
    <span class="n">assigned_transcripts</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="p">[</span><span class="n">df_filtered</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">percent_assigned</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">assigned_transcripts</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">total_transcripts</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">cytoplasmic_transcripts</span> <span class="o">=</span> <span class="n">assigned_transcripts</span><span class="p">[</span><span class="n">assigned_transcripts</span><span class="p">[</span><span class="s2">&quot;overlaps_nucleus&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">percent_cytoplasmic</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cytoplasmic_transcripts</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">assigned_transcripts</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">percent_nucleus</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">-</span> <span class="n">percent_cytoplasmic</span>
    <span class="n">non_assigned_transcripts</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="p">[</span><span class="n">df_filtered</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">non_assigned_cytoplasmic</span> <span class="o">=</span> <span class="n">non_assigned_transcripts</span><span class="p">[</span><span class="n">non_assigned_transcripts</span><span class="p">[</span><span class="s2">&quot;overlaps_nucleus&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">percent_non_assigned_cytoplasmic</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">non_assigned_cytoplasmic</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">non_assigned_transcripts</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">gene_group_assigned</span> <span class="o">=</span> <span class="n">assigned_transcripts</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)</span>
    <span class="n">gene_group_all</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)</span>
    <span class="n">gene_percent_assigned</span> <span class="o">=</span> <span class="p">(</span><span class="n">gene_group_assigned</span><span class="o">.</span><span class="n">size</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="n">gene_group_all</span><span class="o">.</span><span class="n">size</span><span class="p">()</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span>
        <span class="n">names</span><span class="o">=</span><span class="s2">&quot;percent_assigned&quot;</span>
    <span class="p">)</span>
    <span class="n">cytoplasmic_gene_group</span> <span class="o">=</span> <span class="n">cytoplasmic_transcripts</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)</span>
    <span class="n">gene_percent_cytoplasmic</span> <span class="o">=</span> <span class="p">(</span><span class="n">cytoplasmic_gene_group</span><span class="o">.</span><span class="n">size</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cytoplasmic_transcripts</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;percent_cytoplasmic&quot;</span>
    <span class="p">)</span>
    <span class="n">gene_metrics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">gene_percent_assigned</span><span class="p">,</span> <span class="n">gene_percent_cytoplasmic</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;feature_name&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;percent_assigned&quot;</span><span class="p">:</span> <span class="n">percent_assigned</span><span class="p">,</span>
        <span class="s2">&quot;percent_cytoplasmic&quot;</span><span class="p">:</span> <span class="n">percent_cytoplasmic</span><span class="p">,</span>
        <span class="s2">&quot;percent_nucleus&quot;</span><span class="p">:</span> <span class="n">percent_nucleus</span><span class="p">,</span>
        <span class="s2">&quot;percent_non_assigned_cytoplasmic&quot;</span><span class="p">:</span> <span class="n">percent_non_assigned_cytoplasmic</span><span class="p">,</span>
        <span class="s2">&quot;gene_metrics&quot;</span><span class="p">:</span> <span class="n">gene_metrics</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.data.io.create_anndata" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">create_anndata</span>


<a href="#segger.data.io.create_anndata" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">create_anndata</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">panel_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_transcripts</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">cell_id_col</span><span class="o">=</span><span class="s1">&#39;cell_id&#39;</span><span class="p">,</span> <span class="n">qv_threshold</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">min_cell_area</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span> <span class="n">max_cell_area</span><span class="o">=</span><span class="mf">1000.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Generates an AnnData object from a dataframe of segmented transcriptomics data.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing segmented transcriptomics data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>panel_df</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="pandas.DataFrame">DataFrame</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing panel information.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_transcripts</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum number of transcripts required for a cell to be included.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_id_col</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The column name representing the cell ID in the input dataframe.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cell_id&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>qv_threshold</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>30</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_cell_area</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum cell area to include a cell.</p>
              </div>
            </td>
            <td>
                  <code>10.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>max_cell_area</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The maximum cell area to include a cell.</p>
              </div>
            </td>
            <td>
                  <code>1000.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ad.AnnData: The generated AnnData object containing the transcriptomics data and metadata.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_anndata</span><span class="p">(</span>
    <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">panel_df</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">min_transcripts</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">cell_id_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_id&quot;</span><span class="p">,</span>
    <span class="n">qv_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span>
    <span class="n">min_cell_area</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">10.0</span><span class="p">,</span>
    <span class="n">max_cell_area</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1000.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates an AnnData object from a dataframe of segmented transcriptomics data.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        df (pd.DataFrame): The dataframe containing segmented transcriptomics data.</span>
<span class="sd">        panel_df (Optional[pd.DataFrame]): The dataframe containing panel information.</span>
<span class="sd">        min_transcripts (int): The minimum number of transcripts required for a cell to be included.</span>
<span class="sd">        cell_id_col (str): The column name representing the cell ID in the input dataframe.</span>
<span class="sd">        qv_threshold (float): The quality value threshold for filtering transcripts.</span>
<span class="sd">        min_cell_area (float): The minimum cell area to include a cell.</span>
<span class="sd">        max_cell_area (float): The maximum cell area to include a cell.</span>

<span class="sd">    Returns:</span>
<span class="sd">        ad.AnnData: The generated AnnData object containing the transcriptomics data and metadata.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter out unassigned cells</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;UNASSIGNED&quot;</span><span class="p">]</span>

    <span class="c1"># Create pivot table for gene expression counts per cell</span>
    <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">cell_id_col</span><span class="p">:</span> <span class="s2">&quot;cell&quot;</span><span class="p">,</span> <span class="s2">&quot;feature_name&quot;</span><span class="p">:</span> <span class="s2">&quot;gene&quot;</span><span class="p">})[[</span><span class="s2">&quot;cell&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="s2">&quot;cell&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">aggfunc</span><span class="o">=</span><span class="s2">&quot;size&quot;</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span>
    <span class="p">)</span>
    <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="p">[</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">min_transcripts</span><span class="p">]</span>

    <span class="c1"># Summarize cell metrics</span>
    <span class="n">cell_summary</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">cell_id</span><span class="p">,</span> <span class="n">cell_data</span> <span class="ow">in</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">cell_id_col</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_data</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_transcripts</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">cell_convex_hull</span> <span class="o">=</span> <span class="n">ConvexHull</span><span class="p">(</span><span class="n">cell_data</span><span class="p">[[</span><span class="s2">&quot;x_location&quot;</span><span class="p">,</span> <span class="s2">&quot;y_location&quot;</span><span class="p">]],</span> <span class="n">qhull_options</span><span class="o">=</span><span class="s2">&quot;QJ&quot;</span><span class="p">)</span>
        <span class="n">cell_area</span> <span class="o">=</span> <span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">area</span>
        <span class="k">if</span> <span class="n">cell_area</span> <span class="o">&lt;</span> <span class="n">min_cell_area</span> <span class="ow">or</span> <span class="n">cell_area</span> <span class="o">&gt;</span> <span class="n">max_cell_area</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">cell_summary</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;cell&quot;</span><span class="p">:</span> <span class="n">cell_id</span><span class="p">,</span>
                <span class="s2">&quot;cell_centroid_x&quot;</span><span class="p">:</span> <span class="n">cell_data</span><span class="p">[</span><span class="s2">&quot;x_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;cell_centroid_y&quot;</span><span class="p">:</span> <span class="n">cell_data</span><span class="p">[</span><span class="s2">&quot;y_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;cell_area&quot;</span><span class="p">:</span> <span class="n">cell_area</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="n">cell_summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_summary</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;cell&quot;</span><span class="p">)</span>

    <span class="c1"># Add genes from panel_df (if provided) to the pivot table</span>
    <span class="k">if</span> <span class="n">panel_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">panel_df</span> <span class="o">=</span> <span class="n">panel_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">panel_df</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pivot_df</span><span class="p">:</span>
                <span class="n">pivot_df</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="p">[</span><span class="n">genes</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>

    <span class="c1"># Create var DataFrame</span>
    <span class="k">if</span> <span class="n">panel_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">var_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">{</span><span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;feature_types&quot;</span><span class="p">:</span> <span class="s2">&quot;Gene Expression&quot;</span><span class="p">,</span> <span class="s2">&quot;genome&quot;</span><span class="p">:</span> <span class="s2">&quot;Unknown&quot;</span><span class="p">}</span>
                <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">var_df</span> <span class="o">=</span> <span class="n">panel_df</span><span class="p">[[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;ensembl&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;ensembl&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_ids&quot;</span><span class="p">})</span>
        <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;feature_types&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Gene Expression&quot;</span>
        <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Unknown&quot;</span>
        <span class="n">var_df</span> <span class="o">=</span> <span class="n">var_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>

    <span class="c1"># Compute total assigned and unassigned transcript counts for each gene</span>
    <span class="n">assigned_counts</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">unassigned_counts</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;UNASSIGNED&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;total_assigned&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">var_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">assigned_counts</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;total_unassigned&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">var_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">unassigned_counts</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="c1"># Filter cells and create the AnnData object</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">cell_summary</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">cell_summary</span> <span class="o">=</span> <span class="n">cell_summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">var</span> <span class="o">=</span> <span class="n">var_df</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;unique_transcripts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pivot_df</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">cell_summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="p">:],</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.data.io.filter_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">filter_transcripts</span>


<a href="#segger.data.io.filter_transcripts" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">filter_transcripts</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">min_qv</span><span class="o">=</span><span class="mf">20.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Filters transcripts based on quality value and removes unwanted transcripts.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>transcripts_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_qv</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>20.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: The filtered dataframe.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_transcripts</span><span class="p">(</span>  <span class="c1"># ONLY FOR XENIUM</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">min_qv</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">20.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters transcripts based on quality value and removes unwanted transcripts.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        transcripts_df (pd.DataFrame): The dataframe containing transcript data.</span>
<span class="sd">        min_qv (float): The minimum quality value threshold for filtering transcripts.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The filtered dataframe.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filter_codewords</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;NegControlProbe_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;antisense_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NegControlCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;BLANK_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;DeprecatedCodeword_&quot;</span><span class="p">,</span>
        <span class="s2">&quot;UnassignedCodeword_&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">transcripts_df</span><span class="p">[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span>
    <span class="p">)</span>
    <span class="n">mask_quality</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="s2">&quot;qv&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_qv</span>

    <span class="c1"># Apply the filter for unwanted codewords using Dask string functions</span>
    <span class="n">mask_codewords</span> <span class="o">=</span> <span class="o">~</span><span class="n">transcripts_df</span><span class="p">[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="n">filter_codewords</span><span class="p">)</span>

    <span class="c1"># Combine the filters and return the filtered Dask DataFrame</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">mask_quality</span> <span class="o">&amp;</span> <span class="n">mask_codewords</span>
    <span class="k">return</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.data.io.format_time" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">format_time</span>


<a href="#segger.data.io.format_time" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">format_time</span><span class="p">(</span><span class="n">elapsed</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Format elapsed time to h<img alt="" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/24c2.svg" title=":m:" />s.</p>
<h5 id="segger.data.io.format_time--parameters">Parameters:<a class="headerlink" href="#segger.data.io.format_time--parameters" title="Permanent link">&para;</a></h5>
<p>elapsed : float
    Elapsed time in seconds.</p>
<h5 id="segger.data.io.format_time--returns">Returns:<a class="headerlink" href="#segger.data.io.format_time--returns" title="Permanent link">&para;</a></h5>
<p>str
    Formatted time in h<img alt="" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/24c2.svg" title=":m:" />s.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">format_time</span><span class="p">(</span><span class="n">elapsed</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Format elapsed time to h:m:s.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    ----------</span>
<span class="sd">    elapsed : float</span>
<span class="sd">        Elapsed time in seconds.</span>

<span class="sd">    Returns:</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Formatted time in h:m:s.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">timedelta</span><span class="p">(</span><span class="n">seconds</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">elapsed</span><span class="p">)))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.data.io.get_edge_index" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_edge_index</span>


<a href="#segger.data.io.get_edge_index" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_edge_index</span><span class="p">(</span><span class="n">coords_1</span><span class="p">,</span> <span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;kd_tree&#39;</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Computes edge indices using KD-Tree.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>coords_1</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>First set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>coords_2</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Second set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>k</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dist</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distance threshold.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>method</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The method to use. Only 'kd_tree' is supported now.</p>
              </div>
            </td>
            <td>
                  <code>&#39;kd_tree&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="torch.Tensor">Tensor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: Edge indices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_edge_index</span><span class="p">(</span>
    <span class="n">coords_1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">coords_2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;kd_tree&quot;</span><span class="p">,</span>
    <span class="n">workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes edge indices using KD-Tree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        coords_1 (np.ndarray): First set of coordinates.</span>
<span class="sd">        coords_2 (np.ndarray): Second set of coordinates.</span>
<span class="sd">        k (int, optional): Number of nearest neighbors.</span>
<span class="sd">        dist (int, optional): Distance threshold.</span>
<span class="sd">        method (str, optional): The method to use. Only &#39;kd_tree&#39; is supported now.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: Edge indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;kd_tree&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">get_edge_index_kdtree</span><span class="p">(</span><span class="n">coords_1</span><span class="p">,</span> <span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="n">workers</span><span class="p">)</span>
    <span class="c1"># elif method == &quot;cuda&quot;:</span>
    <span class="c1">#     return get_edge_index_cuda(coords_1, coords_2, k=k, dist=dist)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Unknown method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">. The only supported method is &#39;kd_tree&#39; now.&quot;</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.data.io.get_edge_index_kdtree" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_edge_index_kdtree</span>


<a href="#segger.data.io.get_edge_index_kdtree" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_edge_index_kdtree</span><span class="p">(</span><span class="n">coords_1</span><span class="p">,</span> <span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Computes edge indices using KDTree.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>coords_1</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>First set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>coords_2</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Second set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>k</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dist</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distance threshold.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="torch.Tensor">Tensor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: Edge indices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_edge_index_kdtree</span><span class="p">(</span>
    <span class="n">coords_1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">coords_2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes edge indices using KDTree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        coords_1 (np.ndarray): First set of coordinates.</span>
<span class="sd">        coords_2 (np.ndarray): Second set of coordinates.</span>
<span class="sd">        k (int, optional): Number of nearest neighbors.</span>
<span class="sd">        dist (int, optional): Distance threshold.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: Edge indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">coords_1</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
        <span class="n">coords_1</span> <span class="o">=</span> <span class="n">coords_1</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">coords_2</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
        <span class="n">coords_2</span> <span class="o">=</span> <span class="n">coords_2</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">cKDTree</span><span class="p">(</span><span class="n">coords_1</span><span class="p">)</span>
    <span class="n">d_kdtree</span><span class="p">,</span> <span class="n">idx_out</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">distance_upper_bound</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="n">workers</span><span class="p">)</span>
    <span class="n">valid_mask</span> <span class="o">=</span> <span class="n">d_kdtree</span> <span class="o">&lt;</span> <span class="n">dist</span>
    <span class="n">edges</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">valid</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">valid_mask</span><span class="p">):</span>
        <span class="n">valid_indices</span> <span class="o">=</span> <span class="n">idx_out</span><span class="p">[</span><span class="n">idx</span><span class="p">][</span><span class="n">valid</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">valid_indices</span><span class="o">.</span><span class="n">size</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">valid_indices</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">idx</span><span class="p">),</span> <span class="n">valid_indices</span><span class="p">))</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

    <span class="n">edge_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="n">edges</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">)</span><span class="o">.</span><span class="n">contiguous</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">edge_index</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.data.io.get_xy_extents" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_xy_extents</span>


<a href="#segger.data.io.get_xy_extents" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_xy_extents</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Get the bounding box of the x and y coordinates from a Parquet file.</p>
<h5 id="segger.data.io.get_xy_extents--parameters">Parameters<a class="headerlink" href="#segger.data.io.get_xy_extents--parameters" title="Permanent link">&para;</a></h5>
<p>filepath : str
    The path to the Parquet file.
x : str
    The name of the column representing the x-coordinate.
y : str
    The name of the column representing the y-coordinate.</p>
<h5 id="segger.data.io.get_xy_extents--returns">Returns<a class="headerlink" href="#segger.data.io.get_xy_extents--returns" title="Permanent link">&para;</a></h5>
<p>shapely.Polygon
    A polygon representing the bounding box of the x and y coordinates.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/data/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_xy_extents</span><span class="p">(</span>
    <span class="n">filepath</span><span class="p">,</span>
    <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the bounding box of the x and y coordinates from a Parquet file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    filepath : str</span>
<span class="sd">        The path to the Parquet file.</span>
<span class="sd">    x : str</span>
<span class="sd">        The name of the column representing the x-coordinate.</span>
<span class="sd">    y : str</span>
<span class="sd">        The name of the column representing the y-coordinate.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    shapely.Polygon</span>
<span class="sd">        A polygon representing the bounding box of the x and y coordinates.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get index of columns of parquet file</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="n">pq</span><span class="o">.</span><span class="n">read_metadata</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="n">schema_idx</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">reversed</span><span class="p">,</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metadata</span><span class="o">.</span><span class="n">schema</span><span class="o">.</span><span class="n">names</span><span class="p">)))</span>

    <span class="c1"># Find min and max values across all row groups</span>
    <span class="n">x_max</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">x_min</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">maxsize</span>
    <span class="n">y_max</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">y_min</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">maxsize</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">metadata</span><span class="o">.</span><span class="n">num_row_groups</span><span class="p">):</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">row_group</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">x_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">min</span><span class="p">)</span>
        <span class="n">x_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">x_max</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">max</span><span class="p">)</span>
        <span class="n">y_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">y</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">min</span><span class="p">)</span>
        <span class="n">y_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">y_max</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">y</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">max</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">x_min</span><span class="p">,</span> <span class="n">y_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">,</span> <span class="n">y_max</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.sections", "navigation.instant", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate", "content.tabs.link", "navigation.tracking"], "search": "../../../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/termynal@0.0.1/termynal.js"></script>
      
        <script src="../../../termynal.js"></script>
      
    
  </body>
</html>