
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Fast and accurate cell segmentation for single-molecule spatial omics">
      
      
        <meta name="author" content="Elyas Heidari">
      
      
        <link rel="canonical" href="https://EliHei2.github.io/segger_dev/api/validation/">
      
      
        <link rel="prev" href="../prediction/">
      
      
      
      <link rel="icon" href="../../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.11">
    
    
      
        <title>Validation - segger Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=ubuntu:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"ubuntu";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css">
    
      <link rel="stylesheet" href="https://unpkg.com/termynal@0.0.1/termynal.css">
    
      <link rel="stylesheet" href="../../termynal.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#seggervalidation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="segger Documentation" class="md-header__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            segger Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Validation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="segger Documentation" class="md-nav__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../images/logo.png" alt="logo">

    </a>
    segger Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user_guide/data_creation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Dataset Creation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user_guide/training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../user_guide/validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../notebooks/segger_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Segger
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    CLI
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../models/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Models
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../prediction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Prediction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Validation
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Validation
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#submodules" class="md-nav__link">
    <span class="md-ellipsis">
      Submodules
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#api-documentation" class="md-nav__link">
    <span class="md-ellipsis">
      API Documentation
    </span>
  </a>
  
    <nav class="md-nav" aria-label="API Documentation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#segger.validation" class="md-nav__link">
    <span class="md-ellipsis">
      validation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.annotate_query_with_reference" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_query_with_reference
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.calculate_contamination" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_contamination
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.calculate_sensitivity" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_sensitivity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.compute_MECR" class="md-nav__link">
    <span class="md-ellipsis">
      compute_MECR
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.compute_clustering_scores" class="md-nav__link">
    <span class="md-ellipsis">
      compute_clustering_scores
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.compute_neighborhood_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_neighborhood_metrics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.compute_quantized_mecr_area" class="md-nav__link">
    <span class="md-ellipsis">
      compute_quantized_mecr_area
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.compute_quantized_mecr_counts" class="md-nav__link">
    <span class="md-ellipsis">
      compute_quantized_mecr_counts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.compute_transcript_density" class="md-nav__link">
    <span class="md-ellipsis">
      compute_transcript_density
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.draw_umap" class="md-nav__link">
    <span class="md-ellipsis">
      draw_umap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.find_markers" class="md-nav__link">
    <span class="md-ellipsis">
      find_markers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.find_mutually_exclusive_genes" class="md-nav__link">
    <span class="md-ellipsis">
      find_mutually_exclusive_genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.generate_experiment_file" class="md-nav__link">
    <span class="md-ellipsis">
      generate_experiment_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.get_flatten_version" class="md-nav__link">
    <span class="md-ellipsis">
      get_flatten_version
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.get_indices_indptr" class="md-nav__link">
    <span class="md-ellipsis">
      get_indices_indptr
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.get_leiden_umap" class="md-nav__link">
    <span class="md-ellipsis">
      get_leiden_umap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.get_median_expression_table" class="md-nav__link">
    <span class="md-ellipsis">
      get_median_expression_table
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.load_segmentations" class="md-nav__link">
    <span class="md-ellipsis">
      load_segmentations
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_cell_area" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cell_area
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_cell_counts" class="md-nav__link">
    <span class="md-ellipsis">
      plot_cell_counts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_contamination_boxplots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_contamination_boxplots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_contamination_results" class="md-nav__link">
    <span class="md-ellipsis">
      plot_contamination_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_counts_per_cell" class="md-nav__link">
    <span class="md-ellipsis">
      plot_counts_per_cell
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_entropy_boxplots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_entropy_boxplots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_gene_counts" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gene_counts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_general_statistics_plots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_general_statistics_plots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_mecr_results" class="md-nav__link">
    <span class="md-ellipsis">
      plot_mecr_results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_metric_comparison" class="md-nav__link">
    <span class="md-ellipsis">
      plot_metric_comparison
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_percent_assigned" class="md-nav__link">
    <span class="md-ellipsis">
      plot_percent_assigned
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_quantized_mecr_area" class="md-nav__link">
    <span class="md-ellipsis">
      plot_quantized_mecr_area
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_quantized_mecr_counts" class="md-nav__link">
    <span class="md-ellipsis">
      plot_quantized_mecr_counts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_sensitivity_boxplots" class="md-nav__link">
    <span class="md-ellipsis">
      plot_sensitivity_boxplots
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_transcript_density" class="md-nav__link">
    <span class="md-ellipsis">
      plot_transcript_density
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.plot_umaps_with_scores" class="md-nav__link">
    <span class="md-ellipsis">
      plot_umaps_with_scores
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.save_cell_clustering" class="md-nav__link">
    <span class="md-ellipsis">
      save_cell_clustering
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.seg2explorer" class="md-nav__link">
    <span class="md-ellipsis">
      seg2explorer
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#segger.validation.str_to_uint32" class="md-nav__link">
    <span class="md-ellipsis">
      str_to_uint32
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="seggervalidation">segger.validation<a class="headerlink" href="#seggervalidation" title="Permanent link">&para;</a></h1>
<p>This module handles validation utilities for the Segger tool.</p>
<h2 id="submodules">Submodules<a class="headerlink" href="#submodules" title="Permanent link">&para;</a></h2>
<ul>
<li><a href="utils/index.md">Utils</a></li>
<li><a href="xenium_explorer/index.md">Xenium Explorer</a></li>
</ul>
<h2 id="api-documentation">API Documentation<a class="headerlink" href="#api-documentation" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-module">



<a id="segger.validation"></a>
    <div class="doc doc-contents first">









  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="segger.validation.annotate_query_with_reference" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">annotate_query_with_reference</span>


<a href="#segger.validation.annotate_query_with_reference" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">annotate_query_with_reference</span><span class="p">(</span><span class="n">reference_adata</span><span class="p">,</span> <span class="n">query_adata</span><span class="p">,</span> <span class="n">transfer_column</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Annotate query AnnData object using a scRNA-seq reference atlas.</p>
        <ul>
<li>reference_adata: ad.AnnData
    Reference AnnData object containing the scRNA-seq atlas data.</li>
<li>query_adata: ad.AnnData
    Query AnnData object containing the data to be annotated.</li>
<li>transfer_column: str
    The name of the column in the reference atlas's <code>obs</code> to transfer to the query dataset.</li>
</ul>
        <ul>
<li>query_adata: ad.AnnData
    Annotated query AnnData object with transferred labels and UMAP coordinates from the reference.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">annotate_query_with_reference</span><span class="p">(</span>
    <span class="n">reference_adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">query_adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">transfer_column</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate query AnnData object using a scRNA-seq reference atlas.</span>

<span class="sd">    Args:</span>
<span class="sd">    - reference_adata: ad.AnnData</span>
<span class="sd">        Reference AnnData object containing the scRNA-seq atlas data.</span>
<span class="sd">    - query_adata: ad.AnnData</span>
<span class="sd">        Query AnnData object containing the data to be annotated.</span>
<span class="sd">    - transfer_column: str</span>
<span class="sd">        The name of the column in the reference atlas&#39;s `obs` to transfer to the query dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - query_adata: ad.AnnData</span>
<span class="sd">        Annotated query AnnData object with transferred labels and UMAP coordinates from the reference.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">common_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">reference_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">query_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">))</span>
    <span class="n">reference_adata</span> <span class="o">=</span> <span class="n">reference_adata</span><span class="p">[:,</span> <span class="n">common_genes</span><span class="p">]</span>
    <span class="n">query_adata</span> <span class="o">=</span> <span class="n">query_adata</span><span class="p">[:,</span> <span class="n">common_genes</span><span class="p">]</span>
    <span class="n">query_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;raw&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">query_adata</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">X</span> <span class="k">if</span> <span class="n">query_adata</span><span class="o">.</span><span class="n">raw</span> <span class="k">else</span> <span class="n">query_adata</span><span class="o">.</span><span class="n">X</span>
    <span class="n">query_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;raw_counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">query_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;raw&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">query_adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">query_adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">reference_adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">reference_adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">reference_adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ingest</span><span class="p">(</span><span class="n">query_adata</span><span class="p">,</span> <span class="n">reference_adata</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">transfer_column</span><span class="p">)</span>
    <span class="n">query_adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">query_adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">query_adata</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.calculate_contamination" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">calculate_contamination</span>


<a href="#segger.validation.calculate_contamination" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">calculate_contamination</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">markers</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">n_neighs</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">celltype_column</span><span class="o">=</span><span class="s1">&#39;celltype_major&#39;</span><span class="p">,</span> <span class="n">num_cells</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Calculate normalized contamination from neighboring cells of different cell types based on positive markers.</p>
        <ul>
<li>adata: ad.AnnData
    Annotated data object with raw counts and cell type information.</li>
<li>markers: dict
    Dictionary where keys are cell types and values are dictionaries containing:
        'positive': list of top x% highly expressed genes
        'negative': list of top x% lowly expressed genes.</li>
<li>radius: float, default=15
    Radius for spatial neighbor calculation.</li>
<li>n_neighs: int, default=10
    Maximum number of neighbors to consider.</li>
<li>celltype_column: str, default='celltype_major'
    Column name in the AnnData object representing cell types.</li>
<li>num_cells: int, default=10000
    Number of cells to randomly select for the calculation.</li>
</ul>
        <ul>
<li>contamination_df: pd.DataFrame
    DataFrame containing the normalized level of contamination from each cell type to each other cell type.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">calculate_contamination</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">markers</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]],</span>
    <span class="n">radius</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">15</span><span class="p">,</span>
    <span class="n">n_neighs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">celltype_column</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;celltype_major&quot;</span><span class="p">,</span>
    <span class="n">num_cells</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate normalized contamination from neighboring cells of different cell types based on positive markers.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: ad.AnnData</span>
<span class="sd">        Annotated data object with raw counts and cell type information.</span>
<span class="sd">    - markers: dict</span>
<span class="sd">        Dictionary where keys are cell types and values are dictionaries containing:</span>
<span class="sd">            &#39;positive&#39;: list of top x% highly expressed genes</span>
<span class="sd">            &#39;negative&#39;: list of top x% lowly expressed genes.</span>
<span class="sd">    - radius: float, default=15</span>
<span class="sd">        Radius for spatial neighbor calculation.</span>
<span class="sd">    - n_neighs: int, default=10</span>
<span class="sd">        Maximum number of neighbors to consider.</span>
<span class="sd">    - celltype_column: str, default=&#39;celltype_major&#39;</span>
<span class="sd">        Column name in the AnnData object representing cell types.</span>
<span class="sd">    - num_cells: int, default=10000</span>
<span class="sd">        Number of cells to randomly select for the calculation.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - contamination_df: pd.DataFrame</span>
<span class="sd">        DataFrame containing the normalized level of contamination from each cell type to each other cell type.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">celltype_column</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Column celltype_column must be present in adata.obs.&quot;</span><span class="p">)</span>
    <span class="n">positive_markers</span> <span class="o">=</span> <span class="p">{</span><span class="n">ct</span><span class="p">:</span> <span class="n">markers</span><span class="p">[</span><span class="n">ct</span><span class="p">][</span><span class="s2">&quot;positive&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">markers</span><span class="p">}</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;spatial&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="s2">&quot;cell_centroid_x&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_centroid_y&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">sq</span><span class="o">.</span><span class="n">gr</span><span class="o">.</span><span class="n">spatial_neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="n">radius</span><span class="p">,</span> <span class="n">n_neighs</span><span class="o">=</span><span class="n">n_neighs</span><span class="p">,</span> <span class="n">coord_type</span><span class="o">=</span><span class="s2">&quot;generic&quot;</span><span class="p">)</span>
    <span class="n">neighbors</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s2">&quot;spatial_connectivities&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolil</span><span class="p">()</span>
    <span class="n">raw_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;raw&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="n">cell_types</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_column</span><span class="p">]</span>
    <span class="n">selected_cells</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="n">num_cells</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">),</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">contamination</span> <span class="o">=</span> <span class="p">{</span><span class="n">ct</span><span class="p">:</span> <span class="p">{</span><span class="n">ct2</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">ct2</span> <span class="ow">in</span> <span class="n">positive_markers</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">positive_markers</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="n">negighborings</span> <span class="o">=</span> <span class="p">{</span><span class="n">ct</span><span class="p">:</span> <span class="p">{</span><span class="n">ct2</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">ct2</span> <span class="ow">in</span> <span class="n">positive_markers</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span> <span class="k">for</span> <span class="n">ct</span> <span class="ow">in</span> <span class="n">positive_markers</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="k">for</span> <span class="n">cell_idx</span> <span class="ow">in</span> <span class="n">selected_cells</span><span class="p">:</span>
        <span class="n">cell_type</span> <span class="o">=</span> <span class="n">cell_types</span><span class="p">[</span><span class="n">cell_idx</span><span class="p">]</span>
        <span class="n">own_markers</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">positive_markers</span><span class="p">[</span><span class="n">cell_type</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">marker</span> <span class="ow">in</span> <span class="n">own_markers</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">marker</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">:</span>
                <span class="n">total_counts_in_neighborhood</span> <span class="o">=</span> <span class="n">raw_counts</span><span class="p">[</span><span class="n">cell_idx</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">marker</span><span class="p">)]</span>
                <span class="k">for</span> <span class="n">neighbor_idx</span> <span class="ow">in</span> <span class="n">neighbors</span><span class="o">.</span><span class="n">rows</span><span class="p">[</span><span class="n">cell_idx</span><span class="p">]:</span>
                    <span class="n">total_counts_in_neighborhood</span> <span class="o">+=</span> <span class="n">raw_counts</span><span class="p">[</span><span class="n">neighbor_idx</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">marker</span><span class="p">)]</span>
                <span class="k">for</span> <span class="n">neighbor_idx</span> <span class="ow">in</span> <span class="n">neighbors</span><span class="o">.</span><span class="n">rows</span><span class="p">[</span><span class="n">cell_idx</span><span class="p">]:</span>
                    <span class="n">neighbor_type</span> <span class="o">=</span> <span class="n">cell_types</span><span class="p">[</span><span class="n">neighbor_idx</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">cell_type</span> <span class="o">==</span> <span class="n">neighbor_type</span><span class="p">:</span>
                        <span class="k">continue</span>
                    <span class="n">neighbor_markers</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">positive_markers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">neighbor_type</span><span class="p">,</span> <span class="p">[]))</span>
                    <span class="n">contamination_markers</span> <span class="o">=</span> <span class="n">own_markers</span> <span class="o">-</span> <span class="n">neighbor_markers</span>
                    <span class="k">for</span> <span class="n">marker</span> <span class="ow">in</span> <span class="n">contamination_markers</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">marker</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">:</span>
                            <span class="n">marker_counts_in_neighbor</span> <span class="o">=</span> <span class="n">raw_counts</span><span class="p">[</span><span class="n">neighbor_idx</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">marker</span><span class="p">)]</span>
                            <span class="k">if</span> <span class="n">total_counts_in_neighborhood</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                                <span class="n">contamination</span><span class="p">[</span><span class="n">cell_type</span><span class="p">][</span><span class="n">neighbor_type</span><span class="p">]</span> <span class="o">+=</span> <span class="p">(</span>
                                    <span class="n">marker_counts_in_neighbor</span> <span class="o">/</span> <span class="n">total_counts_in_neighborhood</span>
                                <span class="p">)</span>
                                <span class="n">negighborings</span><span class="p">[</span><span class="n">cell_type</span><span class="p">][</span><span class="n">neighbor_type</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="n">contamination_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">contamination</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">negighborings_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">negighborings</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">contamination_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;Source Cell Type&quot;</span>
    <span class="n">contamination_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;Target Cell Type&quot;</span>
    <span class="k">return</span> <span class="n">contamination_df</span> <span class="o">/</span> <span class="p">(</span><span class="n">negighborings_df</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.calculate_sensitivity" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">calculate_sensitivity</span>


<a href="#segger.validation.calculate_sensitivity" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">calculate_sensitivity</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">purified_markers</span><span class="p">,</span> <span class="n">max_cells_per_type</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Calculate the sensitivity of the purified markers for each cell type.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing gene expression data.</li>
<li>purified_markers: dict
    Dictionary where keys are cell types and values are lists of purified marker genes.</li>
<li>max_cells_per_type: int, default=1000
    Maximum number of cells to consider per cell type.</li>
</ul>
        <ul>
<li>sensitivity_results: dict
    Dictionary with cell types as keys and lists of sensitivity values for each cell.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">calculate_sensitivity</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">purified_markers</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="n">max_cells_per_type</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the sensitivity of the purified markers for each cell type.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing gene expression data.</span>
<span class="sd">    - purified_markers: dict</span>
<span class="sd">        Dictionary where keys are cell types and values are lists of purified marker genes.</span>
<span class="sd">    - max_cells_per_type: int, default=1000</span>
<span class="sd">        Maximum number of cells to consider per cell type.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - sensitivity_results: dict</span>
<span class="sd">        Dictionary with cell types as keys and lists of sensitivity values for each cell.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sensitivity_results</span> <span class="o">=</span> <span class="p">{</span><span class="n">cell_type</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">purified_markers</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">markers</span> <span class="ow">in</span> <span class="n">purified_markers</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">markers</span> <span class="o">=</span> <span class="n">markers</span><span class="p">[</span><span class="s2">&quot;positive&quot;</span><span class="p">]</span>
        <span class="n">subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;celltype_major&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">subset</span><span class="o">.</span><span class="n">n_obs</span> <span class="o">&gt;</span> <span class="n">max_cells_per_type</span><span class="p">:</span>
            <span class="n">cell_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">subset</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="n">max_cells_per_type</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">subset</span> <span class="o">=</span> <span class="n">subset</span><span class="p">[</span><span class="n">cell_indices</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">cell_counts</span> <span class="ow">in</span> <span class="n">subset</span><span class="o">.</span><span class="n">X</span><span class="p">:</span>
            <span class="n">expressed_markers</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">((</span><span class="n">cell_counts</span><span class="p">[</span><span class="n">subset</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">get_indexer</span><span class="p">(</span><span class="n">markers</span><span class="p">)]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
            <span class="n">sensitivity</span> <span class="o">=</span> <span class="n">expressed_markers</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">markers</span><span class="p">)</span> <span class="k">if</span> <span class="n">markers</span> <span class="k">else</span> <span class="mi">0</span>
            <span class="n">sensitivity_results</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sensitivity</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sensitivity_results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.compute_MECR" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_MECR</span>


<a href="#segger.validation.compute_MECR" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_MECR</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_pairs</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Compute the Mutually Exclusive Co-expression Rate (MECR) for each gene pair in an AnnData object.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing gene expression data.</li>
<li>gene_pairs: List[Tuple[str, str]]
    List of tuples representing gene pairs to evaluate.</li>
</ul>
        <ul>
<li>mecr_dict: Dict[Tuple[str, str], float]
    Dictionary where keys are gene pairs (tuples) and values are MECR values.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_MECR</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">gene_pairs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the Mutually Exclusive Co-expression Rate (MECR) for each gene pair in an AnnData object.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing gene expression data.</span>
<span class="sd">    - gene_pairs: List[Tuple[str, str]]</span>
<span class="sd">        List of tuples representing gene pairs to evaluate.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - mecr_dict: Dict[Tuple[str, str], float]</span>
<span class="sd">        Dictionary where keys are gene pairs (tuples) and values are MECR values.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mecr_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span> <span class="ow">in</span> <span class="n">gene_pairs</span><span class="p">:</span>
        <span class="n">expr_gene1</span> <span class="o">=</span> <span class="n">gene_expression</span><span class="p">[</span><span class="n">gene1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="n">expr_gene2</span> <span class="o">=</span> <span class="n">gene_expression</span><span class="p">[</span><span class="n">gene2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="n">both_expressed</span> <span class="o">=</span> <span class="p">(</span><span class="n">expr_gene1</span> <span class="o">&amp;</span> <span class="n">expr_gene2</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">at_least_one_expressed</span> <span class="o">=</span> <span class="p">(</span><span class="n">expr_gene1</span> <span class="o">|</span> <span class="n">expr_gene2</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">mecr</span> <span class="o">=</span> <span class="n">both_expressed</span> <span class="o">/</span> <span class="n">at_least_one_expressed</span> <span class="k">if</span> <span class="n">at_least_one_expressed</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="n">mecr_dict</span><span class="p">[(</span><span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span><span class="p">)]</span> <span class="o">=</span> <span class="n">mecr</span>
    <span class="k">return</span> <span class="n">mecr_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.compute_clustering_scores" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_clustering_scores</span>


<a href="#segger.validation.compute_clustering_scores" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_clustering_scores</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type_column</span><span class="o">=</span><span class="s1">&#39;celltype_major&#39;</span><span class="p">,</span> <span class="n">use_pca</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Compute the Calinski-Harabasz and Silhouette scores for an AnnData object based on the assigned cell types.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing gene expression data and cell type assignments.</li>
<li>cell_type_column: str, default='celltype_major'
    Column name in <code>adata.obs</code> that specifies cell types.</li>
<li>use_pca: bool, default=True
    Whether to use PCA components as features. If False, use the raw data.</li>
</ul>
        <ul>
<li>ch_score: float
    The Calinski-Harabasz score.</li>
<li>sh_score: float
    The Silhouette score.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_clustering_scores</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">cell_type_column</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;celltype_major&quot;</span><span class="p">,</span> <span class="n">use_pca</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the Calinski-Harabasz and Silhouette scores for an AnnData object based on the assigned cell types.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing gene expression data and cell type assignments.</span>
<span class="sd">    - cell_type_column: str, default=&#39;celltype_major&#39;</span>
<span class="sd">        Column name in `adata.obs` that specifies cell types.</span>
<span class="sd">    - use_pca: bool, default=True</span>
<span class="sd">        Whether to use PCA components as features. If False, use the raw data.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - ch_score: float</span>
<span class="sd">        The Calinski-Harabasz score.</span>
<span class="sd">    - sh_score: float</span>
<span class="sd">        The Silhouette score.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">cell_type_column</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Column &#39;</span><span class="si">{</span><span class="n">cell_type_column</span><span class="si">}</span><span class="s2">&#39; must be present in adata.obs.&quot;</span><span class="p">)</span>
    <span class="n">features</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
    <span class="n">cell_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="mi">10000</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">features</span> <span class="o">=</span> <span class="n">features</span><span class="p">[</span><span class="n">cell_indices</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">cell_indices</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_column</span><span class="p">]</span>
    <span class="n">ch_score</span> <span class="o">=</span> <span class="n">calinski_harabasz_score</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span>
    <span class="n">sh_score</span> <span class="o">=</span> <span class="n">silhouette_score</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="n">labels</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ch_score</span><span class="p">,</span> <span class="n">sh_score</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.compute_neighborhood_metrics" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_neighborhood_metrics</span>


<a href="#segger.validation.compute_neighborhood_metrics" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_neighborhood_metrics</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">celltype_column</span><span class="o">=</span><span class="s1">&#39;celltype_major&#39;</span><span class="p">,</span> <span class="n">n_neighs</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">subset_size</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Compute neighborhood entropy and number of neighbors for each cell in the AnnData object.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing spatial information and cell type assignments.</li>
<li>radius: int, default=10
    Radius for spatial neighbor calculation.</li>
<li>celltype_column: str, default='celltype_major'
    Column name in <code>adata.obs</code> that specifies cell types.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_neighborhood_metrics</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">radius</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">celltype_column</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;celltype_major&quot;</span><span class="p">,</span>
    <span class="n">n_neighs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">subset_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10000</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute neighborhood entropy and number of neighbors for each cell in the AnnData object.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing spatial information and cell type assignments.</span>
<span class="sd">    - radius: int, default=10</span>
<span class="sd">        Radius for spatial neighbor calculation.</span>
<span class="sd">    - celltype_column: str, default=&#39;celltype_major&#39;</span>
<span class="sd">        Column name in `adata.obs` that specifies cell types.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute neighborhood entropy and number of neighbors for a random subset of cells in the AnnData object.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing spatial information and cell type assignments.</span>
<span class="sd">    - radius: int, default=10</span>
<span class="sd">        Radius for spatial neighbor calculation.</span>
<span class="sd">    - celltype_column: str, default=&#39;celltype_major&#39;</span>
<span class="sd">        Column name in `adata.obs` that specifies cell types.</span>
<span class="sd">    - subset_size: int, default=10000</span>
<span class="sd">        Number of cells to randomly select for the calculation.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Ensure the subset size does not exceed the number of cells</span>
    <span class="n">subset_size</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">subset_size</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)</span>
    <span class="c1"># Randomly select a subset of cells</span>
    <span class="n">subset_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="n">subset_size</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Compute spatial neighbors for the entire dataset</span>
    <span class="n">sq</span><span class="o">.</span><span class="n">gr</span><span class="o">.</span><span class="n">spatial_neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="n">radius</span><span class="p">,</span> <span class="n">coord_type</span><span class="o">=</span><span class="s2">&quot;generic&quot;</span><span class="p">,</span> <span class="n">n_neighs</span><span class="o">=</span><span class="n">n_neighs</span><span class="p">)</span>
    <span class="n">neighbors</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obsp</span><span class="p">[</span><span class="s2">&quot;spatial_distances&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolil</span><span class="p">()</span><span class="o">.</span><span class="n">rows</span>
    <span class="n">entropies</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">num_neighbors</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Calculate entropy and number of neighbors only for the selected subset</span>
    <span class="k">for</span> <span class="n">cell_index</span> <span class="ow">in</span> <span class="n">subset_indices</span><span class="p">:</span>
        <span class="n">neighbor_indices</span> <span class="o">=</span> <span class="n">neighbors</span><span class="p">[</span><span class="n">cell_index</span><span class="p">]</span>
        <span class="n">neighbor_cell_types</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_column</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">neighbor_indices</span><span class="p">]</span>
        <span class="n">cell_type_counts</span> <span class="o">=</span> <span class="n">neighbor_cell_types</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
        <span class="n">total_neighbors</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">neighbor_cell_types</span><span class="p">)</span>
        <span class="n">num_neighbors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">total_neighbors</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">total_neighbors</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cell_type_probs</span> <span class="o">=</span> <span class="n">cell_type_counts</span> <span class="o">/</span> <span class="n">total_neighbors</span>
            <span class="n">cell_type_entropy</span> <span class="o">=</span> <span class="n">entropy</span><span class="p">(</span><span class="n">cell_type_probs</span><span class="p">)</span>
            <span class="n">entropies</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_type_entropy</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">entropies</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Store the results back into the original AnnData object</span>
    <span class="c1"># We fill with NaN for cells not in the subset</span>
    <span class="n">entropy_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">neighbors_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">entropy_full</span><span class="p">[</span><span class="n">subset_indices</span><span class="p">]</span> <span class="o">=</span> <span class="n">entropies</span>
    <span class="n">neighbors_full</span><span class="p">[</span><span class="n">subset_indices</span><span class="p">]</span> <span class="o">=</span> <span class="n">num_neighbors</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;neighborhood_entropy&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">entropy_full</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;number_of_neighbors&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">neighbors_full</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.compute_quantized_mecr_area" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_quantized_mecr_area</span>


<a href="#segger.validation.compute_quantized_mecr_area" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_quantized_mecr_area</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_pairs</span><span class="p">,</span> <span class="n">quantiles</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Compute the average MECR, variance of MECR, and average cell area for quantiles of cell areas.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing gene expression data.</li>
<li>gene_pairs: List[Tuple[str, str]]
    List of tuples representing gene pairs to evaluate.</li>
<li>quantiles: int, default=10
    Number of quantiles to divide the data into.</li>
</ul>
        <ul>
<li>quantized_data: pd.DataFrame
    DataFrame containing quantile information, average MECR, variance of MECR, average area, and number of cells.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_quantized_mecr_area</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">gene_pairs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span> <span class="n">quantiles</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the average MECR, variance of MECR, and average cell area for quantiles of cell areas.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing gene expression data.</span>
<span class="sd">    - gene_pairs: List[Tuple[str, str]]</span>
<span class="sd">        List of tuples representing gene pairs to evaluate.</span>
<span class="sd">    - quantiles: int, default=10</span>
<span class="sd">        Number of quantiles to divide the data into.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - quantized_data: pd.DataFrame</span>
<span class="sd">        DataFrame containing quantile information, average MECR, variance of MECR, average area, and number of cells.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;quantile&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">qcut</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_area&quot;</span><span class="p">],</span> <span class="n">quantiles</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">quantized_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">quantile</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">quantiles</span><span class="p">):</span>
        <span class="n">cells_in_quantile</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;quantile&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">quantile</span>
        <span class="n">mecr</span> <span class="o">=</span> <span class="n">compute_MECR</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">cells_in_quantile</span><span class="p">,</span> <span class="p">:],</span> <span class="n">gene_pairs</span><span class="p">)</span>
        <span class="n">average_mecr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mecr</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
        <span class="n">variance_mecr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mecr</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
        <span class="n">average_area</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells_in_quantile</span><span class="p">,</span> <span class="s2">&quot;cell_area&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">quantized_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;quantile&quot;</span><span class="p">:</span> <span class="n">quantile</span> <span class="o">/</span> <span class="n">quantiles</span><span class="p">,</span>
                <span class="s2">&quot;average_mecr&quot;</span><span class="p">:</span> <span class="n">average_mecr</span><span class="p">,</span>
                <span class="s2">&quot;variance_mecr&quot;</span><span class="p">:</span> <span class="n">variance_mecr</span><span class="p">,</span>
                <span class="s2">&quot;average_area&quot;</span><span class="p">:</span> <span class="n">average_area</span><span class="p">,</span>
                <span class="s2">&quot;num_cells&quot;</span><span class="p">:</span> <span class="n">cells_in_quantile</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">quantized_data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.compute_quantized_mecr_counts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_quantized_mecr_counts</span>


<a href="#segger.validation.compute_quantized_mecr_counts" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_quantized_mecr_counts</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">gene_pairs</span><span class="p">,</span> <span class="n">quantiles</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Compute the average MECR, variance of MECR, and average transcript counts for quantiles of transcript counts.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing gene expression data.</li>
<li>gene_pairs: List[Tuple[str, str]]
    List of tuples representing gene pairs to evaluate.</li>
<li>quantiles: int, default=10
    Number of quantiles to divide the data into.</li>
</ul>
        <ul>
<li>quantized_data: pd.DataFrame
    DataFrame containing quantile information, average MECR, variance of MECR, average counts, and number of cells.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_quantized_mecr_counts</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">gene_pairs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span> <span class="n">quantiles</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the average MECR, variance of MECR, and average transcript counts for quantiles of transcript counts.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing gene expression data.</span>
<span class="sd">    - gene_pairs: List[Tuple[str, str]]</span>
<span class="sd">        List of tuples representing gene pairs to evaluate.</span>
<span class="sd">    - quantiles: int, default=10</span>
<span class="sd">        Number of quantiles to divide the data into.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - quantized_data: pd.DataFrame</span>
<span class="sd">        DataFrame containing quantile information, average MECR, variance of MECR, average counts, and number of cells.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;quantile&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">qcut</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">],</span> <span class="n">quantiles</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">quantized_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">quantile</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">quantiles</span><span class="p">):</span>
        <span class="n">cells_in_quantile</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;quantile&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">quantile</span>
        <span class="n">mecr</span> <span class="o">=</span> <span class="n">compute_MECR</span><span class="p">(</span><span class="n">adata</span><span class="p">[</span><span class="n">cells_in_quantile</span><span class="p">,</span> <span class="p">:],</span> <span class="n">gene_pairs</span><span class="p">)</span>
        <span class="n">average_mecr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mecr</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
        <span class="n">variance_mecr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">mecr</span><span class="o">.</span><span class="n">values</span><span class="p">()])</span>
        <span class="n">average_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells_in_quantile</span><span class="p">,</span> <span class="s2">&quot;transcripts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">quantized_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;quantile&quot;</span><span class="p">:</span> <span class="n">quantile</span> <span class="o">/</span> <span class="n">quantiles</span><span class="p">,</span>
                <span class="s2">&quot;average_mecr&quot;</span><span class="p">:</span> <span class="n">average_mecr</span><span class="p">,</span>
                <span class="s2">&quot;variance_mecr&quot;</span><span class="p">:</span> <span class="n">variance_mecr</span><span class="p">,</span>
                <span class="s2">&quot;average_counts&quot;</span><span class="p">:</span> <span class="n">average_counts</span><span class="p">,</span>
                <span class="s2">&quot;num_cells&quot;</span><span class="p">:</span> <span class="n">cells_in_quantile</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">quantized_data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.compute_transcript_density" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_transcript_density</span>


<a href="#segger.validation.compute_transcript_density" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_transcript_density</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Compute the transcript density for each cell in the AnnData object.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing transcript and cell area information.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">compute_transcript_density</span><span class="p">(</span><span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the transcript density for each cell in the AnnData object.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing transcript and cell area information.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">transcript_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcript_counts&quot;</span><span class="p">]</span>
    <span class="k">except</span><span class="p">:</span>
        <span class="n">transcript_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span>
    <span class="n">cell_areas</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_area&quot;</span><span class="p">]</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcript_density&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">transcript_counts</span> <span class="o">/</span> <span class="n">cell_areas</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.draw_umap" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">draw_umap</span>


<a href="#segger.validation.draw_umap" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">draw_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Draw UMAP plots for the given AnnData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The AnnData object containing the data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>column</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The column to color the UMAP plot by.</p>
              </div>
            </td>
            <td>
                  <code>&#39;leiden&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">draw_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">column</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;leiden&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Draw UMAP plots for the given AnnData object.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): The AnnData object containing the data.</span>
<span class="sd">        column (str): The column to color the UMAP plot by.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">column</span><span class="p">])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;KRT5&quot;</span><span class="p">,</span> <span class="s2">&quot;KRT7&quot;</span><span class="p">],</span> <span class="n">vmax</span><span class="o">=</span><span class="s2">&quot;p95&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ACTA2&quot;</span><span class="p">,</span> <span class="s2">&quot;PTPRC&quot;</span><span class="p">],</span> <span class="n">vmax</span><span class="o">=</span><span class="s2">&quot;p95&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.find_markers" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">find_markers</span>


<a href="#segger.validation.find_markers" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">find_markers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type_column</span><span class="p">,</span> <span class="n">pos_percentile</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">neg_percentile</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">percentage</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Identify positive and negative markers for each cell type based on gene expression and filter by expression percentage.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing gene expression data.</li>
<li>cell_type_column: str
    Column name in <code>adata.obs</code> that specifies cell types.</li>
<li>pos_percentile: float, default=5
    Percentile threshold to determine top x% expressed genes.</li>
<li>neg_percentile: float, default=10
    Percentile threshold to determine top x% lowly expressed genes.</li>
<li>percentage: float, default=50
    Minimum percentage of cells expressing the marker within a cell type for it to be considered.</li>
</ul>
        <ul>
<li>markers: dict
    Dictionary where keys are cell types and values are dictionaries containing:
        'positive': list of top x% highly expressed genes
        'negative': list of top x% lowly expressed genes.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">find_markers</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">cell_type_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">pos_percentile</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">neg_percentile</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">percentage</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify positive and negative markers for each cell type based on gene expression and filter by expression percentage.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing gene expression data.</span>
<span class="sd">    - cell_type_column: str</span>
<span class="sd">        Column name in `adata.obs` that specifies cell types.</span>
<span class="sd">    - pos_percentile: float, default=5</span>
<span class="sd">        Percentile threshold to determine top x% expressed genes.</span>
<span class="sd">    - neg_percentile: float, default=10</span>
<span class="sd">        Percentile threshold to determine top x% lowly expressed genes.</span>
<span class="sd">    - percentage: float, default=50</span>
<span class="sd">        Minimum percentage of cells expressing the marker within a cell type for it to be considered.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - markers: dict</span>
<span class="sd">        Dictionary where keys are cell types and values are dictionaries containing:</span>
<span class="sd">            &#39;positive&#39;: list of top x% highly expressed genes</span>
<span class="sd">            &#39;negative&#39;: list of top x% lowly expressed genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">markers</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">cell_type_column</span><span class="p">)</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
    <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_column</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_column</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">]</span>
        <span class="n">mean_expression</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">subset</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">cutoff_high</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">mean_expression</span><span class="p">,</span> <span class="mi">100</span> <span class="o">-</span> <span class="n">pos_percentile</span><span class="p">)</span>
        <span class="n">cutoff_low</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">mean_expression</span><span class="p">,</span> <span class="n">neg_percentile</span><span class="p">)</span>
        <span class="n">pos_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mean_expression</span> <span class="o">&gt;=</span> <span class="n">cutoff_high</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">neg_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">mean_expression</span> <span class="o">&lt;=</span> <span class="n">cutoff_low</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">expr_frac</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">((</span><span class="n">subset</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="n">pos_indices</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">valid_pos_indices</span> <span class="o">=</span> <span class="n">pos_indices</span><span class="p">[</span><span class="n">expr_frac</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">percentage</span> <span class="o">/</span> <span class="mi">100</span><span class="p">)]</span>
        <span class="n">positive_markers</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="n">valid_pos_indices</span><span class="p">]</span>
        <span class="n">negative_markers</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="n">neg_indices</span><span class="p">]</span>
        <span class="n">markers</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;positive&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">positive_markers</span><span class="p">),</span> <span class="s2">&quot;negative&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">negative_markers</span><span class="p">)}</span>
    <span class="k">return</span> <span class="n">markers</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.find_mutually_exclusive_genes" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">find_mutually_exclusive_genes</span>


<a href="#segger.validation.find_mutually_exclusive_genes" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">find_mutually_exclusive_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">markers</span><span class="p">,</span> <span class="n">cell_type_column</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Identify mutually exclusive genes based on expression criteria.</p>
        <ul>
<li>adata: AnnData
    Annotated data object containing gene expression data.</li>
<li>markers: dict
    Dictionary where keys are cell types and values are dictionaries containing:
        'positive': list of top x% highly expressed genes
        'negative': list of top x% lowly expressed genes.</li>
<li>cell_type_column: str
    Column name in <code>adata.obs</code> that specifies cell types.</li>
</ul>
        <ul>
<li>exclusive_pairs: list
    List of mutually exclusive gene pairs.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">find_mutually_exclusive_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">markers</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]],</span> <span class="n">cell_type_column</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify mutually exclusive genes based on expression criteria.</span>

<span class="sd">    Args:</span>
<span class="sd">    - adata: AnnData</span>
<span class="sd">        Annotated data object containing gene expression data.</span>
<span class="sd">    - markers: dict</span>
<span class="sd">        Dictionary where keys are cell types and values are dictionaries containing:</span>
<span class="sd">            &#39;positive&#39;: list of top x% highly expressed genes</span>
<span class="sd">            &#39;negative&#39;: list of top x% lowly expressed genes.</span>
<span class="sd">    - cell_type_column: str</span>
<span class="sd">        Column name in `adata.obs` that specifies cell types.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - exclusive_pairs: list</span>
<span class="sd">        List of mutually exclusive gene pairs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exclusive_genes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">all_exclusive</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">marker_sets</span> <span class="ow">in</span> <span class="n">markers</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">positive_markers</span> <span class="o">=</span> <span class="n">marker_sets</span><span class="p">[</span><span class="s2">&quot;positive&quot;</span><span class="p">]</span>
        <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">positive_markers</span><span class="p">:</span>
            <span class="n">gene_expr</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">X</span>
            <span class="n">cell_type_mask</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_column</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span>
            <span class="n">non_cell_type_mask</span> <span class="o">=</span> <span class="o">~</span><span class="n">cell_type_mask</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">gene_expr</span><span class="p">[</span><span class="n">cell_type_mask</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mf">0.2</span> <span class="ow">and</span> <span class="p">(</span><span class="n">gene_expr</span><span class="p">[</span><span class="n">non_cell_type_mask</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">:</span>
                <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">all_exclusive</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">unique_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">({</span><span class="n">gene</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">all_exclusive</span><span class="p">})</span>
    <span class="n">filtered_exclusive_genes</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">i</span><span class="p">:</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">unique_genes</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="p">}</span>
    <span class="n">mutually_exclusive_gene_pairs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key1</span><span class="p">,</span> <span class="n">key2</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="n">filtered_exclusive_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">gene1</span> <span class="ow">in</span> <span class="n">filtered_exclusive_genes</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">gene2</span> <span class="ow">in</span> <span class="n">filtered_exclusive_genes</span><span class="p">[</span><span class="n">key2</span><span class="p">]</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">mutually_exclusive_gene_pairs</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.generate_experiment_file" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">generate_experiment_file</span>


<a href="#segger.validation.generate_experiment_file" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">generate_experiment_file</span><span class="p">(</span><span class="n">template_path</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">cells_name</span><span class="o">=</span><span class="s1">&#39;seg_cells&#39;</span><span class="p">,</span> <span class="n">analysis_name</span><span class="o">=</span><span class="s1">&#39;seg_analysis&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Generate the experiment file for Xenium.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>template_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the template file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the output file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cells_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the cells file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;seg_cells&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>analysis_name</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the analysis file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;seg_analysis&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">generate_experiment_file</span><span class="p">(</span>
    <span class="n">template_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">cells_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;seg_cells&quot;</span><span class="p">,</span> <span class="n">analysis_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;seg_analysis&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate the experiment file for Xenium.</span>

<span class="sd">    Args:</span>
<span class="sd">        template_path (str): The path to the template file.</span>
<span class="sd">        output_path (str): The path to the output file.</span>
<span class="sd">        cells_name (str): The name of the cells file.</span>
<span class="sd">        analysis_name (str): The name of the analysis file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">json</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">template_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">experiment</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

    <span class="n">experiment</span><span class="p">[</span><span class="s2">&quot;images&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;morphology_filepath&quot;</span><span class="p">)</span>
    <span class="n">experiment</span><span class="p">[</span><span class="s2">&quot;images&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;morphology_focus_filepath&quot;</span><span class="p">)</span>

    <span class="n">experiment</span><span class="p">[</span><span class="s2">&quot;xenium_explorer_files&quot;</span><span class="p">][</span><span class="s2">&quot;cells_zarr_filepath&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cells_name</span><span class="si">}</span><span class="s2">.zarr.zip&quot;</span>
    <span class="n">experiment</span><span class="p">[</span><span class="s2">&quot;xenium_explorer_files&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s2">&quot;cell_features_zarr_filepath&quot;</span><span class="p">)</span>
    <span class="n">experiment</span><span class="p">[</span><span class="s2">&quot;xenium_explorer_files&quot;</span><span class="p">][</span><span class="s2">&quot;analysis_zarr_filepath&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">analysis_name</span><span class="si">}</span><span class="s2">.zarr.zip&quot;</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">experiment</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.get_flatten_version" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_flatten_version</span>


<a href="#segger.validation.get_flatten_version" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_flatten_version</span><span class="p">(</span><span class="n">polygons</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">21</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Get the flattened version of polygon vertices.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>polygons</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[<span title="numpy.ndarray">ndarray</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of polygon vertices.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>max_value</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The maximum number of vertices to keep.</p>
              </div>
            </td>
            <td>
                  <code>21</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>np.ndarray: The flattened array of polygon vertices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_flatten_version</span><span class="p">(</span><span class="n">polygons</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">],</span> <span class="n">max_value</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">21</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the flattened version of polygon vertices.</span>

<span class="sd">    Args:</span>
<span class="sd">        polygons (List[np.ndarray]): List of polygon vertices.</span>
<span class="sd">        max_value (int): The maximum number of vertices to keep.</span>

<span class="sd">    Returns:</span>
<span class="sd">        np.ndarray: The flattened array of polygon vertices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n</span> <span class="o">=</span> <span class="n">max_value</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">polygons</span><span class="p">),</span> <span class="n">n</span> <span class="o">*</span> <span class="mi">2</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">polygon</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">polygons</span><span class="p">),</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">polygons</span><span class="p">)):</span>
        <span class="n">num_points</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">polygon</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">num_points</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n</span> <span class="o">*</span> <span class="mi">2</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="k">elif</span> <span class="n">num_points</span> <span class="o">&lt;</span> <span class="n">max_value</span><span class="p">:</span>
            <span class="n">repeated_points</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">polygon</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="p">(</span><span class="n">n</span> <span class="o">-</span> <span class="n">num_points</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
            <span class="n">padded_polygon</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">polygon</span><span class="p">,</span> <span class="n">repeated_points</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">padded_polygon</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
            <span class="n">padded_polygon</span><span class="p">[:</span> <span class="nb">min</span><span class="p">(</span><span class="n">num_points</span><span class="p">,</span> <span class="n">n</span><span class="p">)]</span> <span class="o">=</span> <span class="n">polygon</span><span class="p">[:</span> <span class="nb">min</span><span class="p">(</span><span class="n">num_points</span><span class="p">,</span> <span class="n">n</span><span class="p">)]</span>
            <span class="n">padded_polygon</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygon</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">result</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">padded_polygon</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.get_indices_indptr" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_indices_indptr</span>


<a href="#segger.validation.get_indices_indptr" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_indices_indptr</span><span class="p">(</span><span class="n">input_array</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Get the indices and indptr arrays for sparse matrix representation.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>input_array</code>
            </td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The input array containing cluster labels.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Tuple">Tuple</span>[<span title="numpy.ndarray">ndarray</span>, <span title="numpy.ndarray">ndarray</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Tuple[np.ndarray, np.ndarray]: The indices and indptr arrays.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_indices_indptr</span><span class="p">(</span><span class="n">input_array</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the indices and indptr arrays for sparse matrix representation.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_array (np.ndarray): The input array containing cluster labels.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[np.ndarray, np.ndarray]: The indices and indptr arrays.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">clusters</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">input_array</span><span class="p">[</span><span class="n">input_array</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]))</span>
    <span class="n">indptr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">clusters</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters</span><span class="p">:</span>
        <span class="n">cluster_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">input_array</span> <span class="o">==</span> <span class="n">cluster</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">indptr</span><span class="p">[</span><span class="n">cluster</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">indices</span><span class="p">)</span>
        <span class="n">indices</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">cluster_indices</span><span class="p">)</span>

    <span class="n">indices</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">input_array</span><span class="p">[</span><span class="n">input_array</span> <span class="o">==</span> <span class="mi">0</span><span class="p">])))</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">indices</span><span class="p">,</span> <span class="n">indptr</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.get_leiden_umap" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_leiden_umap</span>


<a href="#segger.validation.get_leiden_umap" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_leiden_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">draw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Perform Leiden clustering and UMAP visualization on the given AnnData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The AnnData object containing the data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>draw</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to draw the UMAP plots.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>AnnData</code></td>            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The AnnData object with Leiden clustering and UMAP results.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_leiden_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">draw</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform Leiden clustering and UMAP visualization on the given AnnData object.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): The AnnData object containing the data.</span>
<span class="sd">        draw (bool): Whether to draw the UMAP plots.</span>

<span class="sd">    Returns:</span>
<span class="sd">        AnnData: The AnnData object with Leiden clustering and UMAP results.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

    <span class="n">gene_names</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
    <span class="n">mean_expression_values</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">gene_mean_expression_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;gene_name&quot;</span><span class="p">:</span> <span class="n">gene_names</span><span class="p">,</span> <span class="s2">&quot;mean_expression&quot;</span><span class="p">:</span> <span class="n">mean_expression_values</span><span class="p">})</span>
    <span class="n">top_genes</span> <span class="o">=</span> <span class="n">gene_mean_expression_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;mean_expression&quot;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">30</span><span class="p">)</span>
    <span class="n">top_gene_names</span> <span class="o">=</span> <span class="n">top_genes</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">draw</span><span class="p">:</span>
        <span class="n">draw_umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;leiden&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.get_median_expression_table" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_median_expression_table</span>


<a href="#segger.validation.get_median_expression_table" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_median_expression_table</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">column</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Get the median expression table for the given AnnData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The AnnData object containing the data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>column</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The column to group by.</p>
              </div>
            </td>
            <td>
                  <code>&#39;leiden&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: The median expression table.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_median_expression_table</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">column</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;leiden&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the median expression table for the given AnnData object.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): The AnnData object containing the data.</span>
<span class="sd">        column (str): The column to group by.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The median expression table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">top_genes</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;GATA3&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ACTA2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;KRT7&quot;</span><span class="p">,</span>
        <span class="s2">&quot;KRT8&quot;</span><span class="p">,</span>
        <span class="s2">&quot;KRT5&quot;</span><span class="p">,</span>
        <span class="s2">&quot;AQP1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SERPINA3&quot;</span><span class="p">,</span>
        <span class="s2">&quot;PTGDS&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CXCR4&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SFRP1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ENAH&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MYH11&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SVIL&quot;</span><span class="p">,</span>
        <span class="s2">&quot;KRT14&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CD4&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">top_gene_indices</span> <span class="o">=</span> <span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">top_genes</span><span class="p">]</span>

    <span class="n">clusters</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">column</span><span class="p">]</span>
    <span class="n">cluster_data</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">cluster_cells</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">clusters</span> <span class="o">==</span> <span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">X</span>
        <span class="n">cluster_expression</span> <span class="o">=</span> <span class="n">cluster_cells</span><span class="p">[:,</span> <span class="n">top_gene_indices</span><span class="p">]</span>
        <span class="n">gene_medians</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">cluster_expression</span><span class="p">[:,</span> <span class="n">gene_idx</span><span class="p">])</span><span class="o">.</span><span class="n">median</span><span class="p">()</span> <span class="k">for</span> <span class="n">gene_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">top_gene_indices</span><span class="p">))</span>
        <span class="p">]</span>
        <span class="n">cluster_data</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Cluster_</span><span class="si">{</span><span class="n">cluster</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_medians</span>

    <span class="n">cluster_expression_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cluster_data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">top_genes</span><span class="p">)</span>
    <span class="n">sorted_columns</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">cluster_expression_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">cluster_expression_df</span> <span class="o">=</span> <span class="n">cluster_expression_df</span><span class="p">[</span><span class="n">sorted_columns</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">cluster_expression_df</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">background_gradient</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;Greens&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.load_segmentations" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">load_segmentations</span>


<a href="#segger.validation.load_segmentations" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">load_segmentations</span><span class="p">(</span><span class="n">segmentation_paths</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Load segmentation data from provided paths and handle special cases like separating 'segger' into 'segger_n0' and 'segger_n1'.</p>
<p>Args:
segmentation_paths (Dict[str, Path]): Dictionary mapping segmentation method names to their file paths.</p>
<p>Returns:
Dict[str, sc.AnnData]: Dictionary mapping segmentation method names to loaded AnnData objects.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">load_segmentations</span><span class="p">(</span><span class="n">segmentation_paths</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load segmentation data from provided paths and handle special cases like separating &#39;segger&#39; into &#39;segger_n0&#39; and &#39;segger_n1&#39;.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentation_paths (Dict[str, Path]): Dictionary mapping segmentation method names to their file paths.</span>

<span class="sd">    Returns:</span>
<span class="sd">    Dict[str, sc.AnnData]: Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">segmentations_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">segmentation_paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
        <span class="c1"># Special handling for &#39;segger&#39; to separate into &#39;segger_n0&#39; and &#39;segger_n1&#39;</span>
        <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;segger&quot;</span><span class="p">:</span>
            <span class="n">cells_n1</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;-nx&quot;</span><span class="p">)]</span>
            <span class="n">cells_n0</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;-nx&quot;</span><span class="p">)]</span>
            <span class="n">segmentations_dict</span><span class="p">[</span><span class="s2">&quot;segger_n1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">cells_n1</span><span class="p">,</span> <span class="p">:]</span>
            <span class="n">segmentations_dict</span><span class="p">[</span><span class="s2">&quot;segger_n0&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">cells_n0</span><span class="p">,</span> <span class="p">:]</span>
        <span class="n">segmentations_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span>
    <span class="k">return</span> <span class="n">segmentations_dict</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_cell_area" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_cell_area</span>


<a href="#segger.validation.plot_cell_area" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_cell_area</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the cell area (log2) for each segmentation method.</p>
<p>Args:
segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.
output_path (Path): Path to the directory where the plot will be saved.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_cell_area</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the cell area (log2) for each segmentation method.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Prepare the data for the violin plot</span>
    <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;Cell Area (log2)&quot;</span><span class="p">:</span> <span class="p">[]})</span>
    <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">if</span> <span class="s2">&quot;cell_area&quot;</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">method_area</span> <span class="o">=</span> <span class="n">segmentations_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_area&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">method_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_area</span><span class="p">),</span> <span class="s2">&quot;Cell Area (log2)&quot;</span><span class="p">:</span> <span class="n">method_area</span><span class="o">.</span><span class="n">values</span><span class="p">}</span>
            <span class="p">)</span>
            <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">method_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">violin_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;cell_area_log2_data.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Plot the violin plots</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Cell Area (log2)&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span>
    <span class="c1"># Add a dashed line for the 10X-nucleus median</span>
    <span class="k">if</span> <span class="s2">&quot;10X-nucleus&quot;</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="p">:</span>
        <span class="n">median_10X_nucleus_area</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">[</span><span class="s2">&quot;10X-nucleus&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_area&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">median_10X_nucleus_area</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;10X-nucleus Median&quot;</span><span class="p">)</span>
    <span class="c1"># Set plot titles and labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Cell Area (log2)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Save the figure as a PDF</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;cell_area_log2_violin_plot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_cell_counts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_cell_counts</span>


<a href="#segger.validation.plot_cell_counts" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_cell_counts</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the number of cells per segmentation method and save the cell count data as a CSV.</p>
<p>Args:
segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.
output_path (Path): Path to the directory where the plot will be saved.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_cell_counts</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the number of cells per segmentation method and save the cell count data as a CSV.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate the number of cells in each segmentation method</span>
    <span class="n">cell_counts</span> <span class="o">=</span> <span class="p">{</span><span class="n">method</span><span class="p">:</span> <span class="n">seg</span><span class="o">.</span><span class="n">n_obs</span> <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">seg</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="c1"># Create a DataFrame for the bar plot</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_counts</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Number of Cells&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">T</span>

    <span class="c1"># Save the DataFrame to CSV</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;cell_counts_data.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Generate the bar plot</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
        <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">,</span> <span class="n">stacked</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="n">palette</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="s2">&quot;#333333&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">],</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.9</span>
    <span class="p">)</span>

    <span class="c1"># Add a dashed line for the 10X baseline</span>
    <span class="k">if</span> <span class="s2">&quot;10X&quot;</span> <span class="ow">in</span> <span class="n">cell_counts</span><span class="p">:</span>
        <span class="n">baseline_height</span> <span class="o">=</span> <span class="n">cell_counts</span><span class="p">[</span><span class="s2">&quot;10X&quot;</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">baseline_height</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;10X Baseline&quot;</span><span class="p">)</span>

    <span class="c1"># Set plot titles and labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Number of Cells per Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Number of Cells&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>

    <span class="c1"># Save the figure as a PDF</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;cell_counts_bar_plot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_contamination_boxplots" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_contamination_boxplots</span>


<a href="#segger.validation.plot_contamination_boxplots" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_contamination_boxplots</span><span class="p">(</span><span class="n">boxplot_data</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot boxplots for contamination values across different segmentation methods.</p>
<p>Args:
boxplot_data (pd.DataFrame): DataFrame containing contamination data for all segmentation methods.
output_path (Path): Path to the directory where the plot will be saved.
palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_contamination_boxplots</span><span class="p">(</span><span class="n">boxplot_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot boxplots for contamination values across different segmentation methods.</span>

<span class="sd">    Args:</span>
<span class="sd">    boxplot_data (pd.DataFrame): DataFrame containing contamination data for all segmentation methods.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">boxplot_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;contamination_box_results.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;Source Cell Type&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Contamination&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">boxplot_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Neighborhood Contamination&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Source Cell Type&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Contamination&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;contamination_boxplots.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_contamination_results" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_contamination_results</span>


<a href="#segger.validation.plot_contamination_results" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_contamination_results</span><span class="p">(</span><span class="n">contamination_results</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot contamination results for each segmentation method.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>contamination_results</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="pandas.DataFrame">DataFrame</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of contamination data for each segmentation method.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the directory where the plot will be saved.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>palette</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping segmentation method names to color codes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">980</span>
<span class="normal">981</span>
<span class="normal">982</span>
<span class="normal">983</span>
<span class="normal">984</span>
<span class="normal">985</span>
<span class="normal">986</span>
<span class="normal">987</span>
<span class="normal">988</span>
<span class="normal">989</span>
<span class="normal">990</span>
<span class="normal">991</span>
<span class="normal">992</span>
<span class="normal">993</span>
<span class="normal">994</span>
<span class="normal">995</span>
<span class="normal">996</span>
<span class="normal">997</span>
<span class="normal">998</span>
<span class="normal">999</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_contamination_results</span><span class="p">(</span>
    <span class="n">contamination_results</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot contamination results for each segmentation method.</span>

<span class="sd">    Args:</span>
<span class="sd">        contamination_results (Dict[str, pd.DataFrame]): Dictionary of contamination data for each segmentation method.</span>
<span class="sd">        output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">        palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">contamination_results</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;contamination_results.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">contamination_results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span> <span class="n">linewidths</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Contamination Matrix for </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Target Cell Type&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Source Cell Type&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_contamination_matrix.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_counts_per_cell" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_counts_per_cell</span>


<a href="#segger.validation.plot_counts_per_cell" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_counts_per_cell</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the counts per cell (log2) for each segmentation method.</p>
<p>Args:
segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.
output_path (Path): Path to the directory where the plot will be saved.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_counts_per_cell</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the counts per cell (log2) for each segmentation method.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Prepare the data for the violin plot</span>
    <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;Counts per Cell (log2)&quot;</span><span class="p">:</span> <span class="p">[]})</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">adata</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">method_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">method_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_counts</span><span class="p">),</span> <span class="s2">&quot;Counts per Cell (log2)&quot;</span><span class="p">:</span> <span class="n">method_counts</span><span class="o">.</span><span class="n">values</span><span class="p">}</span>
        <span class="p">)</span>
        <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">method_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">violin_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;counts_per_cell_data.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Plot the violin plots</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Counts per Cell (log2)&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylim</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">300</span><span class="p">))</span>
    <span class="c1"># Add a dashed line for the 10X-nucleus median</span>
    <span class="k">if</span> <span class="s2">&quot;10X-nucleus&quot;</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="p">:</span>
        <span class="n">median_10X_nucleus</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">[</span><span class="s2">&quot;10X-nucleus&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">median_10X_nucleus</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;10X-nucleus Median&quot;</span><span class="p">)</span>
    <span class="c1"># Set plot titles and labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Counts per Cell (log2)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="c1"># Save the figure as a PDF</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;counts_per_cell_violin_plot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_entropy_boxplots" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_entropy_boxplots</span>


<a href="#segger.validation.plot_entropy_boxplots" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_entropy_boxplots</span><span class="p">(</span><span class="n">entropy_boxplot_data</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot boxplots for neighborhood entropy across different segmentation methods by cell type.</p>
<p>Args:
entropy_boxplot_data (pd.DataFrame): DataFrame containing neighborhood entropy data for all segmentation methods.
output_path (Path): Path to the directory where the plot will be saved.
palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_entropy_boxplots</span><span class="p">(</span><span class="n">entropy_boxplot_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot boxplots for neighborhood entropy across different segmentation methods by cell type.</span>

<span class="sd">    Args:</span>
<span class="sd">    entropy_boxplot_data (pd.DataFrame): DataFrame containing neighborhood entropy data for all segmentation methods.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Neighborhood Entropy&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">entropy_boxplot_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Neighborhood Entropy&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Neighborhood Entropy&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;neighborhood_entropy_boxplots.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_gene_counts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_gene_counts</span>


<a href="#segger.validation.plot_gene_counts" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_gene_counts</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the normalized gene counts for each segmentation method.</p>
<p>Args:
segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.
output_path (Path): Path to the directory where the plot will be saved.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_gene_counts</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the normalized gene counts for each segmentation method.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate total counts per gene for each segmentation method</span>
    <span class="n">total_counts_per_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">adata</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">gene_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">gene_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">gene_counts</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">method</span><span class="p">)</span>
        <span class="n">total_counts_per_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">total_counts_per_gene</span><span class="p">,</span> <span class="n">gene_counts</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Normalize by the maximum count per gene across all segmentations</span>
    <span class="n">max_counts_per_gene</span> <span class="o">=</span> <span class="n">total_counts_per_gene</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">normalized_counts_per_gene</span> <span class="o">=</span> <span class="n">total_counts_per_gene</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">max_counts_per_gene</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Prepare the data for the box plot</span>
    <span class="n">boxplot_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;Normalized Counts&quot;</span><span class="p">:</span> <span class="p">[]})</span>

    <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">method_counts</span> <span class="o">=</span> <span class="n">normalized_counts_per_gene</span><span class="p">[</span><span class="n">method</span><span class="p">]</span>
        <span class="n">method_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_counts</span><span class="p">),</span> <span class="s2">&quot;Normalized Counts&quot;</span><span class="p">:</span> <span class="n">method_counts</span><span class="o">.</span><span class="n">values</span><span class="p">}</span>
        <span class="p">)</span>
        <span class="n">boxplot_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">boxplot_data</span><span class="p">,</span> <span class="n">method_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">boxplot_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;gene_counts_normalized_data.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Plot the box plots</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Normalized Counts&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">boxplot_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.9</span><span class="p">)</span>

    <span class="c1"># Add a dashed line for the 10X baseline</span>
    <span class="k">if</span> <span class="s2">&quot;10X&quot;</span> <span class="ow">in</span> <span class="n">normalized_counts_per_gene</span><span class="p">:</span>
        <span class="n">baseline_height</span> <span class="o">=</span> <span class="n">normalized_counts_per_gene</span><span class="p">[</span><span class="s2">&quot;10X&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">baseline_height</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;10X Baseline&quot;</span><span class="p">)</span>

    <span class="c1"># Set plot titles and labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Normalized Counts&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Save the figure as a PDF</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;gene_counts_normalized_boxplot_by_method.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_general_statistics_plots" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_general_statistics_plots</span>


<a href="#segger.validation.plot_general_statistics_plots" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_general_statistics_plots</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Create a summary plot with all the general statistics subplots.</p>
<p>Args:
segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.
output_path (Path): Path to the directory where the summary plot will be saved.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_general_statistics_plots</span><span class="p">(</span>
    <span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a summary plot with all the general statistics subplots.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    output_path (Path): Path to the directory where the summary plot will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">))</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">plot_cell_counts</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">plot_percent_assigned</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
    <span class="n">plot_gene_counts</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
    <span class="n">plot_counts_per_cell</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
    <span class="n">plot_cell_area</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
    <span class="n">plot_transcript_density</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;general_statistics_plots.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_mecr_results" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_mecr_results</span>


<a href="#segger.validation.plot_mecr_results" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_mecr_results</span><span class="p">(</span><span class="n">mecr_results</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the MECR (Mutually Exclusive Co-expression Rate) results for each segmentation method.</p>
<p>Args:
mecr_results (Dict[str, Dict[Tuple[str, str], float]]): Dictionary of MECR results for each segmentation method.
output_path (Path): Path to the directory where the plot will be saved.
palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_mecr_results</span><span class="p">(</span>
    <span class="n">mecr_results</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span> <span class="nb">float</span><span class="p">]],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the MECR (Mutually Exclusive Co-expression Rate) results for each segmentation method.</span>

<span class="sd">    Args:</span>
<span class="sd">    mecr_results (Dict[str, Dict[Tuple[str, str], float]]): Dictionary of MECR results for each segmentation method.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Prepare the data for plotting</span>
    <span class="n">plot_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">mecr_dict</span> <span class="ow">in</span> <span class="n">mecr_results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">gene_pair</span><span class="p">,</span> <span class="n">mecr_value</span> <span class="ow">in</span> <span class="n">mecr_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">plot_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">{</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="n">method</span><span class="p">,</span> <span class="s2">&quot;Gene Pair&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene_pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> - </span><span class="si">{</span><span class="n">gene_pair</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;MECR&quot;</span><span class="p">:</span> <span class="n">mecr_value</span><span class="p">}</span>
            <span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">plot_data</span><span class="p">)</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;mcer_box.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;MECR&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Mutually Exclusive Co-expression Rate (MECR)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;MECR&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;mecr_results_boxplot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_metric_comparison" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_metric_comparison</span>


<a href="#segger.validation.plot_metric_comparison" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_metric_comparison</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">metric</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">method1</span><span class="p">,</span> <span class="n">method2</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot a comparison of a specific metric between two methods and save the comparison data.</p>
        <ul>
<li>ax: plt.Axes
    Matplotlib axis to plot on.</li>
<li>data: pd.DataFrame
    DataFrame containing the data for plotting.</li>
<li>metric: str
    The metric to compare.</li>
<li>label: str
    Label for the metric.</li>
<li>method1: str
    The first method to compare.</li>
<li>method2: str
    The second method to compare.</li>
<li>output_path: Path
    Path to save the merged DataFrame as a CSV.</li>
</ul>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_metric_comparison</span><span class="p">(</span>
    <span class="n">ax</span><span class="p">:</span> <span class="n">plt</span><span class="o">.</span><span class="n">Axes</span><span class="p">,</span> <span class="n">data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">metric</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">label</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">method1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">method2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot a comparison of a specific metric between two methods and save the comparison data.</span>

<span class="sd">    Args:</span>
<span class="sd">    - ax: plt.Axes</span>
<span class="sd">        Matplotlib axis to plot on.</span>
<span class="sd">    - data: pd.DataFrame</span>
<span class="sd">        DataFrame containing the data for plotting.</span>
<span class="sd">    - metric: str</span>
<span class="sd">        The metric to compare.</span>
<span class="sd">    - label: str</span>
<span class="sd">        Label for the metric.</span>
<span class="sd">    - method1: str</span>
<span class="sd">        The first method to compare.</span>
<span class="sd">    - method2: str</span>
<span class="sd">        The second method to compare.</span>
<span class="sd">    - output_path: Path</span>
<span class="sd">        Path to save the merged DataFrame as a CSV.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">subset1</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;method&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">method1</span><span class="p">]</span>
    <span class="n">subset2</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;method&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">method2</span><span class="p">]</span>
    <span class="n">merged_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">subset1</span><span class="p">,</span> <span class="n">subset2</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;celltype_major&quot;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">method1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;_</span><span class="si">{</span><span class="n">method2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">))</span>

    <span class="c1"># Save the merged data used in the plot to CSV</span>
    <span class="n">merged_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;metric_comparison_</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">method1</span><span class="si">}</span><span class="s2">_vs_</span><span class="si">{</span><span class="n">method2</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">merged_data</span><span class="p">[</span><span class="s2">&quot;celltype_major&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">cell_data</span> <span class="o">=</span> <span class="n">merged_data</span><span class="p">[</span><span class="n">merged_data</span><span class="p">[</span><span class="s2">&quot;celltype_major&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">cell_data</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">method1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span> <span class="n">cell_data</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">method2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">cell_type</span><span class="p">)</span>

    <span class="n">max_value</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">merged_data</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">method1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">merged_data</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">metric</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">method2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_value</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_value</span><span class="p">],</span> <span class="s2">&quot;k--&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">method1</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">method2</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">method1</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">method2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_percent_assigned" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_percent_assigned</span>


<a href="#segger.validation.plot_percent_assigned" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_percent_assigned</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the percentage of assigned transcripts (normalized) for each segmentation method.</p>
<p>Args:
segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.
output_path (Path): Path to the directory where the plot will be saved.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_percent_assigned</span><span class="p">(</span>
    <span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the percentage of assigned transcripts (normalized) for each segmentation method.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate total counts per gene for each segmentation method</span>
    <span class="n">total_counts_per_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">adata</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">gene_counts</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>  <span class="c1"># Sum across cells for each gene and flatten to 1D</span>
        <span class="n">gene_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">gene_counts</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">method</span><span class="p">)</span>
        <span class="n">total_counts_per_gene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">total_counts_per_gene</span><span class="p">,</span> <span class="n">gene_counts</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Normalize by the maximum count per gene across all segmentations</span>
    <span class="n">max_counts_per_gene</span> <span class="o">=</span> <span class="n">total_counts_per_gene</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">percent_assigned_normalized</span> <span class="o">=</span> <span class="n">total_counts_per_gene</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">max_counts_per_gene</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>

    <span class="c1"># Prepare the data for the violin plot</span>
    <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;Percent Assigned (Normalized)&quot;</span><span class="p">:</span> <span class="p">[]})</span>

    <span class="c1"># Add normalized percent_assigned data for each method</span>
    <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">method_data</span> <span class="o">=</span> <span class="n">percent_assigned_normalized</span><span class="p">[</span><span class="n">method</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
        <span class="n">method_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_data</span><span class="p">),</span> <span class="s2">&quot;Percent Assigned (Normalized)&quot;</span><span class="p">:</span> <span class="n">method_data</span><span class="o">.</span><span class="n">values</span><span class="p">}</span>
        <span class="p">)</span>
        <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">method_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">violin_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;percent_assigned_normalized.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Plot the violin plots</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Percent Assigned (Normalized)&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

    <span class="c1"># Add a dashed line for the 10X baseline</span>
    <span class="k">if</span> <span class="s2">&quot;10X&quot;</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="p">:</span>
        <span class="n">baseline_height</span> <span class="o">=</span> <span class="n">percent_assigned_normalized</span><span class="p">[</span><span class="s2">&quot;10X&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">baseline_height</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;10X Baseline&quot;</span><span class="p">)</span>

    <span class="c1"># Set plot titles and labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Percent Assigned (Normalized)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>

    <span class="c1"># Save the figure as a PDF</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;percent_assigned_normalized_violin_plot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_quantized_mecr_area" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_quantized_mecr_area</span>


<a href="#segger.validation.plot_quantized_mecr_area" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_quantized_mecr_area</span><span class="p">(</span><span class="n">quantized_mecr_area</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the quantized MECR values against cell areas for each segmentation method, with point size proportional to the variance of MECR.</p>
<p>Args:
quantized_mecr_area (Dict[str, pd.DataFrame]): Dictionary of quantized MECR area data for each segmentation method.
output_path (Path): Path to the directory where the plot will be saved.
palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span>
<span class="normal">973</span>
<span class="normal">974</span>
<span class="normal">975</span>
<span class="normal">976</span>
<span class="normal">977</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_quantized_mecr_area</span><span class="p">(</span>
    <span class="n">quantized_mecr_area</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the quantized MECR values against cell areas for each segmentation method, with point size proportional to the variance of MECR.</span>

<span class="sd">    Args:</span>
<span class="sd">    quantized_mecr_area (Dict[str, pd.DataFrame]): Dictionary of quantized MECR area data for each segmentation method.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># quantized_mecr_area.to_csv(output_path / &#39;quantized_mecr_area.csv&#39;, index=True)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">quantized_mecr_area</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_area&quot;</span><span class="p">],</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_mecr&quot;</span><span class="p">],</span>
            <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span>
            <span class="c1"># s=df[&#39;variance_mecr&#39;]  * 1e5,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="n">palette</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="s2">&quot;#333333&quot;</span><span class="p">),</span>
            <span class="n">label</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
            <span class="n">markersize</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_area&quot;</span><span class="p">],</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_mecr&quot;</span><span class="p">],</span>
            <span class="n">s</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;variance_mecr&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mf">1e5</span><span class="p">,</span>  <span class="c1"># Size of points based on the variance of MECR</span>
            <span class="n">color</span><span class="o">=</span><span class="n">palette</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="s2">&quot;#333333&quot;</span><span class="p">),</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>  <span class="c1"># Slight transparency for overlapping points</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span>  <span class="c1"># White edge color for better visibility</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>  <span class="c1"># Thin edge line</span>
        <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Quantized MECR by Cell Area&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Average Cell Area&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Average MECR&quot;</span><span class="p">)</span>
    <span class="c1"># Place the legend outside the plot on the top right</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;quantized_mecr_area_plot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_quantized_mecr_counts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_quantized_mecr_counts</span>


<a href="#segger.validation.plot_quantized_mecr_counts" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_quantized_mecr_counts</span><span class="p">(</span><span class="n">quantized_mecr_counts</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the quantized MECR values against transcript counts for each segmentation method, with point size proportional to the variance of MECR.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>quantized_mecr_counts</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="pandas.DataFrame">DataFrame</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of quantized MECR count data for each segmentation method.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_path</code>
            </td>
            <td>
                  <code><span title="pathlib.Path">Path</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the directory where the plot will be saved.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>palette</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[<span title="str">str</span>, <span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary mapping segmentation method names to color codes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_quantized_mecr_counts</span><span class="p">(</span>
    <span class="n">quantized_mecr_counts</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the quantized MECR values against transcript counts for each segmentation method, with point size proportional to the variance of MECR.</span>

<span class="sd">    Args:</span>
<span class="sd">        quantized_mecr_counts (Dict[str, pd.DataFrame]): Dictionary of quantized MECR count data for each segmentation method.</span>
<span class="sd">        output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">        palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># quantized_mecr_counts.to_csv(output_path / &#39;quantized_mecr_counts.csv&#39;, index=True)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">quantized_mecr_counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_counts&quot;</span><span class="p">],</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_mecr&quot;</span><span class="p">],</span>
            <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="n">palette</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="s2">&quot;#333333&quot;</span><span class="p">),</span>
            <span class="n">label</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>
            <span class="n">markersize</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>  <span class="c1"># No markers, only lines</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_counts&quot;</span><span class="p">],</span>
            <span class="n">df</span><span class="p">[</span><span class="s2">&quot;average_mecr&quot;</span><span class="p">],</span>
            <span class="n">s</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;variance_mecr&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mf">1e5</span><span class="p">,</span>  <span class="c1"># Size of points based on the variance of MECR</span>
            <span class="n">color</span><span class="o">=</span><span class="n">palette</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">method</span><span class="p">,</span> <span class="s2">&quot;#333333&quot;</span><span class="p">),</span>
            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>  <span class="c1"># Slight transparency for overlapping points</span>
            <span class="n">edgecolor</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">,</span>  <span class="c1"># White edge color for better visibility</span>
            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>  <span class="c1"># Thin edge line</span>
        <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Quantized MECR by Transcript Counts&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Average Transcript Counts&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Average MECR&quot;</span><span class="p">)</span>
    <span class="c1"># Place the legend outside the plot on the top right</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;quantized_mecr_counts_plot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_sensitivity_boxplots" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_sensitivity_boxplots</span>


<a href="#segger.validation.plot_sensitivity_boxplots" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_sensitivity_boxplots</span><span class="p">(</span><span class="n">sensitivity_boxplot_data</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot boxplots for sensitivity across different segmentation methods by cell type.
Args:
    sensitivity_boxplot_data (pd.DataFrame): DataFrame containing sensitivity data for all segmentation methods.
    output_path (Path): Path to the directory where the plot will be saved.
    palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_sensitivity_boxplots</span><span class="p">(</span>
    <span class="n">sensitivity_boxplot_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot boxplots for sensitivity across different segmentation methods by cell type.</span>
<span class="sd">    Args:</span>
<span class="sd">        sensitivity_boxplot_data (pd.DataFrame): DataFrame containing sensitivity data for all segmentation methods.</span>
<span class="sd">        output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">        palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sensitivity_boxplot_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;sensitivity_results.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Sensitivity&quot;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">sensitivity_boxplot_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Sensitivity Score&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Sensitivity&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;sensitivity_boxplots.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_transcript_density" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_transcript_density</span>


<a href="#segger.validation.plot_transcript_density" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_transcript_density</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot the transcript density (log2) for each segmentation method.</p>
<p>Args:
segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.
output_path (Path): Path to the directory where the plot will be saved.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_transcript_density</span><span class="p">(</span>
    <span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span> <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot the transcript density (log2) for each segmentation method.</span>

<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, sc.AnnData]): Dictionary mapping segmentation method names to loaded AnnData objects.</span>
<span class="sd">    output_path (Path): Path to the directory where the plot will be saved.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Prepare the data for the violin plot</span>
    <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[],</span> <span class="s2">&quot;Transcript Density (log2)&quot;</span><span class="p">:</span> <span class="p">[]})</span>

    <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">if</span> <span class="s2">&quot;cell_area&quot;</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">method_density</span> <span class="o">=</span> <span class="n">segmentations_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">segmentations_dict</span><span class="p">[</span><span class="n">method</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_area&quot;</span><span class="p">]</span>
            <span class="n">method_density_log2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">method_density</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">method_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;Segmentation Method&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_density_log2</span><span class="p">),</span>
                    <span class="s2">&quot;Transcript Density (log2)&quot;</span><span class="p">:</span> <span class="n">method_density_log2</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                <span class="p">}</span>
            <span class="p">)</span>
            <span class="n">violin_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">method_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">violin_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;transcript_density_log2_data.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Plot the violin plots</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;Transcript Density (log2)&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">violin_data</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">)</span>

    <span class="c1"># Add a dashed line for the 10X-nucleus median</span>
    <span class="k">if</span> <span class="s2">&quot;10X-nucleus&quot;</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="p">:</span>
        <span class="n">median_10X_nucleus_density_log2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span>
                <span class="n">segmentations_dict</span><span class="p">[</span><span class="s2">&quot;10X-nucleus&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span>
                <span class="o">/</span> <span class="n">segmentations_dict</span><span class="p">[</span><span class="s2">&quot;10X-nucleus&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_area&quot;</span><span class="p">]</span>
                <span class="o">+</span> <span class="mi">1</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>
            <span class="n">y</span><span class="o">=</span><span class="n">median_10X_nucleus_density_log2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;10X-nucleus Median&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Set plot titles and labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Segmentation Method&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Transcript Density (log2)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Save the figure as a PDF</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;transcript_density_log2_violin_plot.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.plot_umaps_with_scores" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">plot_umaps_with_scores</span>


<a href="#segger.validation.plot_umaps_with_scores" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">plot_umaps_with_scores</span><span class="p">(</span><span class="n">segmentations_dict</span><span class="p">,</span> <span class="n">clustering_scores</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">palette</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Plot UMAPs colored by cell type for each segmentation method and display clustering scores in the title.
Args:
segmentations_dict (Dict[str, AnnData]): Dictionary of AnnData objects for each segmentation method.
clustering_scores (Dict[str, Tuple[float, float]]): Dictionary of clustering scores for each method.
output_path (Path): Path to the directory where the plots will be saved.
palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</p>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">plot_umaps_with_scores</span><span class="p">(</span>
    <span class="n">segmentations_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">],</span>
    <span class="n">clustering_scores</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]],</span>
    <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">palette</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Plot UMAPs colored by cell type for each segmentation method and display clustering scores in the title.</span>
<span class="sd">    Args:</span>
<span class="sd">    segmentations_dict (Dict[str, AnnData]): Dictionary of AnnData objects for each segmentation method.</span>
<span class="sd">    clustering_scores (Dict[str, Tuple[float, float]]): Dictionary of clustering scores for each method.</span>
<span class="sd">    output_path (Path): Path to the directory where the plots will be saved.</span>
<span class="sd">    palette (Dict[str, str]): Dictionary mapping segmentation method names to color codes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">adata</span> <span class="ow">in</span> <span class="n">segmentations_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">method</span><span class="p">)</span>
        <span class="n">adata_copy</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">subsample</span><span class="p">(</span><span class="n">adata_copy</span><span class="p">,</span> <span class="n">n_obs</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_copy</span><span class="p">)</span>
        <span class="c1"># Plot UMAP colored by cell type</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata_copy</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_copy</span><span class="p">,</span> <span class="n">spread</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_copy</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;celltype_major&quot;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="c1"># Add clustering scores to the title</span>
        <span class="n">ch_score</span><span class="p">,</span> <span class="n">sh_score</span> <span class="o">=</span> <span class="n">compute_clustering_scores</span><span class="p">(</span><span class="n">adata_copy</span><span class="p">,</span> <span class="n">cell_type_column</span><span class="o">=</span><span class="s2">&quot;celltype_major&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2"> - UMAP</span><span class="se">\n</span><span class="s2">Calinski-Harabasz: </span><span class="si">{</span><span class="n">ch_score</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">, Silhouette: </span><span class="si">{</span><span class="n">sh_score</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># Save the figure</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">_umap_with_scores.pdf&quot;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.save_cell_clustering" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">save_cell_clustering</span>


<a href="#segger.validation.save_cell_clustering" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">save_cell_clustering</span><span class="p">(</span><span class="n">merged</span><span class="p">,</span> <span class="n">zarr_path</span><span class="p">,</span> <span class="n">columns</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Save cell clustering information to a Zarr file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>merged</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The merged dataframe containing cell clustering information.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>zarr_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the Zarr file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>columns</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The list of columns to save.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">save_cell_clustering</span><span class="p">(</span><span class="n">merged</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">zarr_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">columns</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save cell clustering information to a Zarr file.</span>

<span class="sd">    Args:</span>
<span class="sd">        merged (pd.DataFrame): The merged dataframe containing cell clustering information.</span>
<span class="sd">        zarr_path (str): The path to the Zarr file.</span>
<span class="sd">        columns (List[str]): The list of columns to save.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">zarr</span>

    <span class="n">new_zarr</span> <span class="o">=</span> <span class="n">zarr</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">zarr_path</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">new_zarr</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="s2">&quot;/cell_groups&quot;</span><span class="p">)</span>

    <span class="n">mappings</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">column</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
        <span class="n">classes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">merged</span><span class="p">[</span><span class="n">column</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)))</span>
        <span class="n">mapping_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">classes</span><span class="p">)),</span> <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">classes</span> <span class="k">if</span> <span class="n">k</span> <span class="o">!=</span> <span class="s2">&quot;nan&quot;</span><span class="p">])}</span>
        <span class="n">mapping_dict</span><span class="p">[</span><span class="s2">&quot;nan&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">clusters</span> <span class="o">=</span> <span class="n">merged</span><span class="p">[</span><span class="n">column</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">mapping_dict</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">indices</span><span class="p">,</span> <span class="n">indptr</span> <span class="o">=</span> <span class="n">get_indices_indptr</span><span class="p">(</span><span class="n">clusters</span><span class="p">)</span>

        <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">][</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span><span class="s2">&quot;indices&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">indices</span><span class="p">)</span>
        <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">][</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span><span class="s2">&quot;indptr&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">indptr</span><span class="p">)</span>
        <span class="n">mappings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mapping_dict</span><span class="p">)</span>

    <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;major_version&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;minor_version&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;number_groupings&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">columns</span><span class="p">),</span>
            <span class="s2">&quot;grouping_names&quot;</span><span class="p">:</span> <span class="n">columns</span><span class="p">,</span>
            <span class="s2">&quot;group_names&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="p">[</span><span class="n">k</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mapping_dict</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">item</span><span class="p">:</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">])][</span><span class="mi">1</span><span class="p">:]</span> <span class="k">for</span> <span class="n">mapping_dict</span> <span class="ow">in</span> <span class="n">mappings</span>
            <span class="p">],</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="n">new_zarr</span><span class="o">.</span><span class="n">store</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.seg2explorer" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">seg2explorer</span>


<a href="#segger.validation.seg2explorer" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">seg2explorer</span><span class="p">(</span><span class="n">seg_df</span><span class="p">,</span> <span class="n">source_path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">,</span> <span class="n">cells_filename</span><span class="o">=</span><span class="s1">&#39;seg_cells&#39;</span><span class="p">,</span> <span class="n">analysis_filename</span><span class="o">=</span><span class="s1">&#39;seg_analysis&#39;</span><span class="p">,</span> <span class="n">xenium_filename</span><span class="o">=</span><span class="s1">&#39;seg_experiment.xenium&#39;</span><span class="p">,</span> <span class="n">analysis_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">draw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cell_id_columns</span><span class="o">=</span><span class="s1">&#39;seg_cell_id&#39;</span><span class="p">,</span> <span class="n">area_low</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">area_high</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Convert seg output to a format compatible with Xenium explorer.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>seg_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The seg DataFrame.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>source_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The source path.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>output_dir</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The output directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cells_filename</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The filename for cells.</p>
              </div>
            </td>
            <td>
                  <code>&#39;seg_cells&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>analysis_filename</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The filename for analysis.</p>
              </div>
            </td>
            <td>
                  <code>&#39;seg_analysis&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>xenium_filename</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The filename for Xenium.</p>
              </div>
            </td>
            <td>
                  <code>&#39;seg_experiment.xenium&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>analysis_df</code>
            </td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="pandas.DataFrame">DataFrame</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The analysis DataFrame.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>draw</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to draw the plots.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_id_columns</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The cell ID columns.</p>
              </div>
            </td>
            <td>
                  <code>&#39;seg_cell_id&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>area_low</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The lower area threshold.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>area_high</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The upper area threshold.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">seg2explorer</span><span class="p">(</span>
    <span class="n">seg_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">source_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cells_filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;seg_cells&quot;</span><span class="p">,</span>
    <span class="n">analysis_filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;seg_analysis&quot;</span><span class="p">,</span>
    <span class="n">xenium_filename</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;seg_experiment.xenium&quot;</span><span class="p">,</span>
    <span class="n">analysis_df</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">draw</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">cell_id_columns</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;seg_cell_id&quot;</span><span class="p">,</span>
    <span class="n">area_low</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">area_high</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert seg output to a format compatible with Xenium explorer.</span>

<span class="sd">    Args:</span>
<span class="sd">        seg_df (pd.DataFrame): The seg DataFrame.</span>
<span class="sd">        source_path (str): The source path.</span>
<span class="sd">        output_dir (str): The output directory.</span>
<span class="sd">        cells_filename (str): The filename for cells.</span>
<span class="sd">        analysis_filename (str): The filename for analysis.</span>
<span class="sd">        xenium_filename (str): The filename for Xenium.</span>
<span class="sd">        analysis_df (Optional[pd.DataFrame]): The analysis DataFrame.</span>
<span class="sd">        draw (bool): Whether to draw the plots.</span>
<span class="sd">        cell_id_columns (str): The cell ID columns.</span>
<span class="sd">        area_low (float): The lower area threshold.</span>
<span class="sd">        area_high (float): The upper area threshold.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">zarr</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">json</span>

    <span class="n">source_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">source_path</span><span class="p">)</span>
    <span class="n">storage</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_dir</span><span class="p">)</span>

    <span class="n">cell_id2old_id</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">cell_id</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">cell_summary</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">polygon_num_vertices</span> <span class="o">=</span> <span class="p">[[],</span> <span class="p">[]]</span>
    <span class="n">polygon_vertices</span> <span class="o">=</span> <span class="p">[[],</span> <span class="p">[]]</span>
    <span class="n">seg_mask_value</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">tma_id</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">grouped_by</span> <span class="o">=</span> <span class="n">seg_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">cell_id_columns</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">cell_incremental_id</span><span class="p">,</span> <span class="p">(</span><span class="n">seg_cell_id</span><span class="p">,</span> <span class="n">seg_cell</span><span class="p">)</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">grouped_by</span><span class="p">),</span> <span class="n">total</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">grouped_by</span><span class="p">)):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seg_cell</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">cell_convex_hull</span> <span class="o">=</span> <span class="n">ConvexHull</span><span class="p">(</span><span class="n">seg_cell</span><span class="p">[[</span><span class="s2">&quot;x_location&quot;</span><span class="p">,</span> <span class="s2">&quot;y_location&quot;</span><span class="p">]])</span>
        <span class="k">if</span> <span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">area</span> <span class="o">&gt;</span> <span class="n">area_high</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">area</span> <span class="o">&lt;</span> <span class="n">area_low</span><span class="p">:</span>
            <span class="k">continue</span>

        <span class="n">uint_cell_id</span> <span class="o">=</span> <span class="n">cell_incremental_id</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">cell_id2old_id</span><span class="p">[</span><span class="n">uint_cell_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">seg_cell_id</span>

        <span class="n">seg_nucleous</span> <span class="o">=</span> <span class="n">seg_cell</span><span class="p">[</span><span class="n">seg_cell</span><span class="p">[</span><span class="s2">&quot;overlaps_nucleus&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seg_nucleous</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">nucleus_convex_hull</span> <span class="o">=</span> <span class="n">ConvexHull</span><span class="p">(</span><span class="n">seg_nucleous</span><span class="p">[[</span><span class="s2">&quot;x_location&quot;</span><span class="p">,</span> <span class="s2">&quot;y_location&quot;</span><span class="p">]])</span>

        <span class="n">cell_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">uint_cell_id</span><span class="p">)</span>
        <span class="n">cell_summary</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;cell_centroid_x&quot;</span><span class="p">:</span> <span class="n">seg_cell</span><span class="p">[</span><span class="s2">&quot;x_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;cell_centroid_y&quot;</span><span class="p">:</span> <span class="n">seg_cell</span><span class="p">[</span><span class="s2">&quot;y_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;cell_area&quot;</span><span class="p">:</span> <span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">area</span><span class="p">,</span>
                <span class="s2">&quot;nucleus_centroid_x&quot;</span><span class="p">:</span> <span class="n">seg_cell</span><span class="p">[</span><span class="s2">&quot;x_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;nucleus_centroid_y&quot;</span><span class="p">:</span> <span class="n">seg_cell</span><span class="p">[</span><span class="s2">&quot;y_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;nucleus_area&quot;</span><span class="p">:</span> <span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">area</span><span class="p">,</span>
                <span class="s2">&quot;z_level&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">seg_cell</span><span class="o">.</span><span class="n">z_location</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">//</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">)</span>

        <span class="n">polygon_num_vertices</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">vertices</span><span class="p">))</span>
        <span class="n">polygon_num_vertices</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nucleus_convex_hull</span><span class="o">.</span><span class="n">vertices</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seg_nucleous</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">3</span> <span class="k">else</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">polygon_vertices</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seg_cell</span><span class="p">[[</span><span class="s2">&quot;x_location&quot;</span><span class="p">,</span> <span class="s2">&quot;y_location&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">vertices</span><span class="p">])</span>
        <span class="n">polygon_vertices</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="n">seg_nucleous</span><span class="p">[[</span><span class="s2">&quot;x_location&quot;</span><span class="p">,</span> <span class="s2">&quot;y_location&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">nucleus_convex_hull</span><span class="o">.</span><span class="n">vertices</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">seg_nucleous</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">3</span>
            <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[],</span> <span class="p">[]])</span><span class="o">.</span><span class="n">T</span>
        <span class="p">)</span>
        <span class="n">seg_mask_value</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_incremental_id</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

    <span class="n">cell_polygon_vertices</span> <span class="o">=</span> <span class="n">get_flatten_version</span><span class="p">(</span><span class="n">polygon_vertices</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">21</span><span class="p">)</span>
    <span class="n">nucl_polygon_vertices</span> <span class="o">=</span> <span class="n">get_flatten_version</span><span class="p">(</span><span class="n">polygon_vertices</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">21</span><span class="p">)</span>

    <span class="n">cells</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;cell_id&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">cell_id</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cell_id</span><span class="p">))],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">,</span>
        <span class="s2">&quot;cell_summary&quot;</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_summary</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">),</span>
        <span class="s2">&quot;polygon_num_vertices&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">polygon_num_vertices</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span>
                <span class="p">[</span><span class="nb">min</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">polygon_num_vertices</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span>
            <span class="p">],</span>
            <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="s2">&quot;polygon_vertices&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">nucl_polygon_vertices</span><span class="p">,</span> <span class="n">cell_polygon_vertices</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">),</span>
        <span class="s2">&quot;seg_mask_value&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">seg_mask_value</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">),</span>
    <span class="p">}</span>

    <span class="n">existing_store</span> <span class="o">=</span> <span class="n">zarr</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">source_path</span> <span class="o">/</span> <span class="s2">&quot;cells.zarr.zip&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">)</span>
    <span class="n">new_store</span> <span class="o">=</span> <span class="n">zarr</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">storage</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cells_filename</span><span class="si">}</span><span class="s2">.zarr.zip&quot;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">)</span>

    <span class="n">new_store</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">]</span>
    <span class="n">new_store</span><span class="p">[</span><span class="s2">&quot;polygon_num_vertices&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="s2">&quot;polygon_num_vertices&quot;</span><span class="p">]</span>
    <span class="n">new_store</span><span class="p">[</span><span class="s2">&quot;polygon_vertices&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="s2">&quot;polygon_vertices&quot;</span><span class="p">]</span>
    <span class="n">new_store</span><span class="p">[</span><span class="s2">&quot;seg_mask_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="s2">&quot;seg_mask_value&quot;</span><span class="p">]</span>

    <span class="n">new_store</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">existing_store</span><span class="o">.</span><span class="n">attrs</span><span class="p">)</span>
    <span class="n">new_store</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s2">&quot;number_cells&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cells</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">])</span>
    <span class="n">new_store</span><span class="o">.</span><span class="n">store</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">analysis_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">analysis_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">cell_id2old_id</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cell_id</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="n">cell_id_columns</span><span class="p">])</span>
        <span class="n">analysis_df</span><span class="p">[</span><span class="s2">&quot;default&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;seg&quot;</span>

    <span class="n">zarr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span><span class="n">cell_id2old_id</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cell_id</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="n">cell_id_columns</span><span class="p">])</span>
    <span class="n">clustering_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">zarr_df</span><span class="p">,</span> <span class="n">analysis_df</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="n">cell_id_columns</span><span class="p">)</span>

    <span class="n">clusters_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">analysis_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">cell_id_columns</span><span class="p">]</span>
    <span class="n">clusters_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">cluster</span><span class="p">:</span> <span class="p">{</span>
            <span class="n">j</span><span class="p">:</span> <span class="n">i</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">clustering_df</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">())))</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span>
                <span class="nb">sorted</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">clustering_df</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">dropna</span><span class="p">())),</span>
            <span class="p">)</span>
        <span class="p">}</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters_names</span>
    <span class="p">}</span>

    <span class="n">new_zarr</span> <span class="o">=</span> <span class="n">zarr</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">storage</span> <span class="o">/</span> <span class="p">(</span><span class="n">analysis_filename</span> <span class="o">+</span> <span class="s2">&quot;.zarr.zip&quot;</span><span class="p">),</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">new_zarr</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="s2">&quot;/cell_groups&quot;</span><span class="p">)</span>

    <span class="n">clusters</span> <span class="o">=</span> <span class="p">[[</span><span class="n">clusters_dict</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">clustering_df</span><span class="p">[</span><span class="n">cluster</span><span class="p">])]</span> <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters_names</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">clusters</span><span class="p">)):</span>
        <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">create_group</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">indices</span><span class="p">,</span> <span class="n">indptr</span> <span class="o">=</span> <span class="n">get_indices_indptr</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">clusters</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>
        <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span><span class="s2">&quot;indices&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">indices</span><span class="p">)</span>
        <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">create_dataset</span><span class="p">(</span><span class="s2">&quot;indptr&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">indptr</span><span class="p">)</span>

    <span class="n">new_zarr</span><span class="p">[</span><span class="s2">&quot;cell_groups&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">attrs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;major_version&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;minor_version&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;number_groupings&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">clusters_names</span><span class="p">),</span>
            <span class="s2">&quot;grouping_names&quot;</span><span class="p">:</span> <span class="n">clusters_names</span><span class="p">,</span>
            <span class="s2">&quot;group_names&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">clusters_dict</span><span class="p">[</span><span class="n">cluster</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span> <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters_names</span>
            <span class="p">],</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="n">new_zarr</span><span class="o">.</span><span class="n">store</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">generate_experiment_file</span><span class="p">(</span>
        <span class="n">template_path</span><span class="o">=</span><span class="n">source_path</span> <span class="o">/</span> <span class="s2">&quot;experiment.xenium&quot;</span><span class="p">,</span>
        <span class="n">output_path</span><span class="o">=</span><span class="n">storage</span> <span class="o">/</span> <span class="n">xenium_filename</span><span class="p">,</span>
        <span class="n">cells_name</span><span class="o">=</span><span class="n">cells_filename</span><span class="p">,</span>
        <span class="n">analysis_name</span><span class="o">=</span><span class="n">analysis_filename</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="segger.validation.str_to_uint32" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">str_to_uint32</span>


<a href="#segger.validation.str_to_uint32" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">str_to_uint32</span><span class="p">(</span><span class="n">cell_id_str</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

        <p>Convert a string cell ID back to uint32 format.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cell_id_str</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The cell ID in string format.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Tuple">Tuple</span>[<span title="int">int</span>, <span title="int">int</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Tuple[int, int]: The cell ID in uint32 format and the dataset suffix.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/segger/validation/xenium_explorer.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">str_to_uint32</span><span class="p">(</span><span class="n">cell_id_str</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert a string cell ID back to uint32 format.</span>

<span class="sd">    Args:</span>
<span class="sd">        cell_id_str (str): The cell ID in string format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Tuple[int, int]: The cell ID in uint32 format and the dataset suffix.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prefix</span><span class="p">,</span> <span class="n">suffix</span> <span class="o">=</span> <span class="n">cell_id_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>
    <span class="n">str_to_hex_mapping</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;a&quot;</span><span class="p">:</span> <span class="s2">&quot;0&quot;</span><span class="p">,</span>
        <span class="s2">&quot;b&quot;</span><span class="p">:</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;c&quot;</span><span class="p">:</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;d&quot;</span><span class="p">:</span> <span class="s2">&quot;3&quot;</span><span class="p">,</span>
        <span class="s2">&quot;e&quot;</span><span class="p">:</span> <span class="s2">&quot;4&quot;</span><span class="p">,</span>
        <span class="s2">&quot;f&quot;</span><span class="p">:</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span>
        <span class="s2">&quot;g&quot;</span><span class="p">:</span> <span class="s2">&quot;6&quot;</span><span class="p">,</span>
        <span class="s2">&quot;h&quot;</span><span class="p">:</span> <span class="s2">&quot;7&quot;</span><span class="p">,</span>
        <span class="s2">&quot;i&quot;</span><span class="p">:</span> <span class="s2">&quot;8&quot;</span><span class="p">,</span>
        <span class="s2">&quot;j&quot;</span><span class="p">:</span> <span class="s2">&quot;9&quot;</span><span class="p">,</span>
        <span class="s2">&quot;k&quot;</span><span class="p">:</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span>
        <span class="s2">&quot;l&quot;</span><span class="p">:</span> <span class="s2">&quot;b&quot;</span><span class="p">,</span>
        <span class="s2">&quot;m&quot;</span><span class="p">:</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span>
        <span class="s2">&quot;n&quot;</span><span class="p">:</span> <span class="s2">&quot;d&quot;</span><span class="p">,</span>
        <span class="s2">&quot;o&quot;</span><span class="p">:</span> <span class="s2">&quot;e&quot;</span><span class="p">,</span>
        <span class="s2">&quot;p&quot;</span><span class="p">:</span> <span class="s2">&quot;f&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">hex_prefix</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">str_to_hex_mapping</span><span class="p">[</span><span class="n">char</span><span class="p">]</span> <span class="k">for</span> <span class="n">char</span> <span class="ow">in</span> <span class="n">prefix</span><span class="p">])</span>
    <span class="n">cell_id_uint32</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">hex_prefix</span><span class="p">,</span> <span class="mi">16</span><span class="p">)</span>
    <span class="n">dataset_suffix</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">suffix</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">cell_id_uint32</span><span class="p">,</span> <span class="n">dataset_suffix</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.sections", "navigation.instant", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate", "content.tabs.link", "navigation.tracking"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/termynal@0.0.1/termynal.js"></script>
      
        <script src="../../termynal.js"></script>
      
    
  </body>
</html>