
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Fast and accurate cell segmentation for single-molecule spatial omics">
      
      
        <meta name="author" content="Elyas Heidari">
      
      
        <link rel="canonical" href="https://EliHei2.github.io/segger_dev/cli/">
      
      
        <link rel="prev" href="../notebooks/segger_tutorial/">
      
      
        <link rel="next" href="../api/data/">
      
      
      <link rel="icon" href="../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.11">
    
    
      
        <title>CLI - segger Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=ubuntu:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"ubuntu";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../css/extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css">
    
      <link rel="stylesheet" href="https://unpkg.com/termynal@0.0.1/termynal.css">
    
      <link rel="stylesheet" href="../termynal.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#segger-command-line-interface" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="segger Documentation" class="md-header__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            segger Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              CLI
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="segger Documentation" class="md-nav__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../images/logo.png" alt="logo">

    </a>
    segger Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../user_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../user_guide/data_creation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Dataset Creation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../user_guide/training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../user_guide/validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../notebooks/segger_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction to Segger
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    CLI
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    CLI
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#segger-command-line-interface" class="md-nav__link">
    <span class="md-ellipsis">
      Segger Command Line Interface
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Segger Command Line Interface">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-creating-a-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      1. Creating a Dataset
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Creating a Dataset">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#key-updates" class="md-nav__link">
    <span class="md-ellipsis">
      Key Updates
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-training-a-model" class="md-nav__link">
    <span class="md-ellipsis">
      2. Training a Model
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Training a Model">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters_1" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-making-predictions" class="md-nav__link">
    <span class="md-ellipsis">
      3. Making Predictions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Making Predictions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters_2" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-running-the-entire-pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      4. Running the Entire Pipeline
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Running the Entire Pipeline">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#selecting-pipelines" class="md-nav__link">
    <span class="md-ellipsis">
      Selecting Pipelines
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-the-pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      Running the Pipeline
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-containerization" class="md-nav__link">
    <span class="md-ellipsis">
      5. Containerization
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Containerization">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#using-docker" class="md-nav__link">
    <span class="md-ellipsis">
      Using Docker
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-singularity" class="md-nav__link">
    <span class="md-ellipsis">
      Using Singularity
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-hpc-environments" class="md-nav__link">
    <span class="md-ellipsis">
      6. HPC Environments
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    API Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/models/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Models
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/training/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Training
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/prediction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Prediction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>CLI</h1>

<h2 id="segger-command-line-interface">Segger Command Line Interface<a class="headerlink" href="#segger-command-line-interface" title="Permanent link">&para;</a></h2>
<h3 id="1-creating-a-dataset">1. Creating a Dataset<a class="headerlink" href="#1-creating-a-dataset" title="Permanent link">&para;</a></h3>
<p>The <code>create_dataset</code> command helps you to build a dataset for spatial transcriptomics. Here’s a breakdown of the options available:</p>
<div class="termy" data-termynal data-ty-macos data-ty-title="bash"><span data-ty>// Example: Creating a dataset for spatial transcriptomics<br>python3 src/segger/cli/create_dataset_fast.py \<br>    --base_dir /path/to/raw_data \<br>    --data_dir /path/to/save/processed_data \<br>    --sample_type xenium \<br>    --scrnaseq_file /path/to/scrnaseq_file \<br>    --celltype_column celltype_column_name \<br>    --k_bd 3 \<br>    --dist_bd 15.0 \<br>    --k_tx 3 \<br>    --dist_tx 5.0 \<br>    --tile_width 200 \<br>    --tile_height 200 \<br>    --neg_sampling_ratio 5.0 \<br>    --frac 1.0 \<br>    --val_prob 0.1 \<br>    --test_prob 0.2 \<br>    --n_workers 16<br></span></div>

<h4 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">&para;</a></h4>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>base_dir</code></td>
<td>Directory containing the raw dataset (e.g., transcripts, boundaries).</td>
<td>-</td>
</tr>
<tr>
<td><code>data_dir</code></td>
<td>Directory to save the processed Segger dataset (in PyTorch Geometric format).</td>
<td>-</td>
</tr>
<tr>
<td><code>sample_type</code></td>
<td>The sample type of the raw data, e.g., "xenium" or "merscope".</td>
<td>None</td>
</tr>
<tr>
<td><code>scrnaseq_file</code></td>
<td>Path to the scRNAseq file.</td>
<td>None</td>
</tr>
<tr>
<td><code>celltype_column</code></td>
<td>Column name for cell type annotations in the scRNAseq file.</td>
<td>None</td>
</tr>
<tr>
<td><code>k_bd</code></td>
<td>Number of nearest neighbors for boundary nodes.</td>
<td><code>3</code></td>
</tr>
<tr>
<td><code>dist_bd</code></td>
<td>Maximum distance for boundary neighbors.</td>
<td><code>15.0</code></td>
</tr>
<tr>
<td><code>k_tx</code></td>
<td>Number of nearest neighbors for transcript nodes.</td>
<td><code>3</code></td>
</tr>
<tr>
<td><code>dist_tx</code></td>
<td>Maximum distance for transcript neighbors.</td>
<td><code>5.0</code></td>
</tr>
<tr>
<td><code>tile_width</code></td>
<td>Width of the tiles in pixels (ignored if <code>tile_size</code> is provided).</td>
<td>None</td>
</tr>
<tr>
<td><code>tile_height</code></td>
<td>Height of the tiles in pixels (ignored if <code>tile_size</code> is provided).</td>
<td>None</td>
</tr>
<tr>
<td><code>neg_sampling_ratio</code></td>
<td>Ratio of negative samples.</td>
<td><code>5.0</code></td>
</tr>
<tr>
<td><code>frac</code></td>
<td>Fraction of the dataset to process. Useful for subsampling large datasets.</td>
<td><code>1.0</code></td>
</tr>
<tr>
<td><code>val_prob</code></td>
<td>Proportion of the dataset used for validation split.</td>
<td><code>0.1</code></td>
</tr>
<tr>
<td><code>test_prob</code></td>
<td>Proportion of the dataset used for testing split.</td>
<td><code>0.2</code></td>
</tr>
<tr>
<td><code>n_workers</code></td>
<td>Number of workers for parallel processing.</td>
<td><code>1</code></td>
</tr>
</tbody>
</table>
<h4 id="key-updates">Key Updates<a class="headerlink" href="#key-updates" title="Permanent link">&para;</a></h4>
<ul>
<li><strong>Faster Dataset Creation</strong> This method is way faster due to the use of ND-tree-based partitioning and parallel processing.</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Customizing Your Dataset</p>
<ul>
<li><strong>dataset_type</strong>: Defines the type of spatial transcriptomics data. Currently, <strong>xenium</strong> and <strong>merscope</strong> are supported and have been tested.</li>
<li><strong>val_prob, test_prob</strong>: Control the dataset portions for validation and testing. Adjust based on your dataset size and evaluation needs.</li>
<li><strong>frac</strong>: Specifies the fraction of the dataset to process. Reducing <code>frac</code> can be useful when working with very large datasets, allowing for faster dataset creation by only processing a subset of the data.</li>
</ul>
</div>
<div class="admonition tip">
<p class="admonition-title">Faster Dataset Creation</p>
<p>Increasing the number of workers (<code>n_workers</code>) can significantly accelerate the dataset creation process, especially for large datasets, by taking advantage of parallel processing across multiple CPU cores.</p>
</div>
<div class="admonition tip">
<p class="admonition-title">Enhancing Segmentation Accuracy with scRNA-seq</p>
<p>Incorporating single cell RNA sequencing (scRNA-seq) data as features can provide additional biological context, improving the accuracy of the segger model.</p>
</div>
<hr />
<h3 id="2-training-a-model">2. Training a Model<a class="headerlink" href="#2-training-a-model" title="Permanent link">&para;</a></h3>
<p>The <code>train</code> command initializes and trains a model using the dataset created. Here are the key parameters:</p>
<div class="termy" data-termynal data-ty-macos data-ty-title="bash"><span data-ty>// Example: Training a segger model</span><span data-ty="input" data-ty-prompt="$">python3 src/segger/cli/train_model.py \
    --dataset_dir /path/to/saved/processed_data \
    --models_dir /path/to/save/model/checkpoints \
    --sample_tag first_training \
    --init_emb 8 \
    --hidden_channels 32 \
    --num_tx_tokens 500 \
    --out_channels 8 \
    --heads 2 \
    --num_mid_layers 2 \
    --batch_size 4 \
    --num_workers 2 \
    --accelerator cuda \
    --max_epochs 200 \
    --devices 4 \
    --strategy auto \
    --precision 16-mixed</span><span data-ty></span></div>

<h4 id="parameters_1">Parameters<a class="headerlink" href="#parameters_1" title="Permanent link">&para;</a></h4>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>dataset_dir</code></td>
<td>Directory containing the processed Segger dataset (in PyTorch Geometric format).</td>
<td>-</td>
</tr>
<tr>
<td><code>models_dir</code></td>
<td>Directory to save the trained model and training logs.</td>
<td>-</td>
</tr>
<tr>
<td><code>sample_tag</code></td>
<td>Tag used to identify the dataset during training.</td>
<td>-</td>
</tr>
<tr>
<td><code>init_emb</code></td>
<td>Size of the embedding layer for input data.</td>
<td><code>8</code></td>
</tr>
<tr>
<td><code>hidden_channels</code></td>
<td>Number of hidden units in each layer of the neural network.</td>
<td><code>32</code></td>
</tr>
<tr>
<td><code>num_tx_tokens</code></td>
<td>Number of transcript tokens used during training.</td>
<td><code>500</code></td>
</tr>
<tr>
<td><code>out_channels</code></td>
<td>Number of output channels from the model.</td>
<td><code>8</code></td>
</tr>
<tr>
<td><code>heads</code></td>
<td>Number of attention heads used in graph attention layers.</td>
<td><code>2</code></td>
</tr>
<tr>
<td><code>num_mid_layers</code></td>
<td>Number of mid layers in the model.</td>
<td><code>2</code></td>
</tr>
<tr>
<td><code>batch_size</code></td>
<td>Number of samples to process per training batch.</td>
<td><code>4</code></td>
</tr>
<tr>
<td><code>num_workers</code></td>
<td>Number of workers to use for parallel data loading.</td>
<td><code>2</code></td>
</tr>
<tr>
<td><code>accelerator</code></td>
<td>Device used for training (e.g., <code>cuda</code> for GPU or <code>cpu</code>).</td>
<td><code>cuda</code></td>
</tr>
<tr>
<td><code>max_epochs</code></td>
<td>Number of training epochs.</td>
<td><code>200</code></td>
</tr>
<tr>
<td><code>devices</code></td>
<td>Number of devices (GPUs) to use during training.</td>
<td><code>4</code></td>
</tr>
<tr>
<td><code>strategy</code></td>
<td>Strategy used for training (e.g., <code>ddp</code> for distributed training or <code>auto</code>).</td>
<td><code>auto</code></td>
</tr>
<tr>
<td><code>precision</code></td>
<td>Precision used for training (e.g., <code>16-mixed</code> for mixed precision training).</td>
<td><code>16-mixed</code></td>
</tr>
</tbody>
</table>
<div class="admonition tip">
<p class="admonition-title">Optimizing training time</p>
<ul>
<li><strong>devices</strong>: Use multiple GPUs by increasing the <code>devices</code> parameter to further accelerate training.</li>
<li><strong>batch_size</strong>: A larger batch size can speed up training, but requires more memory. Adjust based on your hardware capabilities.</li>
<li><strong>epochs</strong>: Increasing the number of epochs can improve model performance by allowing more learning cycles, but it will also extend the overall training time. Balance this based on your time constraints and hardware capacity.</li>
</ul>
</div>
<div class="admonition warning">
<p class="admonition-title">Ensure Correct CUDA and PyTorch Setup</p>
<p>Before using the <code>--accelerator cuda</code> flag, ensure your system has CUDA installed and configured correctly. Also, check that the installed CUDA version is compatible with your PyTorch and PyTorch Geometric versions.</p>
</div>
<hr />
<h3 id="3-making-predictions">3. Making Predictions<a class="headerlink" href="#3-making-predictions" title="Permanent link">&para;</a></h3>
<p>After training the model, use the <code>predict</code> command to make predictions on new data:</p>
<div class="termy" data-termynal data-ty-macos data-ty-title="bash"><span data-ty>// Example: Make predictions using a trained model</span><span data-ty="input" data-ty-prompt="$">python3 src/segger/cli/predict_fast.py \
    --segger_data_dir /path/to/saved/processed_data \
    --models_dir /path/to/saved/model/checkpoints \
    --benchmarks_dir /path/to/save/segmentation/results \
    --transcripts_file /path/to/raw_data/transcripts.parquet \
    --batch_size 1 \
    --num_workers 1 \
    --model_version 0 \
    --save_tag segger_embedding_1001 \
    --min_transcripts 5 \
    --cell_id_col segger_cell_id \
    --use_cc false \
    --knn_method cuda \
    --file_format anndata \
    --k_bd 4 \
    --dist_bd 12.0 \
    --k_tx 5 \
    --dist_tx 5.0</span><span data-ty></span></div>

<h4 id="parameters_2">Parameters<a class="headerlink" href="#parameters_2" title="Permanent link">&para;</a></h4>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>segger_data_dir</code></td>
<td>Directory containing the processed Segger dataset (in PyTorch Geometric format).</td>
<td>-</td>
</tr>
<tr>
<td><code>models_dir</code></td>
<td>Directory containing the trained models.</td>
<td>-</td>
</tr>
<tr>
<td><code>benchmarks_dir</code></td>
<td>Directory to save the segmentation results, including cell boundaries and associations.</td>
<td>-</td>
</tr>
<tr>
<td><code>transcripts_file</code></td>
<td>Path to the transcripts.parquet file.</td>
<td>-</td>
</tr>
<tr>
<td><code>batch_size</code></td>
<td>Number of samples to process per batch during prediction.</td>
<td><code>1</code></td>
</tr>
<tr>
<td><code>num_workers</code></td>
<td>Number of workers for parallel data loading.</td>
<td><code>1</code></td>
</tr>
<tr>
<td><code>model_version</code></td>
<td>Model version number to load for predictions, corresponding to the version from training logs.</td>
<td><code>0</code></td>
</tr>
<tr>
<td><code>save_tag</code></td>
<td>Tag used to name and organize the segmentation results.</td>
<td><code>segger_embedding_1001</code></td>
</tr>
<tr>
<td><code>min_transcripts</code></td>
<td>Minimum number of transcripts required for segmentation.</td>
<td><code>5</code></td>
</tr>
<tr>
<td><code>cell_id_col</code></td>
<td>Column name for cell IDs in the output data.</td>
<td><code>segger_cell_id</code></td>
</tr>
<tr>
<td><code>use_cc</code></td>
<td>Whether to use connected components for grouping transcripts without direct nucleus association.</td>
<td><code>False</code></td>
</tr>
<tr>
<td><code>knn_method</code></td>
<td>Method for KNN computation (e.g., <code>cuda</code> for GPU-based computation).</td>
<td><code>cuda</code></td>
</tr>
<tr>
<td><code>file_format</code></td>
<td>Format for the output segmentation data (e.g., <code>anndata</code>).</td>
<td><code>anndata</code></td>
</tr>
<tr>
<td><code>k_bd</code></td>
<td>Number of nearest neighbors for boundary nodes.</td>
<td><code>4</code></td>
</tr>
<tr>
<td><code>dist_bd</code></td>
<td>Maximum distance for boundary nodes.</td>
<td><code>12.0</code></td>
</tr>
<tr>
<td><code>k_tx</code></td>
<td>Number of nearest neighbors for transcript nodes.</td>
<td><code>5</code></td>
</tr>
<tr>
<td><code>dist_tx</code></td>
<td>Maximum distance for transcript nodes.</td>
<td><code>5.0</code></td>
</tr>
</tbody>
</table>
<div class="admonition tip">
<p class="admonition-title">Improving Prediction Pipeline</p>
<ul>
<li><strong>batch_size</strong>: A larger batch size can speed up training, but requires more memory. Adjust based on your hardware capabilities.</li>
<li><strong>use_cc</strong>: Enabling connected component analysis can improve the accuracy of transcript assignments.</li>
</ul>
</div>
<div class="admonition warning">
<p class="admonition-title">Ensure Correct CUDA, cuVS, and PyTorch Setup</p>
<p>Before using the <code>knn_method cuda</code> flag, ensure your system has CUDA installed and configured properly. Also, verify that the installed CUDA version is compatible with your cuPy, cuVS, PyTorch, and PyTorch Geometric versions.</p>
</div>
<hr />
<h3 id="4-running-the-entire-pipeline">4. Running the Entire Pipeline<a class="headerlink" href="#4-running-the-entire-pipeline" title="Permanent link">&para;</a></h3>
<p>The <code>submit_job.py</code> script allows you to run the complete Segger pipeline or specific stages like dataset creation, training, or prediction. The pipeline execution is determined by the configuration provided in a YAML file, supporting various environments like Docker, Singularity, and HPC systems (with LSF, Slurm support is planned).</p>
<h4 id="selecting-pipelines">Selecting Pipelines<a class="headerlink" href="#selecting-pipelines" title="Permanent link">&para;</a></h4>
<p>You can run the three stages—dataset creation, training, and prediction—sequentially or independently by specifying the pipelines in the YAML configuration file:</p>
<div class="codehilite"><pre><span></span><code><span class="o">-</span><span class="w"> </span>`<span class="mi">1</span>`<span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nv">dataset</span><span class="w"> </span><span class="nv">creation</span>
<span class="o">-</span><span class="w"> </span>`<span class="mi">2</span>`<span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nv">model</span><span class="w"> </span><span class="nv">training</span>
<span class="o">-</span><span class="w"> </span>`<span class="mi">3</span>`<span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="nv">prediction</span>
</code></pre></div>

<p>This allows you to run the full pipeline or just specific steps. Set the desired stages under the pipelines field in your YAML file.</p>
<h4 id="running-the-pipeline">Running the Pipeline<a class="headerlink" href="#running-the-pipeline" title="Permanent link">&para;</a></h4>
<p>Use the following command to run the pipeline:</p>
<div class="highlight"><pre><span></span><code><span class="go">python3 submit_job.py --config_file=filename.yaml</span>
</code></pre></div>
<ul>
<li>If no <code>--config_file</code> is provided, the default <code>config.yaml</code> file will be used.</li>
</ul>
<h3 id="5-containerization">5. Containerization<a class="headerlink" href="#5-containerization" title="Permanent link">&para;</a></h3>
<p>For users who want a portable, containerized environment, segger supports both Docker and Singularity containers. These containers provide a consistent runtime environment with all dependencies pre-installed.</p>
<h4 id="using-docker">Using Docker<a class="headerlink" href="#using-docker" title="Permanent link">&para;</a></h4>
<p>You can pull the segger Docker image from Docker Hub with this command:</p>
<div class="highlight"><pre><span></span><code><span class="go">docker pull danielunyi42/segger_dev:cuda121</span>
</code></pre></div>
<p>To run the pipeline in Docker, make sure your YAML configuration includes the following settings:</p>
<ul>
<li><code>use_singularity</code>: false</li>
<li><code>use_lsf</code>: false</li>
</ul>
<p>Afterwards, run the pipeline inside the Docker container with the same <code>submit_job.py</code> command.</p>
<h4 id="using-singularity">Using Singularity<a class="headerlink" href="#using-singularity" title="Permanent link">&para;</a></h4>
<p>For a Singularity environment, pull the image with:</p>
<div class="highlight"><pre><span></span><code><span class="go">singularity pull docker://danielunyi42/segger_dev:cuda121</span>
</code></pre></div>
<p>Ensure <code>use_singularity: true</code> in the YAML file and specify the Singularity image file (e.g., <code>segger_dev_latest.sif</code>) in the <code>singularity_image</code> field.</p>
<div class="admonition note">
<p class="admonition-title">Containerization</p>
<ul>
<li>The segger Docker image currently supports CUDA 11.8 and CUDA 12.1.</li>
</ul>
</div>
<h3 id="6-hpc-environments">6. HPC Environments<a class="headerlink" href="#6-hpc-environments" title="Permanent link">&para;</a></h3>
<p>Segger also supports HPC environments with LSF job scheduling. To run the pipeline on an HPC cluster using LSF, set <code>use_lsf: true</code> in your YAML configuration.</p>
<p>If your HPC system supports Slurm, a similar setup is planned and will be introduced soon.</p>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["navigation.sections", "navigation.instant", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate", "content.tabs.link", "navigation.tracking"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/termynal@0.0.1/termynal.js"></script>
      
        <script src="../termynal.js"></script>
      
    
  </body>
</html>