
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Fast and accurate cell segmentation for single-molecule spatial omics">
      
      
        <meta name="author" content="Elyas Heidari">
      
      
        <link rel="canonical" href="https://EliHei2.github.io/segger_dev/api/data/_utils/">
      
      
        <link rel="prev" href="../pyg_dataset/">
      
      
        <link rel="next" href="../transcript_embedding/">
      
      
      <link rel="icon" href="../../../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.33">
    
    
      
        <title>Utils - segger Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=ubuntu:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"ubuntu";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../css/extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css">
    
      <link rel="stylesheet" href="https://unpkg.com/termynal@0.0.1/termynal.css">
    
      <link rel="stylesheet" href="../../../termynal.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#seggerdata_utils" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="segger Documentation" class="md-header__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            segger Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Utils
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="segger Documentation" class="md-nav__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    segger Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/data_creation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dataset Creation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../notebooks/segger_tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to Segger
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CLI
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Data
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../sample/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Sample
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pyg_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyG Dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Utils
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Utils
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils" class="md-nav__link">
    <span class="md-ellipsis">
      _utils
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.SpatialTranscriptomicsDataset" class="md-nav__link">
    <span class="md-ellipsis">
      SpatialTranscriptomicsDataset
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SpatialTranscriptomicsDataset">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.processed_file_names" class="md-nav__link">
    <span class="md-ellipsis">
      processed_file_names
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.raw_file_names" class="md-nav__link">
    <span class="md-ellipsis">
      raw_file_names
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.download" class="md-nav__link">
    <span class="md-ellipsis">
      download
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.get" class="md-nav__link">
    <span class="md-ellipsis">
      get
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.len" class="md-nav__link">
    <span class="md-ellipsis">
      len
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.process" class="md-nav__link">
    <span class="md-ellipsis">
      process
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.add_transcript_ids" class="md-nav__link">
    <span class="md-ellipsis">
      add_transcript_ids
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.calculate_gene_celltype_abundance_embedding" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_gene_celltype_abundance_embedding
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.compute_nuclear_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      compute_nuclear_transcripts
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.compute_transcript_metrics" class="md-nav__link">
    <span class="md-ellipsis">
      compute_transcript_metrics
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.create_anndata" class="md-nav__link">
    <span class="md-ellipsis">
      create_anndata
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.ensure_transcript_ids" class="md-nav__link">
    <span class="md-ellipsis">
      ensure_transcript_ids
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.filter_boundaries" class="md-nav__link">
    <span class="md-ellipsis">
      filter_boundaries
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.filter_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      filter_transcripts
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.find_markers" class="md-nav__link">
    <span class="md-ellipsis">
      find_markers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.find_mutually_exclusive_genes" class="md-nav__link">
    <span class="md-ellipsis">
      find_mutually_exclusive_genes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.format_time" class="md-nav__link">
    <span class="md-ellipsis">
      format_time
    </span>
  </a>
  
    <nav class="md-nav" aria-label="format_time">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.format_time--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters:
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.format_time--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns:
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.get_edge_index" class="md-nav__link">
    <span class="md-ellipsis">
      get_edge_index
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.get_edge_index_kdtree" class="md-nav__link">
    <span class="md-ellipsis">
      get_edge_index_kdtree
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.get_polygons_from_xy" class="md-nav__link">
    <span class="md-ellipsis">
      get_polygons_from_xy
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.get_xy_extents" class="md-nav__link">
    <span class="md-ellipsis">
      get_xy_extents
    </span>
  </a>
  
    <nav class="md-nav" aria-label="get_xy_extents">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.get_xy_extents--parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data._utils.get_xy_extents--returns" class="md-nav__link">
    <span class="md-ellipsis">
      Returns
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.load_settings" class="md-nav__link">
    <span class="md-ellipsis">
      load_settings
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#src.segger.data._utils.read_parquet_region" class="md-nav__link">
    <span class="md-ellipsis">
      read_parquet_region
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../transcript_embedding/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transcript Embedding
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ndtree/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    NDTree
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../models/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Models
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="seggerdata_utils">segger.data._utils<a class="headerlink" href="#seggerdata_utils" title="Permanent link">&para;</a></h1>
<p>The <code>_utils</code> module provides core utility functions for data processing, filtering, and management in the Segger framework. This module contains essential functions for handling spatial transcriptomics data, including coordinate processing, data filtering, and settings management.</p>


<div class="doc doc-object doc-module">



<a id="src.segger.data._utils"></a>
    <div class="doc doc-contents first">



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="src.segger.data._utils.SpatialTranscriptomicsDataset" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">SpatialTranscriptomicsDataset</span>


<a href="#src.segger.data._utils.SpatialTranscriptomicsDataset" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">SpatialTranscriptomicsDataset</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">transform</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pre_transform</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pre_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><span title="torch_geometric.data.InMemoryDataset">InMemoryDataset</span></code></p>


      <p>A dataset class for handling SpatialTranscriptomics spatial transcriptomics data.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="src.segger.data._utils.SpatialTranscriptomicsDataset.root">root</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The root directory where the dataset is stored.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="src.segger.data._utils.SpatialTranscriptomicsDataset.transform">transform</span></code></td>
            <td>
                  <code>callable</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="src.segger.data._utils.SpatialTranscriptomicsDataset.pre_transform">pre_transform</span></code></td>
            <td>
                  <code>callable</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="src.segger.data._utils.SpatialTranscriptomicsDataset.pre_filter">pre_filter</span></code></td>
            <td>
                  <code>callable</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function that takes in a Data object and returns a boolean indicating whether to keep it.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

      <p>Initialize the SpatialTranscriptomicsDataset.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>root</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Root directory where the dataset is stored.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>transform</code></td>
            <td>
                  <code>callable</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pre_transform</code></td>
            <td>
                  <code>callable</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pre_filter</code></td>
            <td>
                  <code>callable</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A function that takes in a Data object and returns a boolean indicating whether to keep it. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>

                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">root</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">transform</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">pre_transform</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">pre_filter</span><span class="p">:</span> <span class="n">Callable</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialize the SpatialTranscriptomicsDataset.</span>

<span class="sd">    Args:</span>
<span class="sd">        root (str): Root directory where the dataset is stored.</span>
<span class="sd">        transform (callable, optional): A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</span>
<span class="sd">        pre_transform (callable, optional): A function/transform that takes in a Data object and returns a transformed version. Defaults to None.</span>
<span class="sd">        pre_filter (callable, optional): A function that takes in a Data object and returns a boolean indicating whether to keep it. Defaults to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">root</span><span class="p">,</span> <span class="n">transform</span><span class="p">,</span> <span class="n">pre_transform</span><span class="p">,</span> <span class="n">pre_filter</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data._utils.SpatialTranscriptomicsDataset.processed_file_names" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">processed_file_names</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.processed_file_names" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">processed_file_names</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Return a list of processed file names in the processed directory.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of processed file names.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data._utils.SpatialTranscriptomicsDataset.raw_file_names" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">raw_file_names</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.raw_file_names" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">raw_file_names</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Return a list of raw file names in the raw directory.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[str]: List of raw file names.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="src.segger.data._utils.SpatialTranscriptomicsDataset.download" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">download</span>


<a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.download" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">download</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Download the raw data. This method should be overridden if you need to download the data.</p>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">download</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Download the raw data. This method should be overridden if you need to download the data.&quot;&quot;&quot;</span>
    <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data._utils.SpatialTranscriptomicsDataset.get" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get</span>


<a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.get" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Get a processed data object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>idx</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Index of the data object to retrieve.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Data</code></td>            <td>
                  <code><span title="torch_geometric.data.Data">Data</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The processed data object.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span>
<span class="normal">973</span>
<span class="normal">974</span>
<span class="normal">975</span>
<span class="normal">976</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Data</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get a processed data object.</span>

<span class="sd">    Args:</span>
<span class="sd">        idx (int): Index of the data object to retrieve.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Data: The processed data object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">processed_dir</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">processed_file_names</span><span class="p">[</span><span class="n">idx</span><span class="p">])</span>
    <span class="p">)</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">to_dense</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">data</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data._utils.SpatialTranscriptomicsDataset.len" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">len</span>


<a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.len" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nb">len</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Return the number of processed files.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of processed files.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the number of processed files.</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: Number of processed files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">processed_file_names</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data._utils.SpatialTranscriptomicsDataset.process" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">process</span>


<a href="#src.segger.data._utils.SpatialTranscriptomicsDataset.process" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">process</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Process the raw data and save it to the processed directory. This method should be overridden if you need to process the data.</p>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process the raw data and save it to the processed directory. This method should be overridden if you need to process the data.&quot;&quot;&quot;</span>
    <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.add_transcript_ids" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">add_transcript_ids</span>


<a href="#src.segger.data._utils.add_transcript_ids" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">add_transcript_ids</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">x_col</span><span class="p">,</span> <span class="n">y_col</span><span class="p">,</span> <span class="n">id_col</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">,</span> <span class="n">precision</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Add unique transcript IDs to a DataFrame based on x,y coordinates.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>transcripts_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing transcript data with x,y coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>x_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the x-coordinate column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>y_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the y-coordinate column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>id_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the column to store the transcript IDs. Defaults to "transcript_id".</p>
              </div>
            </td>
            <td>
                  <code>&#39;transcript_id&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>precision</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Precision multiplier for coordinate values to handle floating point precision.
Defaults to 1000.</p>
              </div>
            </td>
            <td>
                  <code>1000</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: DataFrame with added transcript_id column.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">add_transcript_ids</span><span class="p">(</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">x_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">id_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;transcript_id&quot;</span><span class="p">,</span>
    <span class="n">precision</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add unique transcript IDs to a DataFrame based on x,y coordinates.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcripts_df: DataFrame containing transcript data with x,y coordinates.</span>
<span class="sd">        x_col: Name of the x-coordinate column.</span>
<span class="sd">        y_col: Name of the y-coordinate column.</span>
<span class="sd">        id_col: Name of the column to store the transcript IDs. Defaults to &quot;transcript_id&quot;.</span>
<span class="sd">        precision: Precision multiplier for coordinate values to handle floating point precision.</span>
<span class="sd">            Defaults to 1000.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: DataFrame with added transcript_id column.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create coordinate strings with specified precision</span>
    <span class="n">x_coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">[</span><span class="n">x_col</span><span class="p">]</span> <span class="o">*</span> <span class="n">precision</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">y_coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">[</span><span class="n">y_col</span><span class="p">]</span> <span class="o">*</span> <span class="n">precision</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">coords_str</span> <span class="o">=</span> <span class="n">x_coords</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">y_coords</span>

    <span class="c1"># Generate unique IDs using a deterministic hash function</span>
    <span class="k">def</span> <span class="nf">hash_coords</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate a deterministic hash for coordinate strings.</span>

<span class="sd">        Args:</span>
<span class="sd">            s: Coordinate string to hash.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: 8-digit integer hash value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Use a fixed seed for reproducibility</span>
        <span class="n">seed</span> <span class="o">=</span> <span class="mi">1996</span>
        <span class="c1"># Combine string with seed and take modulo to get an 8-digit integer</span>
        <span class="k">return</span> <span class="nb">abs</span><span class="p">(</span><span class="nb">hash</span><span class="p">(</span><span class="n">s</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">seed</span><span class="p">)))</span> <span class="o">%</span> <span class="mi">100000000</span>

    <span class="n">tx_ids</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">hash_coords</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">coords_str</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

    <span class="c1"># Add IDs to DataFrame</span>
    <span class="n">transcripts_df</span> <span class="o">=</span> <span class="n">transcripts_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">transcripts_df</span><span class="p">[</span><span class="n">id_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">tx_ids</span>

    <span class="k">return</span> <span class="n">transcripts_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.calculate_gene_celltype_abundance_embedding" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">calculate_gene_celltype_abundance_embedding</span>


<a href="#src.segger.data._utils.calculate_gene_celltype_abundance_embedding" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">calculate_gene_celltype_abundance_embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">celltype_column</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Calculate the cell type abundance embedding for each gene based on the fraction of cells in each cell type
that express the gene (non-zero expression).</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>adata</code></td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An AnnData object containing gene expression data and cell type information.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>celltype_column</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The column name in <code>adata.obs</code> that contains the cell type information.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame where rows are genes and columns are cell types, with each value representing
the fraction of cells in that cell type expressing the gene.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>adata = AnnData(...)  # Load your scRNA-seq AnnData object
celltype_column = 'celltype_major'
abundance_df = calculate_gene_celltype_abundance_embedding(adata, celltype_column)
abundance_df.head()</p>
</blockquote>
</blockquote>
</blockquote>
</details>
            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">calculate_gene_celltype_abundance_embedding</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">celltype_column</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate the cell type abundance embedding for each gene based on the fraction of cells in each cell type</span>
<span class="sd">    that express the gene (non-zero expression).</span>

<span class="sd">    Parameters:</span>
<span class="sd">        adata (ad.AnnData): An AnnData object containing gene expression data and cell type information.</span>
<span class="sd">        celltype_column (str): The column name in `adata.obs` that contains the cell type information.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame where rows are genes and columns are cell types, with each value representing</span>
<span class="sd">            the fraction of cells in that cell type expressing the gene.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; adata = AnnData(...)  # Load your scRNA-seq AnnData object</span>
<span class="sd">        &gt;&gt;&gt; celltype_column = &#39;celltype_major&#39;</span>
<span class="sd">        &gt;&gt;&gt; abundance_df = calculate_gene_celltype_abundance_embedding(adata, celltype_column)</span>
<span class="sd">        &gt;&gt;&gt; abundance_df.head()</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract expression data (cells x genes) and cell type information (cells)</span>
    <span class="n">expression_data</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span> <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="s2">&quot;toarray&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
    <span class="n">cell_types</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">celltype_column</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="c1"># Create a binary matrix for gene expression (1 if non-zero, 0 otherwise)</span>
    <span class="n">gene_expression_binary</span> <span class="o">=</span> <span class="p">(</span><span class="n">expression_data</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="c1"># Convert the binary matrix to a DataFrame</span>
    <span class="n">gene_expression_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">gene_expression_binary</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
    <span class="p">)</span>
    <span class="c1"># Perform one-hot encoding on the cell types</span>
    <span class="n">encoder</span> <span class="o">=</span> <span class="n">OneHotEncoder</span><span class="p">(</span><span class="n">sparse_output</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">cell_type_encoded</span> <span class="o">=</span> <span class="n">encoder</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">cell_types</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
    <span class="c1"># Calculate the fraction of cells expressing each gene per cell type</span>
    <span class="n">cell_type_abundance_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cell_type_encoded</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
        <span class="c1"># Extract cells of the current cell type</span>
        <span class="n">cell_type_mask</span> <span class="o">=</span> <span class="n">cell_type_encoded</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="c1"># Calculate the abundance: sum of non-zero expressions in this cell type / total cells in this cell type</span>
        <span class="n">abundance</span> <span class="o">=</span> <span class="n">gene_expression_df</span><span class="p">[</span><span class="n">cell_type_mask</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">cell_type_abundance_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">abundance</span><span class="p">)</span>
    <span class="c1"># Create a DataFrame for the cell type abundance with gene names as rows and cell types as columns</span>
    <span class="n">cell_type_abundance_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">cell_type_abundance_list</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">encoder</span><span class="o">.</span><span class="n">categories_</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="k">return</span> <span class="n">cell_type_abundance_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.compute_nuclear_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_nuclear_transcripts</span>


<a href="#src.segger.data._utils.compute_nuclear_transcripts" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_nuclear_transcripts</span><span class="p">(</span><span class="n">polygons</span><span class="p">,</span> <span class="n">transcripts</span><span class="p">,</span> <span class="n">x_col</span><span class="p">,</span> <span class="n">y_col</span><span class="p">,</span> <span class="n">nuclear_column</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">nuclear_value</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Compute which transcripts are nuclear based on their coordinates and the
nuclear polygons.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>polygons</code></td>
            <td>
                  <code><span title="geopandas.GeoSeries">GeoSeries</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The nuclear polygons.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>transcripts</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The transcripts DataFrame.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>x_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The x-coordinate column name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>y_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The y-coordinate column name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nuclear_column</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The column name that indicates if a transcript is nuclear. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nuclear_value</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The value in nuclear_column that indicates a nuclear transcript. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.Series: A boolean series indicating which transcripts are nuclear.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_nuclear_transcripts</span><span class="p">(</span>
    <span class="n">polygons</span><span class="p">:</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="p">,</span>
    <span class="n">transcripts</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">x_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">nuclear_column</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">nuclear_value</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute which transcripts are nuclear based on their coordinates and the</span>
<span class="sd">    nuclear polygons.</span>

<span class="sd">    Args:</span>
<span class="sd">        polygons: The nuclear polygons.</span>
<span class="sd">        transcripts: The transcripts DataFrame.</span>
<span class="sd">        x_col: The x-coordinate column name.</span>
<span class="sd">        y_col: The y-coordinate column name.</span>
<span class="sd">        nuclear_column: The column name that indicates if a transcript is nuclear. Defaults to None.</span>
<span class="sd">        nuclear_value: The value in nuclear_column that indicates a nuclear transcript. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: A boolean series indicating which transcripts are nuclear.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># If nuclear_column and nuclear_value are provided, use them</span>
    <span class="k">if</span> <span class="n">nuclear_column</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">nuclear_value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">nuclear_column</span> <span class="ow">in</span> <span class="n">transcripts</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">transcripts</span><span class="p">[</span><span class="n">nuclear_column</span><span class="p">]</span><span class="o">.</span><span class="n">eq</span><span class="p">(</span><span class="n">nuclear_value</span><span class="p">)</span>

    <span class="c1"># Otherwise compute based on coordinates</span>
    <span class="n">points</span> <span class="o">=</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="p">(</span><span class="n">gpd</span><span class="o">.</span><span class="n">points_from_xy</span><span class="p">(</span><span class="n">transcripts</span><span class="p">[</span><span class="n">x_col</span><span class="p">],</span> <span class="n">transcripts</span><span class="p">[</span><span class="n">y_col</span><span class="p">]))</span>
    <span class="k">return</span> <span class="n">points</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">p</span><span class="p">:</span> <span class="nb">any</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">within</span><span class="p">(</span><span class="n">poly</span><span class="p">)</span> <span class="k">for</span> <span class="n">poly</span> <span class="ow">in</span> <span class="n">polygons</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.compute_transcript_metrics" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">compute_transcript_metrics</span>


<a href="#src.segger.data._utils.compute_transcript_metrics" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">compute_transcript_metrics</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">qv_threshold</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">cell_id_col</span><span class="o">=</span><span class="s1">&#39;cell_id&#39;</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Computes various metrics for a given dataframe of transcript data filtered by quality value threshold.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>qv_threshold</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>30</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>cell_id_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the cell ID.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cell_id&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Any">Any</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Any]: A dictionary containing various transcript metrics:
- 'percent_assigned' (float): The percentage of assigned transcripts.
- 'percent_cytoplasmic' (float): The percentage of cytoplasmic transcripts among assigned transcripts.
- 'percent_nucleus' (float): The percentage of nucleus transcripts among assigned transcripts.
- 'percent_non_assigned_cytoplasmic' (float): The percentage of non-assigned cytoplasmic transcripts.
- 'gene_metrics' (pd.DataFrame): A dataframe containing gene-level metrics.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_transcript_metrics</span><span class="p">(</span>
    <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">qv_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">cell_id_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_id&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes various metrics for a given dataframe of transcript data filtered by quality value threshold.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        df (pd.DataFrame): The dataframe containing transcript data.</span>
<span class="sd">        qv_threshold (float): The quality value threshold for filtering transcripts.</span>
<span class="sd">        cell_id_col (str): The name of the column representing the cell ID.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Any]: A dictionary containing various transcript metrics:</span>
<span class="sd">            - &#39;percent_assigned&#39; (float): The percentage of assigned transcripts.</span>
<span class="sd">            - &#39;percent_cytoplasmic&#39; (float): The percentage of cytoplasmic transcripts among assigned transcripts.</span>
<span class="sd">            - &#39;percent_nucleus&#39; (float): The percentage of nucleus transcripts among assigned transcripts.</span>
<span class="sd">            - &#39;percent_non_assigned_cytoplasmic&#39; (float): The percentage of non-assigned cytoplasmic transcripts.</span>
<span class="sd">            - &#39;gene_metrics&#39; (pd.DataFrame): A dataframe containing gene-level metrics.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;qv&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">qv_threshold</span><span class="p">]</span>
    <span class="n">total_transcripts</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_filtered</span><span class="p">)</span>
    <span class="n">assigned_transcripts</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="p">[</span><span class="n">df_filtered</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">percent_assigned</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">assigned_transcripts</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">total_transcripts</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="n">cytoplasmic_transcripts</span> <span class="o">=</span> <span class="n">assigned_transcripts</span><span class="p">[</span>
        <span class="n">assigned_transcripts</span><span class="p">[</span><span class="s2">&quot;overlaps_nucleus&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">1</span>
    <span class="p">]</span>
    <span class="n">percent_cytoplasmic</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">cytoplasmic_transcripts</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">assigned_transcripts</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="p">)</span>
    <span class="n">percent_nucleus</span> <span class="o">=</span> <span class="mi">100</span> <span class="o">-</span> <span class="n">percent_cytoplasmic</span>
    <span class="n">non_assigned_transcripts</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="p">[</span><span class="n">df_filtered</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">non_assigned_cytoplasmic</span> <span class="o">=</span> <span class="n">non_assigned_transcripts</span><span class="p">[</span>
        <span class="n">non_assigned_transcripts</span><span class="p">[</span><span class="s2">&quot;overlaps_nucleus&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">1</span>
    <span class="p">]</span>
    <span class="n">percent_non_assigned_cytoplasmic</span> <span class="o">=</span> <span class="p">(</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">non_assigned_cytoplasmic</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">non_assigned_transcripts</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="p">)</span>
    <span class="n">gene_group_assigned</span> <span class="o">=</span> <span class="n">assigned_transcripts</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)</span>
    <span class="n">gene_group_all</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)</span>
    <span class="n">gene_percent_assigned</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">gene_group_assigned</span><span class="o">.</span><span class="n">size</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="n">gene_group_all</span><span class="o">.</span><span class="n">size</span><span class="p">()</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">names</span><span class="o">=</span><span class="s2">&quot;percent_assigned&quot;</span><span class="p">)</span>
    <span class="n">cytoplasmic_gene_group</span> <span class="o">=</span> <span class="n">cytoplasmic_transcripts</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)</span>
    <span class="n">gene_percent_cytoplasmic</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">cytoplasmic_gene_group</span><span class="o">.</span><span class="n">size</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">cytoplasmic_transcripts</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>
    <span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;percent_cytoplasmic&quot;</span><span class="p">)</span>
    <span class="n">gene_metrics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">gene_percent_assigned</span><span class="p">,</span> <span class="n">gene_percent_cytoplasmic</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;feature_name&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;percent_assigned&quot;</span><span class="p">:</span> <span class="n">percent_assigned</span><span class="p">,</span>
        <span class="s2">&quot;percent_cytoplasmic&quot;</span><span class="p">:</span> <span class="n">percent_cytoplasmic</span><span class="p">,</span>
        <span class="s2">&quot;percent_nucleus&quot;</span><span class="p">:</span> <span class="n">percent_nucleus</span><span class="p">,</span>
        <span class="s2">&quot;percent_non_assigned_cytoplasmic&quot;</span><span class="p">:</span> <span class="n">percent_non_assigned_cytoplasmic</span><span class="p">,</span>
        <span class="s2">&quot;gene_metrics&quot;</span><span class="p">:</span> <span class="n">gene_metrics</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="n">results</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.create_anndata" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">create_anndata</span>


<a href="#src.segger.data._utils.create_anndata" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">create_anndata</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">panel_df</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_transcripts</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">cell_id_col</span><span class="o">=</span><span class="s1">&#39;cell_id&#39;</span><span class="p">,</span> <span class="n">qv_threshold</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">min_cell_area</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span> <span class="n">max_cell_area</span><span class="o">=</span><span class="mf">1000.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Generates an AnnData object from a dataframe of segmented transcriptomics data.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing segmented transcriptomics data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>panel_df</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="pandas.DataFrame">DataFrame</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing panel information.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_transcripts</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum number of transcripts required for a cell to be included.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>cell_id_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The column name representing the cell ID in the input dataframe.</p>
              </div>
            </td>
            <td>
                  <code>&#39;cell_id&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>qv_threshold</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The quality value threshold for filtering transcripts.</p>
              </div>
            </td>
            <td>
                  <code>30</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_cell_area</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum cell area to include a cell.</p>
              </div>
            </td>
            <td>
                  <code>10.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_cell_area</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The maximum cell area to include a cell.</p>
              </div>
            </td>
            <td>
                  <code>1000.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>ad.AnnData: The generated AnnData object containing the transcriptomics data and metadata.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">create_anndata</span><span class="p">(</span>
    <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">panel_df</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">min_transcripts</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">cell_id_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;cell_id&quot;</span><span class="p">,</span>
    <span class="n">qv_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span>
    <span class="n">min_cell_area</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">10.0</span><span class="p">,</span>
    <span class="n">max_cell_area</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1000.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generates an AnnData object from a dataframe of segmented transcriptomics data.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        df (pd.DataFrame): The dataframe containing segmented transcriptomics data.</span>
<span class="sd">        panel_df (Optional[pd.DataFrame]): The dataframe containing panel information.</span>
<span class="sd">        min_transcripts (int): The minimum number of transcripts required for a cell to be included.</span>
<span class="sd">        cell_id_col (str): The column name representing the cell ID in the input dataframe.</span>
<span class="sd">        qv_threshold (float): The quality value threshold for filtering transcripts.</span>
<span class="sd">        min_cell_area (float): The minimum cell area to include a cell.</span>
<span class="sd">        max_cell_area (float): The maximum cell area to include a cell.</span>

<span class="sd">    Returns:</span>
<span class="sd">        ad.AnnData: The generated AnnData object containing the transcriptomics data and metadata.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter out unassigned cells</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;UNASSIGNED&quot;</span><span class="p">]</span>

    <span class="c1"># Create pivot table for gene expression counts per cell</span>
    <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="n">cell_id_col</span><span class="p">:</span> <span class="s2">&quot;cell&quot;</span><span class="p">,</span> <span class="s2">&quot;feature_name&quot;</span><span class="p">:</span> <span class="s2">&quot;gene&quot;</span><span class="p">}</span>
    <span class="p">)[[</span><span class="s2">&quot;cell&quot;</span><span class="p">,</span> <span class="s2">&quot;gene&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span>
        <span class="n">index</span><span class="o">=</span><span class="s2">&quot;cell&quot;</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">aggfunc</span><span class="o">=</span><span class="s2">&quot;size&quot;</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">0</span>
    <span class="p">)</span>
    <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="p">[</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">min_transcripts</span><span class="p">]</span>

    <span class="c1"># Summarize cell metrics</span>
    <span class="n">cell_summary</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">cell_id</span><span class="p">,</span> <span class="n">cell_data</span> <span class="ow">in</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">cell_id_col</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_data</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_transcripts</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">cell_convex_hull</span> <span class="o">=</span> <span class="n">ConvexHull</span><span class="p">(</span>
            <span class="n">cell_data</span><span class="p">[[</span><span class="s2">&quot;x_location&quot;</span><span class="p">,</span> <span class="s2">&quot;y_location&quot;</span><span class="p">]],</span> <span class="n">qhull_options</span><span class="o">=</span><span class="s2">&quot;QJ&quot;</span>
        <span class="p">)</span>
        <span class="n">cell_area</span> <span class="o">=</span> <span class="n">cell_convex_hull</span><span class="o">.</span><span class="n">area</span>
        <span class="k">if</span> <span class="n">cell_area</span> <span class="o">&lt;</span> <span class="n">min_cell_area</span> <span class="ow">or</span> <span class="n">cell_area</span> <span class="o">&gt;</span> <span class="n">max_cell_area</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">cell_summary</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;cell&quot;</span><span class="p">:</span> <span class="n">cell_id</span><span class="p">,</span>
                <span class="s2">&quot;cell_centroid_x&quot;</span><span class="p">:</span> <span class="n">cell_data</span><span class="p">[</span><span class="s2">&quot;x_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;cell_centroid_y&quot;</span><span class="p">:</span> <span class="n">cell_data</span><span class="p">[</span><span class="s2">&quot;y_location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(),</span>
                <span class="s2">&quot;cell_area&quot;</span><span class="p">:</span> <span class="n">cell_area</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="n">cell_summary</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_summary</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;cell&quot;</span><span class="p">)</span>

    <span class="c1"># Add genes from panel_df (if provided) to the pivot table</span>
    <span class="k">if</span> <span class="n">panel_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">panel_df</span> <span class="o">=</span> <span class="n">panel_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">panel_df</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pivot_df</span><span class="p">:</span>
                <span class="n">pivot_df</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="p">[</span><span class="n">genes</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>

    <span class="c1"># Create var DataFrame</span>
    <span class="k">if</span> <span class="n">panel_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">var_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="p">{</span><span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span> <span class="s2">&quot;feature_types&quot;</span><span class="p">:</span> <span class="s2">&quot;Gene Expression&quot;</span><span class="p">,</span> <span class="s2">&quot;genome&quot;</span><span class="p">:</span> <span class="s2">&quot;Unknown&quot;</span><span class="p">}</span>
                <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">var_df</span> <span class="o">=</span> <span class="n">panel_df</span><span class="p">[[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;ensembl&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;ensembl&quot;</span><span class="p">:</span> <span class="s2">&quot;gene_ids&quot;</span><span class="p">})</span>
        <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;feature_types&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Gene Expression&quot;</span>
        <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;genome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Unknown&quot;</span>
        <span class="n">var_df</span> <span class="o">=</span> <span class="n">var_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>

    <span class="c1"># Compute total assigned and unassigned transcript counts for each gene</span>
    <span class="n">assigned_counts</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">unassigned_counts</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">cell_id_col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;UNASSIGNED&quot;</span><span class="p">]</span>
        <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;feature_name&quot;</span><span class="p">)[</span><span class="s2">&quot;feature_name&quot;</span><span class="p">]</span>
        <span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;total_assigned&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">var_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">assigned_counts</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">var_df</span><span class="p">[</span><span class="s2">&quot;total_unassigned&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">var_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">unassigned_counts</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Filter cells and create the AnnData object</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">cell_summary</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">pivot_df</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">cell_summary</span> <span class="o">=</span> <span class="n">cell_summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cells</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">pivot_df</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">var</span> <span class="o">=</span> <span class="n">var_df</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;unique_transcripts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pivot_df</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">pivot_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span>
        <span class="n">cell_summary</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span> <span class="p">:],</span>
        <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.ensure_transcript_ids" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">ensure_transcript_ids</span>


<a href="#src.segger.data._utils.ensure_transcript_ids" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">ensure_transcript_ids</span><span class="p">(</span><span class="n">parquet_path</span><span class="p">,</span> <span class="n">x_col</span><span class="p">,</span> <span class="n">y_col</span><span class="p">,</span> <span class="n">id_col</span><span class="o">=</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">,</span> <span class="n">precision</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Ensure that a parquet file has transcript IDs by adding them if missing.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>parquet_path</code></td>
            <td>
                  <code><span title="os.PathLike">PathLike</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the parquet file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>x_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the x-coordinate column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>y_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the y-coordinate column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>id_col</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the column to store the transcript IDs. Defaults to "transcript_id".</p>
              </div>
            </td>
            <td>
                  <code>&#39;transcript_id&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>precision</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Precision multiplier for coordinate values to handle floating point precision.
Defaults to 1000.</p>
              </div>
            </td>
            <td>
                  <code>1000</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">ensure_transcript_ids</span><span class="p">(</span>
    <span class="n">parquet_path</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">,</span>
    <span class="n">x_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y_col</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">id_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;transcript_id&quot;</span><span class="p">,</span>
    <span class="n">precision</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Ensure that a parquet file has transcript IDs by adding them if missing.</span>

<span class="sd">    Args:</span>
<span class="sd">        parquet_path: Path to the parquet file.</span>
<span class="sd">        x_col: Name of the x-coordinate column.</span>
<span class="sd">        y_col: Name of the y-coordinate column.</span>
<span class="sd">        id_col: Name of the column to store the transcript IDs. Defaults to &quot;transcript_id&quot;.</span>
<span class="sd">        precision: Precision multiplier for coordinate values to handle floating point precision.</span>
<span class="sd">            Defaults to 1000.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First check metadata to see if column exists</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="n">pq</span><span class="o">.</span><span class="n">read_metadata</span><span class="p">(</span><span class="n">parquet_path</span><span class="p">)</span>
    <span class="n">schema_idx</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">reversed</span><span class="p">,</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metadata</span><span class="o">.</span><span class="n">schema</span><span class="o">.</span><span class="n">names</span><span class="p">)))</span>

    <span class="c1"># Only proceed if the column doesn&#39;t exist</span>
    <span class="k">if</span> <span class="n">id_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">schema_idx</span><span class="p">:</span>
        <span class="c1"># Read the parquet file</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">parquet_path</span><span class="p">)</span>

        <span class="c1"># Add transcript IDs</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">add_transcript_ids</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">x_col</span><span class="p">,</span> <span class="n">y_col</span><span class="p">,</span> <span class="n">id_col</span><span class="p">,</span> <span class="n">precision</span><span class="p">)</span>

        <span class="c1"># Convert DataFrame to Arrow table</span>
        <span class="n">table</span> <span class="o">=</span> <span class="n">pa</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

        <span class="c1"># Write back to parquet</span>
        <span class="n">pq</span><span class="o">.</span><span class="n">write_table</span><span class="p">(</span>
            <span class="n">table</span><span class="p">,</span>
            <span class="n">parquet_path</span><span class="p">,</span>
            <span class="n">version</span><span class="o">=</span><span class="s2">&quot;2.6&quot;</span><span class="p">,</span>  <span class="c1"># Use latest stable version</span>
            <span class="n">write_statistics</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># Ensure statistics are written</span>
            <span class="n">compression</span><span class="o">=</span><span class="s2">&quot;snappy&quot;</span><span class="p">,</span>  <span class="c1"># Use snappy compression for better performance</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.filter_boundaries" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">filter_boundaries</span>


<a href="#src.segger.data._utils.filter_boundaries" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">filter_boundaries</span><span class="p">(</span><span class="n">boundaries</span><span class="p">,</span> <span class="n">inset</span><span class="p">,</span> <span class="n">outset</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">label</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Filter boundary polygons based on their overlap with specified inset and
outset regions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>boundaries</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A DataFrame containing the boundary data with x and y coordinates and
identifiers.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>inset</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A polygon representing the inner region to filter the boundaries.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>outset</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A polygon representing the outer region to filter the boundaries.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>x</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the x-coordinate.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>y</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the y-coordinate.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>label</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the cell or nucleus label.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the filtered boundary polygons.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <p>The function determines overlaps of boundary polygons with the specified
inset and outset regions. It creates boolean masks for overlaps with the
top, left, right, and bottom sides of the outset region, as well as the
center region defined by the inset polygon. The filtering logic includes
polygons that:
- Are completely within the center region.
- Overlap with the center and the left side, but not the bottom side.
- Overlap with the center and the top side, but not the right side.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_boundaries</span><span class="p">(</span>
    <span class="n">boundaries</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">inset</span><span class="p">:</span> <span class="n">shapely</span><span class="o">.</span><span class="n">Polygon</span><span class="p">,</span>
    <span class="n">outset</span><span class="p">:</span> <span class="n">shapely</span><span class="o">.</span><span class="n">Polygon</span><span class="p">,</span>
    <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">label</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter boundary polygons based on their overlap with specified inset and</span>
<span class="sd">    outset regions.</span>

<span class="sd">    Args:</span>
<span class="sd">        boundaries: A DataFrame containing the boundary data with x and y coordinates and</span>
<span class="sd">            identifiers.</span>
<span class="sd">        inset: A polygon representing the inner region to filter the boundaries.</span>
<span class="sd">        outset: A polygon representing the outer region to filter the boundaries.</span>
<span class="sd">        x: The name of the column representing the x-coordinate.</span>
<span class="sd">        y: The name of the column representing the y-coordinate.</span>
<span class="sd">        label: The name of the column representing the cell or nucleus label.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the filtered boundary polygons.</span>

<span class="sd">    Note:</span>
<span class="sd">        The function determines overlaps of boundary polygons with the specified</span>
<span class="sd">        inset and outset regions. It creates boolean masks for overlaps with the</span>
<span class="sd">        top, left, right, and bottom sides of the outset region, as well as the</span>
<span class="sd">        center region defined by the inset polygon. The filtering logic includes</span>
<span class="sd">        polygons that:</span>
<span class="sd">        - Are completely within the center region.</span>
<span class="sd">        - Overlap with the center and the left side, but not the bottom side.</span>
<span class="sd">        - Overlap with the center and the top side, but not the right side.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Determine overlaps of boundary polygons</span>
    <span class="k">def</span> <span class="nf">in_region</span><span class="p">(</span><span class="n">region</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if boundaries are within a specified region.</span>

<span class="sd">        Args:</span>
<span class="sd">            region: Shapely polygon defining the region to check.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.Series: Boolean mask indicating which boundaries are in the region.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">in_x</span> <span class="o">=</span> <span class="n">boundaries</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="n">region</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">region</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">in_y</span> <span class="o">=</span> <span class="n">boundaries</span><span class="p">[</span><span class="n">y</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="n">region</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">region</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">in_x</span> <span class="o">&amp;</span> <span class="n">in_y</span>

    <span class="n">x1</span><span class="p">,</span> <span class="n">y1</span><span class="p">,</span> <span class="n">x4</span><span class="p">,</span> <span class="n">y4</span> <span class="o">=</span> <span class="n">outset</span><span class="o">.</span><span class="n">bounds</span>
    <span class="n">x2</span><span class="p">,</span> <span class="n">y2</span><span class="p">,</span> <span class="n">x3</span><span class="p">,</span> <span class="n">y3</span> <span class="o">=</span> <span class="n">inset</span><span class="o">.</span><span class="n">bounds</span>
    <span class="n">boundaries</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">in_region</span><span class="p">(</span><span class="n">shapely</span><span class="o">.</span><span class="n">box</span><span class="p">(</span><span class="n">x1</span><span class="p">,</span> <span class="n">y1</span><span class="p">,</span> <span class="n">x4</span><span class="p">,</span> <span class="n">y2</span><span class="p">))</span>
    <span class="n">boundaries</span><span class="p">[</span><span class="s2">&quot;left&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">in_region</span><span class="p">(</span><span class="n">shapely</span><span class="o">.</span><span class="n">box</span><span class="p">(</span><span class="n">x1</span><span class="p">,</span> <span class="n">y1</span><span class="p">,</span> <span class="n">x2</span><span class="p">,</span> <span class="n">y4</span><span class="p">))</span>
    <span class="n">boundaries</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">in_region</span><span class="p">(</span><span class="n">shapely</span><span class="o">.</span><span class="n">box</span><span class="p">(</span><span class="n">x3</span><span class="p">,</span> <span class="n">y1</span><span class="p">,</span> <span class="n">x4</span><span class="p">,</span> <span class="n">y4</span><span class="p">))</span>
    <span class="n">boundaries</span><span class="p">[</span><span class="s2">&quot;bottom&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">in_region</span><span class="p">(</span><span class="n">shapely</span><span class="o">.</span><span class="n">box</span><span class="p">(</span><span class="n">x1</span><span class="p">,</span> <span class="n">y3</span><span class="p">,</span> <span class="n">x4</span><span class="p">,</span> <span class="n">y4</span><span class="p">))</span>
    <span class="n">boundaries</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">in_region</span><span class="p">(</span><span class="n">inset</span><span class="p">)</span>

    <span class="c1"># Filter boundary polygons</span>
    <span class="c1"># Include overlaps with top and left, not bottom and right</span>
    <span class="n">gb</span> <span class="o">=</span> <span class="n">boundaries</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">total</span> <span class="o">=</span> <span class="n">gb</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s2">&quot;size&quot;</span><span class="p">)</span>
    <span class="n">in_top</span> <span class="o">=</span> <span class="n">gb</span><span class="p">[</span><span class="s2">&quot;top&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s2">&quot;sum&quot;</span><span class="p">)</span>
    <span class="n">in_left</span> <span class="o">=</span> <span class="n">gb</span><span class="p">[</span><span class="s2">&quot;left&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s2">&quot;sum&quot;</span><span class="p">)</span>
    <span class="n">in_right</span> <span class="o">=</span> <span class="n">gb</span><span class="p">[</span><span class="s2">&quot;right&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s2">&quot;sum&quot;</span><span class="p">)</span>
    <span class="n">in_bottom</span> <span class="o">=</span> <span class="n">gb</span><span class="p">[</span><span class="s2">&quot;bottom&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s2">&quot;sum&quot;</span><span class="p">)</span>
    <span class="n">in_center</span> <span class="o">=</span> <span class="n">gb</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s2">&quot;sum&quot;</span><span class="p">)</span>
    <span class="n">keep</span> <span class="o">=</span> <span class="n">in_center</span> <span class="o">==</span> <span class="n">total</span>
    <span class="n">keep</span> <span class="o">|=</span> <span class="p">(</span><span class="n">in_center</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">in_left</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">in_bottom</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">keep</span> <span class="o">|=</span> <span class="p">(</span><span class="n">in_center</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">in_top</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">in_right</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">inset_boundaries</span> <span class="o">=</span> <span class="n">boundaries</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">keep</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">inset_boundaries</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.filter_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">filter_transcripts</span>


<a href="#src.segger.data._utils.filter_transcripts" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">filter_transcripts</span><span class="p">(</span><span class="n">transcripts_df</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">filter_substrings</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">qv_column</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_qv</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Filter transcripts based on quality value and remove unwanted transcripts.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>transcripts_df</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The dataframe containing transcript data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>label</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The label of transcript features. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>filter_substrings</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>[str]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The list of feature substrings to remove. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>qv_column</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the quality value. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_qv</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[float]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum quality value threshold for filtering transcripts. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: The filtered dataframe.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_transcripts</span><span class="p">(</span>
    <span class="n">transcripts_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">label</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">filter_substrings</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">qv_column</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">min_qv</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter transcripts based on quality value and remove unwanted transcripts.</span>

<span class="sd">    Args:</span>
<span class="sd">        transcripts_df: The dataframe containing transcript data.</span>
<span class="sd">        label: The label of transcript features. Defaults to None.</span>
<span class="sd">        filter_substrings: The list of feature substrings to remove. Defaults to None.</span>
<span class="sd">        qv_column: The name of the column representing the quality value. Defaults to None.</span>
<span class="sd">        min_qv: The minimum quality value threshold for filtering transcripts. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The filtered dataframe.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">transcripts_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">filter_substrings</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">label</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">&amp;=</span> <span class="o">~</span><span class="n">transcripts_df</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">filter_substrings</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">min_qv</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">qv_column</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mask</span> <span class="o">&amp;=</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="n">qv_column</span><span class="p">]</span><span class="o">.</span><span class="n">ge</span><span class="p">(</span><span class="n">min_qv</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">transcripts_df</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.find_markers" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">find_markers</span>


<a href="#src.segger.data._utils.find_markers" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">find_markers</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">cell_type_column</span><span class="p">,</span> <span class="n">pos_percentile</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">neg_percentile</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">percentage</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Identify positive and negative marker genes for each cell type in an AnnData object.
Positive markers are top-ranked genes that are expressed in at least
<code>percentage</code> percent of cells in the given cell type.
Args:
    adata: Annotated data object containing gene expression data and cell type annotations.
    cell_type_column: Name of the column in <code>adata.obs</code> specifying cell type identity for each cell.
    pos_percentile: Percentile threshold for selecting top highly expressed genes as positive markers. Defaults to 5.
    neg_percentile: Percentile threshold for selecting lowest expressed genes as negative markers. Defaults to 10.
    percentage: Minimum percent of cells (0-100) in a cell type expressing a gene for it to be a marker. Defaults to 50.
Returns:
    dict: Dictionary mapping cell type names to:
        {
            'positive': [list of positive marker gene names],
            'negative': [list of negative marker gene names]
        }</p>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_markers</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span>
    <span class="n">cell_type_column</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">pos_percentile</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">neg_percentile</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">percentage</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify positive and negative marker genes for each cell type in an AnnData object.</span>
<span class="sd">    Positive markers are top-ranked genes that are expressed in at least</span>
<span class="sd">    `percentage` percent of cells in the given cell type.</span>
<span class="sd">    Args:</span>
<span class="sd">        adata: Annotated data object containing gene expression data and cell type annotations.</span>
<span class="sd">        cell_type_column: Name of the column in `adata.obs` specifying cell type identity for each cell.</span>
<span class="sd">        pos_percentile: Percentile threshold for selecting top highly expressed genes as positive markers. Defaults to 5.</span>
<span class="sd">        neg_percentile: Percentile threshold for selecting lowest expressed genes as negative markers. Defaults to 10.</span>
<span class="sd">        percentage: Minimum percent of cells (0-100) in a cell type expressing a gene for it to be a marker. Defaults to 50.</span>
<span class="sd">    Returns:</span>
<span class="sd">        dict: Dictionary mapping cell type names to:</span>
<span class="sd">            {</span>
<span class="sd">                &#39;positive&#39;: [list of positive marker gene names],</span>
<span class="sd">                &#39;negative&#39;: [list of negative marker gene names]</span>
<span class="sd">            }</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">markers</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="n">cell_type_column</span><span class="p">)</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
    <span class="n">n_genes</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="c1"># Work with a dense matrix for expression fraction calculation</span>
    <span class="c1"># (convert sparse to dense if needed)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">):</span>
        <span class="n">expr_matrix</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">expr_matrix</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>
    <span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_column</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_column</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
        <span class="n">gene_names</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">][</span><span class="s1">&#39;names&#39;</span><span class="p">][</span><span class="n">cell_type</span><span class="p">])</span>
        <span class="n">n_pos</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_genes</span> <span class="o">*</span> <span class="n">pos_percentile</span> <span class="o">//</span> <span class="mi">100</span><span class="p">))</span>
        <span class="n">n_neg</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_genes</span> <span class="o">*</span> <span class="n">neg_percentile</span> <span class="o">//</span> <span class="mi">100</span><span class="p">))</span>
        <span class="c1"># Calculate percent of cells in this cell type expressing each gene</span>
        <span class="n">expr_frac</span> <span class="o">=</span> <span class="p">(</span><span class="n">expr_matrix</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>  <span class="c1"># as percent</span>
        <span class="c1"># Filter positive markers by expression fraction</span>
        <span class="n">pos_indices</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_pos</span><span class="p">):</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="n">gene_names</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
            <span class="n">gene_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">genes</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">expr_frac</span><span class="p">[</span><span class="n">gene_idx</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">percentage</span><span class="p">:</span>
                <span class="n">pos_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
        <span class="n">positive_markers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gene_names</span><span class="p">[</span><span class="n">pos_indices</span><span class="p">])</span>
        <span class="c1"># Negative markers are the lowest-ranked</span>
        <span class="n">negative_markers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gene_names</span><span class="p">[</span><span class="o">-</span><span class="n">n_neg</span><span class="p">:])</span>
        <span class="n">markers</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;positive&quot;</span><span class="p">:</span> <span class="n">positive_markers</span><span class="p">,</span>
            <span class="s2">&quot;negative&quot;</span><span class="p">:</span> <span class="n">negative_markers</span>
        <span class="p">}</span>
    <span class="k">return</span> <span class="n">markers</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.find_mutually_exclusive_genes" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">find_mutually_exclusive_genes</span>


<a href="#src.segger.data._utils.find_mutually_exclusive_genes" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">find_mutually_exclusive_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">markers</span><span class="p">,</span> <span class="n">cell_type_column</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Identify mutually exclusive genes based on expression criteria.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>adata</code></td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Annotated data object containing gene expression data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>markers</code></td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Dict">Dict</span>[str, <span title="typing.List">List</span>[str]]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary where keys are cell types and values are dictionaries containing:
'positive': list of top x% highly expressed genes
'negative': list of top x% lowly expressed genes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>cell_type_column</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column name in <code>adata.obs</code> that specifies cell types.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>list</code></td>            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[str, str]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of mutually exclusive gene pairs.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_mutually_exclusive_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">,</span> <span class="n">markers</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]],</span> <span class="n">cell_type_column</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Identify mutually exclusive genes based on expression criteria.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata: Annotated data object containing gene expression data.</span>
<span class="sd">        markers: Dictionary where keys are cell types and values are dictionaries containing:</span>
<span class="sd">            &#39;positive&#39;: list of top x% highly expressed genes</span>
<span class="sd">            &#39;negative&#39;: list of top x% lowly expressed genes.</span>
<span class="sd">        cell_type_column: Column name in `adata.obs` that specifies cell types.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: List of mutually exclusive gene pairs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exclusive_genes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">all_exclusive</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_expression</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">marker_sets</span> <span class="ow">in</span> <span class="n">markers</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">positive_markers</span> <span class="o">=</span> <span class="n">marker_sets</span><span class="p">[</span><span class="s2">&quot;positive&quot;</span><span class="p">]</span>
        <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">positive_markers</span><span class="p">:</span>
            <span class="n">gene_expr</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">X</span>
            <span class="n">cell_type_mask</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">cell_type_column</span><span class="p">]</span> <span class="o">==</span> <span class="n">cell_type</span>
            <span class="n">non_cell_type_mask</span> <span class="o">=</span> <span class="o">~</span><span class="n">cell_type_mask</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">gene_expr</span><span class="p">[</span><span class="n">cell_type_mask</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mf">0.2</span> <span class="ow">and</span> <span class="p">(</span>
                <span class="n">gene_expr</span><span class="p">[</span><span class="n">non_cell_type_mask</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">:</span>
                <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">cell_type</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
                <span class="n">all_exclusive</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">unique_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="n">gene</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">all_exclusive</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="n">filtered_exclusive_genes</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">i</span><span class="p">:</span> <span class="p">[</span><span class="n">gene</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">unique_genes</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">exclusive_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="p">}</span>
    <span class="n">mutually_exclusive_gene_pairs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">((</span><span class="n">gene1</span><span class="p">,</span> <span class="n">gene2</span><span class="p">)))</span>
        <span class="k">for</span> <span class="n">key1</span><span class="p">,</span> <span class="n">key2</span> <span class="ow">in</span> <span class="n">combinations</span><span class="p">(</span><span class="n">filtered_exclusive_genes</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">key1</span> <span class="o">!=</span> <span class="n">key2</span>
        <span class="k">for</span> <span class="n">gene1</span> <span class="ow">in</span> <span class="n">filtered_exclusive_genes</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">gene2</span> <span class="ow">in</span> <span class="n">filtered_exclusive_genes</span><span class="p">[</span><span class="n">key2</span><span class="p">]</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="nb">set</span><span class="p">(</span><span class="n">mutually_exclusive_gene_pairs</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.format_time" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">format_time</span>


<a href="#src.segger.data._utils.format_time" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">format_time</span><span class="p">(</span><span class="n">elapsed</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Format elapsed time to h<img alt="" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.0.3/assets/svg/24c2.svg" title=":m:" />s.</p>
<h5 id="src.segger.data._utils.format_time--parameters">Parameters:<a class="headerlink" href="#src.segger.data._utils.format_time--parameters" title="Permanent link">&para;</a></h5>
<p>elapsed : float
    Elapsed time in seconds.</p>
<h5 id="src.segger.data._utils.format_time--returns">Returns:<a class="headerlink" href="#src.segger.data._utils.format_time--returns" title="Permanent link">&para;</a></h5>
<p>str
    Formatted time in h<img alt="" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.0.3/assets/svg/24c2.svg" title=":m:" />s.</p>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">format_time</span><span class="p">(</span><span class="n">elapsed</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Format elapsed time to h:m:s.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    ----------</span>
<span class="sd">    elapsed : float</span>
<span class="sd">        Elapsed time in seconds.</span>

<span class="sd">    Returns:</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>
<span class="sd">        Formatted time in h:m:s.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">timedelta</span><span class="p">(</span><span class="n">seconds</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">elapsed</span><span class="p">)))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.get_edge_index" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_edge_index</span>


<a href="#src.segger.data._utils.get_edge_index" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_edge_index</span><span class="p">(</span><span class="n">coords_1</span><span class="p">,</span> <span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;kd_tree&#39;</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Computes edge indices using KD-Tree.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>coords_1</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>First set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>coords_2</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Second set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distance threshold.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>method</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The method to use. Only 'kd_tree' is supported now.</p>
              </div>
            </td>
            <td>
                  <code>&#39;kd_tree&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="torch.Tensor">Tensor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: Edge indices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_edge_index</span><span class="p">(</span>
    <span class="n">coords_1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">coords_2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">method</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;kd_tree&quot;</span><span class="p">,</span>
    <span class="n">workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Computes edge indices using KD-Tree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        coords_1 (np.ndarray): First set of coordinates.</span>
<span class="sd">        coords_2 (np.ndarray): Second set of coordinates.</span>
<span class="sd">        k (int, optional): Number of nearest neighbors.</span>
<span class="sd">        dist (int, optional): Distance threshold.</span>
<span class="sd">        method (str, optional): The method to use. Only &#39;kd_tree&#39; is supported now.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: Edge indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;kd_tree&quot;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">get_edge_index_kdtree</span><span class="p">(</span>
            <span class="n">coords_1</span><span class="p">,</span> <span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="n">workers</span>
        <span class="p">)</span>
    <span class="c1"># elif method == &quot;cuda&quot;:</span>
    <span class="c1">#     return get_edge_index_cuda(coords_1, coords_2, k=k, dist=dist)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Unknown method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s2">. The only supported method is &#39;kd_tree&#39; now.&quot;</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.get_edge_index_kdtree" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_edge_index_kdtree</span>


<a href="#src.segger.data._utils.get_edge_index_kdtree" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_edge_index_kdtree</span><span class="p">(</span><span class="n">coords_1</span><span class="p">,</span> <span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">dist</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Computes edge indices using KDTree.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>coords_1</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>First set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>coords_2</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Second set of coordinates.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distance threshold.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="torch.Tensor">Tensor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: Edge indices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_edge_index_kdtree</span><span class="p">(</span>
    <span class="n">coords_1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">coords_2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">dist</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">workers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Computes edge indices using KDTree.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        coords_1 (np.ndarray): First set of coordinates.</span>
<span class="sd">        coords_2 (np.ndarray): Second set of coordinates.</span>
<span class="sd">        k (int, optional): Number of nearest neighbors.</span>
<span class="sd">        dist (int, optional): Distance threshold.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: Edge indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">coords_1</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
        <span class="n">coords_1</span> <span class="o">=</span> <span class="n">coords_1</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">coords_2</span><span class="p">,</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">):</span>
        <span class="n">coords_2</span> <span class="o">=</span> <span class="n">coords_2</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">cKDTree</span><span class="p">(</span><span class="n">coords_1</span><span class="p">)</span>
    <span class="n">d_kdtree</span><span class="p">,</span> <span class="n">idx_out</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
        <span class="n">coords_2</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">,</span> <span class="n">distance_upper_bound</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span> <span class="n">workers</span><span class="o">=</span><span class="n">workers</span>
    <span class="p">)</span>
    <span class="n">valid_mask</span> <span class="o">=</span> <span class="n">d_kdtree</span> <span class="o">&lt;</span> <span class="n">dist</span>
    <span class="n">edges</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">valid</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">valid_mask</span><span class="p">):</span>
        <span class="n">valid_indices</span> <span class="o">=</span> <span class="n">idx_out</span><span class="p">[</span><span class="n">idx</span><span class="p">][</span><span class="n">valid</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">valid_indices</span><span class="o">.</span><span class="n">size</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">full</span><span class="p">(</span><span class="n">valid_indices</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">idx</span><span class="p">),</span> <span class="n">valid_indices</span><span class="p">))</span><span class="o">.</span><span class="n">T</span>
            <span class="p">)</span>

    <span class="n">edge_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">(</span><span class="n">edges</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">)</span><span class="o">.</span><span class="n">contiguous</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">edge_index</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.get_polygons_from_xy" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_polygons_from_xy</span>


<a href="#src.segger.data._utils.get_polygons_from_xy" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_polygons_from_xy</span><span class="p">(</span><span class="n">boundaries</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Convert boundary coordinates from a DataFrame to a GeoSeries of polygons.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>boundaries</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A DataFrame containing the boundary data with x and y coordinates
and identifiers.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>x</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the x-coordinate.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>y</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the y-coordinate.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>label</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the cell or nucleus label.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>scale_factor</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A ratio to scale the polygons. A value of 1.0 means no change,
greater than 1.0 expands the polygons, and less than 1.0 shrinks the polygons.
Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="geopandas.GeoSeries">GeoSeries</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>gpd.GeoSeries: A GeoSeries containing the polygons created from the boundary
coordinates.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_polygons_from_xy</span><span class="p">(</span>
    <span class="n">boundaries</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">label</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">scale_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert boundary coordinates from a DataFrame to a GeoSeries of polygons.</span>

<span class="sd">    Args:</span>
<span class="sd">        boundaries: A DataFrame containing the boundary data with x and y coordinates</span>
<span class="sd">            and identifiers.</span>
<span class="sd">        x: The name of the column representing the x-coordinate.</span>
<span class="sd">        y: The name of the column representing the y-coordinate.</span>
<span class="sd">        label: The name of the column representing the cell or nucleus label.</span>
<span class="sd">        scale_factor: A ratio to scale the polygons. A value of 1.0 means no change,</span>
<span class="sd">            greater than 1.0 expands the polygons, and less than 1.0 shrinks the polygons.</span>
<span class="sd">            Defaults to 1.0.</span>

<span class="sd">    Returns:</span>
<span class="sd">        gpd.GeoSeries: A GeoSeries containing the polygons created from the boundary</span>
<span class="sd">            coordinates.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Polygon offsets in coords</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="n">boundaries</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">splits</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">ids</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="n">ids</span><span class="p">[</span><span class="mi">1</span><span class="p">:])[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">geometry_offset</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">splits</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ids</span><span class="p">)])</span>
    <span class="n">part_offset</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">ids</span><span class="p">))</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Convert to GeoSeries of polygons</span>
    <span class="n">polygons</span> <span class="o">=</span> <span class="n">shapely</span><span class="o">.</span><span class="n">from_ragged_array</span><span class="p">(</span>
        <span class="n">shapely</span><span class="o">.</span><span class="n">GeometryType</span><span class="o">.</span><span class="n">POLYGON</span><span class="p">,</span>
        <span class="n">coords</span><span class="o">=</span><span class="n">boundaries</span><span class="p">[[</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">order</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">),</span>
        <span class="n">offsets</span><span class="o">=</span><span class="p">(</span><span class="n">geometry_offset</span><span class="p">,</span> <span class="n">part_offset</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">gs</span> <span class="o">=</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="p">(</span><span class="n">polygons</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">ids</span><span class="p">))</span>

    <span class="c1"># print(gs)</span>

    <span class="k">if</span> <span class="n">scale_factor</span> <span class="o">!=</span> <span class="mf">1.0</span><span class="p">:</span>
        <span class="c1"># Scale polygons around their centroid</span>
        <span class="n">gs</span> <span class="o">=</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">scale</span><span class="p">(</span><span class="n">geom</span><span class="p">,</span> <span class="n">xfact</span><span class="o">=</span><span class="n">scale_factor</span><span class="p">,</span> <span class="n">yfact</span><span class="o">=</span><span class="n">scale_factor</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s1">&#39;centroid&#39;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">geom</span> <span class="ow">in</span> <span class="n">gs</span>
            <span class="p">],</span>
            <span class="n">index</span><span class="o">=</span><span class="n">gs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># print(gs)</span>

    <span class="k">return</span> <span class="n">gs</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.get_xy_extents" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_xy_extents</span>


<a href="#src.segger.data._utils.get_xy_extents" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_xy_extents</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Get the bounding box of the x and y coordinates from a Parquet file.</p>
<h5 id="src.segger.data._utils.get_xy_extents--parameters">Parameters<a class="headerlink" href="#src.segger.data._utils.get_xy_extents--parameters" title="Permanent link">&para;</a></h5>
<p>filepath : str
    The path to the Parquet file.
x : str
    The name of the column representing the x-coordinate.
y : str
    The name of the column representing the y-coordinate.</p>
<h5 id="src.segger.data._utils.get_xy_extents--returns">Returns<a class="headerlink" href="#src.segger.data._utils.get_xy_extents--returns" title="Permanent link">&para;</a></h5>
<p>shapely.Polygon
    A polygon representing the bounding box of the x and y coordinates.</p>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_xy_extents</span><span class="p">(</span>
    <span class="n">filepath</span><span class="p">,</span>
    <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the bounding box of the x and y coordinates from a Parquet file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    filepath : str</span>
<span class="sd">        The path to the Parquet file.</span>
<span class="sd">    x : str</span>
<span class="sd">        The name of the column representing the x-coordinate.</span>
<span class="sd">    y : str</span>
<span class="sd">        The name of the column representing the y-coordinate.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    shapely.Polygon</span>
<span class="sd">        A polygon representing the bounding box of the x and y coordinates.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get index of columns of parquet file</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="n">pq</span><span class="o">.</span><span class="n">read_metadata</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="n">schema_idx</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">reversed</span><span class="p">,</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metadata</span><span class="o">.</span><span class="n">schema</span><span class="o">.</span><span class="n">names</span><span class="p">)))</span>

    <span class="c1"># Find min and max values across all row groups</span>
    <span class="n">x_max</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">x_min</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">maxsize</span>
    <span class="n">y_max</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">y_min</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">maxsize</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">metadata</span><span class="o">.</span><span class="n">num_row_groups</span><span class="p">):</span>
        <span class="n">group</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">row_group</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">x_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">x_min</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">min</span><span class="p">)</span>
        <span class="n">x_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">x_max</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">max</span><span class="p">)</span>
        <span class="n">y_min</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">y_min</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">y</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">min</span><span class="p">)</span>
        <span class="n">y_max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">y_max</span><span class="p">,</span> <span class="n">group</span><span class="o">.</span><span class="n">column</span><span class="p">(</span><span class="n">schema_idx</span><span class="p">[</span><span class="n">y</span><span class="p">])</span><span class="o">.</span><span class="n">statistics</span><span class="o">.</span><span class="n">max</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">x_min</span><span class="p">,</span> <span class="n">y_min</span><span class="p">,</span> <span class="n">x_max</span><span class="p">,</span> <span class="n">y_max</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.load_settings" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">load_settings</span>


<a href="#src.segger.data._utils.load_settings" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">load_settings</span><span class="p">(</span><span class="n">sample_type</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Load a matching YAML file from the _settings/ directory and convert its
contents into a SimpleNamespace.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>sample_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the sample type to load (case-insensitive).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>SimpleNamespace</code></td>            <td>
                  <code><span title="types.SimpleNamespace">SimpleNamespace</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The settings loaded from the YAML file as a SimpleNamespace.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>FileNotFoundError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>sample_type</code> does not match any filenames.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_settings</span><span class="p">(</span><span class="n">sample_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SimpleNamespace</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load a matching YAML file from the _settings/ directory and convert its</span>
<span class="sd">    contents into a SimpleNamespace.</span>

<span class="sd">    Args:</span>
<span class="sd">        sample_type: Name of the sample type to load (case-insensitive).</span>

<span class="sd">    Returns:</span>
<span class="sd">        SimpleNamespace: The settings loaded from the YAML file as a SimpleNamespace.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If `sample_type` does not match any filenames.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">settings_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)</span><span class="o">.</span><span class="n">parent</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;_settings&quot;</span>
    <span class="c1"># Get a list of YAML filenames (without extensions) in the _settings dir</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">settings_dir</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s2">&quot;*.yaml&quot;</span><span class="p">)]</span>
    <span class="c1"># Convert sample_type to lowercase and check if it matches any filename</span>
    <span class="n">sample_type</span> <span class="o">=</span> <span class="n">sample_type</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">sample_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">filenames</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Sample type &#39;</span><span class="si">{</span><span class="n">sample_type</span><span class="si">}</span><span class="s2">&#39; not found in settings. &quot;</span>
            <span class="sa">f</span><span class="s2">&quot;Available options: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="c1"># Load the matching YAML file</span>
    <span class="n">yaml_file_path</span> <span class="o">=</span> <span class="n">settings_dir</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_type</span><span class="si">}</span><span class="s2">.yaml&quot;</span>
    <span class="k">with</span> <span class="n">yaml_file_path</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">safe_load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

    <span class="c1"># Convert the YAML data into a SimpleNamespace recursively</span>
    <span class="k">return</span> <span class="n">_dict_to_namespace</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="src.segger.data._utils.read_parquet_region" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">read_parquet_region</span>


<a href="#src.segger.data._utils.read_parquet_region" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">read_parquet_region</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">bounds</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">extra_columns</span><span class="o">=</span><span class="p">[],</span> <span class="n">extra_filters</span><span class="o">=</span><span class="p">[],</span> <span class="n">row_group_chunksize</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Read a region from a Parquet file based on x and y coordinates and optional
filters.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>filepath</code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to the Parquet file.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>x</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the x-coordinate.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>y</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the column representing the y-coordinate.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bounds</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A polygon representing the bounding box to filter the data. If None,
no bounding box filter is applied. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>extra_columns</code></td>
            <td>
                  <code>list[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of additional columns to include in the output DataFrame. Defaults to [].</p>
              </div>
            </td>
            <td>
                  <code>[]</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>extra_filters</code></td>
            <td>
                  <code>list[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of additional filters to apply to the data. Defaults to [].</p>
              </div>
            </td>
            <td>
                  <code>[]</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>row_group_chunksize</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[int]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Chunk size for row group processing. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the filtered data from the Parquet file.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/_utils.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read_parquet_region</span><span class="p">(</span>
    <span class="n">filepath</span><span class="p">,</span>
    <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">y</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">bounds</span><span class="p">:</span> <span class="n">shapely</span><span class="o">.</span><span class="n">Polygon</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">extra_columns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">extra_filters</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[],</span>
    <span class="n">row_group_chunksize</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read a region from a Parquet file based on x and y coordinates and optional</span>
<span class="sd">    filters.</span>

<span class="sd">    Args:</span>
<span class="sd">        filepath: The path to the Parquet file.</span>
<span class="sd">        x: The name of the column representing the x-coordinate.</span>
<span class="sd">        y: The name of the column representing the y-coordinate.</span>
<span class="sd">        bounds: A polygon representing the bounding box to filter the data. If None,</span>
<span class="sd">            no bounding box filter is applied. Defaults to None.</span>
<span class="sd">        extra_columns: A list of additional columns to include in the output DataFrame. Defaults to [].</span>
<span class="sd">        extra_filters: A list of additional filters to apply to the data. Defaults to [].</span>
<span class="sd">        row_group_chunksize: Chunk size for row group processing. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the filtered data from the Parquet file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check backend and load dependencies if not already loaded</span>

    <span class="c1"># Find bounds of full file if not supplied</span>
    <span class="k">if</span> <span class="n">bounds</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">bounds</span> <span class="o">=</span> <span class="n">get_xy_extents</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>

    <span class="c1"># Load pre-filtered data from Parquet file</span>
    <span class="n">filters</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">[</span>
            <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span> <span class="n">bounds</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
            <span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span> <span class="n">bounds</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
            <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="s2">&quot;&lt;&quot;</span><span class="p">,</span> <span class="n">bounds</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">2</span><span class="p">]),</span>
            <span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="s2">&quot;&lt;&quot;</span><span class="p">,</span> <span class="n">bounds</span><span class="o">.</span><span class="n">bounds</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span>
        <span class="p">]</span>
        <span class="o">+</span> <span class="n">extra_filters</span>
    <span class="p">]</span>

    <span class="n">columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">({</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">}</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="n">extra_columns</span><span class="p">))</span>

    <span class="c1"># Check if &#39;Geometry&#39;, &#39;geometry&#39;, &#39;polygon&#39;, or &#39;Polygon&#39; is in the columns</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">col</span> <span class="ow">in</span> <span class="n">columns</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Geometry&#39;</span><span class="p">,</span> <span class="s1">&#39;geometry&#39;</span><span class="p">,</span> <span class="s1">&#39;polygon&#39;</span><span class="p">,</span> <span class="s1">&#39;Polygon&#39;</span><span class="p">]):</span>
        <span class="kn">import</span> <span class="nn">geopandas</span> <span class="k">as</span> <span class="nn">gpd</span>
        <span class="c1"># If geometry columns are present, read with geopandas</span>
        <span class="n">region</span> <span class="o">=</span> <span class="n">gpd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span>
            <span class="n">filepath</span><span class="p">,</span>
            <span class="n">filters</span><span class="o">=</span><span class="n">filters</span><span class="p">,</span>
            <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Otherwise, read with pandas</span>
        <span class="n">region</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span>
            <span class="n">filepath</span><span class="p">,</span>
            <span class="n">filters</span><span class="o">=</span><span class="n">filters</span><span class="p">,</span>
            <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">region</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.sections", "navigation.instant", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate", "content.tabs.link", "navigation.tracking"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.af256bd8.min.js"></script>
      
        <script src="../../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/termynal@0.0.1/termynal.js"></script>
      
        <script src="../../../termynal.js"></script>
      
    
  </body>
</html>