
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Fast and accurate cell segmentation for single-molecule spatial omics">
      
      
        <meta name="author" content="Elyas Heidari">
      
      
        <link rel="canonical" href="https://EliHei2.github.io/segger_dev/api/data/sample/">
      
      
        <link rel="prev" href="../">
      
      
        <link rel="next" href="../pyg_dataset/">
      
      
      <link rel="icon" href="../../../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.33">
    
    
      
        <title>Sample - segger Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=ubuntu:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"ubuntu";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../css/extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css">
    
      <link rel="stylesheet" href="https://unpkg.com/termynal@0.0.1/termynal.css">
    
      <link rel="stylesheet" href="../../../termynal.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#seggerdatasample" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="segger Documentation" class="md-header__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            segger Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Sample
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="segger Documentation" class="md-nav__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    segger Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/data_creation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dataset Creation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../notebooks/segger_tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to Segger
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CLI
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" checked>
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Data
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Sample
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Sample
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#src.segger.data.sample" class="md-nav__link">
    <span class="md-ellipsis">
      sample
    </span>
  </a>
  
    <nav class="md-nav" aria-label="sample">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STInMemoryDataset" class="md-nav__link">
    <span class="md-ellipsis">
      STInMemoryDataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet" class="md-nav__link">
    <span class="md-ellipsis">
      STSampleParquet
    </span>
  </a>
  
    <nav class="md-nav" aria-label="STSampleParquet">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet.boundaries_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      boundaries_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet.extents" class="md-nav__link">
    <span class="md-ellipsis">
      extents
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet.n_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      n_transcripts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet.transcripts_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      transcripts_metadata
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet.save" class="md-nav__link">
    <span class="md-ellipsis">
      save
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet.save_debug" class="md-nav__link">
    <span class="md-ellipsis">
      save_debug
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STSampleParquet.set_transcript_embedding" class="md-nav__link">
    <span class="md-ellipsis">
      set_transcript_embedding
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile" class="md-nav__link">
    <span class="md-ellipsis">
      STTile
    </span>
  </a>
  
    <nav class="md-nav" aria-label="STTile">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.boundaries" class="md-nav__link">
    <span class="md-ellipsis">
      boundaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      transcripts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.uid" class="md-nav__link">
    <span class="md-ellipsis">
      uid
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.canonical_edges" class="md-nav__link">
    <span class="md-ellipsis">
      canonical_edges
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.get_boundary_props" class="md-nav__link">
    <span class="md-ellipsis">
      get_boundary_props
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.get_filtered_boundaries" class="md-nav__link">
    <span class="md-ellipsis">
      get_filtered_boundaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.get_filtered_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      get_filtered_transcripts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.get_kdtree_edge_index" class="md-nav__link">
    <span class="md-ellipsis">
      get_kdtree_edge_index
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.get_polygon_props" class="md-nav__link">
    <span class="md-ellipsis">
      get_polygon_props
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.get_transcript_props" class="md-nav__link">
    <span class="md-ellipsis">
      get_transcript_props
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#src.segger.data.sample.STTile.to_pyg_dataset" class="md-nav__link">
    <span class="md-ellipsis">
      to_pyg_dataset
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#usage-examples" class="md-nav__link">
    <span class="md-ellipsis">
      Usage Examples
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Usage Examples">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-data-loading" class="md-nav__link">
    <span class="md-ellipsis">
      Basic Data Loading
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#spatial-tiling-and-processing" class="md-nav__link">
    <span class="md-ellipsis">
      Spatial Tiling and Processing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#in-memory-dataset-processing" class="md-nav__link">
    <span class="md-ellipsis">
      In-Memory Dataset Processing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#individual-tile-processing" class="md-nav__link">
    <span class="md-ellipsis">
      Individual Tile Processing
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pyg_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyG Dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../transcript_embedding/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transcript Embedding
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ndtree/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    NDTree
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../models/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Models
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="seggerdatasample">segger.data.sample<a class="headerlink" href="#seggerdatasample" title="Permanent link">&para;</a></h1>
<p>The <code>sample</code> module is the core of the Segger data processing framework, providing comprehensive classes for handling spatial transcriptomics data. This module contains the main classes that orchestrate the entire data processing pipeline from raw data to machine learning-ready graphs.</p>


<div class="doc doc-object doc-module">



<h2 id="src.segger.data.sample" class="doc doc-heading">
            <span class="doc doc-object-name doc-module-name">sample</span>


<a href="#src.segger.data.sample" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="src.segger.data.sample.STInMemoryDataset" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">STInMemoryDataset</span>


<a href="#src.segger.data.sample.STInMemoryDataset" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">STInMemoryDataset</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">extents</span><span class="p">,</span> <span class="n">margin</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">


      <p>A class for handling in-memory representations of ST data.</p>
<p>This class is used to load and manage ST sample data from parquet files,
filter boundaries and transcripts, and provide spatial tiling for further
analysis. The class also pre-loads KDTrees for efficient spatial queries.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>sample</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="src.segger.data.sample.STSampleParquet" href="#src.segger.data.sample.STSampleParquet">STSampleParquet</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The ST sample containing paths to the data files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>extents</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The polygon defining the spatial extents for the dataset.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>margin</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The margin to buffer around the extents when filtering data. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>sample</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="src.segger.data.sample.STSampleParquet" href="#src.segger.data.sample.STSampleParquet">STSampleParquet</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The ST sample from which the data is loaded.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>extents</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The spatial extents of the dataset.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>margin</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The buffer margin around the extents for filtering.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>transcripts</code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The filtered transcripts within the dataset extents.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>boundaries</code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The filtered boundaries within the dataset extents.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>kdtree_tx</code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The KDTree for fast spatial queries on the transcripts.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the transcripts or boundaries could not be loaded or filtered.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

      <p>Initialize the STInMemoryDataset instance by loading transcripts
and boundaries from parquet files and pre-loading a KDTree for fast
spatial queries.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>sample</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="src.segger.data.sample.STSampleParquet" href="#src.segger.data.sample.STSampleParquet">STSampleParquet</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The ST sample containing paths to the data files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>extents</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The polygon defining the spatial extents for the dataset.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>margin</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The margin to buffer around the extents when filtering data. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
      </tbody>
    </table>

                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/sample.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">sample</span><span class="p">:</span> <span class="n">STSampleParquet</span><span class="p">,</span>
    <span class="n">extents</span><span class="p">:</span> <span class="n">shapely</span><span class="o">.</span><span class="n">Polygon</span><span class="p">,</span>
    <span class="n">margin</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialize the STInMemoryDataset instance by loading transcripts</span>
<span class="sd">    and boundaries from parquet files and pre-loading a KDTree for fast</span>
<span class="sd">    spatial queries.</span>

<span class="sd">    Args:</span>
<span class="sd">        sample: The ST sample containing paths to the data files.</span>
<span class="sd">        extents: The polygon defining the spatial extents for the dataset.</span>
<span class="sd">        margin: The margin to buffer around the extents when filtering data. Defaults to 10.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Set properties</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">extents</span> <span class="o">=</span> <span class="n">extents</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">margin</span> <span class="o">=</span> <span class="n">margin</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">settings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sample</span><span class="o">.</span><span class="n">settings</span>

    <span class="c1"># Load data from parquet files</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_load_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample</span><span class="o">.</span><span class="n">_transcripts_filepath</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_load_boundaries</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample</span><span class="o">.</span><span class="n">_boundaries_filepath</span><span class="p">)</span>

    <span class="c1"># Pre-load KDTrees</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">kdtree_tx</span> <span class="o">=</span> <span class="n">KDTree</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xy</span><span class="p">],</span> <span class="n">leafsize</span><span class="o">=</span><span class="mi">100</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">











  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="src.segger.data.sample.STSampleParquet" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">STSampleParquet</span>


<a href="#src.segger.data.sample.STSampleParquet" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">STSampleParquet</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="n">n_workers</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">scale_factor</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">sample_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">


      <p>A class to manage spatial transcriptomics data stored in parquet files.</p>
<p>This class provides methods for loading, processing, and saving data related
to ST samples. It supports parallel processing and efficient handling of
transcript and boundary data.</p>

      <p>Initialize the STSampleParquet instance.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>base_dir</code></td>
            <td>
                  <code><span title="os.PathLike">PathLike</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The base directory containing the ST data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>n_workers</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[int]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of workers for parallel processing. Defaults to 1.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>sample_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The sample type of the raw data, e.g., 'xenium' or 'merscope'. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>weights</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing weights for transcript embedding. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>scale_factor</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[float]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The scale factor to be used for expanding the boundary extents
during spatial queries. If not provided, the default from settings
will be used. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>FileNotFoundError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the base directory does not exist or the required files are
missing.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/sample.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">base_dir</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">,</span>
    <span class="n">n_workers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">scale_factor</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="n">sample_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">weights</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialize the STSampleParquet instance.</span>

<span class="sd">    Args:</span>
<span class="sd">        base_dir: The base directory containing the ST data.</span>
<span class="sd">        n_workers: The number of workers for parallel processing. Defaults to 1.</span>
<span class="sd">        sample_type: The sample type of the raw data, e.g., &#39;xenium&#39; or &#39;merscope&#39;. Defaults to None.</span>
<span class="sd">        weights: DataFrame containing weights for transcript embedding. Defaults to None.</span>
<span class="sd">        scale_factor: The scale factor to be used for expanding the boundary extents</span>
<span class="sd">            during spatial queries. If not provided, the default from settings</span>
<span class="sd">            will be used. Defaults to None.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the base directory does not exist or the required files are</span>
<span class="sd">            missing.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Setup paths and resource constraints</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_base_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">base_dir</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">settings</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">load_settings</span><span class="p">(</span><span class="n">sample_type</span><span class="p">)</span>
    <span class="n">transcripts_fn</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">filename</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_transcripts_filepath</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_base_dir</span> <span class="o">/</span> <span class="n">transcripts_fn</span>
    <span class="n">boundaries_fn</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">filename</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_boundaries_filepath</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_base_dir</span> <span class="o">/</span> <span class="n">boundaries_fn</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_workers</span> <span class="o">=</span> <span class="n">n_workers</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">scale_factor</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">nuclear_column</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="p">,</span> <span class="s2">&quot;nuclear_column&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">nuclear_column</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">scale_factor</span> <span class="o">!=</span> <span class="mf">1.0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;Boundary-transcript overlap information has not been pre-computed. It will be calculated during tile generation.&quot;</span>
        <span class="p">)</span>
    <span class="c1"># Set scale factor if provided</span>
    <span class="k">if</span> <span class="n">scale_factor</span> <span class="o">!=</span> <span class="mf">1.0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">scale_factor</span> <span class="o">=</span> <span class="n">scale_factor</span>

    <span class="c1"># Ensure transcript IDs exist</span>
    <span class="n">utils</span><span class="o">.</span><span class="n">ensure_transcript_ids</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_transcripts_filepath</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">x</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">y</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Setup logging</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">Logger</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;STSample@</span><span class="si">{</span><span class="n">base_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Internal caches</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_extents</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_transcripts_metadata</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_boundaries_metadata</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Setup default embedding for transcripts</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_emb_genes</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">weights</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_emb_genes</span> <span class="o">=</span> <span class="n">weights</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
    <span class="n">classes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcripts_metadata</span><span class="p">[</span><span class="s2">&quot;feature_names&quot;</span><span class="p">]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_transcript_embedding</span> <span class="o">=</span> <span class="n">TranscriptEmbedding</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">classes</span><span class="p">),</span> <span class="n">weights</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data.sample.STSampleParquet.boundaries_metadata" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">boundaries_metadata</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-cached"><code>cached</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data.sample.STSampleParquet.boundaries_metadata" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">boundaries_metadata</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Retrieve metadata for the boundaries stored in the sample.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Metadata dictionary for boundaries including column sizes.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>FileNotFoundError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the boundaries parquet file does not exist.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data.sample.STSampleParquet.extents" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">extents</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-cached"><code>cached</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data.sample.STSampleParquet.extents" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">extents</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Get the combined extents (bounding box) of the transcripts and boundaries.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>shapely.Polygon: The bounding box covering all transcripts and boundaries.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data.sample.STSampleParquet.n_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">n_transcripts</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data.sample.STSampleParquet.n_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">n_transcripts</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Get the total number of transcripts in the sample.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of transcripts.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data.sample.STSampleParquet.transcripts_metadata" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">transcripts_metadata</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-cached"><code>cached</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data.sample.STSampleParquet.transcripts_metadata" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">transcripts_metadata</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Retrieve metadata for the transcripts stored in the sample.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Metadata dictionary for transcripts including column sizes and</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>feature names.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>FileNotFoundError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the transcript parquet file does not exist.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STSampleParquet.save" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">save</span>


<a href="#src.segger.data.sample.STSampleParquet.save" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">save</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_bd</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span> <span class="n">k_tx</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span> <span class="n">k_tx_ex</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">dist_tx_ex</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">tile_size</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">tile_width</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">tile_height</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span> <span class="n">frac</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">mutually_exclusive_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Save the tiles of the sample as PyTorch geometric datasets.</p>
<p>See documentation for 'STTile' for more information on dataset contents.</p>
<p>Note: This function requires either 'tile_size' OR both 'tile_width' and
'tile_height' to be provided.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>data_dir</code></td>
            <td>
                  <code><span title="os.PathLike">PathLike</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the dataset should be saved.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_bd</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for boundary nodes. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_bd</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for boundary neighbors. Defaults to 15.0.</p>
              </div>
            </td>
            <td>
                  <code>15.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_tx</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for transcript nodes. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_tx</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for transcript neighbors. Defaults to 5.0.</p>
              </div>
            </td>
            <td>
                  <code>5.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>tile_size</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[int]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If provided, specifies the size of the tile. Overrides <code>tile_width</code>
and <code>tile_height</code>. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>tile_width</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[int]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Width of the tiles in pixels. Ignored if <code>tile_size</code> is provided. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>tile_height</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[int]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Height of the tiles in pixels. Ignored if <code>tile_size</code> is provided. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>neg_sampling_ratio</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ratio of negative samples. Defaults to 5.0.</p>
              </div>
            </td>
            <td>
                  <code>5.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>frac</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of the dataset to process. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>val_prob</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Proportion of data for use for validation split. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>test_prob</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Proportion of data for use for test split. Defaults to 0.2.</p>
              </div>
            </td>
            <td>
                  <code>0.2</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the 'frac' parameter is greater than 1.0 or if the calculated
number of tiles is zero.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code>AssertionError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the specified directory structure is not properly set up.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">save</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">data_dir</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">,</span>
    <span class="n">k_bd</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_bd</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">15.0</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_tx</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5.0</span><span class="p">,</span>
    <span class="n">k_tx_ex</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
    <span class="n">dist_tx_ex</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">tile_size</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">tile_width</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">tile_height</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">neg_sampling_ratio</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5.0</span><span class="p">,</span>
    <span class="n">frac</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="n">val_prob</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">test_prob</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">mutually_exclusive_genes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save the tiles of the sample as PyTorch geometric datasets.</span>

<span class="sd">    See documentation for &#39;STTile&#39; for more information on dataset contents.</span>

<span class="sd">    Note: This function requires either &#39;tile_size&#39; OR both &#39;tile_width&#39; and</span>
<span class="sd">    &#39;tile_height&#39; to be provided.</span>

<span class="sd">    Args:</span>
<span class="sd">        data_dir: The directory where the dataset should be saved.</span>
<span class="sd">        k_bd: Number of nearest neighbors for boundary nodes. Defaults to 3.</span>
<span class="sd">        dist_bd: Maximum distance for boundary neighbors. Defaults to 15.0.</span>
<span class="sd">        k_tx: Number of nearest neighbors for transcript nodes. Defaults to 3.</span>
<span class="sd">        dist_tx: Maximum distance for transcript neighbors. Defaults to 5.0.</span>
<span class="sd">        tile_size: If provided, specifies the size of the tile. Overrides `tile_width`</span>
<span class="sd">            and `tile_height`. Defaults to None.</span>
<span class="sd">        tile_width: Width of the tiles in pixels. Ignored if `tile_size` is provided. Defaults to None.</span>
<span class="sd">        tile_height: Height of the tiles in pixels. Ignored if `tile_size` is provided. Defaults to None.</span>
<span class="sd">        neg_sampling_ratio: Ratio of negative samples. Defaults to 5.0.</span>
<span class="sd">        frac: Fraction of the dataset to process. Defaults to 1.0.</span>
<span class="sd">        val_prob: Proportion of data for use for validation split. Defaults to 0.1.</span>
<span class="sd">        test_prob: Proportion of data for use for test split. Defaults to 0.2.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the &#39;frac&#39; parameter is greater than 1.0 or if the calculated</span>
<span class="sd">            number of tiles is zero.</span>
<span class="sd">        AssertionError: If the specified directory structure is not properly set up.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check inputs</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">frac</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Arg &#39;frac&#39; should be &lt;= 1.0, but got </span><span class="si">{</span><span class="n">frac</span><span class="si">}</span><span class="s2">.&quot;</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">tile_size</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">n_tiles</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_transcripts</span> <span class="o">/</span> <span class="n">tile_size</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_workers</span> <span class="o">*</span> <span class="n">frac</span>
            <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_tiles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Sampling parameters would yield 0 total tiles.&quot;</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="c1"># Propagate errors to logging</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">),</span> <span class="n">exc_info</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">raise</span> <span class="n">e</span>

    <span class="c1"># Setup directory structure to save tiles</span>
    <span class="n">data_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">data_dir</span><span class="p">)</span>
    <span class="n">STSampleParquet</span><span class="o">.</span><span class="n">_setup_directory</span><span class="p">(</span><span class="n">data_dir</span><span class="p">)</span>

    <span class="c1"># Function to parallelize over workers</span>
    <span class="k">def</span> <span class="nf">func</span><span class="p">(</span><span class="n">region</span><span class="p">):</span>
        <span class="n">xm</span> <span class="o">=</span> <span class="n">STInMemoryDataset</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span> <span class="n">extents</span><span class="o">=</span><span class="n">region</span><span class="p">)</span>
        <span class="n">tiles</span> <span class="o">=</span> <span class="n">xm</span><span class="o">.</span><span class="n">_tile</span><span class="p">(</span><span class="n">tile_width</span><span class="p">,</span> <span class="n">tile_height</span><span class="p">,</span> <span class="n">tile_size</span><span class="p">)</span>
        <span class="c1"># print(tiles)</span>
        <span class="k">if</span> <span class="n">frac</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">tiles</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">tiles</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tiles</span><span class="p">)</span> <span class="o">*</span> <span class="n">frac</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">tile</span> <span class="ow">in</span> <span class="n">tiles</span><span class="p">:</span>
            <span class="c1"># Choose training, test, or validation datasets</span>
            <span class="n">data_type</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
                <span class="n">a</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;train_tiles&quot;</span><span class="p">,</span> <span class="s2">&quot;test_tiles&quot;</span><span class="p">,</span> <span class="s2">&quot;val_tiles&quot;</span><span class="p">],</span>
                <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="n">test_prob</span> <span class="o">+</span> <span class="n">val_prob</span><span class="p">),</span> <span class="n">test_prob</span><span class="p">,</span> <span class="n">val_prob</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="n">xt</span> <span class="o">=</span> <span class="n">STTile</span><span class="p">(</span><span class="n">dataset</span><span class="o">=</span><span class="n">xm</span><span class="p">,</span> <span class="n">extents</span><span class="o">=</span><span class="n">tile</span><span class="p">)</span>
            <span class="n">pyg_data</span> <span class="o">=</span> <span class="n">xt</span><span class="o">.</span><span class="n">to_pyg_dataset</span><span class="p">(</span>
                <span class="n">k_bd</span><span class="o">=</span><span class="n">k_bd</span><span class="p">,</span>
                <span class="n">dist_bd</span><span class="o">=</span><span class="n">dist_bd</span><span class="p">,</span>
                <span class="n">k_tx</span><span class="o">=</span><span class="n">k_tx</span><span class="p">,</span>
                <span class="n">dist_tx</span><span class="o">=</span><span class="n">dist_tx</span><span class="p">,</span>
                <span class="n">k_tx_ex</span><span class="o">=</span><span class="n">k_tx_ex</span><span class="p">,</span>
                <span class="n">dist_tx_ex</span><span class="o">=</span><span class="n">dist_tx_ex</span><span class="p">,</span>
                <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="n">neg_sampling_ratio</span><span class="p">,</span>
                <span class="n">mutually_exclusive_genes</span> <span class="o">=</span> <span class="n">mutually_exclusive_genes</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">pyg_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;belongs&quot;</span><span class="p">,</span> <span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># this tile is only for testing</span>
                    <span class="n">data_type</span> <span class="o">=</span> <span class="s2">&quot;test_tiles&quot;</span>
                <span class="n">filepath</span> <span class="o">=</span> <span class="n">data_dir</span> <span class="o">/</span> <span class="n">data_type</span> <span class="o">/</span> <span class="s2">&quot;processed&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">xt</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">.pt&quot;</span>
                <span class="n">torch</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">pyg_data</span><span class="p">,</span> <span class="n">filepath</span><span class="p">)</span>

    <span class="c1"># TODO: Add Dask backend</span>
    <span class="n">regions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_balanced_regions</span><span class="p">()</span>
    <span class="n">outs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">:</span>
        <span class="n">outs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">func</span><span class="p">(</span><span class="n">region</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">outs</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STSampleParquet.save_debug" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">save_debug</span>


<a href="#src.segger.data.sample.STSampleParquet.save_debug" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">save_debug</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_bd</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span> <span class="n">k_tx</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span> <span class="n">k_tx_ex</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">dist_tx_ex</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">tile_width</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">tile_height</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span> <span class="n">frac</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Debug version of save method that processes tiles sequentially and prints
detailed information about the process.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>data_dir</code></td>
            <td>
                  <code><span title="os.PathLike">PathLike</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the dataset should be saved.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_bd</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for boundary nodes. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_bd</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for boundary neighbors. Defaults to 15.0.</p>
              </div>
            </td>
            <td>
                  <code>15.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_tx</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for transcript nodes. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_tx</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for transcript neighbors. Defaults to 5.0.</p>
              </div>
            </td>
            <td>
                  <code>5.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_tx_ex</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for transcript exclusion. Defaults to 100.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_tx_ex</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for transcript exclusion. Defaults to 20.</p>
              </div>
            </td>
            <td>
                  <code>20</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>tile_width</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[float]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Width of the tiles in pixels. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>tile_height</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[float]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Height of the tiles in pixels. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>neg_sampling_ratio</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ratio of negative samples. Defaults to 5.0.</p>
              </div>
            </td>
            <td>
                  <code>5.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>frac</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Fraction of the dataset to process. Defaults to 1.0.</p>
              </div>
            </td>
            <td>
                  <code>1.0</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>val_prob</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Proportion of data for use for validation split. Defaults to 0.1.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>test_prob</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Proportion of data for use for test split. Defaults to 0.2.</p>
              </div>
            </td>
            <td>
                  <code>0.2</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">save_debug</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">data_dir</span><span class="p">:</span> <span class="n">os</span><span class="o">.</span><span class="n">PathLike</span><span class="p">,</span>
    <span class="n">k_bd</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_bd</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">15.0</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_tx</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5.0</span><span class="p">,</span>
    <span class="n">k_tx_ex</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
    <span class="n">dist_tx_ex</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">tile_width</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">tile_height</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">neg_sampling_ratio</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5.0</span><span class="p">,</span>
    <span class="n">frac</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="n">val_prob</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">test_prob</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Debug version of save method that processes tiles sequentially and prints</span>
<span class="sd">    detailed information about the process.</span>

<span class="sd">    Args:</span>
<span class="sd">        data_dir: The directory where the dataset should be saved.</span>
<span class="sd">        k_bd: Number of nearest neighbors for boundary nodes. Defaults to 3.</span>
<span class="sd">        dist_bd: Maximum distance for boundary neighbors. Defaults to 15.0.</span>
<span class="sd">        k_tx: Number of nearest neighbors for transcript nodes. Defaults to 3.</span>
<span class="sd">        dist_tx: Maximum distance for transcript neighbors. Defaults to 5.0.</span>
<span class="sd">        k_tx_ex: Number of nearest neighbors for transcript exclusion. Defaults to 100.</span>
<span class="sd">        dist_tx_ex: Maximum distance for transcript exclusion. Defaults to 20.</span>
<span class="sd">        tile_width: Width of the tiles in pixels. Defaults to None.</span>
<span class="sd">        tile_height: Height of the tiles in pixels. Defaults to None.</span>
<span class="sd">        neg_sampling_ratio: Ratio of negative samples. Defaults to 5.0.</span>
<span class="sd">        frac: Fraction of the dataset to process. Defaults to 1.0.</span>
<span class="sd">        val_prob: Proportion of data for use for validation split. Defaults to 0.1.</span>
<span class="sd">        test_prob: Proportion of data for use for test split. Defaults to 0.2.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Starting Debug Tile Generation ===&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Parameters:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- k_bd: </span><span class="si">{</span><span class="n">k_bd</span><span class="si">}</span><span class="s2"> (boundary neighbors)&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- dist_bd: </span><span class="si">{</span><span class="n">dist_bd</span><span class="si">}</span><span class="s2"> (boundary distance)&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- k_tx: </span><span class="si">{</span><span class="n">k_tx</span><span class="si">}</span><span class="s2"> (transcript neighbors)&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- dist_tx: </span><span class="si">{</span><span class="n">dist_tx</span><span class="si">}</span><span class="s2"> (transcript distance)&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- tile_width: </span><span class="si">{</span><span class="n">tile_width</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- tile_height: </span><span class="si">{</span><span class="n">tile_height</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- frac: </span><span class="si">{</span><span class="n">frac</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- val_prob: </span><span class="si">{</span><span class="n">val_prob</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- test_prob: </span><span class="si">{</span><span class="n">test_prob</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Setup directory structure to save tiles</span>
    <span class="n">data_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">data_dir</span><span class="p">)</span>
    <span class="n">STSampleParquet</span><span class="o">.</span><span class="n">_setup_directory</span><span class="p">(</span><span class="n">data_dir</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Output directory: </span><span class="si">{</span><span class="n">data_dir</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Get regions to process</span>
    <span class="n">regions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_balanced_regions</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Total regions to process: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">regions</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Region bounds:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">region</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regions</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Region </span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">region</span><span class="o">.</span><span class="n">bounds</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Process each region sequentially</span>
    <span class="k">for</span> <span class="n">region_idx</span><span class="p">,</span> <span class="n">region</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regions</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Processing Region </span><span class="si">{</span><span class="n">region_idx</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">regions</span><span class="p">)</span><span class="si">}</span><span class="s2"> ===&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Region bounds: </span><span class="si">{</span><span class="n">region</span><span class="o">.</span><span class="n">bounds</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">xm</span> <span class="o">=</span> <span class="n">STInMemoryDataset</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="bp">self</span><span class="p">,</span> <span class="n">extents</span><span class="o">=</span><span class="n">region</span><span class="p">)</span>
        <span class="n">tiles</span> <span class="o">=</span> <span class="n">xm</span><span class="o">.</span><span class="n">_tile</span><span class="p">(</span><span class="n">tile_width</span><span class="p">,</span> <span class="n">tile_height</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generated </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">tiles</span><span class="p">)</span><span class="si">}</span><span class="s2"> tiles for this region&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">frac</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">tiles</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">tiles</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tiles</span><span class="p">)</span> <span class="o">*</span> <span class="n">frac</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;After sampling: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">tiles</span><span class="p">)</span><span class="si">}</span><span class="s2"> tiles&quot;</span><span class="p">)</span>

        <span class="c1"># Process each tile</span>
        <span class="k">for</span> <span class="n">tile_idx</span><span class="p">,</span> <span class="n">tile</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tiles</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">--- Processing Tile </span><span class="si">{</span><span class="n">tile_idx</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="mi">1</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">tiles</span><span class="p">)</span><span class="si">}</span><span class="s2"> ---&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tile bounds: </span><span class="si">{</span><span class="n">tile</span><span class="o">.</span><span class="n">bounds</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Choose training, test, or validation datasets</span>
            <span class="n">data_type</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span>
                <span class="n">a</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;train_tiles&quot;</span><span class="p">,</span> <span class="s2">&quot;test_tiles&quot;</span><span class="p">,</span> <span class="s2">&quot;val_tiles&quot;</span><span class="p">],</span>
                <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span> <span class="o">-</span> <span class="p">(</span><span class="n">test_prob</span> <span class="o">+</span> <span class="n">val_prob</span><span class="p">),</span> <span class="n">test_prob</span><span class="p">,</span> <span class="n">val_prob</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Assigned to: </span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">xt</span> <span class="o">=</span> <span class="n">STTile</span><span class="p">(</span><span class="n">dataset</span><span class="o">=</span><span class="n">xm</span><span class="p">,</span> <span class="n">extents</span><span class="o">=</span><span class="n">tile</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tile UID: </span><span class="si">{</span><span class="n">xt</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">pyg_data</span> <span class="o">=</span> <span class="n">xt</span><span class="o">.</span><span class="n">to_pyg_dataset</span><span class="p">(</span>
                <span class="n">k_bd</span><span class="o">=</span><span class="n">k_bd</span><span class="p">,</span>
                <span class="n">dist_bd</span><span class="o">=</span><span class="n">dist_bd</span><span class="p">,</span>
                <span class="n">k_tx</span><span class="o">=</span><span class="n">k_tx</span><span class="p">,</span>
                <span class="n">dist_tx</span><span class="o">=</span><span class="n">dist_tx</span><span class="p">,</span>
                <span class="n">k_tx_ex</span><span class="o">=</span><span class="n">k_tx_ex</span><span class="p">,</span>
                <span class="n">dist_tx_ex</span><span class="o">=</span><span class="n">dist_tx_ex</span><span class="p">,</span>
                <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="n">neg_sampling_ratio</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">pyg_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;belongs&quot;</span><span class="p">,</span> <span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">data_type</span> <span class="o">=</span> <span class="s2">&quot;test_tiles&quot;</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No tx-belongs-bd edges found, reassigning to test_tiles&quot;</span><span class="p">)</span>

                <span class="n">filepath</span> <span class="o">=</span> <span class="n">data_dir</span> <span class="o">/</span> <span class="n">data_type</span> <span class="o">/</span> <span class="s2">&quot;processed&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">xt</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">.pt&quot;</span>
                <span class="n">torch</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">pyg_data</span><span class="p">,</span> <span class="n">filepath</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved to: </span><span class="si">{</span><span class="n">filepath</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="c1"># Print some statistics about the generated data</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Data statistics:&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- Number of transcripts: </span><span class="si">{</span><span class="n">pyg_data</span><span class="p">[</span><span class="s1">&#39;tx&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">num_nodes</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- Number of boundaries: </span><span class="si">{</span><span class="n">pyg_data</span><span class="p">[</span><span class="s1">&#39;bd&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">num_nodes</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;- Number of tx-tx edges: </span><span class="si">{</span><span class="n">pyg_data</span><span class="p">[</span><span class="s1">&#39;tx&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;neighbors&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;tx&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;- Number of tx-bd edges: </span><span class="si">{</span><span class="n">pyg_data</span><span class="p">[</span><span class="s1">&#39;tx&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;neighbors&#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;bd&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="c1"># print(f&quot;- Number of tx-belongs-bd edges: {pyg_data[&#39;tx&#39;, &#39;belongs&#39;, &#39;bd&#39;].edge_index.shape[1]}&quot;)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Skipping tile - no valid data generated&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Debug Tile Generation Completed ===&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STSampleParquet.set_transcript_embedding" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">set_transcript_embedding</span>


<a href="#src.segger.data.sample.STSampleParquet.set_transcript_embedding" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">set_transcript_embedding</span><span class="p">(</span><span class="n">weights</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Set the transcript embedding for the sample.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>weights</code></td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A DataFrame containing the weights for each transcript.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If the provided weights do not match the number of transcript
features.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">set_transcript_embedding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">weights</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Set the transcript embedding for the sample.</span>

<span class="sd">    Args:</span>
<span class="sd">        weights: A DataFrame containing the weights for each transcript.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the provided weights do not match the number of transcript</span>
<span class="sd">            features.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">classes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_transcripts_metadata</span><span class="p">[</span><span class="s2">&quot;feature_names&quot;</span><span class="p">]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_transcript_embedding</span> <span class="o">=</span> <span class="n">TranscriptEmbedding</span><span class="p">(</span><span class="n">classes</span><span class="p">,</span> <span class="n">weights</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h3 id="src.segger.data.sample.STTile" class="doc doc-heading">
            <span class="doc doc-object-name doc-class-name">STTile</span>


<a href="#src.segger.data.sample.STTile" class="headerlink" title="Permanent link">&para;</a></h3>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">STTile</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">extents</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">


      <p>A class representing a tile of a ST sample.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>dataset</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="src.segger.data.sample.STInMemoryDataset" href="#src.segger.data.sample.STInMemoryDataset">STInMemoryDataset</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The ST dataset containing data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>extents</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The extents of the tile in the sample.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>boundaries</code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filtered boundaries within the tile extents.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>transcripts</code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filtered transcripts within the tile extents.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

      <p>Initialize a STTile instance.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>dataset</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="src.segger.data.sample.STInMemoryDataset" href="#src.segger.data.sample.STInMemoryDataset">STInMemoryDataset</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The ST dataset containing data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>extents</code></td>
            <td>
                  <code><span title="shapely.Polygon">Polygon</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The extents of the tile in the sample.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <p>The <code>boundaries</code> and <code>transcripts</code> attributes are cached to avoid the
overhead of filtering when tiles are instantiated. This is particularly
useful in multiprocessing settings where generating tiles in parallel
could lead to high overhead.</p>
</details>

<details class="internal-args" open>
  <summary>Internal Args</summary>
  <p>_boundaries: Cached DataFrame of filtered boundaries. Initially set to None.
_transcripts: Cached DataFrame of filtered transcripts. Initially set to None.</p>
</details>
                  <details class="quote">
                    <summary>Source code in <code>src/segger/data/sample.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">dataset</span><span class="p">:</span> <span class="n">STInMemoryDataset</span><span class="p">,</span>
    <span class="n">extents</span><span class="p">:</span> <span class="n">shapely</span><span class="o">.</span><span class="n">Polygon</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Initialize a STTile instance.</span>

<span class="sd">    Args:</span>
<span class="sd">        dataset: The ST dataset containing data.</span>
<span class="sd">        extents: The extents of the tile in the sample.</span>

<span class="sd">    Note:</span>
<span class="sd">        The `boundaries` and `transcripts` attributes are cached to avoid the</span>
<span class="sd">        overhead of filtering when tiles are instantiated. This is particularly</span>
<span class="sd">        useful in multiprocessing settings where generating tiles in parallel</span>
<span class="sd">        could lead to high overhead.</span>

<span class="sd">    Internal Args:</span>
<span class="sd">        _boundaries: Cached DataFrame of filtered boundaries. Initially set to None.</span>
<span class="sd">        _transcripts: Cached DataFrame of filtered transcripts. Initially set to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span> <span class="o">=</span> <span class="n">dataset</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">extents</span> <span class="o">=</span> <span class="n">extents</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">margin</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">margin</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">settings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">settings</span>

    <span class="c1"># Internal caches for filtered data</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_boundaries</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_transcripts</span> <span class="o">=</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">







<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data.sample.STTile.boundaries" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">boundaries</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-cached"><code>cached</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data.sample.STTile.boundaries" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">boundaries</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Return the filtered boundaries within the tile extents, cached for
efficiency.</p>
<p>The boundaries are computed only once and cached. If the boundaries
have not been computed yet, they are computed using
<code>get_filtered_boundaries()</code>.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the filtered boundaries within the tile
extents.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data.sample.STTile.transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">transcripts</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-cached"><code>cached</code></small>
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data.sample.STTile.transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">transcripts</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Return the filtered transcripts within the tile extents, cached for
efficiency.</p>
<p>The transcripts are computed only once and cached. If the transcripts
have not been computed yet, they are computed using
<code>get_filtered_transcripts()</code>.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the filtered transcripts within the tile
extents.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="src.segger.data.sample.STTile.uid" class="doc doc-heading">
            <span class="doc doc-object-name doc-attribute-name">uid</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-property"><code>property</code></small>
  </span>

<a href="#src.segger.data.sample.STTile.uid" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="n">uid</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Generate a unique identifier for the tile based on its extents.</p>
<p>This UID is particularly useful for saving or indexing tiles in distributed
processing environments.</p>
<p>The UID is constructed using the minimum and maximum x and y coordinates
of the tile's bounding box, representing its position and size in the
sample.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A unique identifier string in the format
'x=<x_min>_y=<y_min>_w=<width>_h=<height>' where:
- <code>&lt;x_min&gt;</code>: Minimum x-coordinate of the tile's extents.
- <code>&lt;y_min&gt;</code>: Minimum y-coordinate of the tile's extents.
- <code>&lt;width&gt;</code>: Width of the tile.
- <code>&lt;height&gt;</code>: Height of the tile.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="example" open>
  <summary>Example</summary>
  <p>If the tile's extents are bounded by (x_min, y_min) = (100, 200) and
(x_max, y_max) = (150, 250), the generated UID would be:
'x=100_y=200_w=50_h=50'</p>
</details>    </div>

</div>



<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.canonical_edges" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">canonical_edges</span>


<a href="#src.segger.data.sample.STTile.canonical_edges" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">canonical_edges</span><span class="p">(</span><span class="n">edge_index</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Sort edge indices to ensure canonical ordering.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>edge_index</code></td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The edge index tensor to sort.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: The sorted edge index tensor.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">canonical_edges</span><span class="p">(</span><span class="n">edge_index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sort edge indices to ensure canonical ordering.</span>

<span class="sd">    Args:</span>
<span class="sd">        edge_index: The edge index tensor to sort.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: The sorted edge index tensor.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">torch</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">edge_index</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.get_boundary_props" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_boundary_props</span>


<a href="#src.segger.data.sample.STTile.get_boundary_props" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_boundary_props</span><span class="p">(</span><span class="n">area</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">convexity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">elongation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">circularity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Compute geometric properties of boundary polygons.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>area</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the area of each boundary polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>convexity</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the convexity of each boundary polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>elongation</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the elongation of each boundary polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>circularity</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the circularity of each boundary polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="torch.Tensor">Tensor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: A tensor containing the computed properties for each boundary
polygon.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <p>The intention is for this function to simplify testing new strategies
for 'bd' node representations. You can just change the function body to
return another torch.Tensor without worrying about changes to the rest
of the code.</p>
</details>
            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_boundary_props</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">area</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">convexity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">elongation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">circularity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute geometric properties of boundary polygons.</span>

<span class="sd">    Args:</span>
<span class="sd">        area: If True, compute the area of each boundary polygon. Defaults to True.</span>
<span class="sd">        convexity: If True, compute the convexity of each boundary polygon. Defaults to True.</span>
<span class="sd">        elongation: If True, compute the elongation of each boundary polygon. Defaults to True.</span>
<span class="sd">        circularity: If True, compute the circularity of each boundary polygon. Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: A tensor containing the computed properties for each boundary</span>
<span class="sd">            polygon.</span>

<span class="sd">    Note:</span>
<span class="sd">        The intention is for this function to simplify testing new strategies</span>
<span class="sd">        for &#39;bd&#39; node representations. You can just change the function body to</span>
<span class="sd">        return another torch.Tensor without worrying about changes to the rest</span>
<span class="sd">        of the code.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get polygons from coordinates</span>
    <span class="c1"># Use getattr to check for the geometry column</span>
    <span class="n">geometry_column</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="p">,</span> <span class="s1">&#39;geometry&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">geometry_column</span> <span class="ow">and</span> <span class="n">geometry_column</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">polygons</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="p">[</span><span class="n">geometry_column</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">polygons</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="p">[</span><span class="s1">&#39;geometry&#39;</span><span class="p">]</span>  <span class="c1"># Assign None if the geometry column does not exist</span>
    <span class="c1"># Geometric properties of polygons</span>
    <span class="n">props</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_polygon_props</span><span class="p">(</span><span class="n">polygons</span><span class="p">)</span>
    <span class="n">props</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">as_tensor</span><span class="p">(</span><span class="n">props</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">float</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">props</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.get_filtered_boundaries" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_filtered_boundaries</span>


<a href="#src.segger.data.sample.STTile.get_filtered_boundaries" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_filtered_boundaries</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Filter the boundaries in the sample to include only those within
the specified tile extents.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the filtered boundaries within the tile
extents.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_filtered_boundaries</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter the boundaries in the sample to include only those within</span>
<span class="sd">    the specified tile extents.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the filtered boundaries within the tile</span>
<span class="sd">            extents.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filtered_boundaries</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">filter_boundaries</span><span class="p">(</span>
        <span class="n">boundaries</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">boundaries</span><span class="p">,</span>
        <span class="n">inset</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">extents</span><span class="p">,</span>
        <span class="n">outset</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">extents</span><span class="o">.</span><span class="n">buffer</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">margin</span><span class="p">,</span> <span class="n">join_style</span><span class="o">=</span><span class="s2">&quot;mitre&quot;</span><span class="p">),</span>
        <span class="n">x</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">x</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">y</span><span class="p">,</span>
        <span class="n">label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">filtered_boundaries</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.get_filtered_transcripts" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_filtered_transcripts</span>


<a href="#src.segger.data.sample.STTile.get_filtered_transcripts" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_filtered_transcripts</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Filter the transcripts in the sample to include only those within
the specified tile extents.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the filtered transcripts within the tile
extents.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_filtered_transcripts</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter the transcripts in the sample to include only those within</span>
<span class="sd">    the specified tile extents.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the filtered transcripts within the tile</span>
<span class="sd">            extents.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Buffer tile bounds to include transcripts around boundary</span>
    <span class="n">outset</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">extents</span><span class="o">.</span><span class="n">buffer</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">margin</span><span class="p">,</span> <span class="n">join_style</span><span class="o">=</span><span class="s2">&quot;mitre&quot;</span><span class="p">)</span>
    <span class="n">xmin</span><span class="p">,</span> <span class="n">ymin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">=</span> <span class="n">outset</span><span class="o">.</span><span class="n">bounds</span>

    <span class="c1"># Get transcripts inside buffered region</span>
    <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xy</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">)</span>
    <span class="n">mask</span> <span class="o">&amp;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="n">y</span><span class="p">]</span><span class="o">.</span><span class="n">between</span><span class="p">(</span><span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span><span class="p">)</span>
    <span class="n">filtered_transcripts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">filtered_transcripts</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.get_kdtree_edge_index" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_kdtree_edge_index</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#src.segger.data.sample.STTile.get_kdtree_edge_index" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_kdtree_edge_index</span><span class="p">(</span><span class="n">index_coords</span><span class="p">,</span> <span class="n">query_coords</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">max_distance</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Compute the k-nearest neighbor edge indices using a KDTree.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>index_coords</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An array of shape (n_samples, n_features) representing the
coordinates of the points to be indexed.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>query_coords</code></td>
            <td>
                  <code><span title="numpy.ndarray">ndarray</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An array of shape (m_samples, n_features) representing the
coordinates of the query points.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of nearest neighbors to find for each query point.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_distance</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The maximum distance to consider for neighbors.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="torch.Tensor">Tensor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: An array of shape (2, n_edges) containing the edge indices. Each
column represents an edge between two points, where the first row
contains the source indices and the second row contains the target
indices.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span>
<span class="normal">973</span>
<span class="normal">974</span>
<span class="normal">975</span>
<span class="normal">976</span>
<span class="normal">977</span>
<span class="normal">978</span>
<span class="normal">979</span>
<span class="normal">980</span>
<span class="normal">981</span>
<span class="normal">982</span>
<span class="normal">983</span>
<span class="normal">984</span>
<span class="normal">985</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">get_kdtree_edge_index</span><span class="p">(</span>
    <span class="n">index_coords</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">query_coords</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
    <span class="n">k</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">max_distance</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the k-nearest neighbor edge indices using a KDTree.</span>

<span class="sd">    Args:</span>
<span class="sd">        index_coords: An array of shape (n_samples, n_features) representing the</span>
<span class="sd">            coordinates of the points to be indexed.</span>
<span class="sd">        query_coords: An array of shape (m_samples, n_features) representing the</span>
<span class="sd">            coordinates of the query points.</span>
<span class="sd">        k: The number of nearest neighbors to find for each query point.</span>
<span class="sd">        max_distance: The maximum distance to consider for neighbors.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: An array of shape (2, n_edges) containing the edge indices. Each</span>
<span class="sd">            column represents an edge between two points, where the first row</span>
<span class="sd">            contains the source indices and the second row contains the target</span>
<span class="sd">            indices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># KDTree search</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">KDTree</span><span class="p">(</span><span class="n">index_coords</span><span class="p">)</span>
    <span class="n">dist</span><span class="p">,</span> <span class="n">idx</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query_coords</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">max_distance</span><span class="p">)</span>

    <span class="c1"># To sparse adjacency</span>
    <span class="n">edge_index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argwhere</span><span class="p">(</span><span class="n">dist</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">edge_index</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">idx</span><span class="p">[</span><span class="n">dist</span> <span class="o">!=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">]</span>
    <span class="n">edge_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">edge_index</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">)</span><span class="o">.</span><span class="n">contiguous</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">edge_index</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.get_polygon_props" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_polygon_props</span>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#src.segger.data.sample.STTile.get_polygon_props" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_polygon_props</span><span class="p">(</span><span class="n">polygons</span><span class="p">,</span> <span class="n">area</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">convexity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">elongation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">circularity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Compute geometric properties of polygons.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>polygons</code></td>
            <td>
                  <code><span title="geopandas.GeoSeries">GeoSeries</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A GeoSeries containing polygon geometries.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>area</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the area of each polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>convexity</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the convexity of each polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>elongation</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the elongation of each polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>circularity</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute the circularity of each polygon. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing the computed properties for each polygon.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">get_polygon_props</span><span class="p">(</span>
    <span class="n">polygons</span><span class="p">:</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="p">,</span>
    <span class="n">area</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">convexity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">elongation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">circularity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute geometric properties of polygons.</span>

<span class="sd">    Args:</span>
<span class="sd">        polygons: A GeoSeries containing polygon geometries.</span>
<span class="sd">        area: If True, compute the area of each polygon. Defaults to True.</span>
<span class="sd">        convexity: If True, compute the convexity of each polygon. Defaults to True.</span>
<span class="sd">        elongation: If True, compute the elongation of each polygon. Defaults to True.</span>
<span class="sd">        circularity: If True, compute the circularity of each polygon. Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing the computed properties for each polygon.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">props</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">polygons</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">area</span><span class="p">:</span>
        <span class="n">props</span><span class="p">[</span><span class="s2">&quot;area&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">area</span>
    <span class="k">if</span> <span class="n">convexity</span><span class="p">:</span>
        <span class="n">props</span><span class="p">[</span><span class="s2">&quot;convexity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">convex_hull</span><span class="o">.</span><span class="n">area</span> <span class="o">/</span> <span class="n">polygons</span><span class="o">.</span><span class="n">area</span>
    <span class="k">if</span> <span class="n">elongation</span><span class="p">:</span>
        <span class="n">rects</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">minimum_rotated_rectangle</span><span class="p">()</span>
        <span class="n">props</span><span class="p">[</span><span class="s2">&quot;elongation&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rects</span><span class="o">.</span><span class="n">area</span> <span class="o">/</span> <span class="n">polygons</span><span class="o">.</span><span class="n">envelope</span><span class="o">.</span><span class="n">area</span>
    <span class="k">if</span> <span class="n">circularity</span><span class="p">:</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">minimum_bounding_radius</span><span class="p">()</span>
        <span class="n">props</span><span class="p">[</span><span class="s2">&quot;circularity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">area</span> <span class="o">/</span> <span class="n">r</span><span class="o">**</span><span class="mi">2</span>

    <span class="k">return</span> <span class="n">props</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.get_transcript_props" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">get_transcript_props</span>


<a href="#src.segger.data.sample.STTile.get_transcript_props" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">get_transcript_props</span><span class="p">()</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Encode transcript features in a sparse format.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="torch.Tensor">Tensor</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>torch.Tensor: A sparse tensor containing the encoded transcript features.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <p>The intention is for this function to simplify testing new strategies
for 'tx' node representations. For example, the encoder can be any type
of encoder that transforms the transcript labels into a numerical
matrix (in sparse format).</p>
</details>
            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_transcript_props</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Encode transcript features in a sparse format.</span>

<span class="sd">    Returns:</span>
<span class="sd">        torch.Tensor: A sparse tensor containing the encoded transcript features.</span>

<span class="sd">    Note:</span>
<span class="sd">        The intention is for this function to simplify testing new strategies</span>
<span class="sd">        for &#39;tx&#39; node representations. For example, the encoder can be any type</span>
<span class="sd">        of encoder that transforms the transcript labels into a numerical</span>
<span class="sd">        matrix (in sparse format).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Encode transcript features in sparse format</span>
    <span class="n">embedding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dataset</span><span class="o">.</span><span class="n">sample</span><span class="o">.</span><span class="n">_transcript_embedding</span>
    <span class="n">label</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">label</span>
    <span class="n">props</span> <span class="o">=</span> <span class="n">embedding</span><span class="o">.</span><span class="n">embed</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="n">label</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">props</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="src.segger.data.sample.STTile.to_pyg_dataset" class="doc doc-heading">
            <span class="doc doc-object-name doc-function-name">to_pyg_dataset</span>


<a href="#src.segger.data.sample.STTile.to_pyg_dataset" class="headerlink" title="Permanent link">&para;</a></h4>
<div class="doc-signature highlight"><pre><span></span><code><span class="nf">to_pyg_dataset</span><span class="p">(</span><span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_bd</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">k_tx</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_tx</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">k_tx_ex</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">dist_tx_ex</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">area</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">convexity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">elongation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">circularity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mutually_exclusive_genes</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</code></pre></div>

    <div class="doc doc-contents ">

      <p>Convert the sample data to a PyG HeteroData object.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>neg_sampling_ratio</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ratio of negative samples. Defaults to 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_bd</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for boundary nodes. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_bd</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for boundary neighbors. Defaults to 15.</p>
              </div>
            </td>
            <td>
                  <code>15</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_tx</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for transcript nodes. Defaults to 3.</p>
              </div>
            </td>
            <td>
                  <code>3</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_tx</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for transcript neighbors. Defaults to 5.</p>
              </div>
            </td>
            <td>
                  <code>5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k_tx_ex</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of nearest neighbors for transcript exclusion. Defaults to 100.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>dist_tx_ex</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum distance for transcript exclusion. Defaults to 20.</p>
              </div>
            </td>
            <td>
                  <code>20</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>area</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute area of boundary polygons. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>convexity</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute convexity of boundary polygons. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>elongation</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute elongation of boundary polygons. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>circularity</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, compute circularity of boundary polygons. Defaults to True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mutually_exclusive_genes</code></td>
            <td>
                  <code><span title="typing.Optional">Optional</span>[<span title="typing.List">List</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of mutually exclusive gene pairs. Defaults to None.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>HeteroData</code></td>            <td>
                  <code><span title="torch_geometric.data.HeteroData">HeteroData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A PyTorch Geometric HeteroData object containing the sample data.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/segger/data/sample.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">to_pyg_dataset</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="c1"># train: bool,</span>
    <span class="n">neg_sampling_ratio</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">k_bd</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_bd</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">15</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_tx</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">k_tx_ex</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
    <span class="n">dist_tx_ex</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">area</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">convexity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">elongation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">circularity</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">mutually_exclusive_genes</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">HeteroData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert the sample data to a PyG HeteroData object.</span>

<span class="sd">    Args:</span>
<span class="sd">        neg_sampling_ratio: Ratio of negative samples. Defaults to 10.</span>
<span class="sd">        k_bd: Number of nearest neighbors for boundary nodes. Defaults to 3.</span>
<span class="sd">        dist_bd: Maximum distance for boundary neighbors. Defaults to 15.</span>
<span class="sd">        k_tx: Number of nearest neighbors for transcript nodes. Defaults to 3.</span>
<span class="sd">        dist_tx: Maximum distance for transcript neighbors. Defaults to 5.</span>
<span class="sd">        k_tx_ex: Number of nearest neighbors for transcript exclusion. Defaults to 100.</span>
<span class="sd">        dist_tx_ex: Maximum distance for transcript exclusion. Defaults to 20.</span>
<span class="sd">        area: If True, compute area of boundary polygons. Defaults to True.</span>
<span class="sd">        convexity: If True, compute convexity of boundary polygons. Defaults to True.</span>
<span class="sd">        elongation: If True, compute elongation of boundary polygons. Defaults to True.</span>
<span class="sd">        circularity: If True, compute circularity of boundary polygons. Defaults to True.</span>
<span class="sd">        mutually_exclusive_genes: List of mutually exclusive gene pairs. Defaults to None.</span>

<span class="sd">    Returns:</span>
<span class="sd">        HeteroData: A PyTorch Geometric HeteroData object containing the sample data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initialize an empty HeteroData object</span>
    <span class="n">pyg_data</span> <span class="o">=</span> <span class="n">HeteroData</span><span class="p">()</span>

    <span class="c1"># Set up Transcript nodes</span>
    <span class="c1"># Get transcript IDs - use getattr to safely check for id attribute</span>
    <span class="n">transcript_id_column</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">transcript_id_column</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Transcript IDs not found in DataFrame. Please run add_transcript_ids() &quot;</span>
            <span class="s2">&quot;as a preprocessing step before creating the dataset.&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Assign IDs to PyG data</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="n">transcript_id_column</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span>
    <span class="p">)</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xyz</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
        <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float32</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_transcript_props</span><span class="p">()</span>



    <span class="c1"># Set up Transcript-Transcript neighbor edges</span>
    <span class="n">nbrs_edge_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_kdtree_edge_index</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xyz</span><span class="p">],</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xyz</span><span class="p">],</span>
        <span class="n">k</span><span class="o">=</span><span class="n">k_tx</span><span class="p">,</span>
        <span class="n">max_distance</span><span class="o">=</span><span class="n">dist_tx</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># If there are no tx-neighbors-tx edges, skip saving tile</span>
    <span class="k">if</span> <span class="n">nbrs_edge_idx</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;neighbors&quot;</span><span class="p">,</span> <span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span> <span class="o">=</span> <span class="n">nbrs_edge_idx</span>


    <span class="k">if</span> <span class="n">mutually_exclusive_genes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Get potential repulsive edges (k-nearest neighbors within distance)</span>
        <span class="c1"># --- Step 1: Get repulsive edges (mutually exclusive genes) ---</span>
        <span class="n">repels_edge_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_kdtree_edge_index</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xyz</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xyz</span><span class="p">],</span>
            <span class="n">k</span><span class="o">=</span><span class="n">k_tx_ex</span><span class="p">,</span>
            <span class="n">max_distance</span><span class="o">=</span><span class="n">dist_tx_ex</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">gene_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="c1"># Filter repels_edge_idx to only keep mutually exclusive gene pairs</span>
        <span class="n">src_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">repels_edge_idx</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
        <span class="n">dst_genes</span> <span class="o">=</span> <span class="p">[</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">repels_edge_idx</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span>
        <span class="n">mask</span> <span class="o">=</span> <span class="p">[</span>
            <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">((</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)))</span> <span class="ow">in</span> <span class="n">mutually_exclusive_genes</span> <span class="k">if</span> <span class="n">a</span> <span class="o">!=</span> <span class="n">b</span> <span class="k">else</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">src_genes</span><span class="p">,</span> <span class="n">dst_genes</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="n">repels_edge_idx</span> <span class="o">=</span> <span class="n">repels_edge_idx</span><span class="p">[:,</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">mask</span><span class="p">)]</span>

        <span class="c1"># --- Step 2: Get attractive edges (same gene, at least one node in repels) ---</span>
        <span class="c1"># Nodes involved in repels (for filtering nbrs_edge_idx)</span>
        <span class="n">repels_nodes</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">([</span><span class="n">repels_edge_idx</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">repels_edge_idx</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

        <span class="c1"># Filter nbrs_edge_idx: keep edges where (1) same gene AND (2) at least one node in repels</span>
        <span class="n">attractive_mask</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">nbrs_edge_idx</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">bool</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">src</span><span class="p">,</span> <span class="n">dst</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">nbrs_edge_idx</span><span class="o">.</span><span class="n">t</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()):</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">src</span> <span class="o">!=</span> <span class="n">dst</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">gene_ids</span><span class="p">[</span><span class="n">src</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene_ids</span><span class="p">[</span><span class="n">dst</span><span class="p">])</span> <span class="ow">and</span> <span class="p">(</span><span class="n">src</span> <span class="ow">in</span> <span class="n">repels_nodes</span> <span class="ow">or</span> <span class="n">dst</span> <span class="ow">in</span> <span class="n">repels_nodes</span><span class="p">):</span>
                <span class="n">attractive_mask</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">attractive_edge_idx</span> <span class="o">=</span> <span class="n">nbrs_edge_idx</span><span class="p">[:,</span> <span class="n">attractive_mask</span><span class="p">]</span>

        <span class="c1"># --- Step 3: Combine repels (label=0) and attractive (label=1) edges ---</span>
        <span class="n">edge_label_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">([</span><span class="n">repels_edge_idx</span><span class="p">,</span> <span class="n">attractive_edge_idx</span><span class="p">],</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">edge_label</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">([</span>
            <span class="n">torch</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">repels_edge_idx</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">),</span>  <span class="c1"># 0 for repels</span>
            <span class="n">torch</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">attractive_edge_idx</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span><span class="p">)</span>  <span class="c1"># 1 for attracts</span>
        <span class="p">])</span>

        <span class="c1"># --- Step 4: Store in PyG data object ---</span>
        <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;attracts&quot;</span><span class="p">,</span> <span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_label_index</span> <span class="o">=</span> <span class="n">edge_label_index</span>
        <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;attracts&quot;</span><span class="p">,</span> <span class="s2">&quot;tx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_label</span> <span class="o">=</span> <span class="n">edge_label</span>


    <span class="c1"># Set up Boundary nodes</span>
    <span class="c1"># Check if boundaries have geometries</span>
    <span class="n">geometry_column</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="p">,</span> <span class="s1">&#39;geometry&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">geometry_column</span> <span class="ow">and</span> <span class="n">geometry_column</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">polygons</span> <span class="o">=</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoSeries</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="p">[</span><span class="n">geometry_column</span><span class="p">],</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Fallback: compute polygons</span>
        <span class="n">polygons</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">get_polygons_from_xy</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="p">,</span>
            <span class="n">x</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">x</span><span class="p">,</span>
            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">y</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">label</span><span class="p">,</span>
            <span class="n">scale_factor</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">scale_factor</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Ensure self.boundaries is a GeoDataFrame with correct geometry</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span> <span class="o">=</span> <span class="n">gpd</span><span class="o">.</span><span class="n">GeoDataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span> <span class="n">geometry</span><span class="o">=</span><span class="n">polygons</span><span class="p">)</span>
    <span class="n">centroids</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">centroid</span><span class="o">.</span><span class="n">get_coordinates</span><span class="p">()</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">polygons</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">centroids</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_boundary_props</span><span class="p">(</span>
        <span class="n">area</span><span class="p">,</span> <span class="n">convexity</span><span class="p">,</span> <span class="n">elongation</span><span class="p">,</span> <span class="n">circularity</span>
    <span class="p">)</span>

    <span class="c1"># Set up Boundary-Transcript neighbor edges</span>
    <span class="n">dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">polygons</span><span class="o">.</span><span class="n">area</span><span class="o">.</span><span class="n">max</span><span class="p">())</span> <span class="o">*</span> <span class="mi">10</span>  <span class="c1"># heuristic distance</span>
    <span class="n">nbrs_edge_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_kdtree_edge_index</span><span class="p">(</span>
        <span class="n">centroids</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">xy</span><span class="p">],</span>
        <span class="n">k</span><span class="o">=</span><span class="n">k_bd</span><span class="p">,</span>
        <span class="n">max_distance</span><span class="o">=</span><span class="n">dist</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;neighbors&quot;</span><span class="p">,</span> <span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span> <span class="o">=</span> <span class="n">nbrs_edge_idx</span>

    <span class="c1"># If there are no tx-neighbors-bd edges, we put the tile automatically in test set</span>
    <span class="k">if</span> <span class="n">nbrs_edge_idx</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># logging.warning(f&quot;No tx-neighbors-bd edges found in tile {self.uid}.&quot;)</span>
        <span class="n">pyg_data</span><span class="p">[</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;belongs&quot;</span><span class="p">,</span> <span class="s2">&quot;bd&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span>
            <span class="p">[],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">long</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">pyg_data</span>

    <span class="c1"># Now we identify and split the tx-belongs-bd edges</span>
    <span class="n">edge_type</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;tx&quot;</span><span class="p">,</span> <span class="s2">&quot;belongs&quot;</span><span class="p">,</span> <span class="s2">&quot;bd&quot;</span><span class="p">)</span>

    <span class="c1"># Find nuclear transcripts</span>
    <span class="n">tx_cell_ids</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
    <span class="n">cell_ids_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">idx</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">idx</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">polygons</span><span class="o">.</span><span class="n">index</span><span class="p">)}</span>

    <span class="c1"># Get nuclear column and value from settings</span>
    <span class="n">nuclear_column</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="p">,</span> <span class="s2">&quot;nuclear_column&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">nuclear_value</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="p">,</span> <span class="s2">&quot;nuclear_value&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">nuclear_column</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">boundaries</span><span class="o">.</span><span class="n">scale_factor</span> <span class="o">!=</span> <span class="mf">1.0</span><span class="p">:</span>
        <span class="n">is_nuclear</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">compute_nuclear_transcripts</span><span class="p">(</span>
            <span class="n">polygons</span><span class="o">=</span><span class="n">polygons</span><span class="p">,</span>
            <span class="n">transcripts</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">,</span>
            <span class="n">x_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">x</span><span class="p">,</span>
            <span class="n">y_col</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">transcripts</span><span class="o">.</span><span class="n">y</span><span class="p">,</span>
            <span class="n">nuclear_column</span><span class="o">=</span><span class="n">nuclear_column</span><span class="p">,</span>
            <span class="n">nuclear_value</span><span class="o">=</span><span class="n">nuclear_value</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">is_nuclear</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">transcripts</span><span class="p">[</span><span class="n">nuclear_column</span><span class="p">]</span><span class="o">.</span><span class="n">eq</span><span class="p">(</span><span class="n">nuclear_value</span><span class="p">)</span>
    <span class="n">is_nuclear</span> <span class="o">&amp;=</span> <span class="n">tx_cell_ids</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">polygons</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="c1"># # Set up overlap edges</span>
    <span class="c1"># row_idx = np.where(is_nuclear)[0]</span>
    <span class="c1"># col_idx = tx_cell_ids.iloc[row_idx].map(cell_ids_map)</span>
    <span class="c1"># blng_edge_idx = torch.tensor(np.stack([row_idx, col_idx])).long()</span>
    <span class="c1"># pyg_data[edge_type].edge_index = blng_edge_idx</span>

    <span class="c1"># # If there are no tx-belongs-bd edges, flag tile as test only (cannot be used for training)</span>
    <span class="c1"># if blng_edge_idx.numel() == 0:</span>
    <span class="c1">#     return pyg_data</span>

    <span class="c1">#         # If there are tx-bd edges, add negative edges for training</span>
    <span class="c1"># transform = RandomLinkSplit(</span>
    <span class="c1">#     num_val=0,</span>
    <span class="c1">#     num_test=0,</span>
    <span class="c1">#     is_undirected=True,</span>
    <span class="c1">#     edge_types=[edge_type],</span>
    <span class="c1">#     neg_sampling_ratio=neg_sampling_ratio,</span>
    <span class="c1"># )</span>
    <span class="c1"># pyg_data, _, _ = transform(pyg_data)</span>

    <span class="c1"># # Refilter negative edges to include only transcripts in the</span>
    <span class="c1"># # original positive edges (still need a memory-efficient solution)</span>
    <span class="c1"># edges = pyg_data[edge_type]</span>
    <span class="c1"># mask = edges.edge_label_index[0].unsqueeze(1) == edges.edge_index[0].unsqueeze(</span>
    <span class="c1">#     0</span>
    <span class="c1"># )</span>
    <span class="c1"># mask = torch.nonzero(torch.any(mask, 1)).squeeze()</span>
    <span class="c1"># edges.edge_label_index = edges.edge_label_index[:, mask]</span>
    <span class="c1"># edges.edge_label = edges.edge_label[mask]</span>

    <span class="c1"># return pyg_data</span>


    <span class="c1"># Set up overlap edges</span>
    <span class="n">row_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">is_nuclear</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">col_idx</span> <span class="o">=</span> <span class="n">tx_cell_ids</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">row_idx</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">cell_ids_map</span><span class="p">)</span>
    <span class="n">blng_edge_idx</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">stack</span><span class="p">([</span><span class="n">row_idx</span><span class="p">,</span> <span class="n">col_idx</span><span class="p">]))</span><span class="o">.</span><span class="n">long</span><span class="p">()</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="n">edge_type</span><span class="p">]</span><span class="o">.</span><span class="n">edge_index</span> <span class="o">=</span> <span class="n">blng_edge_idx</span>

    <span class="c1"># If there are no tx-belongs-bd edges, flag tile as test only (cannot be used for training)</span>
    <span class="k">if</span> <span class="n">blng_edge_idx</span><span class="o">.</span><span class="n">numel</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pyg_data</span>

    <span class="c1"># If there are tx-bd edges, add negative edges for training</span>
    <span class="n">pos_edges</span> <span class="o">=</span> <span class="n">blng_edge_idx</span>  <span class="c1"># shape (2, num_pos)</span>
    <span class="n">num_pos</span> <span class="o">=</span> <span class="n">pos_edges</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Negative edges (tx-neighbors-bd) - EXCLUDE positives</span>
    <span class="n">neg_candidates</span> <span class="o">=</span> <span class="n">nbrs_edge_idx</span>  <span class="c1"># shape (2, num_candidates)</span>

    <span class="c1"># --- Fast Negative Filtering (PyTorch-only) ---</span>
    <span class="c1"># Reshape edges for broadcasting: (2, num_pos) vs (2, num_candidates, 1)</span>
    <span class="n">pos_expanded</span> <span class="o">=</span> <span class="n">pos_edges</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>  <span class="c1"># shape (2, num_pos, 1)</span>
    <span class="n">neg_expanded</span> <span class="o">=</span> <span class="n">neg_candidates</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># shape (2, 1, num_candidates)</span>

    <span class="c1"># Compare all edges in one go (broadcasting)</span>
    <span class="n">matches</span> <span class="o">=</span> <span class="p">(</span><span class="n">pos_expanded</span> <span class="o">==</span> <span class="n">neg_expanded</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># shape (num_pos, num_candidates)</span>
    <span class="n">is_negative</span> <span class="o">=</span> <span class="o">~</span><span class="n">matches</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># shape (num_candidates,)</span>

    <span class="c1"># Filter negatives</span>
    <span class="n">neg_edges</span> <span class="o">=</span> <span class="n">neg_candidates</span><span class="p">[:,</span> <span class="n">is_negative</span><span class="p">]</span>  <span class="c1"># shape (2, num_filtered_neg)</span>
    <span class="n">num_neg</span> <span class="o">=</span> <span class="n">neg_edges</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># --- Combine and label ---</span>
    <span class="n">edge_label_index</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">([</span><span class="n">neg_edges</span><span class="p">,</span> <span class="n">pos_edges</span><span class="p">],</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">edge_label</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">([</span>
        <span class="n">torch</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">num_neg</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float</span><span class="p">),</span>
        <span class="n">torch</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">num_pos</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float</span><span class="p">)</span>
    <span class="p">])</span>

    <span class="n">mask</span> <span class="o">=</span> <span class="n">edge_label_index</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="o">==</span> <span class="n">blng_edge_idx</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
    <span class="n">edge_label_index</span> <span class="o">=</span> <span class="n">edge_label_index</span><span class="p">[:,</span> <span class="n">mask</span><span class="p">]</span>
    <span class="n">edge_label</span> <span class="o">=</span> <span class="n">edge_label</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>

    <span class="n">pyg_data</span><span class="p">[</span><span class="n">edge_type</span><span class="p">]</span><span class="o">.</span><span class="n">edge_label_index</span> <span class="o">=</span> <span class="n">edge_label_index</span>
    <span class="n">pyg_data</span><span class="p">[</span><span class="n">edge_type</span><span class="p">]</span><span class="o">.</span><span class="n">edge_label</span> <span class="o">=</span> <span class="n">edge_label</span>

    <span class="k">return</span> <span class="n">pyg_data</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>




  </div>

    </div>

</div><h2 id="usage-examples">Usage Examples<a class="headerlink" href="#usage-examples" title="Permanent link">&para;</a></h2>
<h3 id="basic-data-loading">Basic Data Loading<a class="headerlink" href="#basic-data-loading" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">segger.data.sample</span> <span class="kn">import</span> <span class="n">STSampleParquet</span>

<span class="c1"># Load a spatial transcriptomics sample</span>
<span class="n">sample</span> <span class="o">=</span> <span class="n">STSampleParquet</span><span class="p">(</span>
    <span class="n">base_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/xenium/data&quot;</span><span class="p">,</span>
    <span class="n">n_workers</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
    <span class="n">sample_type</span><span class="o">=</span><span class="s2">&quot;xenium&quot;</span>
<span class="p">)</span>

<span class="c1"># Get sample information</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Transcripts: </span><span class="si">{</span><span class="n">sample</span><span class="o">.</span><span class="n">n_transcripts</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Spatial extents: </span><span class="si">{</span><span class="n">sample</span><span class="o">.</span><span class="n">extents</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Feature names: </span><span class="si">{</span><span class="n">sample</span><span class="o">.</span><span class="n">transcripts_metadata</span><span class="p">[</span><span class="s1">&#39;feature_names&#39;</span><span class="p">][:</span><span class="mi">5</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="spatial-tiling-and-processing">Spatial Tiling and Processing<a class="headerlink" href="#spatial-tiling-and-processing" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Save processed tiles</span>
<span class="n">sample</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
    <span class="n">data_dir</span><span class="o">=</span><span class="s2">&quot;./processed_data&quot;</span><span class="p">,</span>
    <span class="n">tile_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>  <span class="c1"># 1000 transcripts per tile</span>
    <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>          <span class="c1"># 3 boundary neighbors</span>
    <span class="n">k_tx</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>          <span class="c1"># 5 transcript neighbors</span>
    <span class="n">dist_bd</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span>    <span class="c1"># 15 pixel boundary distance</span>
    <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>     <span class="c1"># 5 pixel transcript distance</span>
    <span class="n">frac</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>        <span class="c1"># Process 80% of data</span>
    <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>    <span class="c1"># 10% validation</span>
    <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span>    <span class="c1"># 20% test</span>
<span class="p">)</span>
</code></pre></div>
<h3 id="in-memory-dataset-processing">In-Memory Dataset Processing<a class="headerlink" href="#in-memory-dataset-processing" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">segger.data.sample</span> <span class="kn">import</span> <span class="n">STInMemoryDataset</span>

<span class="c1"># Create dataset for a specific region</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">STInMemoryDataset</span><span class="p">(</span>
    <span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">,</span>
    <span class="n">extents</span><span class="o">=</span><span class="n">region_polygon</span><span class="p">,</span>
    <span class="n">margin</span><span class="o">=</span><span class="mi">10</span>
<span class="p">)</span>

<span class="c1"># Generate tiles</span>
<span class="n">tiles</span> <span class="o">=</span> <span class="n">dataset</span><span class="o">.</span><span class="n">_tile</span><span class="p">(</span>
    <span class="n">width</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>    <span class="c1"># 100 pixel width</span>
    <span class="n">height</span><span class="o">=</span><span class="mi">100</span>    <span class="c1"># 100 pixel height</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generated </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">tiles</span><span class="p">)</span><span class="si">}</span><span class="s2"> tiles&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="individual-tile-processing">Individual Tile Processing<a class="headerlink" href="#individual-tile-processing" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">segger.data.sample</span> <span class="kn">import</span> <span class="n">STTile</span>

<span class="c1"># Process individual tile</span>
<span class="n">tile</span> <span class="o">=</span> <span class="n">STTile</span><span class="p">(</span><span class="n">dataset</span><span class="o">=</span><span class="n">dataset</span><span class="p">,</span> <span class="n">extents</span><span class="o">=</span><span class="n">tile_polygon</span><span class="p">)</span>

<span class="c1"># Get tile data</span>
<span class="n">transcripts</span> <span class="o">=</span> <span class="n">tile</span><span class="o">.</span><span class="n">transcripts</span>
<span class="n">boundaries</span> <span class="o">=</span> <span class="n">tile</span><span class="o">.</span><span class="n">boundaries</span>

<span class="c1"># Convert to PyG format</span>
<span class="n">pyg_data</span> <span class="o">=</span> <span class="n">tile</span><span class="o">.</span><span class="n">to_pyg_dataset</span><span class="p">(</span>
    <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_bd</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">dist_tx</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">area</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">convexity</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">elongation</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">circularity</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Tile UID: </span><span class="si">{</span><span class="n">tile</span><span class="o">.</span><span class="n">uid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Transcripts: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Boundaries: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">boundaries</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.sections", "navigation.instant", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate", "content.tabs.link", "navigation.tracking"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.af256bd8.min.js"></script>
      
        <script src="../../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/termynal@0.0.1/termynal.js"></script>
      
        <script src="../../../termynal.js"></script>
      
    
  </body>
</html>