
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Fast and accurate cell segmentation for single-molecule spatial omics">
      
      
        <meta name="author" content="Elyas Heidari">
      
      
        <link rel="canonical" href="https://EliHei2.github.io/segger_dev/user_guide/data_creation/">
      
      
        <link rel="prev" href="../../installation/">
      
      
        <link rel="next" href="../training/">
      
      
      <link rel="icon" href="../../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.33">
    
    
      
        <title>Dataset Creation - segger Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=ubuntu:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"ubuntu";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css">
    
      <link rel="stylesheet" href="https://unpkg.com/termynal@0.0.1/termynal.css">
    
      <link rel="stylesheet" href="../../termynal.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#data-preparation-for-segger" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="segger Documentation" class="md-header__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            segger Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Dataset Creation
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="segger Documentation" class="md-nav__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../images/logo.png" alt="logo">

    </a>
    segger Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Dataset Creation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Dataset Creation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#steps" class="md-nav__link">
    <span class="md-ellipsis">
      Steps
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#core-components" class="md-nav__link">
    <span class="md-ellipsis">
      Core Components
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Core Components">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-stsampleparquet" class="md-nav__link">
    <span class="md-ellipsis">
      1. STSampleParquet
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. STSampleParquet">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#key-params" class="md-nav__link">
    <span class="md-ellipsis">
      Key Params
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-stinmemorydataset" class="md-nav__link">
    <span class="md-ellipsis">
      2. STInMemoryDataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-sttile" class="md-nav__link">
    <span class="md-ellipsis">
      3. STTile
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Workflow">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-initialize-sample-from-settings" class="md-nav__link">
    <span class="md-ellipsis">
      Step 1: Initialize sample from settings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-region-partitioning-and-tiling" class="md-nav__link">
    <span class="md-ellipsis">
      Step 2: Region partitioning and tiling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-graph-construction-per-tile" class="md-nav__link">
    <span class="md-ellipsis">
      Step 3: Graph construction per tile
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-4-splitting-and-saving" class="md-nav__link">
    <span class="md-ellipsis">
      Step 4: Splitting and saving
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-usage" class="md-nav__link">
    <span class="md-ellipsis">
      Example Usage
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Example Usage">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#xenium-with-optional-scrna-seq-derived-embeddings" class="md-nav__link">
    <span class="md-ellipsis">
      Xenium (with optional scRNA-seq-derived embeddings)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#merscope-fixed-size-tiling" class="md-nav__link">
    <span class="md-ellipsis">
      Merscope (fixed-size tiling)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#debug-mode-step-by-step-logging" class="md-nav__link">
    <span class="md-ellipsis">
      Debug mode (step-by-step logging)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#notes-and-recommendations" class="md-nav__link">
    <span class="md-ellipsis">
      Notes and Recommendations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-scrna-seq-for-embeddings-and-mutually-exclusive-genes" class="md-nav__link">
    <span class="md-ellipsis">
      Using scRNA-seq for embeddings and mutually exclusive genes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using scRNA-seq for embeddings and mutually exclusive genes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-compute-transcript-embeddings-weights-from-scrna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      1) Compute transcript embeddings (weights) from scRNA-seq
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-optional-identify-mutually-exclusive-genes-from-scrna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      2) [OPTIONAL] Identify mutually exclusive genes from scRNA-seq
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-save-tiles-with-both-weights-and-mutually-exclusive-genes" class="md-nav__link">
    <span class="md-ellipsis">
      3) Save tiles with both weights and mutually exclusive genes
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../notebooks/segger_tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to Segger
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CLI
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Data
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/data/sample/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Sample
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/data/pyg_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyG Dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/data/_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/data/transcript_embedding/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transcript Embedding
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/data/ndtree/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    NDTree
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/models/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Models
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="data-preparation-for-segger">Data Preparation for <code>segger</code><a class="headerlink" href="#data-preparation-for-segger" title="Permanent link">&para;</a></h1>
<p>The <code>segger</code> package provides a streamlined, settings-driven pipeline to transform raw spatial transcriptomics outputs (e.g., Xenium, Merscope) into graph tiles ready for model training and evaluation.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently, <code>segger</code> supports <strong>Xenium</strong> and <strong>Merscope</strong> datasets via technology-specific settings.</p>
</div>
<h2 id="steps">Steps<a class="headerlink" href="#steps" title="Permanent link">&para;</a></h2>
<p>The data preparation pipeline includes:</p>
<ol>
<li><strong>Settings-driven I/O</strong>: Uses a <code>sample_type</code> (e.g., <code>xenium</code>, <code>merscope</code>) to resolve input file names and column mappings.</li>
<li><strong>Lazy Loading + Filtering</strong>: Efficiently reads Parquet in spatial regions and filters transcripts/boundaries.</li>
<li><strong>Tiling</strong>: Partitions the whole slide into spatial tiles (fixed size or balanced by transcript count).</li>
<li><strong>Graph Construction</strong>: Builds PyTorch Geometric <code>HeteroData</code> with typed nodes/edges and labels for link prediction.</li>
<li><strong>Splitting</strong>: Writes tiles into <code>train</code>, <code>val</code>, and <code>test</code> subsets.</li>
</ol>
<div class="admonition note">
<p class="admonition-title">Key Technologies</p>
<ul>
<li><strong>PyTorch Geometric (PyG)</strong>: Heterogeneous graphs for GNNs.</li>
<li><strong>Shapely &amp; GeoPandas</strong>: Geometry operations (polygons, centroids, areas).</li>
<li><strong>PyArrow Parquet</strong>: Efficient I/O with schema-aware reads.</li>
</ul>
</div>
<h2 id="core-components">Core Components<a class="headerlink" href="#core-components" title="Permanent link">&para;</a></h2>
<h3 id="1-stsampleparquet">1. <code>STSampleParquet</code><a class="headerlink" href="#1-stsampleparquet" title="Permanent link">&para;</a></h3>
<p>Settings-based entry point for preparing a sample into graph tiles.</p>
<ul>
<li><strong>Constructor</strong>: resolves input file paths and metadata using <code>sample_type</code> settings (e.g., file names, column names, quality fields). Ensures transcript IDs exist.</li>
<li><strong>Embeddings</strong>: optionally accepts a <code>weights</code> DataFrame (index: gene names; columns: embedding dims) to encode transcript features.</li>
<li><strong>Saving</strong>: orchestrates region partitioning, tiling, PyG graph construction, negative sampling, and dataset splitting.</li>
</ul>
<h4 id="key-params">Key Params<a class="headerlink" href="#key-params" title="Permanent link">&para;</a></h4>
<ul>
<li><code>base_dir</code>: folder with required Parquet files (transcripts and boundaries).</li>
<li><code>sample_type</code>: one of <code>xenium</code>, <code>merscope</code> (determines settings such as file names, columns, nuclear flags, scale factors).</li>
<li><code>weights</code>: optional <code>pd.DataFrame</code> of gene embeddings used by <code>TranscriptEmbedding</code>.</li>
<li><code>scale_factor</code>: optional override for boundary scaling used during spatial queries.</li>
</ul>
<h3 id="2-stinmemorydataset">2. <code>STInMemoryDataset</code><a class="headerlink" href="#2-stinmemorydataset" title="Permanent link">&para;</a></h3>
<p>Internal helper that loads filtered transcripts/boundaries for a region, pre-builds a KDTree, and generates tile bounds (fixed-size or balanced by count).</p>
<h3 id="3-sttile">3. <code>STTile</code><a class="headerlink" href="#3-sttile" title="Permanent link">&para;</a></h3>
<p>Per-tile builder that assembles a <code>HeteroData</code> graph:
- Nodes: <code>tx</code> (transcripts) with <code>pos</code>, <code>x</code> (features); <code>bd</code> (boundaries) with <code>pos</code>, <code>x</code> (polygon properties).
- Edges:
  - <code>('tx','neighbors','tx')</code>: transcript proximity (KDTree-based; <code>k_tx</code>, <code>dist_tx</code>).
  - <code>('tx','neighbors','bd')</code>: transcript-to-boundary proximity for receptive field construction.
  - <code>('tx','belongs','bd')</code>: positive labels from nuclear overlap or provided assignment; negative sampling performed from receptive-field candidates.</p>
<h2 id="workflow">Workflow<a class="headerlink" href="#workflow" title="Permanent link">&para;</a></h2>
<h3 id="step-1-initialize-sample-from-settings">Step 1: Initialize sample from settings<a class="headerlink" href="#step-1-initialize-sample-from-settings" title="Permanent link">&para;</a></h3>
<ul>
<li>Provide <code>base_dir</code> containing technology outputs.</li>
<li>Pick <code>sample_type</code> to resolve filenames/columns.</li>
<li>Optionally provide <code>weights</code> for transcript embeddings.</li>
</ul>
<h3 id="step-2-region-partitioning-and-tiling">Step 2: Region partitioning and tiling<a class="headerlink" href="#step-2-region-partitioning-and-tiling" title="Permanent link">&para;</a></h3>
<ul>
<li>If multiple workers are set, extents are split into balanced regions (ND-tree over boundaries).</li>
<li>Tiles are created either by fixed width/height or by a target <code>tile_size</code> (balanced by transcript count).</li>
</ul>
<h3 id="step-3-graph-construction-per-tile">Step 3: Graph construction per tile<a class="headerlink" href="#step-3-graph-construction-per-tile" title="Permanent link">&para;</a></h3>
<ul>
<li>Build <code>HeteroData</code> with transcript (<code>tx</code>) and boundary (<code>bd</code>) nodes.</li>
<li>Add proximity edges and <code>belongs</code> labels (positives + sampled negatives).</li>
</ul>
<h3 id="step-4-splitting-and-saving">Step 4: Splitting and saving<a class="headerlink" href="#step-4-splitting-and-saving" title="Permanent link">&para;</a></h3>
<ul>
<li>Tiles are written to <code>&lt;data_dir&gt;/{train_tiles,val_tiles,test_tiles}/processed/*.pt</code> according to <code>val_prob</code>/<code>test_prob</code>.</li>
</ul>
<h2 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h2>
<ul>
<li>A directory structure with train/val/test tiles in PyG <code>HeteroData</code> format ready for the Segger model and <code>STPyGDataset</code>.</li>
</ul>
<div class="highlight"><pre><span></span><code>&lt;data_dir&gt;/
  train_tiles/
    processed/
      tiles_x=..._y=..._w=..._h=....pt
  val_tiles/
    processed/
      ...
  test_tiles/
    processed/
      ...
</code></pre></div>
<h2 id="example-usage">Example Usage<a class="headerlink" href="#example-usage" title="Permanent link">&para;</a></h2>
<h3 id="xenium-with-optional-scrna-seq-derived-embeddings">Xenium (with optional scRNA-seq-derived embeddings)<a class="headerlink" href="#xenium-with-optional-scrna-seq-derived-embeddings" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Optional: provide transcript embeddings (rows: genes, cols: embedding dims)</span>
<span class="c1"># For example, cell-type abundance embeddings indexed by gene name</span>
<span class="c1"># weights = pd.DataFrame(..., index=gene_names)</span>
<span class="n">weights</span> <span class="o">=</span> <span class="kc">None</span>  <span class="c1"># set to a DataFrame if available</span>

<span class="kn">from</span> <span class="nn">segger.data.sample</span> <span class="kn">import</span> <span class="n">STSampleParquet</span>

<span class="n">base_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/path/to/xenium_output&quot;</span><span class="p">)</span>
<span class="n">data_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/path/to/processed_tiles&quot;</span><span class="p">)</span>

<span class="n">sample</span> <span class="o">=</span> <span class="n">STSampleParquet</span><span class="p">(</span>
    <span class="n">base_dir</span><span class="o">=</span><span class="n">base_dir</span><span class="p">,</span>
    <span class="n">sample_type</span><span class="o">=</span><span class="s2">&quot;xenium&quot;</span><span class="p">,</span>
    <span class="n">n_workers</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>            <span class="c1"># controls parallel tiling across regions</span>
    <span class="c1"># weights=weights,        # optional transcript embeddings</span>
    <span class="n">scale_factor</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>       <span class="c1"># optional override (geometry scaling)</span>
<span class="p">)</span>

<span class="c1"># Save tiles (choose either tile_size OR tile_width+tile_height)</span>
<span class="n">sample</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
    <span class="n">data_dir</span><span class="o">=</span><span class="n">data_dir</span><span class="p">,</span>
    <span class="c1"># Receptive fields (neighbors)</span>
    <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>        <span class="c1"># nearest boundaries per transcript</span>
    <span class="n">dist_bd</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span>  <span class="c1"># max distance for tx-&gt;bd neighbors (µm-equivalent)</span>
    <span class="n">k_tx</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>       <span class="c1"># nearest transcripts per transcript</span>
    <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>   <span class="c1"># max distance for tx-&gt;tx neighbors</span>
    <span class="c1"># Optional broader receptive fields for mutually exclusive genes (if used)</span>
    <span class="c1"># Tiling</span>
    <span class="n">tile_size</span><span class="o">=</span><span class="mi">50000</span><span class="p">,</span>   <span class="c1"># alternative: tile_width=..., tile_height=...</span>
    <span class="c1"># Sampling/splitting</span>
    <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>
    <span class="n">frac</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
    <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
<span class="p">)</span>
</code></pre></div>
<h3 id="merscope-fixed-size-tiling">Merscope (fixed-size tiling)<a class="headerlink" href="#merscope-fixed-size-tiling" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">segger.data.sample</span> <span class="kn">import</span> <span class="n">STSampleParquet</span>

<span class="n">base_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/path/to/merscope_output&quot;</span><span class="p">)</span>
<span class="n">data_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;/path/to/processed_tiles&quot;</span><span class="p">)</span>

<span class="n">sample</span> <span class="o">=</span> <span class="n">STSampleParquet</span><span class="p">(</span>
    <span class="n">base_dir</span><span class="o">=</span><span class="n">base_dir</span><span class="p">,</span>
    <span class="n">sample_type</span><span class="o">=</span><span class="s2">&quot;merscope&quot;</span><span class="p">,</span>
    <span class="n">n_workers</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">sample</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
    <span class="n">data_dir</span><span class="o">=</span><span class="n">data_dir</span><span class="p">,</span>
    <span class="c1"># Nearest neighbors</span>
    <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_bd</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>
    <span class="c1"># Fixed-size tiling in sample units</span>
    <span class="n">tile_width</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
    <span class="n">tile_height</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
    <span class="c1"># Splits</span>
    <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="mf">3.0</span><span class="p">,</span>
    <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
<span class="p">)</span>
</code></pre></div>
<h3 id="debug-mode-step-by-step-logging">Debug mode (step-by-step logging)<a class="headerlink" href="#debug-mode-step-by-step-logging" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="n">sample</span><span class="o">.</span><span class="n">save_debug</span><span class="p">(</span>
    <span class="n">data_dir</span><span class="o">=</span><span class="n">data_dir</span><span class="p">,</span>
    <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">dist_bd</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>
    <span class="n">tile_width</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
    <span class="n">tile_height</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
    <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>
    <span class="n">frac</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
    <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
<span class="p">)</span>
</code></pre></div>
<h2 id="notes-and-recommendations">Notes and Recommendations<a class="headerlink" href="#notes-and-recommendations" title="Permanent link">&para;</a></h2>
<ul>
<li><strong>Settings and columns</strong>: Filenames and columns for transcripts/boundaries are resolved via <code>sample_type</code> settings. See <code>segger.data._settings/*</code> for details.</li>
<li><strong>Transcript IDs</strong>: The constructor ensures an ID column exists in transcripts; if missing, it is added deterministically.</li>
<li><strong>Quality filtering</strong>: Uses settings-defined columns (e.g., QV) and filter substrings. Genes absent from provided <code>weights</code> will be auto-added to filter substrings to avoid OOV embeddings.</li>
<li><strong>Neighbors</strong>: Set <code>k_tx/dist_tx</code> based on typical nuclear radii and transcript densities; <code>k_bd/dist_bd</code> controls candidate boundaries per transcript.</li>
<li><strong>Splits</strong>: Tiles with no <code>('tx','belongs','bd')</code> edges are automatically placed in <code>test_tiles</code>.</li>
<li><strong>Embeddings</strong>: If no <code>weights</code> are provided, transcripts fall back to token/ID-based embeddings.</li>
</ul>
<h2 id="using-scrna-seq-for-embeddings-and-mutually-exclusive-genes">Using scRNA-seq for embeddings and mutually exclusive genes<a class="headerlink" href="#using-scrna-seq-for-embeddings-and-mutually-exclusive-genes" title="Permanent link">&para;</a></h2>
<p>You can leverage scRNA-seq data both to create transcript embeddings (weights) and to identify mutually exclusive gene pairs that guide repulsive/attractive transcript edges.</p>
<h3 id="1-compute-transcript-embeddings-weights-from-scrna-seq">1) Compute transcript embeddings (weights) from scRNA-seq<a class="headerlink" href="#1-compute-transcript-embeddings-weights-from-scrna-seq" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">segger.data._utils</span> <span class="kn">import</span> <span class="n">calculate_gene_celltype_abundance_embedding</span>

<span class="c1"># Load a reference AnnData</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;/path/to/reference_scrnaseq.h5ad&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">subsample</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">)</span>        <span class="c1"># optional downsampling</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>

<span class="c1"># Column in adata.obs with cell-type annotations</span>
<span class="n">celltype_column</span> <span class="o">=</span> <span class="s2">&quot;celltype_minor&quot;</span>

<span class="c1"># Compute gene x cell-type abundance matrix (DataFrame indexed by gene names)</span>
<span class="n">weights</span> <span class="o">=</span> <span class="n">calculate_gene_celltype_abundance_embedding</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">celltype_column</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Pass weights to STSampleParquet to encode transcript features</span>
<span class="kn">from</span> <span class="nn">segger.data.sample</span> <span class="kn">import</span> <span class="n">STSampleParquet</span>
<span class="n">sample</span> <span class="o">=</span> <span class="n">STSampleParquet</span><span class="p">(</span>
    <span class="n">base_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/technology_output&quot;</span><span class="p">,</span>
    <span class="n">sample_type</span><span class="o">=</span><span class="s2">&quot;xenium&quot;</span><span class="p">,</span>      <span class="c1"># or &quot;merscope&quot;</span>
    <span class="n">n_workers</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
    <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">,</span>
<span class="p">)</span>
</code></pre></div>
<h3 id="2-optional-identify-mutually-exclusive-genes-from-scrna-seq">2) [OPTIONAL] Identify mutually exclusive genes from scRNA-seq<a class="headerlink" href="#2-optional-identify-mutually-exclusive-genes-from-scrna-seq" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">segger.data._utils</span> <span class="kn">import</span> <span class="n">find_markers</span><span class="p">,</span> <span class="n">find_mutually_exclusive_genes</span>

<span class="c1"># Optionally restrict to genes present in the sample</span>
<span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">transcripts_metadata</span><span class="p">[</span><span class="s2">&quot;feature_names&quot;</span><span class="p">]))</span>
<span class="n">adata_sub</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">genes</span><span class="p">]</span>

<span class="c1"># Find cell-type markers (tune thresholds as needed)</span>
<span class="n">markers</span> <span class="o">=</span> <span class="n">find_markers</span><span class="p">(</span>
    <span class="n">adata_sub</span><span class="p">,</span>
    <span class="n">cell_type_column</span><span class="o">=</span><span class="n">celltype_column</span><span class="p">,</span>
    <span class="n">pos_percentile</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span>
    <span class="n">neg_percentile</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">percentage</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Compute mutually exclusive gene pairs using markers</span>
<span class="n">exclusive_gene_pairs</span> <span class="o">=</span> <span class="n">find_mutually_exclusive_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
    <span class="n">markers</span><span class="o">=</span><span class="n">markers</span><span class="p">,</span>
    <span class="n">cell_type_column</span><span class="o">=</span><span class="n">celltype_column</span><span class="p">,</span>
<span class="p">)</span>
</code></pre></div>
<h3 id="3-save-tiles-with-both-weights-and-mutually-exclusive-genes">3) Save tiles with both weights and mutually exclusive genes<a class="headerlink" href="#3-save-tiles-with-both-weights-and-mutually-exclusive-genes" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="n">sample</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
    <span class="n">data_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/processed_tiles&quot;</span><span class="p">,</span>
    <span class="c1"># Nearest-neighbor receptive fields</span>
    <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_bd</span><span class="o">=</span><span class="mf">15.0</span><span class="p">,</span>
    <span class="n">k_tx</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>
    <span class="c1"># Optional broader receptive fields used for mutually exclusive genes</span>
    <span class="n">k_tx_ex</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">dist_tx_ex</span><span class="o">=</span><span class="mf">20.0</span><span class="p">,</span>
    <span class="c1"># Tiling and splits</span>
    <span class="n">tile_size</span><span class="o">=</span><span class="mi">50_000</span><span class="p">,</span>
    <span class="n">neg_sampling_ratio</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span>
    <span class="n">val_prob</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">test_prob</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
    <span class="c1"># Use mutually exclusive pairs to add repulsive/attractive tx-tx labels</span>
    <span class="n">mutually_exclusive_genes</span><span class="o">=</span><span class="n">exclusive_gene_pairs</span><span class="p">,</span>
<span class="p">)</span>
</code></pre></div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.sections", "navigation.instant", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate", "content.tabs.link", "navigation.tracking"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.af256bd8.min.js"></script>
      
        <script src="../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/termynal@0.0.1/termynal.js"></script>
      
        <script src="../../termynal.js"></script>
      
    
  </body>
</html>