
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Fast and accurate cell segmentation for single-molecule spatial omics">
      
      
        <meta name="author" content="Elyas Heidari">
      
      
        <link rel="canonical" href="https://EliHei2.github.io/segger_dev/api/models/inference/">
      
      
      
      
      <link rel="icon" href="../../../images/logo.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.33">
    
    
      
        <title>Segger Model Inference - segger Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=ubuntu:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"ubuntu";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../css/extra.css">
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.11/dist/katex.min.css">
    
      <link rel="stylesheet" href="https://unpkg.com/termynal@0.0.1/termynal.css">
    
      <link rel="stylesheet" href="../../../termynal.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#segger-model-inference" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="segger Documentation" class="md-header__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            segger Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Segger Model Inference
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="amber" data-md-color-accent="deep-orange"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="segger Documentation" class="md-nav__button md-logo" aria-label="segger Documentation" data-md-component="logo">
      
  <img src="../../../images/logo.png" alt="logo">

    </a>
    segger Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/EliHei2/segger_dev" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/data_creation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dataset Creation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../user_guide/validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../notebooks/segger_tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction to Segger
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CLI
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
        
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    API Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            API Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Data
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/sample/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Sample
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/pyg_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PyG Dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/_utils/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Utils
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/transcript_embedding/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Transcript Embedding
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data/ndtree/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    NDTree
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Models
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../training/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Training
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Validation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="segger-model-inference">Segger Model Inference<a class="headerlink" href="#segger-model-inference" title="Permanent link">&para;</a></h1>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">&para;</a></h2>
<p>Inference with the trained Segger model involves using the learned Graph Neural Network to predict transcript-to-cell associations in spatial transcriptomics data. The inference process transforms spatial data into cell segmentation results through link prediction, enabling both cell identification and fragment detection.</p>
<h2 id="inference-pipeline">Inference Pipeline<a class="headerlink" href="#inference-pipeline" title="Permanent link">&para;</a></h2>
<h3 id="overall-workflow">Overall Workflow<a class="headerlink" href="#overall-workflow" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code>Trained Model → Spatial Data → Graph Construction → Node Embeddings → Similarity Scores → Cell Assignment → Post-processing
     ↓              ↓              ↓                ↓              ↓              ↓              ↓
Learned Weights   Transcripts    Heterogeneous   GNN Forward    Link Prediction  Thresholding   Final Results
                 Boundaries     Graphs          Pass            Scores           Filtering      + Fragments
</code></pre></div>
<h3 id="key-steps">Key Steps<a class="headerlink" href="#key-steps" title="Permanent link">&para;</a></h3>
<ol>
<li><strong>Model Loading</strong>: Load trained Segger model from checkpoint</li>
<li><strong>Data Preparation</strong>: Construct heterogeneous graphs from spatial data</li>
<li><strong>Embedding Generation</strong>: Generate node embeddings using the trained model</li>
<li><strong>Similarity Computation</strong>: Calculate transcript-to-cell similarity scores</li>
<li><strong>Assignment Decision</strong>: Assign transcripts to cells based on confidence scores</li>
<li><strong>Fragment Detection</strong>: Group unassigned transcripts into fragments</li>
</ol>
<h2 id="model-loading">Model Loading<a class="headerlink" href="#model-loading" title="Permanent link">&para;</a></h2>
<h3 id="loading-trained-model">Loading Trained Model<a class="headerlink" href="#loading-trained-model" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">segger.models.segger_model</span> <span class="kn">import</span> <span class="n">Segger</span>
<span class="kn">import</span> <span class="nn">torch</span>

<span class="c1"># Load trained model</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">Segger</span><span class="p">(</span>
    <span class="n">num_tx_tokens</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
    <span class="n">init_emb</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
    <span class="n">hidden_channels</span><span class="o">=</span><span class="mi">64</span><span class="p">,</span>
    <span class="n">num_mid_layers</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">out_channels</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span>
    <span class="n">heads</span><span class="o">=</span><span class="mi">4</span>
<span class="p">)</span>

<span class="c1"># Load trained weights</span>
<span class="n">checkpoint</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;path/to/checkpoint.ckpt&#39;</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">load_state_dict</span><span class="p">(</span><span class="n">checkpoint</span><span class="p">[</span><span class="s1">&#39;state_dict&#39;</span><span class="p">])</span>
<span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>  <span class="c1"># Set to evaluation mode</span>
</code></pre></div>
<h3 id="model-configuration">Model Configuration<a class="headerlink" href="#model-configuration" title="Permanent link">&para;</a></h3>
<p>Ensure inference parameters match training configuration:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Verify model configuration matches training</span>
<span class="k">assert</span> <span class="n">model</span><span class="o">.</span><span class="n">num_tx_tokens</span> <span class="o">==</span> <span class="mi">5000</span><span class="p">,</span> <span class="s2">&quot;Vocabulary size mismatch&quot;</span>
<span class="k">assert</span> <span class="n">model</span><span class="o">.</span><span class="n">hidden_channels</span> <span class="o">==</span> <span class="mi">64</span><span class="p">,</span> <span class="s2">&quot;Hidden dimension mismatch&quot;</span>
<span class="k">assert</span> <span class="n">model</span><span class="o">.</span><span class="n">out_channels</span> <span class="o">==</span> <span class="mi">32</span><span class="p">,</span> <span class="s2">&quot;Output dimension mismatch&quot;</span>
<span class="k">assert</span> <span class="n">model</span><span class="o">.</span><span class="n">heads</span> <span class="o">==</span> <span class="mi">4</span><span class="p">,</span> <span class="s2">&quot;Attention heads mismatch&quot;</span>
</code></pre></div>
<h2 id="data-preparation">Data Preparation<a class="headerlink" href="#data-preparation" title="Permanent link">&para;</a></h2>
<h3 id="graph-construction">Graph Construction<a class="headerlink" href="#graph-construction" title="Permanent link">&para;</a></h3>
<p>Inference requires the same graph structure used during training:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Construct heterogeneous graph for inference</span>
<span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;tx&quot;</span><span class="p">:</span> <span class="p">{</span>  <span class="c1"># Transcript nodes</span>
        <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="n">transcript_ids</span><span class="p">,</span>
        <span class="s2">&quot;pos&quot;</span><span class="p">:</span> <span class="n">spatial_coordinates</span><span class="p">,</span>
        <span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">feature_vectors</span>
    <span class="p">},</span>
    <span class="s2">&quot;bd&quot;</span><span class="p">:</span> <span class="p">{</span>  <span class="c1"># Boundary nodes</span>
        <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="n">boundary_ids</span><span class="p">,</span>
        <span class="s2">&quot;pos&quot;</span><span class="p">:</span> <span class="n">centroid_coordinates</span><span class="p">,</span>
        <span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">geometric_features</span>
    <span class="p">},</span>
    <span class="s2">&quot;tx,neighbors,tx&quot;</span><span class="p">:</span> <span class="p">{</span>  <span class="c1"># Transcript proximity edges</span>
        <span class="s2">&quot;edge_index&quot;</span><span class="p">:</span> <span class="n">neighbor_connections</span>
    <span class="p">},</span>
    <span class="s2">&quot;tx,belongs,bd&quot;</span><span class="p">:</span> <span class="p">{</span>  <span class="c1"># Transcript-boundary edges (for inference)</span>
        <span class="s2">&quot;edge_index&quot;</span><span class="p">:</span> <span class="n">containment_relationships</span>
    <span class="p">}</span>
<span class="p">}</span>
</code></pre></div>
<h3 id="feature-processing">Feature Processing<a class="headerlink" href="#feature-processing" title="Permanent link">&para;</a></h3>
<h4 id="transcript-features">Transcript Features<a class="headerlink" href="#transcript-features" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Use same feature processing as training</span>
<span class="k">if</span> <span class="n">has_scrnaseq_embeddings</span><span class="p">:</span>
    <span class="n">transcript_features</span> <span class="o">=</span> <span class="n">gene_celltype_embeddings</span><span class="p">[</span><span class="n">gene_labels</span><span class="p">]</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">transcript_features</span> <span class="o">=</span> <span class="n">embedding_layer</span><span class="p">(</span><span class="n">gene_token_ids</span><span class="p">)</span>
</code></pre></div>
<h4 id="boundary-features">Boundary Features<a class="headerlink" href="#boundary-features" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Compute geometric features for boundaries</span>
<span class="k">def</span> <span class="nf">compute_boundary_features</span><span class="p">(</span><span class="n">boundary_polygons</span><span class="p">):</span>
    <span class="n">features</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">polygon</span> <span class="ow">in</span> <span class="n">boundary_polygons</span><span class="p">:</span>
        <span class="n">area_val</span> <span class="o">=</span> <span class="n">polygon</span><span class="o">.</span><span class="n">area</span>
        <span class="n">convex_hull</span> <span class="o">=</span> <span class="n">polygon</span><span class="o">.</span><span class="n">convex_hull</span>
        <span class="n">convexity</span> <span class="o">=</span> <span class="n">convex_hull</span><span class="o">.</span><span class="n">area</span> <span class="o">/</span> <span class="n">area_val</span>

        <span class="c1"># Minimum bounding rectangle</span>
        <span class="n">mbr</span> <span class="o">=</span> <span class="n">polygon</span><span class="o">.</span><span class="n">minimum_rotated_rectangle</span>
        <span class="n">mbr_area</span> <span class="o">=</span> <span class="n">mbr</span><span class="o">.</span><span class="n">area</span>

        <span class="c1"># Envelope (axis-aligned bounding box)</span>
        <span class="n">envelope</span> <span class="o">=</span> <span class="n">polygon</span><span class="o">.</span><span class="n">envelope</span>
        <span class="n">env_area</span> <span class="o">=</span> <span class="n">envelope</span><span class="o">.</span><span class="n">area</span>
        <span class="n">elongation</span> <span class="o">=</span> <span class="n">mbr_area</span> <span class="o">/</span> <span class="n">env_area</span>

        <span class="c1"># Circularity</span>
        <span class="n">min_radius</span> <span class="o">=</span> <span class="n">compute_minimum_bounding_radius</span><span class="p">(</span><span class="n">polygon</span><span class="p">)</span>
        <span class="n">circularity</span> <span class="o">=</span> <span class="n">area_val</span> <span class="o">/</span> <span class="p">(</span><span class="n">min_radius</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">features</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">area_val</span><span class="p">,</span> <span class="n">convexity</span><span class="p">,</span> <span class="n">elongation</span><span class="p">,</span> <span class="n">circularity</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
</code></pre></div>
<h2 id="inference-process">Inference Process<a class="headerlink" href="#inference-process" title="Permanent link">&para;</a></h2>
<h3 id="forward-pass">Forward Pass<a class="headerlink" href="#forward-pass" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Generate node embeddings</span>
<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">():</span>
    <span class="n">embeddings</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">edge_index</span><span class="p">)</span>

<span class="c1"># Separate transcript and boundary embeddings</span>
<span class="n">tx_embeddings</span> <span class="o">=</span> <span class="n">embeddings</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">tx_mask</span><span class="p">]</span>
<span class="n">bd_embeddings</span> <span class="o">=</span> <span class="n">embeddings</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">bd_mask</span><span class="p">]</span>
</code></pre></div>
<h3 id="similarity-score-computation">Similarity Score Computation<a class="headerlink" href="#similarity-score-computation" title="Permanent link">&para;</a></h3>
<h4 id="transcript-to-cell-similarity">Transcript-to-Cell Similarity<a class="headerlink" href="#transcript-to-cell-similarity" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_similarity_scores</span><span class="p">(</span><span class="n">tx_embeddings</span><span class="p">,</span> <span class="n">bd_embeddings</span><span class="p">,</span> <span class="n">edge_index</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute similarity scores between transcripts and boundaries.&quot;&quot;&quot;</span>

    <span class="c1"># Extract source and target indices</span>
    <span class="n">tx_indices</span> <span class="o">=</span> <span class="n">edge_index</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">bd_indices</span> <span class="o">=</span> <span class="n">edge_index</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Get embeddings for connected nodes</span>
    <span class="n">tx_emb</span> <span class="o">=</span> <span class="n">tx_embeddings</span><span class="p">[</span><span class="n">tx_indices</span><span class="p">]</span>
    <span class="n">bd_emb</span> <span class="o">=</span> <span class="n">bd_embeddings</span><span class="p">[</span><span class="n">bd_indices</span><span class="p">]</span>

    <span class="c1"># Compute dot product similarity</span>
    <span class="n">similarity_scores</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">tx_emb</span> <span class="o">*</span> <span class="n">bd_emb</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Apply sigmoid for probability</span>
    <span class="n">probabilities</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sigmoid</span><span class="p">(</span><span class="n">similarity_scores</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">probabilities</span><span class="p">,</span> <span class="n">similarity_scores</span>
</code></pre></div>
<h4 id="receptive-field-construction">Receptive Field Construction<a class="headerlink" href="#receptive-field-construction" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">construct_receptive_field</span><span class="p">(</span><span class="n">transcripts</span><span class="p">,</span> <span class="n">boundaries</span><span class="p">,</span> <span class="n">k_bd</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">dist_bd</span><span class="o">=</span><span class="mf">10.0</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Construct receptive field for transcript-to-cell assignment.&quot;&quot;&quot;</span>

    <span class="c1"># Find nearest boundary cells for each transcript</span>
    <span class="kn">from</span> <span class="nn">sklearn.neighbors</span> <span class="kn">import</span> <span class="n">NearestNeighbors</span>

    <span class="c1"># Extract coordinates</span>
    <span class="n">tx_coords</span> <span class="o">=</span> <span class="n">transcripts</span><span class="p">[[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">bd_coords</span> <span class="o">=</span> <span class="n">boundaries</span><span class="p">[[</span><span class="s1">&#39;centroid_x&#39;</span><span class="p">,</span> <span class="s1">&#39;centroid_y&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># Build nearest neighbor index</span>
    <span class="n">nn</span> <span class="o">=</span> <span class="n">NearestNeighbors</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="n">k_bd</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="n">dist_bd</span><span class="p">)</span>
    <span class="n">nn</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">bd_coords</span><span class="p">)</span>

    <span class="c1"># Find neighbors</span>
    <span class="n">distances</span><span class="p">,</span> <span class="n">indices</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">kneighbors</span><span class="p">(</span><span class="n">tx_coords</span><span class="p">)</span>

    <span class="c1"># Filter by distance threshold</span>
    <span class="n">mask</span> <span class="o">=</span> <span class="n">distances</span> <span class="o">&lt;=</span> <span class="n">dist_bd</span>
    <span class="n">filtered_indices</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">filtered_distances</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">dist</span><span class="p">,</span> <span class="n">idx</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">distances</span><span class="p">,</span> <span class="n">indices</span><span class="p">)):</span>
        <span class="n">valid_mask</span> <span class="o">=</span> <span class="n">mask</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">filtered_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">[</span><span class="n">valid_mask</span><span class="p">])</span>
        <span class="n">filtered_distances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dist</span><span class="p">[</span><span class="n">valid_mask</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">filtered_indices</span><span class="p">,</span> <span class="n">filtered_distances</span>
</code></pre></div>
<h2 id="cell-assignment">Cell Assignment<a class="headerlink" href="#cell-assignment" title="Permanent link">&para;</a></h2>
<h3 id="assignment-decision">Assignment Decision<a class="headerlink" href="#assignment-decision" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">assign_transcripts_to_cells</span><span class="p">(</span><span class="n">similarity_scores</span><span class="p">,</span> <span class="n">score_threshold</span><span class="o">=</span><span class="mf">0.7</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Assign transcripts to cells based on similarity scores.&quot;&quot;&quot;</span>

    <span class="c1"># Find best matching cell for each transcript</span>
    <span class="n">best_scores</span><span class="p">,</span> <span class="n">best_cells</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">similarity_scores</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Apply confidence threshold</span>
    <span class="n">confident_mask</span> <span class="o">=</span> <span class="n">best_scores</span> <span class="o">&gt;=</span> <span class="n">score_threshold</span>

    <span class="c1"># Create assignment results</span>
    <span class="n">assignments</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;transcript_id&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;cell_id&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;confidence_score&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;assigned&#39;</span><span class="p">:</span> <span class="p">[]</span>
    <span class="p">}</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">score</span><span class="p">,</span> <span class="n">cell_id</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">best_scores</span><span class="p">,</span> <span class="n">best_cells</span><span class="p">)):</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_id</span><span class="o">.</span><span class="n">item</span><span class="p">())</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;confidence_score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score</span><span class="o">.</span><span class="n">item</span><span class="p">())</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;assigned&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">confident_mask</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">assignments</span>
</code></pre></div>
<h3 id="confidence-score-analysis">Confidence Score Analysis<a class="headerlink" href="#confidence-score-analysis" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">analyze_confidence_scores</span><span class="p">(</span><span class="n">scores</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Analyze distribution of confidence scores.&quot;&quot;&quot;</span>

    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>

    <span class="c1"># Convert to numpy for analysis</span>
    <span class="n">scores_np</span> <span class="o">=</span> <span class="n">scores</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span>

    <span class="c1"># Basic statistics</span>
    <span class="n">stats_summary</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;mean&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">scores_np</span><span class="p">),</span>
        <span class="s1">&#39;std&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">scores_np</span><span class="p">),</span>
        <span class="s1">&#39;min&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">scores_np</span><span class="p">),</span>
        <span class="s1">&#39;max&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">scores_np</span><span class="p">),</span>
        <span class="s1">&#39;median&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">scores_np</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Percentiles</span>
    <span class="n">percentiles</span> <span class="o">=</span> <span class="p">[</span><span class="mi">25</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">75</span><span class="p">,</span> <span class="mi">90</span><span class="p">,</span> <span class="mi">95</span><span class="p">,</span> <span class="mi">99</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">percentiles</span><span class="p">:</span>
        <span class="n">stats_summary</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;p</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">scores_np</span><span class="p">,</span> <span class="n">p</span><span class="p">)</span>

    <span class="c1"># Find knee point for automatic thresholding</span>
    <span class="c1"># Use the method from the Segger paper</span>
    <span class="n">knee_point</span> <span class="o">=</span> <span class="n">find_knee_point</span><span class="p">(</span><span class="n">scores_np</span><span class="p">)</span>
    <span class="n">stats_summary</span><span class="p">[</span><span class="s1">&#39;knee_point&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">knee_point</span>

    <span class="k">return</span> <span class="n">stats_summary</span>

<span class="k">def</span> <span class="nf">find_knee_point</span><span class="p">(</span><span class="n">scores</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find knee point in score distribution for automatic thresholding.&quot;&quot;&quot;</span>

    <span class="kn">from</span> <span class="nn">kneed</span> <span class="kn">import</span> <span class="n">KneeLocator</span>

    <span class="c1"># Sort scores</span>
    <span class="n">sorted_scores</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>

    <span class="c1"># Create cumulative distribution</span>
    <span class="n">cumulative</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sorted_scores</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">sorted_scores</span><span class="p">)</span>

    <span class="c1"># Find knee point</span>
    <span class="n">kneedle</span> <span class="o">=</span> <span class="n">KneeLocator</span><span class="p">(</span>
        <span class="n">sorted_scores</span><span class="p">,</span> <span class="n">cumulative</span><span class="p">,</span> 
        <span class="n">S</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">curve</span><span class="o">=</span><span class="s1">&#39;concave&#39;</span><span class="p">,</span> <span class="n">direction</span><span class="o">=</span><span class="s1">&#39;increasing&#39;</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">kneedle</span><span class="o">.</span><span class="n">knee</span> <span class="k">if</span> <span class="n">kneedle</span><span class="o">.</span><span class="n">knee</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
</code></pre></div>
<h2 id="fragment-detection">Fragment Detection<a class="headerlink" href="#fragment-detection" title="Permanent link">&para;</a></h2>
<h3 id="unassigned-transcript-handling">Unassigned Transcript Handling<a class="headerlink" href="#unassigned-transcript-handling" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">detect_fragments</span><span class="p">(</span><span class="n">unassigned_transcripts</span><span class="p">,</span> <span class="n">k_tx</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">dist_tx</span><span class="o">=</span><span class="mf">5.0</span><span class="p">,</span> <span class="n">similarity_threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Group unassigned transcripts into fragments.&quot;&quot;&quot;</span>

    <span class="c1"># Construct transcript-transcript similarity graph</span>
    <span class="n">fragment_graph</span> <span class="o">=</span> <span class="n">construct_fragment_graph</span><span class="p">(</span>
        <span class="n">unassigned_transcripts</span><span class="p">,</span> <span class="n">k_tx</span><span class="p">,</span> <span class="n">dist_tx</span><span class="p">,</span> <span class="n">similarity_threshold</span>
    <span class="p">)</span>

    <span class="c1"># Find connected components</span>
    <span class="kn">from</span> <span class="nn">scipy.sparse.csgraph</span> <span class="kn">import</span> <span class="n">connected_components</span>

    <span class="n">n_components</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">connected_components</span><span class="p">(</span><span class="n">fragment_graph</span><span class="p">,</span> <span class="n">directed</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Assign fragment IDs</span>
    <span class="n">fragment_assignments</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">label</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">labels</span><span class="p">):</span>
        <span class="n">transcript_id</span> <span class="o">=</span> <span class="n">unassigned_transcripts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="n">fragment_assignments</span><span class="p">[</span><span class="n">transcript_id</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;fragment_</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="k">return</span> <span class="n">fragment_assignments</span><span class="p">,</span> <span class="n">n_components</span>
</code></pre></div>
<h3 id="fragment-graph-construction">Fragment Graph Construction<a class="headerlink" href="#fragment-graph-construction" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">construct_fragment_graph</span><span class="p">(</span><span class="n">transcripts</span><span class="p">,</span> <span class="n">k_tx</span><span class="p">,</span> <span class="n">dist_tx</span><span class="p">,</span> <span class="n">similarity_threshold</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Construct similarity graph for unassigned transcripts.&quot;&quot;&quot;</span>

    <span class="c1"># Extract coordinates and features</span>
    <span class="n">coords</span> <span class="o">=</span> <span class="n">transcripts</span><span class="p">[[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">features</span> <span class="o">=</span> <span class="n">transcripts</span><span class="p">[</span><span class="s1">&#39;features&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># Build nearest neighbor graph</span>
    <span class="n">nn</span> <span class="o">=</span> <span class="n">NearestNeighbors</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="n">k_tx</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="n">dist_tx</span><span class="p">)</span>
    <span class="n">nn</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">coords</span><span class="p">)</span>

    <span class="c1"># Find neighbors</span>
    <span class="n">distances</span><span class="p">,</span> <span class="n">indices</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">radius_neighbors</span><span class="p">(</span><span class="n">coords</span><span class="p">,</span> <span class="n">radius</span><span class="o">=</span><span class="n">dist_tx</span><span class="p">)</span>

    <span class="c1"># Compute similarity scores</span>
    <span class="n">edges</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">neighbors</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">indices</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">neighbors</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">j</span><span class="p">:</span>
                <span class="c1"># Compute feature similarity</span>
                <span class="n">sim_score</span> <span class="o">=</span> <span class="n">compute_feature_similarity</span><span class="p">(</span><span class="n">features</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">features</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>

                <span class="c1"># Add edge if similarity exceeds threshold</span>
                <span class="k">if</span> <span class="n">sim_score</span> <span class="o">&gt;=</span> <span class="n">similarity_threshold</span><span class="p">:</span>
                    <span class="n">edges</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">,</span> <span class="n">sim_score</span><span class="p">))</span>

    <span class="c1"># Convert to sparse matrix</span>
    <span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">csr_matrix</span>

    <span class="k">if</span> <span class="n">edges</span><span class="p">:</span>
        <span class="n">rows</span><span class="p">,</span> <span class="n">cols</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="n">edges</span><span class="p">)</span>
        <span class="n">n_transcripts</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)</span>
        <span class="n">fragment_graph</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">((</span><span class="n">data</span><span class="p">,</span> <span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">cols</span><span class="p">)),</span> <span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="n">n_transcripts</span><span class="p">,</span> <span class="n">n_transcripts</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fragment_graph</span> <span class="o">=</span> <span class="n">csr_matrix</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)))</span>

    <span class="k">return</span> <span class="n">fragment_graph</span>

<span class="k">def</span> <span class="nf">compute_feature_similarity</span><span class="p">(</span><span class="n">feature1</span><span class="p">,</span> <span class="n">feature2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute similarity between transcript features.&quot;&quot;&quot;</span>

    <span class="c1"># Cosine similarity</span>
    <span class="n">dot_product</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">feature1</span><span class="p">,</span> <span class="n">feature2</span><span class="p">)</span>
    <span class="n">norm1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">feature1</span><span class="p">)</span>
    <span class="n">norm2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">feature2</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">norm1</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">norm2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="mf">0.0</span>

    <span class="n">similarity</span> <span class="o">=</span> <span class="n">dot_product</span> <span class="o">/</span> <span class="p">(</span><span class="n">norm1</span> <span class="o">*</span> <span class="n">norm2</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">similarity</span>
</code></pre></div>
<h2 id="batch-processing">Batch Processing<a class="headerlink" href="#batch-processing" title="Permanent link">&para;</a></h2>
<h3 id="large-dataset-handling">Large Dataset Handling<a class="headerlink" href="#large-dataset-handling" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">batch_inference</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">data_loader</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="s1">&#39;cuda&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform inference on large datasets in batches.&quot;&quot;&quot;</span>

    <span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
    <span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>

    <span class="n">all_results</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">batch</span> <span class="ow">in</span> <span class="n">data_loader</span><span class="p">:</span>
            <span class="c1"># Move batch to device</span>
            <span class="n">batch</span> <span class="o">=</span> <span class="n">batch</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>

            <span class="c1"># Generate embeddings</span>
            <span class="n">embeddings</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">batch</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="n">batch</span><span class="o">.</span><span class="n">edge_index</span><span class="p">)</span>

            <span class="c1"># Compute similarity scores</span>
            <span class="n">scores</span> <span class="o">=</span> <span class="n">compute_batch_similarities</span><span class="p">(</span><span class="n">batch</span><span class="p">,</span> <span class="n">embeddings</span><span class="p">)</span>

            <span class="c1"># Store results</span>
            <span class="n">batch_results</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;transcript_ids&#39;</span><span class="p">:</span> <span class="n">batch</span><span class="o">.</span><span class="n">transcript_ids</span><span class="p">,</span>
                <span class="s1">&#39;cell_ids&#39;</span><span class="p">:</span> <span class="n">batch</span><span class="o">.</span><span class="n">cell_ids</span><span class="p">,</span>
                <span class="s1">&#39;similarity_scores&#39;</span><span class="p">:</span> <span class="n">scores</span><span class="p">,</span>
                <span class="s1">&#39;batch_idx&#39;</span><span class="p">:</span> <span class="n">batch</span><span class="o">.</span><span class="n">batch</span>
            <span class="p">}</span>

            <span class="n">all_results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">batch_results</span><span class="p">)</span>

    <span class="c1"># Combine results from all batches</span>
    <span class="n">combined_results</span> <span class="o">=</span> <span class="n">combine_batch_results</span><span class="p">(</span><span class="n">all_results</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">combined_results</span>
</code></pre></div>
<h3 id="memory-management">Memory Management<a class="headerlink" href="#memory-management" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">optimize_memory_usage</span><span class="p">(</span><span class="n">batch_size</span><span class="p">,</span> <span class="n">model_size</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Optimize memory usage during inference.&quot;&quot;&quot;</span>

    <span class="c1"># Estimate memory requirements</span>
    <span class="n">estimated_memory</span> <span class="o">=</span> <span class="n">estimate_inference_memory</span><span class="p">(</span><span class="n">batch_size</span><span class="p">,</span> <span class="n">model_size</span><span class="p">)</span>

    <span class="c1"># Adjust batch size if needed</span>
    <span class="k">if</span> <span class="n">estimated_memory</span> <span class="o">&gt;</span> <span class="n">available_memory</span><span class="p">:</span>
        <span class="n">optimal_batch_size</span> <span class="o">=</span> <span class="n">find_optimal_batch_size</span><span class="p">(</span><span class="n">model_size</span><span class="p">,</span> <span class="n">available_memory</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Reducing batch size from </span><span class="si">{</span><span class="n">batch_size</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">optimal_batch_size</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">optimal_batch_size</span>

    <span class="k">return</span> <span class="n">batch_size</span>

<span class="k">def</span> <span class="nf">estimate_inference_memory</span><span class="p">(</span><span class="n">batch_size</span><span class="p">,</span> <span class="n">model_size</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Estimate memory usage for inference.&quot;&quot;&quot;</span>

    <span class="c1"># Model parameters</span>
    <span class="n">model_memory</span> <span class="o">=</span> <span class="n">model_size</span> <span class="o">*</span> <span class="mi">4</span>  <span class="c1"># 4 bytes per float32</span>

    <span class="c1"># Activations (approximate)</span>
    <span class="n">activation_memory</span> <span class="o">=</span> <span class="n">batch_size</span> <span class="o">*</span> <span class="mi">1000</span> <span class="o">*</span> <span class="mi">64</span> <span class="o">*</span> <span class="mi">4</span>  <span class="c1"># Assume 1000 nodes, 64 features</span>

    <span class="c1"># Total memory</span>
    <span class="n">total_memory</span> <span class="o">=</span> <span class="n">model_memory</span> <span class="o">+</span> <span class="n">activation_memory</span>

    <span class="k">return</span> <span class="n">total_memory</span>
</code></pre></div>
<h2 id="post-processing">Post-processing<a class="headerlink" href="#post-processing" title="Permanent link">&para;</a></h2>
<h3 id="result-aggregation">Result Aggregation<a class="headerlink" href="#result-aggregation" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">aggregate_inference_results</span><span class="p">(</span><span class="n">assignments</span><span class="p">,</span> <span class="n">fragments</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Aggregate inference results into final segmentation.&quot;&quot;&quot;</span>

    <span class="c1"># Combine cell assignments and fragment assignments</span>
    <span class="n">final_results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;transcript_id&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;final_cell_id&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;assignment_type&#39;</span><span class="p">:</span> <span class="p">[],</span>  <span class="c1"># &#39;cell&#39; or &#39;fragment&#39;</span>
        <span class="s1">&#39;confidence_score&#39;</span><span class="p">:</span> <span class="p">[]</span>
    <span class="p">}</span>

    <span class="c1"># Add cell assignments</span>
    <span class="k">for</span> <span class="n">tx_id</span><span class="p">,</span> <span class="n">cell_id</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">assigned</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">],</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">],</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;confidence_score&#39;</span><span class="p">],</span>
        <span class="n">assignments</span><span class="p">[</span><span class="s1">&#39;assigned&#39;</span><span class="p">]</span>
    <span class="p">):</span>
        <span class="k">if</span> <span class="n">assigned</span><span class="p">:</span>
            <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx_id</span><span class="p">)</span>
            <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;final_cell_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_id</span><span class="p">)</span>
            <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;assignment_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;cell&#39;</span><span class="p">)</span>
            <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;confidence_score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score</span><span class="p">)</span>

    <span class="c1"># Add fragment assignments</span>
    <span class="k">for</span> <span class="n">tx_id</span><span class="p">,</span> <span class="n">fragment_id</span> <span class="ow">in</span> <span class="n">fragments</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tx_id</span><span class="p">)</span>
        <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;final_cell_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fragment_id</span><span class="p">)</span>
        <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;assignment_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;fragment&#39;</span><span class="p">)</span>
        <span class="n">final_results</span><span class="p">[</span><span class="s1">&#39;confidence_score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>  <span class="c1"># No confidence for fragments</span>

    <span class="k">return</span> <span class="n">final_results</span>
</code></pre></div>
<h3 id="quality-control">Quality Control<a class="headerlink" href="#quality-control" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">quality_control_checks</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">min_transcripts_per_cell</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform quality control checks on inference results.&quot;&quot;&quot;</span>

    <span class="c1"># Count transcripts per cell</span>
    <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>
    <span class="n">cell_counts</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;final_cell_id&#39;</span><span class="p">])</span>

    <span class="c1"># Filter cells with too few transcripts</span>
    <span class="n">valid_cells</span> <span class="o">=</span> <span class="p">{</span><span class="n">cell_id</span><span class="p">:</span> <span class="n">count</span> <span class="k">for</span> <span class="n">cell_id</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">cell_counts</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> 
                   <span class="k">if</span> <span class="n">count</span> <span class="o">&gt;=</span> <span class="n">min_transcripts_per_cell</span><span class="p">}</span>

    <span class="c1"># Filter results</span>
    <span class="n">filtered_results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;transcript_id&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;final_cell_id&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;assignment_type&#39;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s1">&#39;confidence_score&#39;</span><span class="p">:</span> <span class="p">[]</span>
    <span class="p">}</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">cell_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;final_cell_id&#39;</span><span class="p">]):</span>
        <span class="k">if</span> <span class="n">cell_id</span> <span class="ow">in</span> <span class="n">valid_cells</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">filtered_results</span><span class="p">:</span>
                <span class="n">filtered_results</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="n">i</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">filtered_results</span><span class="p">,</span> <span class="n">valid_cells</span>
</code></pre></div>
<h2 id="output-formats">Output Formats<a class="headerlink" href="#output-formats" title="Permanent link">&para;</a></h2>
<h3 id="standard-output">Standard Output<a class="headerlink" href="#standard-output" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">save_inference_results</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;parquet&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save inference results in specified format.&quot;&quot;&quot;</span>

    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

    <span class="c1"># Convert to DataFrame</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;parquet&#39;</span><span class="p">:</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;csv&#39;</span><span class="p">:</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;h5ad&#39;</span><span class="p">:</span>
        <span class="c1"># Convert to AnnData format</span>
        <span class="n">adata</span> <span class="o">=</span> <span class="n">convert_to_anndata</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported format: </span><span class="si">{</span><span class="nb">format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Results saved to </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">convert_to_anndata</span><span class="p">(</span><span class="n">results_df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert results to AnnData format for downstream analysis.&quot;&quot;&quot;</span>

    <span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
    <span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>

    <span class="c1"># Group by cell ID</span>
    <span class="n">cell_groups</span> <span class="o">=</span> <span class="n">results_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;final_cell_id&#39;</span><span class="p">)</span>

    <span class="c1"># Create cell-gene matrix</span>
    <span class="n">cell_gene_matrix</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">cell_ids</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">cell_id</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">cell_groups</span><span class="p">:</span>
        <span class="c1"># Count transcripts per gene</span>
        <span class="n">gene_counts</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
        <span class="n">cell_gene_matrix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_counts</span><span class="p">)</span>
        <span class="n">cell_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_id</span><span class="p">)</span>

    <span class="c1"># Convert to DataFrame</span>
    <span class="n">cell_gene_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_gene_matrix</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">cell_ids</span><span class="p">)</span>
    <span class="n">cell_gene_df</span> <span class="o">=</span> <span class="n">cell_gene_df</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Create AnnData object</span>
    <span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">cell_gene_df</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> 
                       <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">cell_ids</span><span class="p">),</span>
                       <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">cell_gene_df</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">adata</span>
</code></pre></div>
<h2 id="performance-optimization">Performance Optimization<a class="headerlink" href="#performance-optimization" title="Permanent link">&para;</a></h2>
<h3 id="gpu-acceleration">GPU Acceleration<a class="headerlink" href="#gpu-acceleration" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">optimize_gpu_inference</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="s1">&#39;cuda&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Optimize GPU inference performance.&quot;&quot;&quot;</span>

    <span class="c1"># Move model and data to GPU</span>
    <span class="n">model</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>

    <span class="c1"># Enable CUDA optimizations</span>
    <span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">benchmark</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="c1"># Use mixed precision if available</span>
    <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">torch</span><span class="p">,</span> <span class="s1">&#39;autocast&#39;</span><span class="p">):</span>
        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">autocast</span><span class="p">(</span><span class="n">device_type</span><span class="o">=</span><span class="s1">&#39;cuda&#39;</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float16</span><span class="p">):</span>
            <span class="n">embeddings</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">edge_index</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">embeddings</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">edge_index</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">embeddings</span>
</code></pre></div>
<h3 id="parallel-processing">Parallel Processing<a class="headerlink" href="#parallel-processing" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">parallel_inference</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">data_list</span><span class="p">,</span> <span class="n">num_workers</span><span class="o">=</span><span class="mi">4</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform inference in parallel across multiple workers.&quot;&quot;&quot;</span>

    <span class="kn">from</span> <span class="nn">concurrent.futures</span> <span class="kn">import</span> <span class="n">ProcessPoolExecutor</span>
    <span class="kn">import</span> <span class="nn">multiprocessing</span> <span class="k">as</span> <span class="nn">mp</span>

    <span class="c1"># Set multiprocessing start method</span>
    <span class="n">mp</span><span class="o">.</span><span class="n">set_start_method</span><span class="p">(</span><span class="s1">&#39;spawn&#39;</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Split data across workers</span>
    <span class="n">chunk_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data_list</span><span class="p">)</span> <span class="o">//</span> <span class="n">num_workers</span>
    <span class="n">data_chunks</span> <span class="o">=</span> <span class="p">[</span><span class="n">data_list</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">chunk_size</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">data_list</span><span class="p">),</span> <span class="n">chunk_size</span><span class="p">)]</span>

    <span class="c1"># Process in parallel</span>
    <span class="k">with</span> <span class="n">ProcessPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="n">num_workers</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
        <span class="n">futures</span> <span class="o">=</span> <span class="p">[</span><span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="n">process_chunk</span><span class="p">,</span> <span class="n">model</span><span class="p">,</span> <span class="n">chunk</span><span class="p">)</span> <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">data_chunks</span><span class="p">]</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[</span><span class="n">future</span><span class="o">.</span><span class="n">result</span><span class="p">()</span> <span class="k">for</span> <span class="n">future</span> <span class="ow">in</span> <span class="n">futures</span><span class="p">]</span>

    <span class="c1"># Combine results</span>
    <span class="n">combined_results</span> <span class="o">=</span> <span class="n">combine_chunk_results</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">combined_results</span>
</code></pre></div>
<h2 id="troubleshooting">Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permanent link">&para;</a></h2>
<h3 id="common-issues">Common Issues<a class="headerlink" href="#common-issues" title="Permanent link">&para;</a></h3>
<h4 id="memory-errors">Memory Errors<a class="headerlink" href="#memory-errors" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Solutions:</span>
<span class="c1"># 1. Reduce batch size</span>
<span class="n">batch_size</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># Reduce from default</span>

<span class="c1"># 2. Use gradient checkpointing</span>
<span class="n">model</span><span class="o">.</span><span class="n">gradient_checkpointing_enable</span><span class="p">()</span>

<span class="c1"># 3. Clear GPU cache</span>
<span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">empty_cache</span><span class="p">()</span>
</code></pre></div>
<h4 id="slow-inference">Slow Inference<a class="headerlink" href="#slow-inference" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Solutions:</span>
<span class="c1"># 1. Enable CUDA optimizations</span>
<span class="n">torch</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">cudnn</span><span class="o">.</span><span class="n">benchmark</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># 2. Use mixed precision</span>
<span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">autocast</span><span class="p">(</span><span class="n">device_type</span><span class="o">=</span><span class="s1">&#39;cuda&#39;</span><span class="p">):</span>
    <span class="n">embeddings</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">x</span><span class="p">,</span> <span class="n">data</span><span class="o">.</span><span class="n">edge_index</span><span class="p">)</span>

<span class="c1"># 3. Optimize data loading</span>
<span class="n">data_loader</span> <span class="o">=</span> <span class="n">DataLoader</span><span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span> 
                        <span class="n">num_workers</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">pin_memory</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<h4 id="poor-quality-results">Poor Quality Results<a class="headerlink" href="#poor-quality-results" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Solutions:</span>
<span class="c1"># 1. Check model configuration matches training</span>
<span class="c1"># 2. Verify data preprocessing is identical</span>
<span class="c1"># 3. Adjust confidence threshold</span>
<span class="c1"># 4. Check for data distribution shifts</span>
</code></pre></div>
<h2 id="best-practices">Best Practices<a class="headerlink" href="#best-practices" title="Permanent link">&para;</a></h2>
<h3 id="inference-configuration">Inference Configuration<a class="headerlink" href="#inference-configuration" title="Permanent link">&para;</a></h3>
<ol>
<li><strong>Model Consistency</strong>: Ensure inference parameters match training exactly</li>
<li><strong>Data Preprocessing</strong>: Use identical preprocessing as training</li>
<li><strong>Confidence Thresholds</strong>: Start with recommended thresholds and adjust based on data</li>
<li><strong>Memory Management</strong>: Monitor GPU memory usage and optimize batch sizes</li>
</ol>
<h3 id="quality-assurance">Quality Assurance<a class="headerlink" href="#quality-assurance" title="Permanent link">&para;</a></h3>
<ol>
<li><strong>Validation Checks</strong>: Verify inference results against known ground truth</li>
<li><strong>Confidence Analysis</strong>: Analyze score distributions for appropriate thresholds</li>
<li><strong>Fragment Detection</strong>: Enable fragment detection for comprehensive coverage</li>
<li><strong>Post-processing</strong>: Apply quality control filters to remove low-quality assignments</li>
</ol>
<h3 id="performance-optimization_1">Performance Optimization<a class="headerlink" href="#performance-optimization_1" title="Permanent link">&para;</a></h3>
<ol>
<li><strong>GPU Utilization</strong>: Maximize GPU usage with appropriate batch sizes</li>
<li><strong>Parallel Processing</strong>: Use multiple workers for data loading</li>
<li><strong>Memory Efficiency</strong>: Optimize memory usage with mixed precision</li>
<li><strong>Batch Processing</strong>: Process large datasets in manageable chunks</li>
</ol>
<h2 id="future-enhancements">Future Enhancements<a class="headerlink" href="#future-enhancements" title="Permanent link">&para;</a></h2>
<p>Planned inference improvements include:</p>
<ul>
<li><strong>Real-time Inference</strong>: Streaming inference for live data</li>
<li><strong>Model Compression</strong>: Quantized models for faster inference</li>
<li><strong>Distributed Inference</strong>: Multi-node inference capabilities</li>
<li><strong>Adaptive Thresholding</strong>: Dynamic confidence thresholds based on data</li>
<li><strong>Uncertainty Quantification</strong>: Confidence intervals for predictions</li>
</ul>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.sections", "navigation.instant", "toc.integrate", "search.suggest", "search.highlight", "content.code.annotate", "content.tabs.link", "navigation.tracking"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.af256bd8.min.js"></script>
      
        <script src="../../../javascripts/config.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="https://unpkg.com/termynal@0.0.1/termynal.js"></script>
      
        <script src="../../../termynal.js"></script>
      
    
  </body>
</html>